{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [ACCOUNTING] post-dbNSFP filtering and local download\n",
    "\n",
    "## confirming that non-detected & detected CK positions are correct/ready to be made into .bed file for VEP processing!\n",
    "\n",
    "12.10.19\n",
    "\n",
    "### Brief reminder of dbNSFP filtering from searching old html files in 'MAPPING/HTML_before_clean/': \n",
    "- only kept rows where stable canonical uniprot ID was found (searched 3903...2 were chr multimapping so total 3901)\n",
    "    - 2 uniprot IDs were mapping to multiple chromosomes (error in bed file from uniprot)\n",
    "- lost rows that did not have valid hg19 coordinates, failed CADDhg19 annotation (2165 rows)\n",
    "- P42166 & P42167 are isoforms on chr12 that caused issues in dbnsfp, ID was combined for some rows because target position found in both IDs..i think i made combo rows into 'P42166'\n",
    "- BUG FIX entailed searching for canonical uniprot ID in dbNSFP col list, then using index, comparing AApos in list based on index\n",
    "    - result was equivalent AA position for found canonical uniprot ID!\n",
    "- final cleaning of merged dbNSFP (all chr merge), filtered out non missense based on aagain column like 'x', also AA positions with negative (i.e. C-1) splice variants\n",
    "\n",
    "## in K markdown:\n",
    "1. check that all uniprot IDs between detected and non detected files match\n",
    "2. check that all targets are found in subset of IDs\n",
    "3. check that all nontarget CK positions are accounted for by comparing to uniprot fasta raw CK totals\n",
    "4. check that all C and K codons have 7 nucleotide changes overlapping...allowed 2 positions to not have 7, both are nontarget positions..notes below. \n",
    "5. make bed files\n",
    "\n",
    "```python\n",
    "# from non-detected file, more detail at end of markdown\n",
    "\n",
    "    pos_ID\tcodon_changes(7)\n",
    "    Q86W56_K115\t  5\n",
    "    Q9Y6J0_C571\t  6\n",
    "    \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FINAL ACCOUNTING of filtered/checked CK positions for .bed file:\n",
    "* total uniprot IDs = 3840\n",
    "* all targets found with 7 changes per codon\n",
    "* only 2 CK non-targets missing and 2 aminos with incorrect codon change count\n",
    "\n",
    "| TYPE | Cysteine | Lysine | \n",
    "|-------|--------------|-----------|\n",
    "| Low | 895 | 3401 | \n",
    "| Medium | 412 | 660 |\n",
    "| High | 94 | 302 |\n",
    "| Panreactive | 5088 | 7804 |\n",
    "| TargetSM | 708 | 115 |\n",
    "| EVER-labeled | 6057 | 8868 |\n",
    "| not-labeled | 34050 | 140652 |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [1] check that detected and detected files with coordinates of Cysteine and Lysine have the same unique uniprot IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# packages\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from statistics import stdev \n",
    "from ast import literal_eval # for mismap_score func\n",
    "from statistics import mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container {width:90% !important;}</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display, HTML\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.options.display.max_seq_items = 2000\n",
    "display(HTML(\"<style>.container {width:90% !important;}</style>\"))\n",
    "sys.path.append(\"/Users/mariapalafox/Desktop/Toolbox\")\n",
    "from all_funx import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/BUG_FIX\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "det = pd.read_csv(\"detected_appeneded_merge.csv\", low_memory=False)\n",
    "notd = pd.read_csv(\"not_detected_appended.csv\", low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos_id19</th>\n",
       "      <th>pos_id38_x</th>\n",
       "      <th>aaref_x</th>\n",
       "      <th>aaalt_x</th>\n",
       "      <th>matched_aapos_x</th>\n",
       "      <th>CADD_phred_hg38</th>\n",
       "      <th>matched_UKBID_x</th>\n",
       "      <th>pos_dict</th>\n",
       "      <th>matched_target</th>\n",
       "      <th>CADD_phred_hg19</th>\n",
       "      <th>CADDdiff_38minus19</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>pos_ID</th>\n",
       "      <th>Cys_reactivity</th>\n",
       "      <th>Cys_react_threshold</th>\n",
       "      <th>Cys_target_label</th>\n",
       "      <th>Lys_reactivity</th>\n",
       "      <th>Lys_react_threshold</th>\n",
       "      <th>Lys_target_label</th>\n",
       "      <th>#chr</th>\n",
       "      <th>pos(1-based)</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>aaref_y</th>\n",
       "      <th>aaalt_y</th>\n",
       "      <th>rs_dbSNP151</th>\n",
       "      <th>hg19_chr</th>\n",
       "      <th>hg19_pos(1-based)</th>\n",
       "      <th>hg18_chr</th>\n",
       "      <th>hg18_pos(1-based)</th>\n",
       "      <th>aapos</th>\n",
       "      <th>genename</th>\n",
       "      <th>Ensembl_geneid</th>\n",
       "      <th>Ensembl_transcriptid</th>\n",
       "      <th>Ensembl_proteinid</th>\n",
       "      <th>Uniprot_acc</th>\n",
       "      <th>Uniprot_entry</th>\n",
       "      <th>HGVSc_ANNOVAR</th>\n",
       "      <th>HGVSp_ANNOVAR</th>\n",
       "      <th>HGVSc_snpEff</th>\n",
       "      <th>HGVSp_snpEff</th>\n",
       "      <th>HGVSc_VEP</th>\n",
       "      <th>HGVSp_VEP</th>\n",
       "      <th>APPRIS</th>\n",
       "      <th>GENCODE_basic</th>\n",
       "      <th>TSL</th>\n",
       "      <th>VEP_canonical</th>\n",
       "      <th>cds_strand</th>\n",
       "      <th>refcodon</th>\n",
       "      <th>codonpos</th>\n",
       "      <th>codon_degeneracy</th>\n",
       "      <th>Ancestral_allele</th>\n",
       "      <th>AltaiNeandertal</th>\n",
       "      <th>Denisova</th>\n",
       "      <th>VindijiaNeandertal</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>SIFT_converted_rankscore</th>\n",
       "      <th>SIFT_pred</th>\n",
       "      <th>SIFT4G_score</th>\n",
       "      <th>SIFT4G_converted_rankscore</th>\n",
       "      <th>SIFT4G_pred</th>\n",
       "      <th>Polyphen2_HDIV_score</th>\n",
       "      <th>Polyphen2_HDIV_rankscore</th>\n",
       "      <th>Polyphen2_HDIV_pred</th>\n",
       "      <th>Polyphen2_HVAR_score</th>\n",
       "      <th>Polyphen2_HVAR_rankscore</th>\n",
       "      <th>Polyphen2_HVAR_pred</th>\n",
       "      <th>LRT_score</th>\n",
       "      <th>LRT_converted_rankscore</th>\n",
       "      <th>LRT_pred</th>\n",
       "      <th>LRT_Omega</th>\n",
       "      <th>MutationTaster_score</th>\n",
       "      <th>MutationTaster_converted_rankscore</th>\n",
       "      <th>MutationTaster_pred</th>\n",
       "      <th>MutationTaster_model</th>\n",
       "      <th>MutationTaster_AAE</th>\n",
       "      <th>MutationAssessor_score</th>\n",
       "      <th>MutationAssessor_rankscore</th>\n",
       "      <th>MutationAssessor_pred</th>\n",
       "      <th>FATHMM_score</th>\n",
       "      <th>FATHMM_converted_rankscore</th>\n",
       "      <th>FATHMM_pred</th>\n",
       "      <th>PROVEAN_score</th>\n",
       "      <th>PROVEAN_converted_rankscore</th>\n",
       "      <th>PROVEAN_pred</th>\n",
       "      <th>VEST4_score</th>\n",
       "      <th>VEST4_rankscore</th>\n",
       "      <th>MetaSVM_score</th>\n",
       "      <th>MetaSVM_rankscore</th>\n",
       "      <th>MetaSVM_pred</th>\n",
       "      <th>MetaLR_score</th>\n",
       "      <th>MetaLR_rankscore</th>\n",
       "      <th>MetaLR_pred</th>\n",
       "      <th>Reliability_index</th>\n",
       "      <th>M-CAP_score</th>\n",
       "      <th>M-CAP_rankscore</th>\n",
       "      <th>M-CAP_pred</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>REVEL_rankscore</th>\n",
       "      <th>MutPred_score</th>\n",
       "      <th>MutPred_rankscore</th>\n",
       "      <th>MutPred_protID</th>\n",
       "      <th>MutPred_AAchange</th>\n",
       "      <th>MutPred_Top5features</th>\n",
       "      <th>MVP_score</th>\n",
       "      <th>MVP_rankscore</th>\n",
       "      <th>MPC_score</th>\n",
       "      <th>MPC_rankscore</th>\n",
       "      <th>PrimateAI_score</th>\n",
       "      <th>PrimateAI_rankscore</th>\n",
       "      <th>PrimateAI_pred</th>\n",
       "      <th>DEOGEN2_score</th>\n",
       "      <th>DEOGEN2_rankscore</th>\n",
       "      <th>DEOGEN2_pred</th>\n",
       "      <th>Aloft_Fraction_transcripts_affected</th>\n",
       "      <th>Aloft_prob_Tolerant</th>\n",
       "      <th>Aloft_prob_Recessive</th>\n",
       "      <th>Aloft_prob_Dominant</th>\n",
       "      <th>Aloft_pred</th>\n",
       "      <th>Aloft_Confidence</th>\n",
       "      <th>CADD_raw</th>\n",
       "      <th>CADD_raw_rankscore</th>\n",
       "      <th>CADD_phred</th>\n",
       "      <th>DANN_score</th>\n",
       "      <th>DANN_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_score</th>\n",
       "      <th>fathmm-MKL_coding_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_pred</th>\n",
       "      <th>fathmm-MKL_coding_group</th>\n",
       "      <th>fathmm-XF_coding_score</th>\n",
       "      <th>fathmm-XF_coding_rankscore</th>\n",
       "      <th>fathmm-XF_coding_pred</th>\n",
       "      <th>Eigen-raw_coding</th>\n",
       "      <th>Eigen-raw_coding_rankscore</th>\n",
       "      <th>Eigen-pred_coding</th>\n",
       "      <th>Eigen-PC-raw_coding</th>\n",
       "      <th>Eigen-PC-raw_coding_rankscore</th>\n",
       "      <th>Eigen-PC-phred_coding</th>\n",
       "      <th>GenoCanyon_score</th>\n",
       "      <th>GenoCanyon_rankscore</th>\n",
       "      <th>integrated_fitCons_score</th>\n",
       "      <th>integrated_fitCons_rankscore</th>\n",
       "      <th>integrated_confidence_value</th>\n",
       "      <th>GM12878_fitCons_score</th>\n",
       "      <th>GM12878_fitCons_rankscore</th>\n",
       "      <th>GM12878_confidence_value</th>\n",
       "      <th>H1-hESC_fitCons_score</th>\n",
       "      <th>H1-hESC_fitCons_rankscore</th>\n",
       "      <th>H1-hESC_confidence_value</th>\n",
       "      <th>HUVEC_fitCons_score</th>\n",
       "      <th>HUVEC_fitCons_rankscore</th>\n",
       "      <th>HUVEC_confidence_value</th>\n",
       "      <th>LINSIGHT</th>\n",
       "      <th>LINSIGHT_rankscore</th>\n",
       "      <th>GERP++_NR</th>\n",
       "      <th>GERP++_RS</th>\n",
       "      <th>GERP++_RS_rankscore</th>\n",
       "      <th>phyloP100way_vertebrate</th>\n",
       "      <th>phyloP100way_vertebrate_rankscore</th>\n",
       "      <th>phyloP30way_mammalian</th>\n",
       "      <th>phyloP30way_mammalian_rankscore</th>\n",
       "      <th>phyloP17way_primate</th>\n",
       "      <th>phyloP17way_primate_rankscore</th>\n",
       "      <th>phastCons100way_vertebrate</th>\n",
       "      <th>phastCons100way_vertebrate_rankscore</th>\n",
       "      <th>phastCons30way_mammalian</th>\n",
       "      <th>phastCons30way_mammalian_rankscore</th>\n",
       "      <th>phastCons17way_primate</th>\n",
       "      <th>phastCons17way_primate_rankscore</th>\n",
       "      <th>SiPhy_29way_pi</th>\n",
       "      <th>SiPhy_29way_logOdds</th>\n",
       "      <th>SiPhy_29way_logOdds_rankscore</th>\n",
       "      <th>bStatistic</th>\n",
       "      <th>bStatistic_rankscore</th>\n",
       "      <th>1000Gp3_AC</th>\n",
       "      <th>1000Gp3_AF</th>\n",
       "      <th>1000Gp3_AFR_AC</th>\n",
       "      <th>1000Gp3_AFR_AF</th>\n",
       "      <th>1000Gp3_EUR_AC</th>\n",
       "      <th>1000Gp3_EUR_AF</th>\n",
       "      <th>1000Gp3_AMR_AC</th>\n",
       "      <th>1000Gp3_AMR_AF</th>\n",
       "      <th>1000Gp3_EAS_AC</th>\n",
       "      <th>1000Gp3_EAS_AF</th>\n",
       "      <th>1000Gp3_SAS_AC</th>\n",
       "      <th>1000Gp3_SAS_AF</th>\n",
       "      <th>TWINSUK_AC</th>\n",
       "      <th>TWINSUK_AF</th>\n",
       "      <th>ALSPAC_AC</th>\n",
       "      <th>ALSPAC_AF</th>\n",
       "      <th>UK10K_AC</th>\n",
       "      <th>UK10K_AF</th>\n",
       "      <th>ESP6500_AA_AC</th>\n",
       "      <th>ESP6500_AA_AF</th>\n",
       "      <th>ESP6500_EA_AC</th>\n",
       "      <th>ESP6500_EA_AF</th>\n",
       "      <th>ExAC_AC</th>\n",
       "      <th>ExAC_AF</th>\n",
       "      <th>ExAC_Adj_AC</th>\n",
       "      <th>ExAC_Adj_AF</th>\n",
       "      <th>ExAC_AFR_AC</th>\n",
       "      <th>ExAC_AFR_AF</th>\n",
       "      <th>ExAC_AMR_AC</th>\n",
       "      <th>ExAC_AMR_AF</th>\n",
       "      <th>ExAC_EAS_AC</th>\n",
       "      <th>ExAC_EAS_AF</th>\n",
       "      <th>ExAC_FIN_AC</th>\n",
       "      <th>ExAC_FIN_AF</th>\n",
       "      <th>ExAC_NFE_AC</th>\n",
       "      <th>ExAC_NFE_AF</th>\n",
       "      <th>ExAC_SAS_AC</th>\n",
       "      <th>ExAC_SAS_AF</th>\n",
       "      <th>ExAC_nonTCGA_AC</th>\n",
       "      <th>ExAC_nonTCGA_AF</th>\n",
       "      <th>ExAC_nonTCGA_Adj_AC</th>\n",
       "      <th>ExAC_nonTCGA_Adj_AF</th>\n",
       "      <th>ExAC_nonTCGA_AFR_AC</th>\n",
       "      <th>ExAC_nonTCGA_AFR_AF</th>\n",
       "      <th>ExAC_nonTCGA_AMR_AC</th>\n",
       "      <th>ExAC_nonTCGA_AMR_AF</th>\n",
       "      <th>ExAC_nonTCGA_EAS_AC</th>\n",
       "      <th>ExAC_nonTCGA_EAS_AF</th>\n",
       "      <th>ExAC_nonTCGA_FIN_AC</th>\n",
       "      <th>ExAC_nonTCGA_FIN_AF</th>\n",
       "      <th>ExAC_nonTCGA_NFE_AC</th>\n",
       "      <th>ExAC_nonTCGA_NFE_AF</th>\n",
       "      <th>ExAC_nonTCGA_SAS_AC</th>\n",
       "      <th>ExAC_nonTCGA_SAS_AF</th>\n",
       "      <th>ExAC_nonpsych_AC</th>\n",
       "      <th>ExAC_nonpsych_AF</th>\n",
       "      <th>ExAC_nonpsych_Adj_AC</th>\n",
       "      <th>ExAC_nonpsych_Adj_AF</th>\n",
       "      <th>ExAC_nonpsych_AFR_AC</th>\n",
       "      <th>ExAC_nonpsych_AFR_AF</th>\n",
       "      <th>ExAC_nonpsych_AMR_AC</th>\n",
       "      <th>ExAC_nonpsych_AMR_AF</th>\n",
       "      <th>ExAC_nonpsych_EAS_AC</th>\n",
       "      <th>ExAC_nonpsych_EAS_AF</th>\n",
       "      <th>ExAC_nonpsych_FIN_AC</th>\n",
       "      <th>ExAC_nonpsych_FIN_AF</th>\n",
       "      <th>ExAC_nonpsych_NFE_AC</th>\n",
       "      <th>ExAC_nonpsych_NFE_AF</th>\n",
       "      <th>ExAC_nonpsych_SAS_AC</th>\n",
       "      <th>ExAC_nonpsych_SAS_AF</th>\n",
       "      <th>gnomAD_exomes_flag</th>\n",
       "      <th>gnomAD_exomes_AC</th>\n",
       "      <th>gnomAD_exomes_AN</th>\n",
       "      <th>gnomAD_exomes_AF</th>\n",
       "      <th>gnomAD_exomes_nhomalt</th>\n",
       "      <th>gnomAD_exomes_AFR_AC</th>\n",
       "      <th>gnomAD_exomes_AFR_AN</th>\n",
       "      <th>gnomAD_exomes_AFR_AF</th>\n",
       "      <th>gnomAD_exomes_AFR_nhomalt</th>\n",
       "      <th>gnomAD_exomes_AMR_AC</th>\n",
       "      <th>gnomAD_exomes_AMR_AN</th>\n",
       "      <th>gnomAD_exomes_AMR_AF</th>\n",
       "      <th>gnomAD_exomes_AMR_nhomalt</th>\n",
       "      <th>gnomAD_exomes_ASJ_AC</th>\n",
       "      <th>gnomAD_exomes_ASJ_AN</th>\n",
       "      <th>gnomAD_exomes_ASJ_AF</th>\n",
       "      <th>gnomAD_exomes_ASJ_nhomalt</th>\n",
       "      <th>gnomAD_exomes_EAS_AC</th>\n",
       "      <th>gnomAD_exomes_EAS_AN</th>\n",
       "      <th>gnomAD_exomes_EAS_AF</th>\n",
       "      <th>gnomAD_exomes_EAS_nhomalt</th>\n",
       "      <th>gnomAD_exomes_FIN_AC</th>\n",
       "      <th>gnomAD_exomes_FIN_AN</th>\n",
       "      <th>gnomAD_exomes_FIN_AF</th>\n",
       "      <th>gnomAD_exomes_FIN_nhomalt</th>\n",
       "      <th>gnomAD_exomes_NFE_AC</th>\n",
       "      <th>gnomAD_exomes_NFE_AN</th>\n",
       "      <th>gnomAD_exomes_NFE_AF</th>\n",
       "      <th>gnomAD_exomes_NFE_nhomalt</th>\n",
       "      <th>gnomAD_exomes_SAS_AC</th>\n",
       "      <th>gnomAD_exomes_SAS_AN</th>\n",
       "      <th>gnomAD_exomes_SAS_AF</th>\n",
       "      <th>gnomAD_exomes_SAS_nhomalt</th>\n",
       "      <th>gnomAD_exomes_POPMAX_AC</th>\n",
       "      <th>gnomAD_exomes_POPMAX_AN</th>\n",
       "      <th>gnomAD_exomes_POPMAX_AF</th>\n",
       "      <th>gnomAD_exomes_POPMAX_nhomalt</th>\n",
       "      <th>gnomAD_exomes_controls_AC</th>\n",
       "      <th>gnomAD_exomes_controls_AN</th>\n",
       "      <th>gnomAD_exomes_controls_AF</th>\n",
       "      <th>gnomAD_exomes_controls_nhomalt</th>\n",
       "      <th>gnomAD_exomes_controls_AFR_AC</th>\n",
       "      <th>gnomAD_exomes_controls_AFR_AN</th>\n",
       "      <th>gnomAD_exomes_controls_AFR_AF</th>\n",
       "      <th>gnomAD_exomes_controls_AFR_nhomalt</th>\n",
       "      <th>gnomAD_exomes_controls_AMR_AC</th>\n",
       "      <th>gnomAD_exomes_controls_AMR_AN</th>\n",
       "      <th>gnomAD_exomes_controls_AMR_AF</th>\n",
       "      <th>gnomAD_exomes_controls_AMR_nhomalt</th>\n",
       "      <th>gnomAD_exomes_controls_ASJ_AC</th>\n",
       "      <th>gnomAD_exomes_controls_ASJ_AN</th>\n",
       "      <th>gnomAD_exomes_controls_ASJ_AF</th>\n",
       "      <th>gnomAD_exomes_controls_ASJ_nhomalt</th>\n",
       "      <th>gnomAD_exomes_controls_EAS_AC</th>\n",
       "      <th>gnomAD_exomes_controls_EAS_AN</th>\n",
       "      <th>gnomAD_exomes_controls_EAS_AF</th>\n",
       "      <th>gnomAD_exomes_controls_EAS_nhomalt</th>\n",
       "      <th>gnomAD_exomes_controls_FIN_AC</th>\n",
       "      <th>gnomAD_exomes_controls_FIN_AN</th>\n",
       "      <th>gnomAD_exomes_controls_FIN_AF</th>\n",
       "      <th>gnomAD_exomes_controls_FIN_nhomalt</th>\n",
       "      <th>gnomAD_exomes_controls_NFE_AC</th>\n",
       "      <th>gnomAD_exomes_controls_NFE_AN</th>\n",
       "      <th>gnomAD_exomes_controls_NFE_AF</th>\n",
       "      <th>gnomAD_exomes_controls_NFE_nhomalt</th>\n",
       "      <th>gnomAD_exomes_controls_SAS_AC</th>\n",
       "      <th>gnomAD_exomes_controls_SAS_AN</th>\n",
       "      <th>gnomAD_exomes_controls_SAS_AF</th>\n",
       "      <th>gnomAD_exomes_controls_SAS_nhomalt</th>\n",
       "      <th>gnomAD_exomes_controls_POPMAX_AC</th>\n",
       "      <th>gnomAD_exomes_controls_POPMAX_AN</th>\n",
       "      <th>gnomAD_exomes_controls_POPMAX_AF</th>\n",
       "      <th>gnomAD_exomes_controls_POPMAX_nhomalt</th>\n",
       "      <th>gnomAD_genomes_flag</th>\n",
       "      <th>gnomAD_genomes_AC</th>\n",
       "      <th>gnomAD_genomes_AN</th>\n",
       "      <th>gnomAD_genomes_AF</th>\n",
       "      <th>gnomAD_genomes_nhomalt</th>\n",
       "      <th>gnomAD_genomes_AFR_AC</th>\n",
       "      <th>gnomAD_genomes_AFR_AN</th>\n",
       "      <th>gnomAD_genomes_AFR_AF</th>\n",
       "      <th>gnomAD_genomes_AFR_nhomalt</th>\n",
       "      <th>gnomAD_genomes_AMR_AC</th>\n",
       "      <th>gnomAD_genomes_AMR_AN</th>\n",
       "      <th>gnomAD_genomes_AMR_AF</th>\n",
       "      <th>gnomAD_genomes_AMR_nhomalt</th>\n",
       "      <th>gnomAD_genomes_ASJ_AC</th>\n",
       "      <th>gnomAD_genomes_ASJ_AN</th>\n",
       "      <th>gnomAD_genomes_ASJ_AF</th>\n",
       "      <th>gnomAD_genomes_ASJ_nhomalt</th>\n",
       "      <th>gnomAD_genomes_EAS_AC</th>\n",
       "      <th>gnomAD_genomes_EAS_AN</th>\n",
       "      <th>gnomAD_genomes_EAS_AF</th>\n",
       "      <th>gnomAD_genomes_EAS_nhomalt</th>\n",
       "      <th>gnomAD_genomes_FIN_AC</th>\n",
       "      <th>gnomAD_genomes_FIN_AN</th>\n",
       "      <th>gnomAD_genomes_FIN_AF</th>\n",
       "      <th>gnomAD_genomes_FIN_nhomalt</th>\n",
       "      <th>gnomAD_genomes_NFE_AC</th>\n",
       "      <th>gnomAD_genomes_NFE_AN</th>\n",
       "      <th>gnomAD_genomes_NFE_AF</th>\n",
       "      <th>gnomAD_genomes_NFE_nhomalt</th>\n",
       "      <th>gnomAD_genomes_POPMAX_AC</th>\n",
       "      <th>gnomAD_genomes_POPMAX_AN</th>\n",
       "      <th>gnomAD_genomes_POPMAX_AF</th>\n",
       "      <th>gnomAD_genomes_POPMAX_nhomalt</th>\n",
       "      <th>gnomAD_genomes_controls_AC</th>\n",
       "      <th>gnomAD_genomes_controls_AN</th>\n",
       "      <th>gnomAD_genomes_controls_AF</th>\n",
       "      <th>gnomAD_genomes_controls_nhomalt</th>\n",
       "      <th>gnomAD_genomes_controls_AFR_AC</th>\n",
       "      <th>gnomAD_genomes_controls_AFR_AN</th>\n",
       "      <th>gnomAD_genomes_controls_AFR_AF</th>\n",
       "      <th>gnomAD_genomes_controls_AFR_nhomalt</th>\n",
       "      <th>gnomAD_genomes_controls_AMR_AC</th>\n",
       "      <th>gnomAD_genomes_controls_AMR_AN</th>\n",
       "      <th>gnomAD_genomes_controls_AMR_AF</th>\n",
       "      <th>gnomAD_genomes_controls_AMR_nhomalt</th>\n",
       "      <th>gnomAD_genomes_controls_ASJ_AC</th>\n",
       "      <th>gnomAD_genomes_controls_ASJ_AN</th>\n",
       "      <th>gnomAD_genomes_controls_ASJ_AF</th>\n",
       "      <th>gnomAD_genomes_controls_ASJ_nhomalt</th>\n",
       "      <th>gnomAD_genomes_controls_EAS_AC</th>\n",
       "      <th>gnomAD_genomes_controls_EAS_AN</th>\n",
       "      <th>gnomAD_genomes_controls_EAS_AF</th>\n",
       "      <th>gnomAD_genomes_controls_EAS_nhomalt</th>\n",
       "      <th>gnomAD_genomes_controls_FIN_AC</th>\n",
       "      <th>gnomAD_genomes_controls_FIN_AN</th>\n",
       "      <th>gnomAD_genomes_controls_FIN_AF</th>\n",
       "      <th>gnomAD_genomes_controls_FIN_nhomalt</th>\n",
       "      <th>gnomAD_genomes_controls_NFE_AC</th>\n",
       "      <th>gnomAD_genomes_controls_NFE_AN</th>\n",
       "      <th>gnomAD_genomes_controls_NFE_AF</th>\n",
       "      <th>gnomAD_genomes_controls_NFE_nhomalt</th>\n",
       "      <th>gnomAD_genomes_controls_POPMAX_AC</th>\n",
       "      <th>gnomAD_genomes_controls_POPMAX_AN</th>\n",
       "      <th>gnomAD_genomes_controls_POPMAX_AF</th>\n",
       "      <th>gnomAD_genomes_controls_POPMAX_nhomalt</th>\n",
       "      <th>clinvar_id</th>\n",
       "      <th>clinvar_clnsig</th>\n",
       "      <th>clinvar_trait</th>\n",
       "      <th>clinvar_review</th>\n",
       "      <th>clinvar_hgvs</th>\n",
       "      <th>clinvar_var_source</th>\n",
       "      <th>clinvar_MedGen_id</th>\n",
       "      <th>clinvar_OMIM_id</th>\n",
       "      <th>clinvar_Orphanet_id</th>\n",
       "      <th>Interpro_domain</th>\n",
       "      <th>GTEx_V7_gene</th>\n",
       "      <th>GTEx_V7_tissue</th>\n",
       "      <th>Geuvadis_eQTL_target_gene</th>\n",
       "      <th>matched_UKBID_y</th>\n",
       "      <th>matched_aapos_y</th>\n",
       "      <th>matched_index</th>\n",
       "      <th>pos_id38_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>10_000093270_A_C</td>\n",
       "      <td>10_000047330_A_C</td>\n",
       "      <td>C</td>\n",
       "      <td>W</td>\n",
       "      <td>354</td>\n",
       "      <td>22.6</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>C354</td>\n",
       "      <td>22.30</td>\n",
       "      <td>0.30</td>\n",
       "      <td>Cys/Trp</td>\n",
       "      <td>Q3ZCM7_C354</td>\n",
       "      <td>5.37</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>47330</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>W</td>\n",
       "      <td>.</td>\n",
       "      <td>10</td>\n",
       "      <td>93270.0</td>\n",
       "      <td>10</td>\n",
       "      <td>83270</td>\n",
       "      <td>320;317;354</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>ENSG00000261456;ENSG00000261456;ENSG00000261456</td>\n",
       "      <td>ENST00000564130;ENST00000568866;ENST00000568584</td>\n",
       "      <td>ENSP00000457610;ENSP00000457062;ENSP00000456206</td>\n",
       "      <td>Q5SQY0;A0A075B736;Q3ZCM7</td>\n",
       "      <td>Q5SQY0_HUMAN;A0A075B736_HUMAN;TBB8_HUMAN</td>\n",
       "      <td>c.960T&gt;G;c.951T&gt;G;c.1062T&gt;G</td>\n",
       "      <td>p.C320W;p.C317W;p.C354W</td>\n",
       "      <td>c.960T&gt;G;c.951T&gt;G;c.1062T&gt;G</td>\n",
       "      <td>p.Cys320Trp;p.Cys317Trp;p.Cys354Trp</td>\n",
       "      <td>c.960T&gt;G;c.951T&gt;G;c.1062T&gt;G</td>\n",
       "      <td>p.Cys320Trp;p.Cys317Trp;p.Cys354Trp</td>\n",
       "      <td>.;.;principal1</td>\n",
       "      <td>Y;Y;Y</td>\n",
       "      <td>5;5;1</td>\n",
       "      <td>.;.;YES</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>./.</td>\n",
       "      <td>./.</td>\n",
       "      <td>./.</td>\n",
       "      <td>0.0;.;.</td>\n",
       "      <td>0.91255</td>\n",
       "      <td>D;.;.</td>\n",
       "      <td>0.0;0.0;0.0</td>\n",
       "      <td>0.92824</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>.;.;1.0</td>\n",
       "      <td>0.90584</td>\n",
       "      <td>.;.;D</td>\n",
       "      <td>.;.;0.998</td>\n",
       "      <td>0.88582</td>\n",
       "      <td>.;.;D</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.999882;0.999882;0.999882;0.999882;1</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D;D;D;D;D</td>\n",
       "      <td>simple_aae;simple_aae;simple_aae;simple_aae;wi...</td>\n",
       "      <td>C320W;C317W;C354W;C282W;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>-1.98;-1.98;-1.98</td>\n",
       "      <td>0.85247</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>-7.21;-6.69;-6.84</td>\n",
       "      <td>0.94249</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>0.753;0.756;0.751</td>\n",
       "      <td>0.75466</td>\n",
       "      <td>0.3131</td>\n",
       "      <td>0.87764</td>\n",
       "      <td>D</td>\n",
       "      <td>0.8180</td>\n",
       "      <td>0.93873</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.020603</td>\n",
       "      <td>0.43227</td>\n",
       "      <td>T</td>\n",
       "      <td>0.437</td>\n",
       "      <td>0.74446</td>\n",
       "      <td>0.868</td>\n",
       "      <td>0.96212</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>C354W</td>\n",
       "      <td>Gain of MoRF binding (P = 0.0706); Loss of she...</td>\n",
       "      <td>0.823156120182;0.823156120182;0.823156120182</td>\n",
       "      <td>0.82148</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.76919400692</td>\n",
       "      <td>0.77304</td>\n",
       "      <td>T</td>\n",
       "      <td>.;.;0.780824</td>\n",
       "      <td>0.94212</td>\n",
       "      <td>.;.;D</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>2.794161</td>\n",
       "      <td>0.41057</td>\n",
       "      <td>22.6</td>\n",
       "      <td>0.851523</td>\n",
       "      <td>0.15717</td>\n",
       "      <td>0.19276</td>\n",
       "      <td>0.20597</td>\n",
       "      <td>N</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.422079</td>\n",
       "      <td>0.48823</td>\n",
       "      <td>N</td>\n",
       "      <td>0.0406181179914726</td>\n",
       "      <td>0.43713</td>\n",
       "      <td>2.658815</td>\n",
       "      <td>-0.363070660159128</td>\n",
       "      <td>0.26108</td>\n",
       "      <td>1.439964</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.02871</td>\n",
       "      <td>0.553676</td>\n",
       "      <td>0.25195</td>\n",
       "      <td>0</td>\n",
       "      <td>0.573888</td>\n",
       "      <td>0.26702</td>\n",
       "      <td>0</td>\n",
       "      <td>0.618467</td>\n",
       "      <td>0.43123</td>\n",
       "      <td>0</td>\n",
       "      <td>0.564101</td>\n",
       "      <td>0.26826</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.135</td>\n",
       "      <td>0.15775</td>\n",
       "      <td>-0.446</td>\n",
       "      <td>0.05944</td>\n",
       "      <td>0.300</td>\n",
       "      <td>0.18932</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.42656</td>\n",
       "      <td>0.054</td>\n",
       "      <td>0.17266</td>\n",
       "      <td>0.107</td>\n",
       "      <td>0.18608</td>\n",
       "      <td>0.5123:0.0:0.4877:0.0</td>\n",
       "      <td>2.6478</td>\n",
       "      <td>0.04686</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Tubulin/FtsZ, 2-layer sandwich domain|Tubulin/...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>354</td>\n",
       "      <td>2</td>\n",
       "      <td>10_000047330_A_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>10_000093271_C_A</td>\n",
       "      <td>10_000047331_C_A</td>\n",
       "      <td>C</td>\n",
       "      <td>F</td>\n",
       "      <td>354</td>\n",
       "      <td>22.6</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>C354</td>\n",
       "      <td>21.80</td>\n",
       "      <td>0.80</td>\n",
       "      <td>Cys/Phe</td>\n",
       "      <td>Q3ZCM7_C354</td>\n",
       "      <td>5.37</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>47331</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>F</td>\n",
       "      <td>.</td>\n",
       "      <td>10</td>\n",
       "      <td>93271.0</td>\n",
       "      <td>10</td>\n",
       "      <td>83271</td>\n",
       "      <td>320;317;354</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>ENSG00000261456;ENSG00000261456;ENSG00000261456</td>\n",
       "      <td>ENST00000564130;ENST00000568866;ENST00000568584</td>\n",
       "      <td>ENSP00000457610;ENSP00000457062;ENSP00000456206</td>\n",
       "      <td>Q5SQY0;A0A075B736;Q3ZCM7</td>\n",
       "      <td>Q5SQY0_HUMAN;A0A075B736_HUMAN;TBB8_HUMAN</td>\n",
       "      <td>c.959G&gt;T;c.950G&gt;T;c.1061G&gt;T</td>\n",
       "      <td>p.C320F;p.C317F;p.C354F</td>\n",
       "      <td>c.959G&gt;T;c.950G&gt;T;c.1061G&gt;T</td>\n",
       "      <td>p.Cys320Phe;p.Cys317Phe;p.Cys354Phe</td>\n",
       "      <td>c.959G&gt;T;c.950G&gt;T;c.1061G&gt;T</td>\n",
       "      <td>p.Cys320Phe;p.Cys317Phe;p.Cys354Phe</td>\n",
       "      <td>.;.;principal1</td>\n",
       "      <td>Y;Y;Y</td>\n",
       "      <td>5;5;1</td>\n",
       "      <td>.;.;YES</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>./.</td>\n",
       "      <td>./.</td>\n",
       "      <td>./.</td>\n",
       "      <td>0.001;.;.</td>\n",
       "      <td>0.78490</td>\n",
       "      <td>D;.;.</td>\n",
       "      <td>0.009;0.013;0.013</td>\n",
       "      <td>0.66756</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>.;.;1.0</td>\n",
       "      <td>0.90584</td>\n",
       "      <td>.;.;D</td>\n",
       "      <td>.;.;0.998</td>\n",
       "      <td>0.88582</td>\n",
       "      <td>.;.;D</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.986821;0.986821;0.986821;0.986821;1</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D;D;D;D;D</td>\n",
       "      <td>simple_aae;simple_aae;simple_aae;simple_aae;wi...</td>\n",
       "      <td>C320F;C317F;C354F;C282F;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>-1.95;-1.95;-1.95</td>\n",
       "      <td>0.85003</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>-7.2;-6.67;-6.83</td>\n",
       "      <td>0.94223</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>0.779;0.782;0.782</td>\n",
       "      <td>0.77883</td>\n",
       "      <td>0.4271</td>\n",
       "      <td>0.89500</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7835</td>\n",
       "      <td>0.92649</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.015962</td>\n",
       "      <td>0.36975</td>\n",
       "      <td>T</td>\n",
       "      <td>0.374</td>\n",
       "      <td>0.69594</td>\n",
       "      <td>0.808</td>\n",
       "      <td>0.92509</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>C354F</td>\n",
       "      <td>Gain of sheet (P = 0.0827); Loss of methylatio...</td>\n",
       "      <td>0.788852735829;0.788852735829;0.788852735829</td>\n",
       "      <td>0.78689</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.727896571159</td>\n",
       "      <td>0.71199</td>\n",
       "      <td>T</td>\n",
       "      <td>.;.;0.744225</td>\n",
       "      <td>0.92941</td>\n",
       "      <td>.;.;D</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>2.814052</td>\n",
       "      <td>0.41528</td>\n",
       "      <td>22.6</td>\n",
       "      <td>0.895918</td>\n",
       "      <td>0.18929</td>\n",
       "      <td>0.57914</td>\n",
       "      <td>0.30502</td>\n",
       "      <td>D</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.537556</td>\n",
       "      <td>0.55531</td>\n",
       "      <td>D</td>\n",
       "      <td>0.0988420731411684</td>\n",
       "      <td>0.46409</td>\n",
       "      <td>2.88164</td>\n",
       "      <td>-0.263815814260368</td>\n",
       "      <td>0.29300</td>\n",
       "      <td>1.640415</td>\n",
       "      <td>0.006508</td>\n",
       "      <td>0.11235</td>\n",
       "      <td>0.553676</td>\n",
       "      <td>0.25195</td>\n",
       "      <td>0</td>\n",
       "      <td>0.573888</td>\n",
       "      <td>0.26702</td>\n",
       "      <td>0</td>\n",
       "      <td>0.618467</td>\n",
       "      <td>0.43123</td>\n",
       "      <td>0</td>\n",
       "      <td>0.564101</td>\n",
       "      <td>0.26826</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>5.571</td>\n",
       "      <td>0.67105</td>\n",
       "      <td>0.335</td>\n",
       "      <td>0.29199</td>\n",
       "      <td>0.245</td>\n",
       "      <td>0.18292</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.71638</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.17390</td>\n",
       "      <td>0.107</td>\n",
       "      <td>0.18608</td>\n",
       "      <td>0.0:0.9992:0.0:8.0E-4</td>\n",
       "      <td>5.9051</td>\n",
       "      <td>0.18235</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Tubulin/FtsZ, 2-layer sandwich domain|Tubulin/...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>354</td>\n",
       "      <td>2</td>\n",
       "      <td>10_000047331_C_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>10_000093271_C_G</td>\n",
       "      <td>10_000047331_C_G</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>354</td>\n",
       "      <td>22.4</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>C354</td>\n",
       "      <td>19.67</td>\n",
       "      <td>2.73</td>\n",
       "      <td>Cys/Ser</td>\n",
       "      <td>Q3ZCM7_C354</td>\n",
       "      <td>5.37</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>47331</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>.</td>\n",
       "      <td>10</td>\n",
       "      <td>93271.0</td>\n",
       "      <td>10</td>\n",
       "      <td>83271</td>\n",
       "      <td>320;317;354</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>ENSG00000261456;ENSG00000261456;ENSG00000261456</td>\n",
       "      <td>ENST00000564130;ENST00000568866;ENST00000568584</td>\n",
       "      <td>ENSP00000457610;ENSP00000457062;ENSP00000456206</td>\n",
       "      <td>Q5SQY0;A0A075B736;Q3ZCM7</td>\n",
       "      <td>Q5SQY0_HUMAN;A0A075B736_HUMAN;TBB8_HUMAN</td>\n",
       "      <td>c.959G&gt;C;c.950G&gt;C;c.1061G&gt;C</td>\n",
       "      <td>p.C320S;p.C317S;p.C354S</td>\n",
       "      <td>c.959G&gt;C;c.950G&gt;C;c.1061G&gt;C</td>\n",
       "      <td>p.Cys320Ser;p.Cys317Ser;p.Cys354Ser</td>\n",
       "      <td>c.959G&gt;C;c.950G&gt;C;c.1061G&gt;C</td>\n",
       "      <td>p.Cys320Ser;p.Cys317Ser;p.Cys354Ser</td>\n",
       "      <td>.;.;principal1</td>\n",
       "      <td>Y;Y;Y</td>\n",
       "      <td>5;5;1</td>\n",
       "      <td>.;.;YES</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>./.</td>\n",
       "      <td>./.</td>\n",
       "      <td>./.</td>\n",
       "      <td>0.0;.;.</td>\n",
       "      <td>0.91255</td>\n",
       "      <td>D;.;.</td>\n",
       "      <td>0.004;0.005;0.026</td>\n",
       "      <td>0.74150</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>.;.;0.997</td>\n",
       "      <td>0.70673</td>\n",
       "      <td>.;.;D</td>\n",
       "      <td>.;.;0.968</td>\n",
       "      <td>0.71741</td>\n",
       "      <td>.;.;D</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.930492;0.930492;0.930492;0.930492;1</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D;D;D;D;D</td>\n",
       "      <td>simple_aae;simple_aae;simple_aae;simple_aae;wi...</td>\n",
       "      <td>C320S;C317S;C354S;C282S;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>-1.87;-1.87;-1.87</td>\n",
       "      <td>0.84415</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>-6.55;-6.06;-6.21</td>\n",
       "      <td>0.91827</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>0.74;0.748;0.739</td>\n",
       "      <td>0.74735</td>\n",
       "      <td>0.4621</td>\n",
       "      <td>0.90007</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7769</td>\n",
       "      <td>0.92419</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.022348</td>\n",
       "      <td>0.45230</td>\n",
       "      <td>T</td>\n",
       "      <td>0.377</td>\n",
       "      <td>0.69846</td>\n",
       "      <td>0.781</td>\n",
       "      <td>0.90537</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>C354S</td>\n",
       "      <td>Gain of disorder (P = 0.0216); Gain of loop (P...</td>\n",
       "      <td>0.812824006922;0.812824006922;0.812824006922</td>\n",
       "      <td>0.81106</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.713165700436</td>\n",
       "      <td>0.69052</td>\n",
       "      <td>T</td>\n",
       "      <td>.;.;0.555385</td>\n",
       "      <td>0.85020</td>\n",
       "      <td>.;.;D</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>2.697199</td>\n",
       "      <td>0.38776</td>\n",
       "      <td>22.4</td>\n",
       "      <td>0.799045</td>\n",
       "      <td>0.12954</td>\n",
       "      <td>0.59480</td>\n",
       "      <td>0.30927</td>\n",
       "      <td>D</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.537556</td>\n",
       "      <td>0.55531</td>\n",
       "      <td>D</td>\n",
       "      <td>0.0966724300357427</td>\n",
       "      <td>0.46307</td>\n",
       "      <td>2.873107</td>\n",
       "      <td>-0.265973651554631</td>\n",
       "      <td>0.29227</td>\n",
       "      <td>1.635795</td>\n",
       "      <td>0.006508</td>\n",
       "      <td>0.11235</td>\n",
       "      <td>0.553676</td>\n",
       "      <td>0.25195</td>\n",
       "      <td>0</td>\n",
       "      <td>0.573888</td>\n",
       "      <td>0.26702</td>\n",
       "      <td>0</td>\n",
       "      <td>0.618467</td>\n",
       "      <td>0.43123</td>\n",
       "      <td>0</td>\n",
       "      <td>0.564101</td>\n",
       "      <td>0.26826</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>5.571</td>\n",
       "      <td>0.67105</td>\n",
       "      <td>0.335</td>\n",
       "      <td>0.29199</td>\n",
       "      <td>0.245</td>\n",
       "      <td>0.18292</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.71638</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.17390</td>\n",
       "      <td>0.107</td>\n",
       "      <td>0.18608</td>\n",
       "      <td>0.0:0.9992:0.0:8.0E-4</td>\n",
       "      <td>5.9051</td>\n",
       "      <td>0.18235</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Tubulin/FtsZ, 2-layer sandwich domain|Tubulin/...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>354</td>\n",
       "      <td>2</td>\n",
       "      <td>10_000047331_C_G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos_id19        pos_id38_x aaref_x aaalt_x  matched_aapos_x  \\\n",
       "0  10_000093270_A_C  10_000047330_A_C       C       W              354   \n",
       "1  10_000093271_C_A  10_000047331_C_A       C       F              354   \n",
       "2  10_000093271_C_G  10_000047331_C_G       C       S              354   \n",
       "\n",
       "   CADD_phred_hg38 matched_UKBID_x                                 pos_dict  \\\n",
       "0             22.6          Q3ZCM7  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}   \n",
       "1             22.6          Q3ZCM7  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}   \n",
       "2             22.4          Q3ZCM7  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}   \n",
       "\n",
       "  matched_target  CADD_phred_hg19  CADDdiff_38minus19 Amino_acids  \\\n",
       "0           C354            22.30                0.30     Cys/Trp   \n",
       "1           C354            21.80                0.80     Cys/Phe   \n",
       "2           C354            19.67                2.73     Cys/Ser   \n",
       "\n",
       "        pos_ID  Cys_reactivity Cys_react_threshold Cys_target_label  \\\n",
       "0  Q3ZCM7_C354            5.37                 Low              NaN   \n",
       "1  Q3ZCM7_C354            5.37                 Low              NaN   \n",
       "2  Q3ZCM7_C354            5.37                 Low              NaN   \n",
       "\n",
       "   Lys_reactivity Lys_react_threshold Lys_target_label #chr  pos(1-based) ref  \\\n",
       "0             NaN                 NaN              NaN   10         47330   A   \n",
       "1             NaN                 NaN              NaN   10         47331   C   \n",
       "2             NaN                 NaN              NaN   10         47331   C   \n",
       "\n",
       "  alt aaref_y aaalt_y rs_dbSNP151 hg19_chr  hg19_pos(1-based) hg18_chr  \\\n",
       "0   C       C       W           .       10            93270.0       10   \n",
       "1   A       C       F           .       10            93271.0       10   \n",
       "2   G       C       S           .       10            93271.0       10   \n",
       "\n",
       "  hg18_pos(1-based)        aapos genename  \\\n",
       "0             83270  320;317;354    TUBB8   \n",
       "1             83271  320;317;354    TUBB8   \n",
       "2             83271  320;317;354    TUBB8   \n",
       "\n",
       "                                    Ensembl_geneid  \\\n",
       "0  ENSG00000261456;ENSG00000261456;ENSG00000261456   \n",
       "1  ENSG00000261456;ENSG00000261456;ENSG00000261456   \n",
       "2  ENSG00000261456;ENSG00000261456;ENSG00000261456   \n",
       "\n",
       "                              Ensembl_transcriptid  \\\n",
       "0  ENST00000564130;ENST00000568866;ENST00000568584   \n",
       "1  ENST00000564130;ENST00000568866;ENST00000568584   \n",
       "2  ENST00000564130;ENST00000568866;ENST00000568584   \n",
       "\n",
       "                                 Ensembl_proteinid               Uniprot_acc  \\\n",
       "0  ENSP00000457610;ENSP00000457062;ENSP00000456206  Q5SQY0;A0A075B736;Q3ZCM7   \n",
       "1  ENSP00000457610;ENSP00000457062;ENSP00000456206  Q5SQY0;A0A075B736;Q3ZCM7   \n",
       "2  ENSP00000457610;ENSP00000457062;ENSP00000456206  Q5SQY0;A0A075B736;Q3ZCM7   \n",
       "\n",
       "                              Uniprot_entry                HGVSc_ANNOVAR  \\\n",
       "0  Q5SQY0_HUMAN;A0A075B736_HUMAN;TBB8_HUMAN  c.960T>G;c.951T>G;c.1062T>G   \n",
       "1  Q5SQY0_HUMAN;A0A075B736_HUMAN;TBB8_HUMAN  c.959G>T;c.950G>T;c.1061G>T   \n",
       "2  Q5SQY0_HUMAN;A0A075B736_HUMAN;TBB8_HUMAN  c.959G>C;c.950G>C;c.1061G>C   \n",
       "\n",
       "             HGVSp_ANNOVAR                 HGVSc_snpEff  \\\n",
       "0  p.C320W;p.C317W;p.C354W  c.960T>G;c.951T>G;c.1062T>G   \n",
       "1  p.C320F;p.C317F;p.C354F  c.959G>T;c.950G>T;c.1061G>T   \n",
       "2  p.C320S;p.C317S;p.C354S  c.959G>C;c.950G>C;c.1061G>C   \n",
       "\n",
       "                          HGVSp_snpEff                    HGVSc_VEP  \\\n",
       "0  p.Cys320Trp;p.Cys317Trp;p.Cys354Trp  c.960T>G;c.951T>G;c.1062T>G   \n",
       "1  p.Cys320Phe;p.Cys317Phe;p.Cys354Phe  c.959G>T;c.950G>T;c.1061G>T   \n",
       "2  p.Cys320Ser;p.Cys317Ser;p.Cys354Ser  c.959G>C;c.950G>C;c.1061G>C   \n",
       "\n",
       "                             HGVSp_VEP          APPRIS GENCODE_basic    TSL  \\\n",
       "0  p.Cys320Trp;p.Cys317Trp;p.Cys354Trp  .;.;principal1         Y;Y;Y  5;5;1   \n",
       "1  p.Cys320Phe;p.Cys317Phe;p.Cys354Phe  .;.;principal1         Y;Y;Y  5;5;1   \n",
       "2  p.Cys320Ser;p.Cys317Ser;p.Cys354Ser  .;.;principal1         Y;Y;Y  5;5;1   \n",
       "\n",
       "  VEP_canonical cds_strand refcodon  codonpos  codon_degeneracy  \\\n",
       "0       .;.;YES          -      TGT         3                 2   \n",
       "1       .;.;YES          -      TGT         2                 0   \n",
       "2       .;.;YES          -      TGT         2                 0   \n",
       "\n",
       "  Ancestral_allele AltaiNeandertal Denisova VindijiaNeandertal SIFT_score  \\\n",
       "0                -             ./.      ./.                ./.    0.0;.;.   \n",
       "1                -             ./.      ./.                ./.  0.001;.;.   \n",
       "2                -             ./.      ./.                ./.    0.0;.;.   \n",
       "\n",
       "  SIFT_converted_rankscore SIFT_pred       SIFT4G_score  \\\n",
       "0                  0.91255     D;.;.        0.0;0.0;0.0   \n",
       "1                  0.78490     D;.;.  0.009;0.013;0.013   \n",
       "2                  0.91255     D;.;.  0.004;0.005;0.026   \n",
       "\n",
       "  SIFT4G_converted_rankscore SIFT4G_pred Polyphen2_HDIV_score  \\\n",
       "0                    0.92824       D;D;D              .;.;1.0   \n",
       "1                    0.66756       D;D;D              .;.;1.0   \n",
       "2                    0.74150       D;D;D            .;.;0.997   \n",
       "\n",
       "  Polyphen2_HDIV_rankscore Polyphen2_HDIV_pred Polyphen2_HVAR_score  \\\n",
       "0                  0.90584               .;.;D            .;.;0.998   \n",
       "1                  0.90584               .;.;D            .;.;0.998   \n",
       "2                  0.70673               .;.;D            .;.;0.968   \n",
       "\n",
       "  Polyphen2_HVAR_rankscore Polyphen2_HVAR_pred LRT_score  \\\n",
       "0                  0.88582               .;.;D  0.000003   \n",
       "1                  0.88582               .;.;D  0.000003   \n",
       "2                  0.71741               .;.;D  0.000003   \n",
       "\n",
       "  LRT_converted_rankscore LRT_pred LRT_Omega  \\\n",
       "0                 0.62929        U  0.000000   \n",
       "1                 0.62929        U  0.000000   \n",
       "2                 0.62929        U  0.000000   \n",
       "\n",
       "                    MutationTaster_score MutationTaster_converted_rankscore  \\\n",
       "0  0.999882;0.999882;0.999882;0.999882;1                            0.81001   \n",
       "1  0.986821;0.986821;0.986821;0.986821;1                            0.81001   \n",
       "2  0.930492;0.930492;0.930492;0.930492;1                            0.81001   \n",
       "\n",
       "  MutationTaster_pred                               MutationTaster_model  \\\n",
       "0           D;D;D;D;D  simple_aae;simple_aae;simple_aae;simple_aae;wi...   \n",
       "1           D;D;D;D;D  simple_aae;simple_aae;simple_aae;simple_aae;wi...   \n",
       "2           D;D;D;D;D  simple_aae;simple_aae;simple_aae;simple_aae;wi...   \n",
       "\n",
       "          MutationTaster_AAE MutationAssessor_score  \\\n",
       "0  C320W;C317W;C354W;C282W;.                  .;.;.   \n",
       "1  C320F;C317F;C354F;C282F;.                  .;.;.   \n",
       "2  C320S;C317S;C354S;C282S;.                  .;.;.   \n",
       "\n",
       "  MutationAssessor_rankscore MutationAssessor_pred       FATHMM_score  \\\n",
       "0                          .                 .;.;.  -1.98;-1.98;-1.98   \n",
       "1                          .                 .;.;.  -1.95;-1.95;-1.95   \n",
       "2                          .                 .;.;.  -1.87;-1.87;-1.87   \n",
       "\n",
       "  FATHMM_converted_rankscore FATHMM_pred      PROVEAN_score  \\\n",
       "0                    0.85247       D;D;D  -7.21;-6.69;-6.84   \n",
       "1                    0.85003       D;D;D   -7.2;-6.67;-6.83   \n",
       "2                    0.84415       D;D;D  -6.55;-6.06;-6.21   \n",
       "\n",
       "  PROVEAN_converted_rankscore PROVEAN_pred        VEST4_score VEST4_rankscore  \\\n",
       "0                     0.94249        D;D;D  0.753;0.756;0.751         0.75466   \n",
       "1                     0.94223        D;D;D  0.779;0.782;0.782         0.77883   \n",
       "2                     0.91827        D;D;D   0.74;0.748;0.739         0.74735   \n",
       "\n",
       "  MetaSVM_score MetaSVM_rankscore MetaSVM_pred MetaLR_score MetaLR_rankscore  \\\n",
       "0        0.3131           0.87764            D       0.8180          0.93873   \n",
       "1        0.4271           0.89500            D       0.7835          0.92649   \n",
       "2        0.4621           0.90007            D       0.7769          0.92419   \n",
       "\n",
       "  MetaLR_pred Reliability_index M-CAP_score M-CAP_rankscore M-CAP_pred  \\\n",
       "0           D                 9    0.020603         0.43227          T   \n",
       "1           D                 9    0.015962         0.36975          T   \n",
       "2           D                 9    0.022348         0.45230          T   \n",
       "\n",
       "  REVEL_score REVEL_rankscore MutPred_score MutPred_rankscore MutPred_protID  \\\n",
       "0       0.437         0.74446         0.868           0.96212         Q3ZCM7   \n",
       "1       0.374         0.69594         0.808           0.92509         Q3ZCM7   \n",
       "2       0.377         0.69846         0.781           0.90537         Q3ZCM7   \n",
       "\n",
       "  MutPred_AAchange                               MutPred_Top5features  \\\n",
       "0            C354W  Gain of MoRF binding (P = 0.0706); Loss of she...   \n",
       "1            C354F  Gain of sheet (P = 0.0827); Loss of methylatio...   \n",
       "2            C354S  Gain of disorder (P = 0.0216); Gain of loop (P...   \n",
       "\n",
       "                                      MVP_score MVP_rankscore MPC_score  \\\n",
       "0  0.823156120182;0.823156120182;0.823156120182       0.82148     .;.;.   \n",
       "1  0.788852735829;0.788852735829;0.788852735829       0.78689     .;.;.   \n",
       "2  0.812824006922;0.812824006922;0.812824006922       0.81106     .;.;.   \n",
       "\n",
       "  MPC_rankscore PrimateAI_score PrimateAI_rankscore PrimateAI_pred  \\\n",
       "0             .   0.76919400692             0.77304              T   \n",
       "1             .  0.727896571159             0.71199              T   \n",
       "2             .  0.713165700436             0.69052              T   \n",
       "\n",
       "  DEOGEN2_score DEOGEN2_rankscore DEOGEN2_pred  \\\n",
       "0  .;.;0.780824           0.94212        .;.;D   \n",
       "1  .;.;0.744225           0.92941        .;.;D   \n",
       "2  .;.;0.555385           0.85020        .;.;D   \n",
       "\n",
       "  Aloft_Fraction_transcripts_affected Aloft_prob_Tolerant  \\\n",
       "0                               .;.;.               .;.;.   \n",
       "1                               .;.;.               .;.;.   \n",
       "2                               .;.;.               .;.;.   \n",
       "\n",
       "  Aloft_prob_Recessive Aloft_prob_Dominant Aloft_pred Aloft_Confidence  \\\n",
       "0                .;.;.               .;.;.      .;.;.            .;.;.   \n",
       "1                .;.;.               .;.;.      .;.;.            .;.;.   \n",
       "2                .;.;.               .;.;.      .;.;.            .;.;.   \n",
       "\n",
       "   CADD_raw  CADD_raw_rankscore  CADD_phred  DANN_score  DANN_rankscore  \\\n",
       "0  2.794161             0.41057        22.6    0.851523         0.15717   \n",
       "1  2.814052             0.41528        22.6    0.895918         0.18929   \n",
       "2  2.697199             0.38776        22.4    0.799045         0.12954   \n",
       "\n",
       "   fathmm-MKL_coding_score  fathmm-MKL_coding_rankscore  \\\n",
       "0                  0.19276                      0.20597   \n",
       "1                  0.57914                      0.30502   \n",
       "2                  0.59480                      0.30927   \n",
       "\n",
       "  fathmm-MKL_coding_pred fathmm-MKL_coding_group fathmm-XF_coding_score  \\\n",
       "0                      N                   AEFBI               0.422079   \n",
       "1                      D                   AEFBI               0.537556   \n",
       "2                      D                   AEFBI               0.537556   \n",
       "\n",
       "  fathmm-XF_coding_rankscore fathmm-XF_coding_pred    Eigen-raw_coding  \\\n",
       "0                    0.48823                     N  0.0406181179914726   \n",
       "1                    0.55531                     D  0.0988420731411684   \n",
       "2                    0.55531                     D  0.0966724300357427   \n",
       "\n",
       "  Eigen-raw_coding_rankscore Eigen-pred_coding Eigen-PC-raw_coding  \\\n",
       "0                    0.43713          2.658815  -0.363070660159128   \n",
       "1                    0.46409           2.88164  -0.263815814260368   \n",
       "2                    0.46307          2.873107  -0.265973651554631   \n",
       "\n",
       "  Eigen-PC-raw_coding_rankscore Eigen-PC-phred_coding  GenoCanyon_score  \\\n",
       "0                       0.26108              1.439964          0.000013   \n",
       "1                       0.29300              1.640415          0.006508   \n",
       "2                       0.29227              1.635795          0.006508   \n",
       "\n",
       "   GenoCanyon_rankscore integrated_fitCons_score integrated_fitCons_rankscore  \\\n",
       "0               0.02871                 0.553676                      0.25195   \n",
       "1               0.11235                 0.553676                      0.25195   \n",
       "2               0.11235                 0.553676                      0.25195   \n",
       "\n",
       "  integrated_confidence_value GM12878_fitCons_score GM12878_fitCons_rankscore  \\\n",
       "0                           0              0.573888                   0.26702   \n",
       "1                           0              0.573888                   0.26702   \n",
       "2                           0              0.573888                   0.26702   \n",
       "\n",
       "  GM12878_confidence_value H1-hESC_fitCons_score H1-hESC_fitCons_rankscore  \\\n",
       "0                        0              0.618467                   0.43123   \n",
       "1                        0              0.618467                   0.43123   \n",
       "2                        0              0.618467                   0.43123   \n",
       "\n",
       "  H1-hESC_confidence_value HUVEC_fitCons_score HUVEC_fitCons_rankscore  \\\n",
       "0                        0            0.564101                 0.26826   \n",
       "1                        0            0.564101                 0.26826   \n",
       "2                        0            0.564101                 0.26826   \n",
       "\n",
       "  HUVEC_confidence_value LINSIGHT LINSIGHT_rankscore GERP++_NR GERP++_RS  \\\n",
       "0                      0        .                  .         .         .   \n",
       "1                      0        .                  .         .         .   \n",
       "2                      0        .                  .         .         .   \n",
       "\n",
       "  GERP++_RS_rankscore  phyloP100way_vertebrate  \\\n",
       "0                   .                    0.135   \n",
       "1                   .                    5.571   \n",
       "2                   .                    5.571   \n",
       "\n",
       "   phyloP100way_vertebrate_rankscore  phyloP30way_mammalian  \\\n",
       "0                            0.15775                 -0.446   \n",
       "1                            0.67105                  0.335   \n",
       "2                            0.67105                  0.335   \n",
       "\n",
       "   phyloP30way_mammalian_rankscore  phyloP17way_primate  \\\n",
       "0                          0.05944                0.300   \n",
       "1                          0.29199                0.245   \n",
       "2                          0.29199                0.245   \n",
       "\n",
       "   phyloP17way_primate_rankscore  phastCons100way_vertebrate  \\\n",
       "0                        0.18932                       0.999   \n",
       "1                        0.18292                       1.000   \n",
       "2                        0.18292                       1.000   \n",
       "\n",
       "   phastCons100way_vertebrate_rankscore  phastCons30way_mammalian  \\\n",
       "0                               0.42656                     0.054   \n",
       "1                               0.71638                     0.056   \n",
       "2                               0.71638                     0.056   \n",
       "\n",
       "   phastCons30way_mammalian_rankscore  phastCons17way_primate  \\\n",
       "0                             0.17266                   0.107   \n",
       "1                             0.17390                   0.107   \n",
       "2                             0.17390                   0.107   \n",
       "\n",
       "   phastCons17way_primate_rankscore         SiPhy_29way_pi  \\\n",
       "0                           0.18608  0.5123:0.0:0.4877:0.0   \n",
       "1                           0.18608  0.0:0.9992:0.0:8.0E-4   \n",
       "2                           0.18608  0.0:0.9992:0.0:8.0E-4   \n",
       "\n",
       "  SiPhy_29way_logOdds SiPhy_29way_logOdds_rankscore bStatistic  \\\n",
       "0              2.6478                       0.04686          .   \n",
       "1              5.9051                       0.18235          .   \n",
       "2              5.9051                       0.18235          .   \n",
       "\n",
       "  bStatistic_rankscore 1000Gp3_AC 1000Gp3_AF 1000Gp3_AFR_AC 1000Gp3_AFR_AF  \\\n",
       "0                    .          .          .              .              .   \n",
       "1                    .          .          .              .              .   \n",
       "2                    .          .          .              .              .   \n",
       "\n",
       "  1000Gp3_EUR_AC 1000Gp3_EUR_AF 1000Gp3_AMR_AC 1000Gp3_AMR_AF 1000Gp3_EAS_AC  \\\n",
       "0              .              .              .              .              .   \n",
       "1              .              .              .              .              .   \n",
       "2              .              .              .              .              .   \n",
       "\n",
       "  1000Gp3_EAS_AF 1000Gp3_SAS_AC 1000Gp3_SAS_AF TWINSUK_AC TWINSUK_AF  \\\n",
       "0              .              .              .          .          .   \n",
       "1              .              .              .          .          .   \n",
       "2              .              .              .          .          .   \n",
       "\n",
       "  ALSPAC_AC ALSPAC_AF UK10K_AC UK10K_AF ESP6500_AA_AC ESP6500_AA_AF  \\\n",
       "0         .         .        .        .             .             .   \n",
       "1         .         .        .        .             .             .   \n",
       "2         .         .        .        .             .             .   \n",
       "\n",
       "  ESP6500_EA_AC ESP6500_EA_AF ExAC_AC ExAC_AF ExAC_Adj_AC ExAC_Adj_AF  \\\n",
       "0             .             .       .       .           .           .   \n",
       "1             .             .       .       .           .           .   \n",
       "2             .             .       .       .           .           .   \n",
       "\n",
       "  ExAC_AFR_AC ExAC_AFR_AF ExAC_AMR_AC ExAC_AMR_AF ExAC_EAS_AC ExAC_EAS_AF  \\\n",
       "0           .           .           .           .           .           .   \n",
       "1           .           .           .           .           .           .   \n",
       "2           .           .           .           .           .           .   \n",
       "\n",
       "  ExAC_FIN_AC ExAC_FIN_AF ExAC_NFE_AC ExAC_NFE_AF ExAC_SAS_AC ExAC_SAS_AF  \\\n",
       "0           .           .           .           .           .           .   \n",
       "1           .           .           .           .           .           .   \n",
       "2           .           .           .           .           .           .   \n",
       "\n",
       "  ExAC_nonTCGA_AC ExAC_nonTCGA_AF ExAC_nonTCGA_Adj_AC ExAC_nonTCGA_Adj_AF  \\\n",
       "0               .               .                   .                   .   \n",
       "1               .               .                   .                   .   \n",
       "2               .               .                   .                   .   \n",
       "\n",
       "  ExAC_nonTCGA_AFR_AC ExAC_nonTCGA_AFR_AF ExAC_nonTCGA_AMR_AC  \\\n",
       "0                   .                   .                   .   \n",
       "1                   .                   .                   .   \n",
       "2                   .                   .                   .   \n",
       "\n",
       "  ExAC_nonTCGA_AMR_AF ExAC_nonTCGA_EAS_AC ExAC_nonTCGA_EAS_AF  \\\n",
       "0                   .                   .                   .   \n",
       "1                   .                   .                   .   \n",
       "2                   .                   .                   .   \n",
       "\n",
       "  ExAC_nonTCGA_FIN_AC ExAC_nonTCGA_FIN_AF ExAC_nonTCGA_NFE_AC  \\\n",
       "0                   .                   .                   .   \n",
       "1                   .                   .                   .   \n",
       "2                   .                   .                   .   \n",
       "\n",
       "  ExAC_nonTCGA_NFE_AF ExAC_nonTCGA_SAS_AC ExAC_nonTCGA_SAS_AF  \\\n",
       "0                   .                   .                   .   \n",
       "1                   .                   .                   .   \n",
       "2                   .                   .                   .   \n",
       "\n",
       "  ExAC_nonpsych_AC ExAC_nonpsych_AF ExAC_nonpsych_Adj_AC ExAC_nonpsych_Adj_AF  \\\n",
       "0                .                .                    .                    .   \n",
       "1                .                .                    .                    .   \n",
       "2                .                .                    .                    .   \n",
       "\n",
       "  ExAC_nonpsych_AFR_AC ExAC_nonpsych_AFR_AF ExAC_nonpsych_AMR_AC  \\\n",
       "0                    .                    .                    .   \n",
       "1                    .                    .                    .   \n",
       "2                    .                    .                    .   \n",
       "\n",
       "  ExAC_nonpsych_AMR_AF ExAC_nonpsych_EAS_AC ExAC_nonpsych_EAS_AF  \\\n",
       "0                    .                    .                    .   \n",
       "1                    .                    .                    .   \n",
       "2                    .                    .                    .   \n",
       "\n",
       "  ExAC_nonpsych_FIN_AC ExAC_nonpsych_FIN_AF ExAC_nonpsych_NFE_AC  \\\n",
       "0                    .                    .                    .   \n",
       "1                    .                    .                    .   \n",
       "2                    .                    .                    .   \n",
       "\n",
       "  ExAC_nonpsych_NFE_AF ExAC_nonpsych_SAS_AC ExAC_nonpsych_SAS_AF  \\\n",
       "0                    .                    .                    .   \n",
       "1                    .                    .                    .   \n",
       "2                    .                    .                    .   \n",
       "\n",
       "  gnomAD_exomes_flag gnomAD_exomes_AC gnomAD_exomes_AN gnomAD_exomes_AF  \\\n",
       "0                  .                .                .                .   \n",
       "1                  .                .                .                .   \n",
       "2                  .                .                .                .   \n",
       "\n",
       "  gnomAD_exomes_nhomalt gnomAD_exomes_AFR_AC gnomAD_exomes_AFR_AN  \\\n",
       "0                     .                    .                    .   \n",
       "1                     .                    .                    .   \n",
       "2                     .                    .                    .   \n",
       "\n",
       "  gnomAD_exomes_AFR_AF gnomAD_exomes_AFR_nhomalt gnomAD_exomes_AMR_AC  \\\n",
       "0                    .                         .                    .   \n",
       "1                    .                         .                    .   \n",
       "2                    .                         .                    .   \n",
       "\n",
       "  gnomAD_exomes_AMR_AN gnomAD_exomes_AMR_AF gnomAD_exomes_AMR_nhomalt  \\\n",
       "0                    .                    .                         .   \n",
       "1                    .                    .                         .   \n",
       "2                    .                    .                         .   \n",
       "\n",
       "  gnomAD_exomes_ASJ_AC gnomAD_exomes_ASJ_AN gnomAD_exomes_ASJ_AF  \\\n",
       "0                    .                    .                    .   \n",
       "1                    .                    .                    .   \n",
       "2                    .                    .                    .   \n",
       "\n",
       "  gnomAD_exomes_ASJ_nhomalt gnomAD_exomes_EAS_AC gnomAD_exomes_EAS_AN  \\\n",
       "0                         .                    .                    .   \n",
       "1                         .                    .                    .   \n",
       "2                         .                    .                    .   \n",
       "\n",
       "  gnomAD_exomes_EAS_AF gnomAD_exomes_EAS_nhomalt gnomAD_exomes_FIN_AC  \\\n",
       "0                    .                         .                    .   \n",
       "1                    .                         .                    .   \n",
       "2                    .                         .                    .   \n",
       "\n",
       "  gnomAD_exomes_FIN_AN gnomAD_exomes_FIN_AF gnomAD_exomes_FIN_nhomalt  \\\n",
       "0                    .                    .                         .   \n",
       "1                    .                    .                         .   \n",
       "2                    .                    .                         .   \n",
       "\n",
       "  gnomAD_exomes_NFE_AC gnomAD_exomes_NFE_AN gnomAD_exomes_NFE_AF  \\\n",
       "0                    .                    .                    .   \n",
       "1                    .                    .                    .   \n",
       "2                    .                    .                    .   \n",
       "\n",
       "  gnomAD_exomes_NFE_nhomalt gnomAD_exomes_SAS_AC gnomAD_exomes_SAS_AN  \\\n",
       "0                         .                    .                    .   \n",
       "1                         .                    .                    .   \n",
       "2                         .                    .                    .   \n",
       "\n",
       "  gnomAD_exomes_SAS_AF gnomAD_exomes_SAS_nhomalt gnomAD_exomes_POPMAX_AC  \\\n",
       "0                    .                         .                       .   \n",
       "1                    .                         .                       .   \n",
       "2                    .                         .                       .   \n",
       "\n",
       "  gnomAD_exomes_POPMAX_AN gnomAD_exomes_POPMAX_AF  \\\n",
       "0                       .                       .   \n",
       "1                       .                       .   \n",
       "2                       .                       .   \n",
       "\n",
       "  gnomAD_exomes_POPMAX_nhomalt gnomAD_exomes_controls_AC  \\\n",
       "0                            .                         .   \n",
       "1                            .                         .   \n",
       "2                            .                         .   \n",
       "\n",
       "  gnomAD_exomes_controls_AN gnomAD_exomes_controls_AF  \\\n",
       "0                         .                         .   \n",
       "1                         .                         .   \n",
       "2                         .                         .   \n",
       "\n",
       "  gnomAD_exomes_controls_nhomalt gnomAD_exomes_controls_AFR_AC  \\\n",
       "0                              .                             .   \n",
       "1                              .                             .   \n",
       "2                              .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_AFR_AN gnomAD_exomes_controls_AFR_AF  \\\n",
       "0                             .                             .   \n",
       "1                             .                             .   \n",
       "2                             .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_AFR_nhomalt gnomAD_exomes_controls_AMR_AC  \\\n",
       "0                                  .                             .   \n",
       "1                                  .                             .   \n",
       "2                                  .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_AMR_AN gnomAD_exomes_controls_AMR_AF  \\\n",
       "0                             .                             .   \n",
       "1                             .                             .   \n",
       "2                             .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_AMR_nhomalt gnomAD_exomes_controls_ASJ_AC  \\\n",
       "0                                  .                             .   \n",
       "1                                  .                             .   \n",
       "2                                  .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_ASJ_AN gnomAD_exomes_controls_ASJ_AF  \\\n",
       "0                             .                             .   \n",
       "1                             .                             .   \n",
       "2                             .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_ASJ_nhomalt gnomAD_exomes_controls_EAS_AC  \\\n",
       "0                                  .                             .   \n",
       "1                                  .                             .   \n",
       "2                                  .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_EAS_AN gnomAD_exomes_controls_EAS_AF  \\\n",
       "0                             .                             .   \n",
       "1                             .                             .   \n",
       "2                             .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_EAS_nhomalt gnomAD_exomes_controls_FIN_AC  \\\n",
       "0                                  .                             .   \n",
       "1                                  .                             .   \n",
       "2                                  .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_FIN_AN gnomAD_exomes_controls_FIN_AF  \\\n",
       "0                             .                             .   \n",
       "1                             .                             .   \n",
       "2                             .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_FIN_nhomalt gnomAD_exomes_controls_NFE_AC  \\\n",
       "0                                  .                             .   \n",
       "1                                  .                             .   \n",
       "2                                  .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_NFE_AN gnomAD_exomes_controls_NFE_AF  \\\n",
       "0                             .                             .   \n",
       "1                             .                             .   \n",
       "2                             .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_NFE_nhomalt gnomAD_exomes_controls_SAS_AC  \\\n",
       "0                                  .                             .   \n",
       "1                                  .                             .   \n",
       "2                                  .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_SAS_AN gnomAD_exomes_controls_SAS_AF  \\\n",
       "0                             .                             .   \n",
       "1                             .                             .   \n",
       "2                             .                             .   \n",
       "\n",
       "  gnomAD_exomes_controls_SAS_nhomalt gnomAD_exomes_controls_POPMAX_AC  \\\n",
       "0                                  .                                .   \n",
       "1                                  .                                .   \n",
       "2                                  .                                .   \n",
       "\n",
       "  gnomAD_exomes_controls_POPMAX_AN gnomAD_exomes_controls_POPMAX_AF  \\\n",
       "0                                .                                .   \n",
       "1                                .                                .   \n",
       "2                                .                                .   \n",
       "\n",
       "  gnomAD_exomes_controls_POPMAX_nhomalt gnomAD_genomes_flag gnomAD_genomes_AC  \\\n",
       "0                                     .                   .                 .   \n",
       "1                                     .                   .                 .   \n",
       "2                                     .                   .                 .   \n",
       "\n",
       "  gnomAD_genomes_AN gnomAD_genomes_AF gnomAD_genomes_nhomalt  \\\n",
       "0                 .                 .                      .   \n",
       "1                 .                 .                      .   \n",
       "2                 .                 .                      .   \n",
       "\n",
       "  gnomAD_genomes_AFR_AC gnomAD_genomes_AFR_AN gnomAD_genomes_AFR_AF  \\\n",
       "0                     .                     .                     .   \n",
       "1                     .                     .                     .   \n",
       "2                     .                     .                     .   \n",
       "\n",
       "  gnomAD_genomes_AFR_nhomalt gnomAD_genomes_AMR_AC gnomAD_genomes_AMR_AN  \\\n",
       "0                          .                     .                     .   \n",
       "1                          .                     .                     .   \n",
       "2                          .                     .                     .   \n",
       "\n",
       "  gnomAD_genomes_AMR_AF gnomAD_genomes_AMR_nhomalt gnomAD_genomes_ASJ_AC  \\\n",
       "0                     .                          .                     .   \n",
       "1                     .                          .                     .   \n",
       "2                     .                          .                     .   \n",
       "\n",
       "  gnomAD_genomes_ASJ_AN gnomAD_genomes_ASJ_AF gnomAD_genomes_ASJ_nhomalt  \\\n",
       "0                     .                     .                          .   \n",
       "1                     .                     .                          .   \n",
       "2                     .                     .                          .   \n",
       "\n",
       "  gnomAD_genomes_EAS_AC gnomAD_genomes_EAS_AN gnomAD_genomes_EAS_AF  \\\n",
       "0                     .                     .                     .   \n",
       "1                     .                     .                     .   \n",
       "2                     .                     .                     .   \n",
       "\n",
       "  gnomAD_genomes_EAS_nhomalt gnomAD_genomes_FIN_AC gnomAD_genomes_FIN_AN  \\\n",
       "0                          .                     .                     .   \n",
       "1                          .                     .                     .   \n",
       "2                          .                     .                     .   \n",
       "\n",
       "  gnomAD_genomes_FIN_AF gnomAD_genomes_FIN_nhomalt gnomAD_genomes_NFE_AC  \\\n",
       "0                     .                          .                     .   \n",
       "1                     .                          .                     .   \n",
       "2                     .                          .                     .   \n",
       "\n",
       "  gnomAD_genomes_NFE_AN gnomAD_genomes_NFE_AF gnomAD_genomes_NFE_nhomalt  \\\n",
       "0                     .                     .                          .   \n",
       "1                     .                     .                          .   \n",
       "2                     .                     .                          .   \n",
       "\n",
       "  gnomAD_genomes_POPMAX_AC gnomAD_genomes_POPMAX_AN gnomAD_genomes_POPMAX_AF  \\\n",
       "0                        .                        .                        .   \n",
       "1                        .                        .                        .   \n",
       "2                        .                        .                        .   \n",
       "\n",
       "  gnomAD_genomes_POPMAX_nhomalt gnomAD_genomes_controls_AC  \\\n",
       "0                             .                          .   \n",
       "1                             .                          .   \n",
       "2                             .                          .   \n",
       "\n",
       "  gnomAD_genomes_controls_AN gnomAD_genomes_controls_AF  \\\n",
       "0                          .                          .   \n",
       "1                          .                          .   \n",
       "2                          .                          .   \n",
       "\n",
       "  gnomAD_genomes_controls_nhomalt gnomAD_genomes_controls_AFR_AC  \\\n",
       "0                               .                              .   \n",
       "1                               .                              .   \n",
       "2                               .                              .   \n",
       "\n",
       "  gnomAD_genomes_controls_AFR_AN gnomAD_genomes_controls_AFR_AF  \\\n",
       "0                              .                              .   \n",
       "1                              .                              .   \n",
       "2                              .                              .   \n",
       "\n",
       "  gnomAD_genomes_controls_AFR_nhomalt gnomAD_genomes_controls_AMR_AC  \\\n",
       "0                                   .                              .   \n",
       "1                                   .                              .   \n",
       "2                                   .                              .   \n",
       "\n",
       "  gnomAD_genomes_controls_AMR_AN gnomAD_genomes_controls_AMR_AF  \\\n",
       "0                              .                              .   \n",
       "1                              .                              .   \n",
       "2                              .                              .   \n",
       "\n",
       "  gnomAD_genomes_controls_AMR_nhomalt gnomAD_genomes_controls_ASJ_AC  \\\n",
       "0                                   .                              .   \n",
       "1                                   .                              .   \n",
       "2                                   .                              .   \n",
       "\n",
       "  gnomAD_genomes_controls_ASJ_AN gnomAD_genomes_controls_ASJ_AF  \\\n",
       "0                              .                              .   \n",
       "1                              .                              .   \n",
       "2                              .                              .   \n",
       "\n",
       "  gnomAD_genomes_controls_ASJ_nhomalt gnomAD_genomes_controls_EAS_AC  \\\n",
       "0                                   .                              .   \n",
       "1                                   .                              .   \n",
       "2                                   .                              .   \n",
       "\n",
       "  gnomAD_genomes_controls_EAS_AN gnomAD_genomes_controls_EAS_AF  \\\n",
       "0                              .                              .   \n",
       "1                              .                              .   \n",
       "2                              .                              .   \n",
       "\n",
       "  gnomAD_genomes_controls_EAS_nhomalt gnomAD_genomes_controls_FIN_AC  \\\n",
       "0                                   .                              .   \n",
       "1                                   .                              .   \n",
       "2                                   .                              .   \n",
       "\n",
       "  gnomAD_genomes_controls_FIN_AN gnomAD_genomes_controls_FIN_AF  \\\n",
       "0                              .                              .   \n",
       "1                              .                              .   \n",
       "2                              .                              .   \n",
       "\n",
       "  gnomAD_genomes_controls_FIN_nhomalt gnomAD_genomes_controls_NFE_AC  \\\n",
       "0                                   .                              .   \n",
       "1                                   .                              .   \n",
       "2                                   .                              .   \n",
       "\n",
       "  gnomAD_genomes_controls_NFE_AN gnomAD_genomes_controls_NFE_AF  \\\n",
       "0                              .                              .   \n",
       "1                              .                              .   \n",
       "2                              .                              .   \n",
       "\n",
       "  gnomAD_genomes_controls_NFE_nhomalt gnomAD_genomes_controls_POPMAX_AC  \\\n",
       "0                                   .                                 .   \n",
       "1                                   .                                 .   \n",
       "2                                   .                                 .   \n",
       "\n",
       "  gnomAD_genomes_controls_POPMAX_AN gnomAD_genomes_controls_POPMAX_AF  \\\n",
       "0                                 .                                 .   \n",
       "1                                 .                                 .   \n",
       "2                                 .                                 .   \n",
       "\n",
       "  gnomAD_genomes_controls_POPMAX_nhomalt clinvar_id clinvar_clnsig  \\\n",
       "0                                      .          .              .   \n",
       "1                                      .          .              .   \n",
       "2                                      .          .              .   \n",
       "\n",
       "  clinvar_trait clinvar_review clinvar_hgvs clinvar_var_source  \\\n",
       "0             .              .            .                  .   \n",
       "1             .              .            .                  .   \n",
       "2             .              .            .                  .   \n",
       "\n",
       "  clinvar_MedGen_id clinvar_OMIM_id clinvar_Orphanet_id  \\\n",
       "0                 .               .                   .   \n",
       "1                 .               .                   .   \n",
       "2                 .               .                   .   \n",
       "\n",
       "                                     Interpro_domain GTEx_V7_gene  \\\n",
       "0  Tubulin/FtsZ, 2-layer sandwich domain|Tubulin/...            .   \n",
       "1  Tubulin/FtsZ, 2-layer sandwich domain|Tubulin/...            .   \n",
       "2  Tubulin/FtsZ, 2-layer sandwich domain|Tubulin/...            .   \n",
       "\n",
       "  GTEx_V7_tissue Geuvadis_eQTL_target_gene matched_UKBID_y  matched_aapos_y  \\\n",
       "0              .                         .          Q3ZCM7              354   \n",
       "1              .                         .          Q3ZCM7              354   \n",
       "2              .                         .          Q3ZCM7              354   \n",
       "\n",
       "   matched_index        pos_id38_y  \n",
       "0              2  10_000047330_A_C  \n",
       "1              2  10_000047331_C_A  \n",
       "2              2  10_000047331_C_G  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "det.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos_id19</th>\n",
       "      <th>pos_id38</th>\n",
       "      <th>aaref</th>\n",
       "      <th>aaalt</th>\n",
       "      <th>matched_aapos</th>\n",
       "      <th>HGVSp_VEP</th>\n",
       "      <th>CADD_phred_hg38</th>\n",
       "      <th>matched_UKBID</th>\n",
       "      <th>pos_dict</th>\n",
       "      <th>matched_target</th>\n",
       "      <th>CADD_phred_hg19</th>\n",
       "      <th>CADDdiff_38minus19</th>\n",
       "      <th>lost_amino</th>\n",
       "      <th>gained_amino</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>pos_ID_falseCKtarget</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>10_000093156_C_A</td>\n",
       "      <td>10_000047216_C_A</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>392</td>\n",
       "      <td>p.Lys358Asn;p.Lys355Asn;p.Lys392Asn</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>False</td>\n",
       "      <td>22.9</td>\n",
       "      <td>0.1</td>\n",
       "      <td>Lys</td>\n",
       "      <td>Asn</td>\n",
       "      <td>Lys/Asn</td>\n",
       "      <td>Q3ZCM7_K392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>10_000093156_C_G</td>\n",
       "      <td>10_000047216_C_G</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>392</td>\n",
       "      <td>p.Lys358Asn;p.Lys355Asn;p.Lys392Asn</td>\n",
       "      <td>23.1</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>False</td>\n",
       "      <td>22.8</td>\n",
       "      <td>0.3</td>\n",
       "      <td>Lys</td>\n",
       "      <td>Asn</td>\n",
       "      <td>Lys/Asn</td>\n",
       "      <td>Q3ZCM7_K392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>10_000093157_T_A</td>\n",
       "      <td>10_000047217_T_A</td>\n",
       "      <td>K</td>\n",
       "      <td>M</td>\n",
       "      <td>392</td>\n",
       "      <td>p.Lys358Met;p.Lys355Met;p.Lys392Met</td>\n",
       "      <td>23.2</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>False</td>\n",
       "      <td>23.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Lys</td>\n",
       "      <td>Met</td>\n",
       "      <td>Lys/Met</td>\n",
       "      <td>Q3ZCM7_K392</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos_id19          pos_id38 aaref aaalt  matched_aapos  \\\n",
       "0  10_000093156_C_A  10_000047216_C_A     K     N            392   \n",
       "1  10_000093156_C_G  10_000047216_C_G     K     N            392   \n",
       "2  10_000093157_T_A  10_000047217_T_A     K     M            392   \n",
       "\n",
       "                             HGVSp_VEP  CADD_phred_hg38 matched_UKBID  \\\n",
       "0  p.Lys358Asn;p.Lys355Asn;p.Lys392Asn             23.0        Q3ZCM7   \n",
       "1  p.Lys358Asn;p.Lys355Asn;p.Lys392Asn             23.1        Q3ZCM7   \n",
       "2  p.Lys358Met;p.Lys355Met;p.Lys392Met             23.2        Q3ZCM7   \n",
       "\n",
       "                                  pos_dict  matched_target  CADD_phred_hg19  \\\n",
       "0  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}           False             22.9   \n",
       "1  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}           False             22.8   \n",
       "2  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}           False             23.2   \n",
       "\n",
       "   CADDdiff_38minus19 lost_amino gained_amino Amino_acids pos_ID_falseCKtarget  \n",
       "0                 0.1        Lys          Asn     Lys/Asn          Q3ZCM7_K392  \n",
       "1                 0.3        Lys          Asn     Lys/Asn          Q3ZCM7_K392  \n",
       "2                 0.0        Lys          Met     Lys/Met          Q3ZCM7_K392  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "notd.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len of col:  106064\n",
      "len of col set:  3880\n",
      "\n",
      "len of col:  106064\n",
      "len of col set:  15152\n",
      "\n",
      "len of col:  1259801\n",
      "len of col set:  3878\n",
      "\n",
      "len of col:  1259801\n",
      "len of col set:  179957\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# DETECTED FILE\n",
    "# how many unique uniprot IDs: \n",
    "uniqueCount(det, 'matched_UKBID_x')\n",
    "# how many unique c and k positions\n",
    "uniqueCount(det, 'pos_ID')\n",
    "\n",
    "# not detected file\n",
    "uniqueCount(notd, 'matched_UKBID')\n",
    "uniqueCount(notd, 'pos_ID_falseCKtarget')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "different number of IDs :  4\n",
      "['Q9Y4H4', 'P63220', 'O95777', 'P62857']\n",
      "3876\n",
      "3876\n"
     ]
    }
   ],
   "source": [
    "# controlling for same uniprot IDs in both detected and non detected files\n",
    "detID = list(det.matched_UKBID_x)\n",
    "notdID = list(notd.matched_UKBID)\n",
    "\n",
    "difflist = diffList(detID, notdID)\n",
    "intersec = intersectionCheck(detID, notdID)\n",
    "\n",
    "print(difflist)\n",
    "print(len(intersec))\n",
    "# result of detectecd - notdetected, so these IDs are in the det set only\n",
    "\n",
    "'''\n",
    "['Q9Y4H4', 'P63220', 'O95777', 'P62857']\n",
    "3876\n",
    "'''\n",
    "intersec = list(intersec)\n",
    "print(len(intersec))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  shared3876   Count\n",
      "0       True  195046\n",
      "1      False      63\n",
      "dropping false\n",
      "df shape post drop:  (195046, 3)\n",
      "  shared3876   Count\n",
      "0       True  105959\n",
      "1      False     105\n",
      "dropping false\n",
      "df shape post drop:  (105959, 3)\n",
      "(195046, 3)\n",
      "(105959, 3)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariapalafox/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:376: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = _infer_fill_value(value)\n",
      "/Users/mariapalafox/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[item] = s\n",
      "/Users/mariapalafox/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4102: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  errors=errors,\n"
     ]
    }
   ],
   "source": [
    "# combining det and not det CK positions and entries to compare to uniprot fasta file\n",
    "sim_det = det[['matched_UKBID_x', 'pos_ID']]\n",
    "sim_not = notd[['matched_UKBID', 'pos_ID_falseCKtarget']]\n",
    "# rename col\n",
    "sim_det.columns = ['entry', 'pos_ID']\n",
    "sim_not.columns = ['entry', 'pos_ID']\n",
    "# combine \n",
    "dbnsfp_ck = pd.concat([sim_det, sim_not])\n",
    "dbnsfp_ck.drop_duplicates(inplace=True)\n",
    "\n",
    "# drop rows with IDs not in shared 3876 set\n",
    "dbnsfp_3876 = addcolumnconditionalDrop(intersec, dbnsfp_ck, 'entry', 'shared3876')\n",
    "detected3876 = addcolumnconditionalDrop(intersec, sim_det, 'entry', 'shared3876')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3876\n",
      "     entry       pos_ID shared3876\n",
      "0   Q3ZCM7  Q3ZCM7_C354       True\n",
      "7   Q3ZCM7  Q3ZCM7_C303       True\n",
      "14  Q3ZCM7  Q3ZCM7_C239       True\n",
      "21  Q3ZCM7   Q3ZCM7_C12       True\n",
      "28  Q92615  Q92615_C633       True\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariapalafox/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "print(len(detected3876.entry.unique()))\n",
    "detected3876.drop_duplicates(inplace=True)\n",
    "print(detected3876.head(5))\n",
    "# ONLY DETECTED POSITIONS IN DBNSFP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "# grouping dbnsfp detected file only by uniprot ID\n",
    "detected2 = detected3876.groupby('entry')['pos_ID'].apply(lambda x: len(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from chemoproteomics data processing:\n",
    "targetcounts = pd.read_csv(\"/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CYS_LYS_2012_to2018UKBpositions/REFERENCE_2018_mappedCK/SUBSETS/detection_freq_3953_CK_all.csv\")\n",
    "c_target = pd.read_csv(\"/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CYS_LYS_2012_to2018UKBpositions/REFERENCE_2018_mappedCK/SUBSETS/Cys2018_ever_labeled_6234.csv\")\n",
    "k_target = pd.read_csv(\"/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CYS_LYS_2012_to2018UKBpositions/REFERENCE_2018_mappedCK/SUBSETS/Lys2018_ever_labeled_9042.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbNSFP_target_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>A0AVT1</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>A0JNW5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>A0MZ66</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>A1KXE4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>A1L0T0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Q9Y6M9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Q9Y6W3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Q9Y6W5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Q9Y6W6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3876 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        dbNSFP_target_count\n",
       "ID                         \n",
       "A0AVT1                   11\n",
       "A0JNW5                    1\n",
       "A0MZ66                    1\n",
       "A1KXE4                    1\n",
       "A1L0T0                    1\n",
       "...                     ...\n",
       "Q9Y6M9                    1\n",
       "Q9Y6W3                    2\n",
       "Q9Y6W5                    1\n",
       "Q9Y6W6                    1\n",
       "Q9Y6Y8                    1\n",
       "\n",
       "[3876 rows x 1 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "detected2 = pd.DataFrame(detected2)\n",
    "detected2.reset_index(inplace=True)\n",
    "detected2.columns = ['ID', 'dbNSFP_target_count']\n",
    "detected2.set_index('ID', inplace=True)\n",
    "targetcounts.sort_values('ID', inplace=True)\n",
    "targetcounts.set_index('ID', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>times_detected</th>\n",
       "      <th>count_C_targets</th>\n",
       "      <th>count_K_targets</th>\n",
       "      <th>dbNSFP_target_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>A0AVT1</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>A0JNW5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>A0MZ66</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>A1KXE4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>A1L0T0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3837</td>\n",
       "      <td>Q9Y6M9</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3838</td>\n",
       "      <td>Q9Y6W3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3839</td>\n",
       "      <td>Q9Y6W5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3840</td>\n",
       "      <td>Q9Y6W6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3841</td>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3842 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ID  times_detected  count_C_targets  count_K_targets  \\\n",
       "0     A0AVT1              11                6                5   \n",
       "1     A0JNW5               1                0                1   \n",
       "2     A0MZ66               1                1                0   \n",
       "3     A1KXE4               1                1                0   \n",
       "4     A1L0T0               1                0                1   \n",
       "...      ...             ...              ...              ...   \n",
       "3837  Q9Y6M9               1                1                0   \n",
       "3838  Q9Y6W3               2                2                0   \n",
       "3839  Q9Y6W5               1                1                0   \n",
       "3840  Q9Y6W6               1                1                0   \n",
       "3841  Q9Y6Y8               1                1                0   \n",
       "\n",
       "      dbNSFP_target_count  \n",
       "0                      11  \n",
       "1                       1  \n",
       "2                       1  \n",
       "3                       1  \n",
       "4                       1  \n",
       "...                   ...  \n",
       "3837                    1  \n",
       "3838                    2  \n",
       "3839                    1  \n",
       "3840                    1  \n",
       "3841                    1  \n",
       "\n",
       "[3842 rows x 5 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "detected2.reset_index(inplace=True)\n",
    "targetcounts.reset_index(inplace=True)\n",
    "mergetargets = pd.merge(targetcounts,detected2,how='inner',on=['ID'])\n",
    "mergetargets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   check_target_total  Count\n",
      "0                   0   3841\n",
      "1                   1      1\n"
     ]
    }
   ],
   "source": [
    "mergetargets['check_target_total'] = mergetargets['times_detected'] - mergetargets['dbNSFP_target_count']\n",
    "checkColumnValues(mergetargets, 'check_target_total')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "different number of IDs :  34\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['P54136',\n",
       " 'P86791',\n",
       " 'P68371',\n",
       " 'Q71DI3',\n",
       " 'Q9Y2S7',\n",
       " 'O75044',\n",
       " 'P58107',\n",
       " 'Q9NXV6',\n",
       " 'Q99873',\n",
       " 'P63261',\n",
       " 'P62826',\n",
       " 'Q8WUH1',\n",
       " 'Q4G0P3',\n",
       " 'P0DMV9',\n",
       " 'Q8WUH6',\n",
       " 'P82921',\n",
       " 'P19105',\n",
       " 'P06753',\n",
       " 'Q8IYB9',\n",
       " 'P04406',\n",
       " 'Q9BTY7',\n",
       " 'P68363',\n",
       " 'Q9BQE3',\n",
       " 'P80303',\n",
       " 'Q9GZM3',\n",
       " 'Q99613',\n",
       " 'P60174',\n",
       " 'Q96TA1',\n",
       " 'P08238',\n",
       " 'Q86YV5',\n",
       " 'P04350',\n",
       " 'P26640',\n",
       " 'Q96GY3',\n",
       " 'P63241']"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many IDs are in dbNSFP detected file and not in newly filtered step 1 of mapping data (post 2012 positions mapping)\n",
    "targetlist = targetcounts.ID.tolist()\n",
    "detectedlist = detected2.ID.tolist()\n",
    "# checking list of IDs not found in chemoproteomic filtered files\n",
    "difflist = diffList(detectedlist, targetlist)\n",
    "difflist"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## why are some uniprot IDs in the dbNSFP detected file and not in uniprot chemoproteomics ever labeled files??\n",
    "\n",
    "**these uniprot IDs were dropped when i corrected the CK position checked script for 2012 chemoposition mapping to 2018 reference sequences**\n",
    "- only keep IDs where all detected positions found (n=3841)...final count 3840"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [2] check that detected and detected files have all C and K positions for uniprot ID "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>times_detected</th>\n",
       "      <th>count_C_targets</th>\n",
       "      <th>count_K_targets</th>\n",
       "      <th>dbNSFP_target_count</th>\n",
       "      <th>check_target_total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>A0AVT1</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>A0JNW5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>A0MZ66</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>A1KXE4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>A1L0T0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3837</td>\n",
       "      <td>Q9Y6M9</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3838</td>\n",
       "      <td>Q9Y6W3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3839</td>\n",
       "      <td>Q9Y6W5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3840</td>\n",
       "      <td>Q9Y6W6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3841</td>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3841 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ID  times_detected  count_C_targets  count_K_targets  \\\n",
       "0     A0AVT1              11                6                5   \n",
       "1     A0JNW5               1                0                1   \n",
       "2     A0MZ66               1                1                0   \n",
       "3     A1KXE4               1                1                0   \n",
       "4     A1L0T0               1                0                1   \n",
       "...      ...             ...              ...              ...   \n",
       "3837  Q9Y6M9               1                1                0   \n",
       "3838  Q9Y6W3               2                2                0   \n",
       "3839  Q9Y6W5               1                1                0   \n",
       "3840  Q9Y6W6               1                1                0   \n",
       "3841  Q9Y6Y8               1                1                0   \n",
       "\n",
       "      dbNSFP_target_count  check_target_total  \n",
       "0                      11                   0  \n",
       "1                       1                   0  \n",
       "2                       1                   0  \n",
       "3                       1                   0  \n",
       "4                       1                   0  \n",
       "...                   ...                 ...  \n",
       "3837                    1                   0  \n",
       "3838                    2                   0  \n",
       "3839                    1                   0  \n",
       "3840                    1                   0  \n",
       "3841                    1                   0  \n",
       "\n",
       "[3841 rows x 6 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_targets = mergetargets[mergetargets['check_target_total'] != 1].copy()\n",
    "all_targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ID                     A0AVT1A0JNW5A0MZ66A1KXE4A1L0T0A2A288A2RUC4A3KN...\n",
       "times_detected                                                     14926\n",
       "count_C_targets                                                     6058\n",
       "count_K_targets                                                     8868\n",
       "dbNSFP_target_count                                                14926\n",
       "check_target_total                                                     0\n",
       "dtype: object"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_targets.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  shared3841   Count\n",
      "0       True  104475\n",
      "1      False       7\n",
      "dropping false\n",
      "df shape post drop:  (104475, 400)\n",
      "  shared3841    Count\n",
      "0       True  1222911\n",
      "1      False    24297\n",
      "dropping false\n",
      "df shape post drop:  (1222911, 17)\n",
      "len of col:  104475\n",
      "len of col set:  3840\n",
      "\n",
      "len of col:  104475\n",
      "len of col set:  14925\n",
      "\n",
      "len of col:  1222911\n",
      "len of col set:  3840\n",
      "\n",
      "len of col:  1222911\n",
      "len of col set:  174702\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# dropped id found with weird K codon missense counts and checking again post drop\n",
    "# drop rows with IDs not in shared 3876 set\n",
    "all3840 = all_targets.ID.tolist()\n",
    "all3840.remove('Q8WZ42')\n",
    "print(len(all3840))\n",
    "\n",
    "det_final = addcolumnconditionalDrop(all3840, det, 'matched_UKBID_x', 'shared3841')\n",
    "not_final = addcolumnconditionalDrop(all3840, notd, 'matched_UKBID', 'shared3841')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_targets.to_csv(\"all_targetCK_found_in_UKBIDs3840_dbNSFP.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [3] checking that all missense for C and K codon have 7 reported alterations\n",
    "- all target positions had 7 changes for codon (correct)\n",
    "- non targets had 3 IDs with not 7 count for codon changes, dropped worst ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   pos_id19  Count\n",
      "0         7  14925\n",
      "   pos_id19   Count\n",
      "0         7  174700\n",
      "1         6       1\n",
      "2         5       1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\nBEFORE DROPPING TTN gene\\npos_id19   Count\\n0         7  178139\\n1        14      16\\n2         6       1\\n3         5       1\\n# detected file has 3 uniprot IDs with amino acids containing incorrect missense codon count\\n'"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# checking that all missense and in file\n",
    "det_final_group = det_final.groupby('pos_ID')['pos_id19'].apply(lambda x:len(x))\n",
    "not_final_group = not_final.groupby('pos_ID_falseCKtarget')['pos_id19'].apply(lambda x:len(x))\n",
    "\n",
    "det_final_group = pd.DataFrame(det_final_group)\n",
    "checkColumnValues(det_final_group, 'pos_id19') # all target codons had 7\n",
    "\n",
    "not_final_group = pd.DataFrame(not_final_group)\n",
    "checkColumnValues(not_final_group, 'pos_id19')\n",
    "'''\n",
    "BEFORE DROPPING TTN gene\n",
    "pos_id19   Count\n",
    "0         7  178139\n",
    "1        14      16\n",
    "2         6       1\n",
    "3         5       1\n",
    "# detected file has 3 uniprot IDs with amino acids containing incorrect missense codon count\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos_id19</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos_ID_falseCKtarget</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Q86W56_K115</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Q9Y6J0_C571</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      pos_id19\n",
       "pos_ID_falseCKtarget          \n",
       "Q86W56_K115                  5\n",
       "Q9Y6J0_C571                  6"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# BEFORE DROPPING Q8WZ42 ID\n",
    "#INCORRECT CODON NUMBER IN NOT detected file from dbNSFP\n",
    "not_final_group[not_final_group['pos_id19'] != 7]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## keeping: \n",
    "Q9Y6J0 has 2 targets is Calcineurin-binding protein cabin-1 important for T cell receptor signaling\n",
    "\n",
    "Q86W56 has 3 target is PolyADP-ribose glycohydrolase with PCNA binding domain dont delete!\n",
    "\n",
    "## dropping: \n",
    "Q8WZ42 has 1 target TTN important for straited muscle\n",
    "\n",
    "```\n",
    "pos_ID_falseCKtarget\t\n",
    "Q8WZ42_K11429\t14\n",
    "Q8WZ42_K11435\t14\n",
    "Q8WZ42_K11436\t14\n",
    "Q8WZ42_K11443\t14\n",
    "Q8WZ42_K11444\t14\n",
    "Q8WZ42_K11451\t14\n",
    "Q8WZ42_K11457\t14\n",
    "Q8WZ42_K11463\t14\n",
    "Q8WZ42_K11464\t14\n",
    "Q8WZ42_K11471\t14\n",
    "Q8WZ42_K11472\t14\n",
    "Q8WZ42_K11479\t14\n",
    "Q8WZ42_K11485\t14\n",
    "Q8WZ42_K11491\t14\n",
    "Q8WZ42_K11492\t14\n",
    "Q8WZ42_K11500\t14\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ALL targets in 3840 uniprot IDs found!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len of col:  104475\n",
      "len of col set:  3840\n",
      "\n",
      "len of col:  104475\n",
      "len of col set:  14925\n",
      "\n",
      "\n",
      "len of col:  1222911\n",
      "len of col set:  3840\n",
      "\n",
      "len of col:  1222911\n",
      "len of col set:  174702\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# DETECTED FILE \n",
    "uniqueCount(det_final, 'matched_UKBID_x')\n",
    "uniqueCount(det_final, 'pos_ID')\n",
    "print()\n",
    "# not detected file\n",
    "uniqueCount(not_final, 'matched_UKBID')\n",
    "uniqueCount(not_final, 'pos_ID_falseCKtarget')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [4] now checking CK total counts for these IDs (uniprot fasta) vs total CK not detected from dbNSFP\n",
    "- all CK found exceptp 2 postions, great!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  shared3840  Count\n",
      "0      False  14592\n",
      "1       True   3840\n",
      "dropping false\n",
      "df shape post drop:  (3840, 30)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Unnamed: 0                                              33973922\n",
       "ID             Q9HAS0Q86X76Q9NQR4Q9Y314Q04721Q9UHY1P49116Q166...\n",
       "entryName      NJMU_HUMANNIT1_HUMANNIT2_HUMANNOSIP_HUMANNOTC2...\n",
       "Length                                                   2276173\n",
       "proSequence    MLPSLQESMDGDEKELESSEEGGSAEERRLEPPSSSHYCLYSYRGS...\n",
       "A                                                         164566\n",
       "B                                                              0\n",
       "C                                                          40115\n",
       "D                                                         119895\n",
       "E                                                         179808\n",
       "F                                                          77005\n",
       "G                                                         143874\n",
       "H                                                          54491\n",
       "I                                                         100800\n",
       "K                                                         149550\n",
       "L                                                         220532\n",
       "M                                                          52137\n",
       "N                                                          84008\n",
       "P                                                         135395\n",
       "Q                                                         113370\n",
       "R                                                         127436\n",
       "S                                                         178605\n",
       "T                                                         116020\n",
       "U                                                              5\n",
       "V                                                         137380\n",
       "W                                                          22868\n",
       "X                                                              0\n",
       "Y                                                          58313\n",
       "Z                                                              0\n",
       "shared3840     TrueTrueTrueTrueTrueTrueTrueTrueTrueTrueTrueTr...\n",
       "dtype: object"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ukb = pd.read_csv(\"/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CYS_LYS_2012_to2018UKBpositions/REFERENCE_2018_mappedCK/SUBSETS/countsfromfastaCCDS_filter_counts_all_AA.csv\")\n",
    "ukb_final = addcolumnconditionalDrop(all3840, ukb, 'ID', 'shared3840')\n",
    "ukb_final.sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## results of non detected CK accounting: \n",
    "\n",
    "* Total C = 40,115\n",
    "* Total K = 149,550\n",
    "* Sum in ukb fasta = 189,665\n",
    "* sumUKB - dbNSFPnontarget = 14,963...which is 2 less than total number of target C and K...so only short 2 amino acids from total count!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "189665\n",
      "14963\n"
     ]
    }
   ],
   "source": [
    "print(149550 + 40115) # CK sum in ukb\n",
    "print(189665 - 174702) # difference btw ukb and dbNSFP non detected CK positions\n",
    "# 14963 , which is 2 less than it is suppose to be"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariapalafox/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aaref_x</th>\n",
       "      <th>pos_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>Q3ZCM7_C354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>C</td>\n",
       "      <td>Q3ZCM7_C303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>Q3ZCM7_C239</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   aaref_x       pos_ID\n",
       "0        C  Q3ZCM7_C354\n",
       "7        C  Q3ZCM7_C303\n",
       "14       C  Q3ZCM7_C239"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# det_final_accounting = det_final[['matched_UKBID_x', 'pos_ID','Cys_reactivity', 'Cys_react_threshold', 'Cys_target_label', 'Lys_reactivity', 'Lys_react_threshold', 'Lys_target_label']]\n",
    "det_final_accounting = det_final[['aaref_x', 'pos_ID']]\n",
    "det_final_accounting.drop_duplicates(inplace=True)\n",
    "det_final_accounting.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  aaref_x  Count\n",
      "0       K   8868\n",
      "1       C   6057\n"
     ]
    }
   ],
   "source": [
    "# everlabeled C or K count\n",
    "checkColumnValues(det_final_accounting, 'aaref_x')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Cys_react_threshold  Count\n",
      "0                 Low    895\n",
      "1              Medium    412\n",
      "2                High     94\n",
      "\n",
      "  Cys_target_label  Count\n",
      "0      panReactive   5088\n",
      "1           Target    708\n",
      "\n",
      "  Lys_react_threshold  Count\n",
      "0                 Low   3401\n",
      "1              Medium    660\n",
      "2                High    302\n",
      "\n",
      "  Lys_target_label  Count\n",
      "0      panReactive   7840\n",
      "1           Target    115\n"
     ]
    }
   ],
   "source": [
    "checkColumnValues(det_final_accounting, 'Cys_react_threshold')\n",
    "print()\n",
    "checkColumnValues(det_final_accounting, 'Cys_target_label')\n",
    "print()\n",
    "checkColumnValues(det_final_accounting, 'Lys_react_threshold')\n",
    "print()\n",
    "checkColumnValues(det_final_accounting, 'Lys_target_label')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariapalafox/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aaref</th>\n",
       "      <th>pos_ID_falseCKtarget</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>K</td>\n",
       "      <td>Q3ZCM7_K392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>K</td>\n",
       "      <td>Q3ZCM7_K379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>K</td>\n",
       "      <td>Q3ZCM7_K362</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   aaref pos_ID_falseCKtarget\n",
       "0      K          Q3ZCM7_K392\n",
       "7      K          Q3ZCM7_K379\n",
       "14     K          Q3ZCM7_K362"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_final_accounting = not_final[[\"aaref\", \"pos_ID_falseCKtarget\"]]\n",
    "not_final_accounting.drop_duplicates(inplace=True)\n",
    "not_final_accounting.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  aaref   Count\n",
      "0     K  140652\n",
      "1     C   34050\n"
     ]
    }
   ],
   "source": [
    "checkColumnValues(not_final_accounting, 'aaref')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FINAL ACCOUNTING of filtered/checked CK positions for .bed file:\n",
    "* total uniprot IDs = 3840\n",
    "* all targets found with 7 changes per codon\n",
    "* only 2 CK non-targets missing and 2 aminos with incorrect codon change count\n",
    "\n",
    "| TYPE | Cysteine | Lysine | \n",
    "|-------|--------------|-----------|\n",
    "| Low | 895 | 3401 | \n",
    "| Medium | 412 | 660 |\n",
    "| High | 94 | 302 |\n",
    "| Panreactive | 5088 | 7804 |\n",
    "| TargetSM | 708 | 115 |\n",
    "| EVER-labeled | 6057 | 8868 |\n",
    "| not-labeled | 34050 | 140652 |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [MAKE BED FILES WITH hg19 COORDINATES]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampledet25 = det_final[0:25]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampledet25.to_csv(\"detected25_sample.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
