{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container {width:90% !important;}</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import argparse\n",
    "import pandas as pd\n",
    "import csv\n",
    "import numpy as np\n",
    "\n",
    "from IPython.display import display, HTML\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.options.display.max_seq_items = 2000\n",
    "display(HTML(\"<style>.container {width:90% !important;}</style>\"))\n",
    "sys.path.append(\"/Users/mariapalafox/Desktop/Toolbox\")\n",
    "from all_funx import *\n",
    "from maplib import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MAP CADDv1.4 annotations (for detected & non-detected CK pos) to dbNSFP score files\n",
    "\n",
    "(12.27.19 mfpfox on local, other M part on hoffman)\n",
    "\n",
    "### VEP VERSION NOTE: \n",
    "    - CADDv1.4 using VEPv92\n",
    "    - dbNSFP using VEPv94\n",
    "\n",
    "**hoffman PATH : /u/home/m/mfpalafo/project-arboleda/CADD/PULL_annotations**\n",
    "    * DIR downloaded locally for part 3 QC \"RESULT_pos_overlap_dbNSFPcoordinates/\"\n",
    "    * hoffman files have all consequences\n",
    "\n",
    "**hoffman M markdown has code for:** \n",
    "- [1] chr chunks\n",
    "- [2] search pos against CADD files\n",
    "\n",
    "**this M markdwon has code for:** \n",
    "- [3] filter CADDv1.4 files for MISSENSE consequence only, format AA col and pos_id col\n",
    "- [4] simplify dbNSFP cols\n",
    "- [5] merge missense cadd files & merge with dbNSFP score files\n",
    "- [6] QC merge CADD+dbNSFP annotation files\n",
    "\n",
    "| [1] | [2] | [3] | \n",
    "|----|----|----- |\n",
    "| break up search files with det / not det coordinates into chr chunks | search dbNSFP derived coordinates for CK codon positions against CADDannotations(GRCh37 and GRCh38 model) | filter CADDv1.4 files w/ dbNSFP CK positions for MISSENSE consequence only, add A/A column, and pos_id |\n",
    "\n",
    "\n",
    "| [4] | [5] | [6] |\n",
    "| ----|-----|----|\n",
    " |simplify dbNSFP detected and notdect files for select columns | merge missense cadd files & merge with dbNSFP score files | QC merge CADD+dbNSFP annotation files |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [5] CADDv1.4 Missense merge chr files together \n",
    "\n",
    "#### from append.sh\n",
    "```bash\n",
    "#!/bin/bash\n",
    "# keeps header and appends all rows together\n",
    "# for Pmap dbNSFP CADD mapped detected and non detected chr merge\n",
    "\n",
    "# 19 NOT DETECTED FILES\n",
    "awk -F ',' 'FNR==1 && NR!=1 { while (/^pos_hg19/) getline; } 1 {print}' MISSENSE_chr*_CADD_GRCh37_NOT_DETECTED_CK.csv > SCORE_CADD_37_NOT_detected_CK.csv\n",
    "\n",
    "# 19 DETECTED FILES\n",
    "awk -F ',' 'FNR==1 && NR!=1 { while (/^pos_hg19/) getline; } 1 {print}' MISSENSE_chr*_CADD_GRCh37_DETECTED_CK.csv > SCORE_CADD_37_detected_CK.csv\n",
    "\n",
    "# same for hg38 files just replace hg19 with hg38 and file names\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "\n",
    "# CADD results {line count} {filename}\n",
    "\n",
    "### DETECTED hg19 : \n",
    "* 106,176 SCORE_CADD_37_detected_CK.csv (1,700 more rows than dbNSFP)\n",
    "\n",
    "### DETECTED hg38  : \n",
    "* 106,816 SCORE_CADD_38_detected_CK.csv (2,340 more rows than dbNSFP)\n",
    "\n",
    "### NOT DECT hg19 : \n",
    "* 1,234,967 SCORE_CADD_37_NOT_detected_CK.csv (12,055 more rows than dbNSFP)\n",
    "\n",
    "### NOT DECT hg38 :\n",
    "* 1,240,470 SCORE_CADD_38_NOT_detected_CK.csv (17,558 more rows than dbNSFP)\n",
    "\n",
    "\n",
    "# dbNSFP results:\n",
    "### DETECTED : \n",
    "* 104,476 SCORE_dbNSFP_selectcols_detected_CK_104475.csv\n",
    "\n",
    "### NOT DECT : \n",
    "* 1,222,912 SCORE_dbNSFP_selectcols_NOT_detected_CK_1222911.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SCORE2_dbNSFP_selectcols_NOT_detected_CK_1222911.csv', 'SCORE_dbNSFP_selectcols_detected_CK_104475.csv', 'SCORE2_CADD_37_NOT_detected_CK.csv', 'DUPLICATE_great_posid_detected19_cadd_1786.csv', 'SCORE2_CADD_38_NOT_detected_CK.csv', 'SCORE_CADD_37_detected_CK.csv', 'SCORE_CADD_38_detected_CK.csv', 'SCORE2_CADD_37_detected_CK.csv', 'dbnsfp_featureID_notdetected_description.csv', 'SCORE_dbNSFP_selectcols_NOT_detected_CK_1222911.csv', 'SCORE_CADD_38_NOT_detected_CK.csv', 'SCORE2_CADD_38_detected_CK.csv', 'dbnsfp_featureID_detected_description.csv', 'SCORE_CADD_37_NOT_detected_CK.csv', 'SCORE2_dbNSFP_selectcols_detected_CK_104475.csv']\n"
     ]
    }
   ],
   "source": [
    "os.chdir(\"/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/RESULT_CADDv14_pos_overlap_dbNSFPcoordinates/MISSENSE_FILTERED/MERGE_CHR\")\n",
    "print(os.listdir())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "---\n",
    "\n",
    "# REDUNDANCY in pos_id CADD columns is problem for merging with dbNSFP files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving files with correct pos_id\n",
    "det19=pd.read_csv('SCORE_CADD_37_detected_CK.csv', low_memory=False)\n",
    "det38=pd.read_csv('SCORE_CADD_38_detected_CK.csv',  low_memory=False)\n",
    "not19=pd.read_csv('SCORE_CADD_37_NOT_detected_CK.csv', low_memory=False)\n",
    "not38=pd.read_csv('SCORE_CADD_38_NOT_detected_CK.csv',  low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "106175\n",
      "104225\n",
      "1234966\n",
      "1217997\n"
     ]
    }
   ],
   "source": [
    "dfls = [det19, not19]\n",
    "for f in dfls:\n",
    "    print(len(list(f.pos_id19)))\n",
    "    print(len(set(f.pos_id19)))\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "106815\n",
      "104185\n",
      "1240469\n",
      "1218415\n"
     ]
    }
   ],
   "source": [
    "dfls = [det38,  not38]\n",
    "for f in dfls:\n",
    "    print(len(list(f.pos_id38)))\n",
    "    print(len(set(f.pos_id38)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['pos_hg19', 'chr', 'pos_ID', 'Ref', 'Alt', 'Type', 'Length', 'AnnoType',\n",
      "       'Consequence', 'ConsScore', 'ConsDetail', 'GC', 'CpG', 'motifECount',\n",
      "       'motifEName', 'motifEHIPos', 'motifEScoreChng', 'oAA', 'nAA', 'GeneID',\n",
      "       'FeatureID', 'GeneName', 'CCDS', 'Intron', 'Exon', 'cDNApos',\n",
      "       'relcDNApos', 'CDSpos', 'relCDSpos', 'protPos', 'relProtPos', 'Domain',\n",
      "       'Dst2Splice', 'Dst2SplType', 'minDistTSS', 'minDistTSE', 'SIFTcat',\n",
      "       'SIFTval', 'PolyPhenCat', 'PolyPhenVal', 'priPhCons', 'mamPhCons',\n",
      "       'verPhCons', 'priPhyloP', 'mamPhyloP', 'verPhyloP', 'GerpRS',\n",
      "       'GerpRSpval', 'GerpN', 'GerpS', 'Grantham', 'Dist2Mutation',\n",
      "       'Freq100bp', 'Rare100bp', 'Sngl100bp', 'Freq1000bp', 'Rare1000bp',\n",
      "       'Sngl1000bp', 'Freq10000bp', 'Rare10000bp', 'Sngl10000bp', 'RawScore',\n",
      "       'PHRED', 'pos_id19', 'Amino_acids'],\n",
      "      dtype='object')\n",
      "\n",
      "Index(['pos_hg38', 'chr', 'pos_ID', 'Ref', 'Alt', 'Type', 'Length', 'AnnoType',\n",
      "       'Consequence', 'ConsScore', 'ConsDetail', 'GC', 'CpG', 'motifECount',\n",
      "       'motifEName', 'motifEHIPos', 'motifEScoreChng', 'oAA', 'nAA', 'GeneID',\n",
      "       'FeatureID', 'GeneName', 'CCDS', 'Intron', 'Exon', 'cDNApos',\n",
      "       'relcDNApos', 'CDSpos', 'relCDSpos', 'protPos', 'relProtPos', 'Domain',\n",
      "       'Dst2Splice', 'Dst2SplType', 'minDistTSS', 'minDistTSE', 'SIFTcat',\n",
      "       'SIFTval', 'PolyPhenCat', 'PolyPhenVal', 'priPhCons', 'mamPhCons',\n",
      "       'verPhCons', 'priPhyloP', 'mamPhyloP', 'verPhyloP', 'GerpRS',\n",
      "       'GerpRSpval', 'GerpN', 'GerpS', 'Grantham', 'Dist2Mutation',\n",
      "       'Freq100bp', 'Rare100bp', 'Sngl100bp', 'Freq1000bp', 'Rare1000bp',\n",
      "       'Sngl1000bp', 'Freq10000bp', 'Rare10000bp', 'Sngl10000bp', 'RawScore',\n",
      "       'PHRED', 'pos_id38', 'Amino_acids', 'super_pos_id38', 'great_pos_id38'],\n",
      "      dtype='object')\n",
      "\n",
      "Index(['pos_hg19', 'chr', 'pos_ID', 'Ref', 'Alt', 'Type', 'Length', 'AnnoType',\n",
      "       'Consequence', 'ConsScore', 'ConsDetail', 'GC', 'CpG', 'motifECount',\n",
      "       'motifEName', 'motifEHIPos', 'motifEScoreChng', 'oAA', 'nAA', 'GeneID',\n",
      "       'FeatureID', 'GeneName', 'CCDS', 'Intron', 'Exon', 'cDNApos',\n",
      "       'relcDNApos', 'CDSpos', 'relCDSpos', 'protPos', 'relProtPos', 'Domain',\n",
      "       'Dst2Splice', 'Dst2SplType', 'minDistTSS', 'minDistTSE', 'SIFTcat',\n",
      "       'SIFTval', 'PolyPhenCat', 'PolyPhenVal', 'priPhCons', 'mamPhCons',\n",
      "       'verPhCons', 'priPhyloP', 'mamPhyloP', 'verPhyloP', 'GerpRS',\n",
      "       'GerpRSpval', 'GerpN', 'GerpS', 'Grantham', 'Dist2Mutation',\n",
      "       'Freq100bp', 'Rare100bp', 'Sngl100bp', 'Freq1000bp', 'Rare1000bp',\n",
      "       'Sngl1000bp', 'Freq10000bp', 'Rare10000bp', 'Sngl10000bp', 'RawScore',\n",
      "       'PHRED', 'pos_id19', 'Amino_acids'],\n",
      "      dtype='object')\n",
      "\n",
      "Index(['pos_hg38', 'chr', 'pos_ID', 'Ref', 'Alt', 'Type', 'Length', 'AnnoType',\n",
      "       'Consequence', 'ConsScore', 'ConsDetail', 'GC', 'CpG', 'motifECount',\n",
      "       'motifEName', 'motifEHIPos', 'motifEScoreChng', 'oAA', 'nAA', 'GeneID',\n",
      "       'FeatureID', 'GeneName', 'CCDS', 'Intron', 'Exon', 'cDNApos',\n",
      "       'relcDNApos', 'CDSpos', 'relCDSpos', 'protPos', 'relProtPos', 'Domain',\n",
      "       'Dst2Splice', 'Dst2SplType', 'minDistTSS', 'minDistTSE', 'SIFTcat',\n",
      "       'SIFTval', 'PolyPhenCat', 'PolyPhenVal', 'priPhCons', 'mamPhCons',\n",
      "       'verPhCons', 'priPhyloP', 'mamPhyloP', 'verPhyloP', 'GerpRS',\n",
      "       'GerpRSpval', 'GerpN', 'GerpS', 'Grantham', 'Dist2Mutation',\n",
      "       'Freq100bp', 'Rare100bp', 'Sngl100bp', 'Freq1000bp', 'Rare1000bp',\n",
      "       'Sngl1000bp', 'Freq10000bp', 'Rare10000bp', 'Sngl10000bp', 'RawScore',\n",
      "       'PHRED', 'pos_id38', 'Amino_acids'],\n",
      "      dtype='object')\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dfls = [det19, det38, not19, not38]\n",
    "\n",
    "dropme = ['Type', 'Length', 'AnnoType','ConsScore', 'ConsDetail','oAA', 'nAA', 'GeneID',\n",
    "       'FeatureID', 'GeneName', 'CCDS', 'Intron', 'Exon', 'cDNApos',\n",
    "       'relcDNApos', 'CDSpos', 'relCDSpos', 'protPos', 'relProtPos', 'Domain',\n",
    "       'Dst2Splice', 'Dst2SplType', 'minDistTSS', 'minDistTSE', 'SIFTcat',\n",
    "       'SIFTval', 'PolyPhenCat', 'PolyPhenVal', 'Amino_acids']\n",
    "\n",
    "for i in dfls:\n",
    "    print(i.shape)\n",
    "    i.drop(dropme, axis=1, inplace=True)\n",
    "    print(i.shape)\n",
    "    i.drop_duplicates(keep='first', inplace = True)\n",
    "    print(i.shape)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "---\n",
    "---\n",
    "\n",
    "\n",
    "# key versions\n",
    "### super format == { posid# _ AApos _ ENST }\n",
    "### great format == { posid# _ AApos _ aminoacids }\n",
    "### alright format == { posid# _ aminoacids }\n",
    "\n",
    "**goal 1:1 mapping btw CADD and dbnsfp, recovering all positions in dbNSFP file too**\n",
    "\n",
    "### column names:\n",
    "\n",
    "**cadd files:**\n",
    "- 'pos_id19' or 'pos_id38' & 'protPos' & 'FeatureID'\n",
    "\n",
    "**dbNSFP files:**\n",
    "- 'pos_id19' or 'pos_id38' & 'matched_aapos' & 'FeatureID'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def super_pos_id(df, posid, aapos, assembly):\n",
    "    # format = posid _ aapos _ ENST\n",
    "    if assembly == 37:\n",
    "        df.loc[:,'super_pos_id19'] = df[posid].astype(str) + '_' + df[aapos].astype(str) + \\\n",
    "        '_' + df['FeatureID']\n",
    "        return df\n",
    "    if assembly == 38:\n",
    "        df.loc[:,'super_pos_id38'] = df[posid].astype(str) + '_' + df[aapos].astype(str) + \\\n",
    "        '_' + df['FeatureID']\n",
    "        return df\n",
    "\n",
    "def great_pos_id(df, posid, aapos, assembly):\n",
    "    # format = posid _ aapos\n",
    "    if assembly == 37:\n",
    "        df.loc[:,'great_pos_id19'] = df[posid].astype(str) + '_' + df[aapos].astype(str) + \\\n",
    "        '_' + df['Amino_acids']\n",
    "        return df\n",
    "    if assembly == 38:\n",
    "        df.loc[:,'great_pos_id38'] = df[posid].astype(str) + '_' + df[aapos].astype(str) + \\\n",
    "        '_' + df['Amino_acids']\n",
    "        return df\n",
    "\n",
    "def alright_pos_id(df, posid, assembly):\n",
    "    # format = posid _ aapos\n",
    "    if assembly == 37:\n",
    "        df.loc[:,'alright_pos_id19'] = df[posid].astype(str) + '_' + df['Amino_acids']\n",
    "        return df\n",
    "    if assembly == 38:\n",
    "        df.loc[:,'alright_pos_id38'] = df[posid].astype(str) + '_' + df['Amino_acids']\n",
    "        return df\n",
    "    \n",
    "def unique_df(df19, df38, dbnsfp, colname19, colname38):\n",
    "    print(\"hg19: \")\n",
    "    uniqueCount(df19, colname19)\n",
    "    print(\"hg38: \")\n",
    "    uniqueCount(df38, colname38)\n",
    "    print(\"dbNSFP file both assemblies, 19 and 38: \")\n",
    "    uniqueCount(dbnsfp, colname19)\n",
    "    uniqueCount(dbnsfp, colname38)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## dbNSFP adding KEYS :\n",
    "\n",
    "```python\n",
    "dbnsfp = pd.read_csv('SCORE2_dbNSFP_selectcols_detected_CK_104475.csv', low_memory=False)\n",
    "print(dbnsfp.shape)\n",
    "notdbnsfp = pd.read_csv('SCORE2_dbNSFP_selectcols_NOT_detected_CK_1222911.csv', low_memory=False)\n",
    "print(notdbnsfp.shape)\n",
    "\n",
    "# SUPER DETECTED\n",
    "dbnsfp = super_pos_id(dbnsfp, 'pos_id19', 'matched_aapos', 37)\n",
    "dbnsfp = super_pos_id(dbnsfp, 'pos_id38', 'matched_aapos', 38)\n",
    "# SUPER NOTDETECTED\n",
    "notdbnsfp = super_pos_id(notdbnsfp, 'pos_id19', 'matched_aapos', 37)\n",
    "notdbnsfp = super_pos_id(notdbnsfp, 'pos_id38', 'matched_aapos', 38)\n",
    "\n",
    "# GREAT DETECTED\n",
    "dbnsfp = great_pos_id(dbnsfp, 'pos_id19', 'matched_aapos', 37)\n",
    "dbnsfp = great_pos_id(dbnsfp, 'pos_id38', 'matched_aapos', 38)\n",
    "# GREAT NOTDETECTED\n",
    "notdbnsfp = great_pos_id(notdbnsfp, 'pos_id19', 'matched_aapos', 37)\n",
    "notdbnsfp = great_pos_id(notdbnsfp, 'pos_id38', 'matched_aapos', 38)\n",
    "\n",
    "# ALRIGHT DETECTED\n",
    "dbnsfp = alright_pos_id(dbnsfp, 'pos_id19', 37)\n",
    "dbnsfp = alright_pos_id(dbnsfp, 'pos_id38', 38)\n",
    "# ALRIGHT NOTDETECTED\n",
    "notdbnsfp = alright_pos_id(notdbnsfp, 'pos_id19', 37)\n",
    "notdbnsfp = alright_pos_id(notdbnsfp, 'pos_id38', 38)\n",
    "\n",
    "dbnsfp.to_csv('SCORE2_dbNSFP_selectcols_detected_CK_104475.csv', index=False)\n",
    "notdbnsfp.to_csv('SCORE2_dbNSFP_selectcols_NOT_detected_CK_1222911.csv', index=False)\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CADD adding keys:\n",
    "\n",
    "### DETECTED FILES\n",
    "```python\n",
    "det19 = pd.read_csv('SCORE_CADD_37_detected_CK.csv', low_memory=False)\n",
    "det38 = pd.read_csv('SCORE_CADD_38_detected_CK.csv',  low_memory=False)\n",
    "# aa pos as int\n",
    "det19.protPos = det19.protPos.astype(int)\n",
    "det38.protPos = det38.protPos.astype(int)\n",
    "\n",
    "# CADD DETECTED GREAT\n",
    "det19 = great_pos_id(det19, 'pos_id19', 'protPos', 37)\n",
    "det38 = great_pos_id(det38, 'pos_id38', 'protPos', 38)\n",
    "\n",
    "# CADD ALRIGHT GREAT\n",
    "det19 = alright_pos_id(det19, 'pos_id19', 37)\n",
    "det38 = alright_pos_id(det38, 'pos_id38', 38)\n",
    "\n",
    "# saving \n",
    "det19.to_csv('SCORE2_CADD_37_detected_CK.csv', index=False)\n",
    "det38.to_csv('SCORE2_CADD_38_detected_CK.csv',  index=False)\n",
    "```\n",
    "\n",
    "### NOT DETECTED FILES\n",
    "```python\n",
    "not19 = pd.read_csv('SCORE_CADD_37_NOT_detected_CK.csv', low_memory=False)\n",
    "not38 = pd.read_csv('SCORE_CADD_38_NOT_detected_CK.csv',  low_memory=False)\n",
    "not19.protPos = not19.protPos.astype(int)\n",
    "not38.protPos = not38.protPos.astype(int)\n",
    "\n",
    "# CADD NOT GREAT\n",
    "not19 = great_pos_id(not19, 'pos_id19', 'protPos', 37)\n",
    "not38 = great_pos_id(not38, 'pos_id38', 'protPos', 38)\n",
    "\n",
    "# CADD NOT ALRIGHT\n",
    "not19 = alright_pos_id(not19, 'pos_id19',  37)\n",
    "not38 = alright_pos_id(not38, 'pos_id38',  38)\n",
    "\n",
    "# saving \n",
    "not19.to_csv('SCORE2_CADD_37_NOT_detected_CK.csv', index=False)\n",
    "not38.to_csv('SCORE2_CADD_38_NOT_detected_CK.csv',  index=False)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "detected 19 merge\n",
      "(104475, 1)\n",
      "(105091, 1)\n",
      "\n",
      "detected 38 merge\n",
      "(104475, 1)\n",
      "(105623, 1)\n"
     ]
    }
   ],
   "source": [
    "# simple df for merge\n",
    "dbnsfpkey19 = dbnsfp[['alright_pos_id19']]\n",
    "dbnsfpkey38 = dbnsfp[['alright_pos_id38']]\n",
    "det19key = det19[['alright_pos_id19']]\n",
    "det38key = det38[['alright_pos_id38']]\n",
    "\n",
    "#Now we can just merge both dataframes with an outer join:\n",
    "det19merge = pd.merge(dbnsfpkey19, det19key, on=['alright_pos_id19'], how='inner')\n",
    "print('detected 19 merge')\n",
    "print(dbnsfpkey19.shape)\n",
    "print(det19merge.shape)\n",
    "print()\n",
    "print('detected 38 merge')\n",
    "det38merge = pd.merge(dbnsfpkey38, det38key, on=['alright_pos_id38'], how='inner')\n",
    "print(dbnsfpkey38.shape)\n",
    "print(det38merge.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GREAT KEY MERGER Resulted in \n",
    "\n",
    "### detected CK dbnsfp - cadd19\n",
    "###104475 - 95543 = 8932\n",
    "\n",
    "### detected CK dbnsfp - cadd38\n",
    "##104475-92834 = 11641\n",
    "\n",
    "### CADD key results (SUCCESS that all great keys are unique):\n",
    "```python\n",
    "unique_df(det19, det38, dbnsfp, 'great_pos_id19', 'great_pos_id38')\n",
    "# all unique same for notdetected set\n",
    "```\n",
    "\n",
    "### MERGE dbNSFP + CADD on 'super_pos_id19' | 'super_pos_id38'\n",
    "```python\n",
    "# simple df for merge\n",
    "dbnsfpkey19 = dbnsfp[['great_pos_id19']]\n",
    "dbnsfpkey38 = dbnsfp[['great_pos_id38']]\n",
    "det19key = det19[['great_pos_id19']]\n",
    "det38key = det38[['great_pos_id38']]\n",
    "#Now we can just merge both dataframes with an outer join:\n",
    "det19merge = pd.merge(dbnsfpkey19, det19key, on=['great_pos_id19'], how='inner')\n",
    "print('det 19 merge')\n",
    "print(dbnsfpkey19.shape)\n",
    "print(det19merge.shape)\n",
    "print()\n",
    "print('det 38 merge')\n",
    "det38merge = pd.merge(dbnsfpkey38, det38key, on=['great_pos_id38'], how='inner')\n",
    "print(dbnsfpkey38.shape)\n",
    "print(det38merge.shape)\n",
    "det 19 merge\n",
    "(104475, 1)\n",
    "(95543, 1)\n",
    "\n",
    "det 38 merge\n",
    "(104475, 1)\n",
    "(92834, 1)\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SUPER KEY MERGER Resulted in 20k positions lost\n",
    "\n",
    "### DETECTED CADD to dbNSFP merge inner on super_pos_id KEY:\n",
    "    #### 25,158 less rows with det19 cadd merge to dbnsfp (79317 rows in merge)\n",
    "    #### 26,131 less rows with det38 cadd merge to dbnsfp (78344 rows in merge)\n",
    "    \n",
    "### CADD key results (SUCCESS that all super keys are unique):\n",
    "```python\n",
    "unique_df(det19, det38, dbnsfp, 'super_pos_id19', 'super_pos_id38')\n",
    "# output:\n",
    "    hg19: \n",
    "    len of col:  106175\n",
    "    len of col set:  106175\n",
    "    hg38: \n",
    "    len of col:  106815\n",
    "    len of col set:  106815\n",
    "    dbNSFP file both assemblies, 19 and 38: \n",
    "    len of col:  104475\n",
    "    len of col set:  104475\n",
    "```\n",
    "\n",
    "### MERGE dbNSFP + CADD on 'super_pos_id19' | 'super_pos_id38'\n",
    "```python\n",
    "# simple df for merge\n",
    "dbkey19 = dbnsfp[['super_pos_id19']]\n",
    "dbkey38 = dbnsfp[['super_pos_id38']]\n",
    "det19key = det19[['super_pos_id19']]\n",
    "det38key = det38[['super_pos_id38']]\n",
    "#Now we can just merge both dataframes with an outer join:\n",
    "det_19_mer = pd.merge(dbkey19, det19key, how='inner')\n",
    "det_38_mer = pd.merge(dbkey38, det38key, how='inner')\n",
    "print(det_19_mer)\n",
    "print()\n",
    "print(det_38_mer)\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CADD processing- check for duplicate values:\n",
    "```python\n",
    "\n",
    "# all rows with duplicated values\n",
    "dup38 = pd.concat(d for _, d in det38.groupby(\"great_pos_id38\") if len(d) > 1)\n",
    "dup38\n",
    "# all rows with duplicated values\n",
    "dup = pd.concat(d for _, d in det19.groupby(\"great_pos_id19\") if len(d) > 1)\n",
    "dup\n",
    "#dup.to_csv(\"DUPLICATE_great_posid_detected19_cadd_1786.csv\", index=False)\n",
    "```\n",
    "\n",
    "```python\n",
    "## CADD adding keys:\n",
    "dropme19 = ['AnnoType','ConsScore', 'ConsDetail', 'GeneID','FeatureID', 'GeneName', 'CCDS', 'Intron', 'Exon', 'cDNApos',\n",
    "       'relcDNApos', 'CDSpos', 'relCDSpos', 'relProtPos', 'Domain','Dst2Splice', 'Dst2SplType', 'minDistTSS', 'minDistTSE', 'SIFTcat',\n",
    "       'SIFTval', 'PolyPhenCat', 'PolyPhenVal', 'super_pos_id19']\n",
    "\n",
    "dropme38 = ['AnnoType','ConsScore', 'ConsDetail', 'GeneID','FeatureID', 'GeneName', 'CCDS', 'Intron', 'Exon', 'cDNApos',\n",
    "       'relcDNApos', 'CDSpos', 'relCDSpos', 'relProtPos', 'Domain','Dst2Splice', 'Dst2SplType', 'minDistTSS', 'minDistTSE', 'SIFTcat',\n",
    "       'SIFTval', 'PolyPhenCat', 'PolyPhenVal', 'super_pos_id38']\n",
    "\n",
    "det19 = pd.read_csv('SCORE2_CADD_37_detected_CK.csv', low_memory=False)\n",
    "det19.protPos = det19.protPos.astype(int)\n",
    "det19 = great_pos_id(det19, 'pos_id19', 'protPos', 37)\n",
    "det38 = pd.read_csv('SCORE2_CADD_38_detected_CK.csv',  low_memory=False)\n",
    "det38.protPos = det38.protPos.astype(int)\n",
    "det38 = great_pos_id(det38, 'pos_id38', 'protPos', 38)\n",
    "\n",
    "dfls = [det19]\n",
    "for i in dfls:\n",
    "    i.drop(dropme19, axis = 1, inplace=True) \n",
    "    print(i.shape)\n",
    "    # dropping ALL duplicte values \n",
    "    i.drop_duplicates(keep='first', inplace = True)\n",
    "    print(i.shape)\n",
    "    print()\n",
    "dfls = [det38]\n",
    "for i in dfls:\n",
    "    i.drop(dropme38, axis = 1, inplace=True) \n",
    "    print(i.shape)\n",
    "    # dropping ALL duplicte values \n",
    "    i.drop_duplicates(keep='first', inplace = True)\n",
    "    print(i.shape)\n",
    "    print()\n",
    "\n",
    "### NOT DETECTED FILES\n",
    "not19 = pd.read_csv('SCORE2_CADD_37_NOT_detected_CK.csv', low_memory=False)\n",
    "not19.protPos = not19.protPos.astype(int)\n",
    "not19 = great_pos_id(not19, 'pos_id19', 'protPos', 37)\n",
    "not38 = pd.read_csv('SCORE2_CADD_38_NOT_detected_CK.csv',  low_memory=False)\n",
    "not38.protPos = not38.protPos.astype(int)\n",
    "not38 = great_pos_id(not38, 'pos_id38', 'protPos', 38)\n",
    "\n",
    "dfls = [not19]\n",
    "for i in dfls:\n",
    "    i.drop(dropme19, axis = 1, inplace=True) \n",
    "    print(i.shape)\n",
    "    # dropping ALL duplicte values \n",
    "    i.drop_duplicates(keep='first', inplace = True)\n",
    "    print(i.shape)\n",
    "    print()\n",
    "dfls = [not38]\n",
    "for i in dfls:\n",
    "    i.drop(dropme38, axis = 1, inplace=True) \n",
    "    print(i.shape)\n",
    "    # dropping ALL duplicte values \n",
    "    i.drop_duplicates(keep='first', inplace = True)\n",
    "    print(i.shape)\n",
    "    print()\n",
    "\n",
    "# SAVING ALL FILES\n",
    "det19.to_csv('SCORE2_CADD_37_detected_CK.csv', index=False)\n",
    "det38.to_csv('SCORE2_CADD_38_detected_CK.csv',  index=False)\n",
    "not19.to_csv('SCORE2_CADD_37_NOT_detected_CK.csv', index=False)\n",
    "not38.to_csv('SCORE2_CADD_38_NOT_detected_CK.csv',  index=False)\n",
    "```\n",
    "\n",
    "## unique great_pos_id counts post dropping 'dropme' columns and dropping duplicate rows\n",
    "\n",
    "|  | dbNSFP count | CADD19 | CADD38 |\n",
    "|--|---------------| -------| -----|\n",
    "| DETECTED | 104475 | 105250 | 105793 |\n",
    "| NOT DECT | 1222911 | 1226678 | 1231490 |\n",
    "\n",
    "- all CADD rows greater than dbNSFP row count\n",
    "- all df's from CADD and dbNSFP have unique great_pos_id, ready to merge!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# (done) CADD file correcting pos_id # format (add leading 0's)\n",
    "- 9 digits with leading zeros for pos_id19 or pos_id38\n",
    "\n",
    "```python \n",
    "def create_pos_id19(df, chrr, pos, ref, alt):\n",
    "    # variables are colnames of df\n",
    "    df.loc[:,'pos_id19'] = df[chrr].astype(str) + '_' + \\\n",
    "    df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "    '_' + df[alt].astype(str)\n",
    "    return df\n",
    "\n",
    "def create_pos_id38(df, chrr, pos, ref, alt):\n",
    "    # variables are colnames of df\n",
    "    df.loc[:,'pos_id38'] = df[chrr].astype(str) + '_' + \\\n",
    "    df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "    '_' + df[alt].astype(str)\n",
    "    return df\n",
    "\n",
    "# dbNSFP FILES\n",
    "dbnsfp = pd.read_csv('SCORE_dbNSFP_selectcols_detected_CK_104475.csv', low_memory=False)\n",
    "notdbnsfp = pd.read_csv('SCORE_dbNSFP_selectcols_NOT_detected_CK_1222911.csv', low_memory=False)\n",
    "\n",
    "# CADD FILES\n",
    "det19 = pd.read_csv('SCORE_CADD_37_detected_CK.csv', low_memory=False, converters={'pos_hg19': '{:0>9}'.format})\n",
    "det38 = pd.read_csv('SCORE_CADD_38_detected_CK.csv', low_memory=False, converters={'pos_hg38': '{:0>9}'.format})\n",
    "not19 = pd.read_csv('SCORE_CADD_37_NOT_detected_CK.csv', low_memory=False, converters={'pos_hg19': '{:0>9}'.format})\n",
    "not38 = pd.read_csv('SCORE_CADD_38_NOT_detected_CK.csv', low_memory=False, converters={'pos_hg38': '{:0>9}'.format})\n",
    "```\n",
    "\n",
    "## calling function to create pos id with correct # format:\n",
    "```python\n",
    "# cadd detected files\n",
    "det19 = create_pos_id19(det19, 'chr', 'pos_hg19', 'Ref', 'Alt')\n",
    "det38 = create_pos_id38(det38, 'chr', 'pos_hg38', 'Ref', 'Alt')\n",
    "\n",
    "# cadd NOTdetected files\n",
    "not19 = create_pos_id19(not19, 'chr', 'pos_hg19', 'Ref', 'Alt')\n",
    "not38 = create_pos_id38(not38, 'chr', 'pos_hg38', 'Ref', 'Alt')\n",
    "\n",
    "# saving files with correct pos_id\n",
    "det19.to_csv('SCORE_CADD_37_detected_CK.csv', index=False)\n",
    "det38.to_csv('SCORE_CADD_38_detected_CK.csv',  index=False)\n",
    "not19.to_csv('SCORE_CADD_37_NOT_detected_CK.csv', index=False)\n",
    "not38.to_csv('SCORE_CADD_38_NOT_detected_CK.csv',  index=False)\n",
    "```\n",
    "\n",
    "---\n",
    "---\n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# (done) pt.5  1st making ENST column for dbNSFP files, \n",
    "### 'FeatureID' col already in rows of CADD file and may help make super_posid key a 1:1 mapping\n",
    "\n",
    "\n",
    "```python\n",
    "# header col index positions\n",
    "header = dbnsfp.columns\n",
    "header2 = notdbnsfp.columns\n",
    "d = {header[i] : i for i in range(0, len(header))}\n",
    "# print(d) # 'matched_index': 7, 'Ensembl_transcriptid': 8,\n",
    "d2 = {header2[i] : i for i in range(0, len(header2))}\n",
    "# print(d2) # 'matched_index': 5, 'Ensembl_transcriptid': 6, \n",
    "\n",
    "\n",
    "def feature_id_col(filename, outfile, subset):\n",
    "    if subset == 'dect':\n",
    "        # adding feature_ID column from parsing enst list col from dbNSFP with matched_index col\n",
    "        # modeled after function from Pmap_parseID_correction.py\n",
    "        with open(filename, newline='') as file:\n",
    "            # read in file, save header\n",
    "            csvReader = csv.reader(file)\n",
    "            header = next(csvReader)\n",
    "            # add new col\n",
    "            header.append('FeatureID')\n",
    "            # create and write to outfile\n",
    "            os.system(\"touch %s\" % (outfile))\n",
    "            with open(outfile, 'w') as out:\n",
    "                csvWriter = csv.writer(out)\n",
    "                csvWriter.writerow(header)\n",
    "            # loop over rows\n",
    "            for row in csvReader:\n",
    "                # create list\n",
    "                matchI = int(row[7])\n",
    "                ENST = row[8]\n",
    "                lsID = ENST.split(\";\")\n",
    "                featureid = lsID[matchI]\n",
    "                row.append(featureid)\n",
    "                with open(outfile, 'a') as out:\n",
    "                        csvWriter = csv.writer(out)\n",
    "                        csvWriter.writerow(row)\n",
    "        print(\"done with : \", outfile)\n",
    "    if subset == 'notdect':\n",
    "        # adding feature_ID column from parsing enst list col from dbNSFP with matched_index col\n",
    "        # modeled after function from Pmap_parseID_correction.py\n",
    "        with open(filename, newline='') as file:\n",
    "            # read in file, save header\n",
    "            csvReader = csv.reader(file)\n",
    "            header = next(csvReader)\n",
    "            # add new col\n",
    "            header.append('FeatureID')\n",
    "            # create and write to outfile\n",
    "            os.system(\"touch %s\" % (outfile))\n",
    "            with open(outfile, 'w') as out:\n",
    "                csvWriter = csv.writer(out)\n",
    "                csvWriter.writerow(header)\n",
    "            # loop over rows\n",
    "            for row in csvReader:\n",
    "                # create list\n",
    "                matchI = int(row[5])\n",
    "                ENST = row[6]\n",
    "                lsID = ENST.split(\";\")\n",
    "                featureid = lsID[matchI]\n",
    "                row.append(featureid)\n",
    "                with open(outfile, 'a') as out:\n",
    "                        csvWriter = csv.writer(out)\n",
    "                        csvWriter.writerow(row)\n",
    "        print(\"done with : \", outfile)\n",
    "        \n",
    "        \n",
    "# calling function\n",
    "feature_id_col('SCORE_dbNSFP_selectcols_detected_CK_104475.csv', 'SCORE2_dbNSFP_selectcols_detected_CK_104475.csv', 'dect')\n",
    "feature_id_col('SCORE_dbNSFP_selectcols_NOT_detected_CK_1222911.csv', 'SCORE2_dbNSFP_selectcols_NOT_detected_CK_1222911.csv', 'notdect')\n",
    "```\n",
    "\n",
    "## result of feature_id_col() =\n",
    "- all dbnsfp file rows have 1 ENST ID matching index of canonical ukb ID\n",
    "\n",
    "description files for new featureID column added created with below code:\n",
    "```python\n",
    "saveColumnValues(dbnsfp, 'FeatureID', 'dbnsfp_featureID_detected_description.csv')\n",
    "saveColumnValues(notdbnsfp, 'FeatureID', 'dbnsfp_featureID_notdetected_description.csv')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "\n",
    "\n",
    "# [4] simplifying dbNSFP score files by COLUMNS OF INTEREST\n",
    "\n",
    "rules: \n",
    "1. merge on pos_idcoordinates\n",
    "2. match on Amino_acids 'A/A'\n",
    "3. protein positions match- not required since most annotations refer to coordinates for annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/BUG_FIX/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(104475, 400)\n"
     ]
    }
   ],
   "source": [
    "##  detected dbNSFP score file: \n",
    "det_scores = 'SCORE_annotation_3840_CYS_LYS_detected.csv'\n",
    "detscore = pd.read_csv(det_scores, low_memory=False)\n",
    "print(detscore.shape)\n",
    "# (104475, 400)\n",
    "\n",
    "### simplify df with select columns: \n",
    "det_cols = ['pos_id19', 'pos_id38_x', 'aaref_x', 'aaalt_x', 'matched_aapos_x',\n",
    "       'CADD_phred_hg38', 'matched_UKBID_x', 'pos_dict', 'matched_target',\n",
    "       'CADD_phred_hg19', 'CADDdiff_38minus19', 'Amino_acids', 'pos_ID',\n",
    "       'Cys_reactivity', 'Cys_react_threshold', 'Cys_target_label',\n",
    "       'Lys_reactivity', 'Lys_react_threshold', 'Lys_target_label', 'rs_dbSNP151', 'genename','Ensembl_transcriptid','cds_strand', 'refcodon', 'codonpos', 'codon_degeneracy','LRT_score', 'LRT_converted_rankscore',\n",
    "       'LRT_pred', 'LRT_Omega', 'MutationTaster_score',\n",
    "       'FATHMM_score','FATHMM_converted_rankscore', 'FATHMM_pred', 'PROVEAN_score','PROVEAN_converted_rankscore', 'PROVEAN_pred', 'VEST4_score','VEST4_rankscore', 'MetaSVM_score', 'MetaSVM_rankscore', 'MetaSVM_pred','MetaLR_score', 'MetaLR_rankscore', 'MetaLR_pred', 'Reliability_index','M-CAP_score', 'M-CAP_rankscore', 'M-CAP_pred', 'REVEL_score','REVEL_rankscore', 'MutPred_score', 'MutPred_rankscore',\n",
    "       'MutPred_protID', 'MutPred_AAchange', 'MutPred_Top5features',\n",
    "        'MPC_score','MPC_rankscore','PrimateAI_score', 'PrimateAI_rankscore', 'PrimateAI_pred',\n",
    "       'DANN_score', 'DANN_rankscore', 'fathmm-MKL_coding_score',\n",
    "       'fathmm-MKL_coding_rankscore', 'fathmm-MKL_coding_pred',\n",
    "       'fathmm-MKL_coding_group', 'fathmm-XF_coding_score',\n",
    "       'fathmm-XF_coding_rankscore', 'fathmm-XF_coding_pred',\n",
    "       'Eigen-raw_coding', 'Eigen-raw_coding_rankscore', 'Eigen-pred_coding',\n",
    "       'Eigen-PC-raw_coding', 'Eigen-PC-raw_coding_rankscore',\n",
    "       'Eigen-PC-phred_coding', 'matched_index']\n",
    "# filter\n",
    "simpleD = detscore[det_cols].copy()\n",
    "rename = ['pos_id19', 'pos_id38', 'aaref', 'aaalt', 'matched_aapos',\n",
    "       'CADD_phred_hg38', 'matched_UKBID', 'pos_dict', 'matched_target',\n",
    "       'CADD_phred_hg19', 'CADDdiff_38minus19', 'Amino_acids', 'pos_ID',\n",
    "       'Cys_reactivity', 'Cys_react_threshold', 'Cys_target_label',\n",
    "       'Lys_reactivity', 'Lys_react_threshold', 'Lys_target_label', 'rs_dbSNP151', 'genename','Ensembl_transcriptid', 'cds_strand', 'refcodon', 'codonpos', 'codon_degeneracy','LRT_score', 'LRT_converted_rankscore',\n",
    "       'LRT_pred', 'LRT_Omega', 'MutationTaster_score',\n",
    "       'FATHMM_score','FATHMM_converted_rankscore', 'FATHMM_pred', 'PROVEAN_score','PROVEAN_converted_rankscore', 'PROVEAN_pred', 'VEST4_score','VEST4_rankscore', 'MetaSVM_score', 'MetaSVM_rankscore', 'MetaSVM_pred','MetaLR_score', 'MetaLR_rankscore', 'MetaLR_pred', 'Reliability_index','M-CAP_score', 'M-CAP_rankscore', 'M-CAP_pred', 'REVEL_score','REVEL_rankscore', 'MutPred_score', 'MutPred_rankscore',\n",
    "       'MutPred_protID', 'MutPred_AAchange', 'MutPred_Top5features',\n",
    "       'MPC_score','MPC_rankscore','PrimateAI_score', 'PrimateAI_rankscore', 'PrimateAI_pred',\n",
    "       'DANN_score', 'DANN_rankscore', 'fathmm-MKL_coding_score',\n",
    "       'fathmm-MKL_coding_rankscore', 'fathmm-MKL_coding_pred',\n",
    "       'fathmm-MKL_coding_group', 'fathmm-XF_coding_score',\n",
    "       'fathmm-XF_coding_rankscore', 'fathmm-XF_coding_pred',\n",
    "       'Eigen-raw_coding', 'Eigen-raw_coding_rankscore', 'Eigen-pred_coding',\n",
    "       'Eigen-PC-raw_coding', 'Eigen-PC-raw_coding_rankscore',\n",
    "       'Eigen-PC-phred_coding', 'matched_index']\n",
    "simpleD.columns = rename\n",
    "# change order\n",
    "neworder = ['pos_id19', 'pos_id38', 'pos_ID', 'matched_UKBID', 'pos_dict', 'matched_target', 'matched_aapos', 'matched_index', 'Ensembl_transcriptid',\n",
    "        'Amino_acids', 'aaref', 'aaalt', 'CADD_phred_hg38',  'CADD_phred_hg19', 'CADDdiff_38minus19', \n",
    "       'Cys_reactivity', 'Cys_react_threshold', 'Cys_target_label',\n",
    "       'Lys_reactivity', 'Lys_react_threshold', 'Lys_target_label', \n",
    "       'rs_dbSNP151', 'genename','cds_strand', 'refcodon', 'codonpos', 'codon_degeneracy','LRT_score', 'LRT_converted_rankscore',\n",
    "       'LRT_pred', 'LRT_Omega', 'MutationTaster_score',\n",
    "       'FATHMM_score','FATHMM_converted_rankscore', 'FATHMM_pred', 'PROVEAN_score','PROVEAN_converted_rankscore', 'PROVEAN_pred', 'VEST4_score','VEST4_rankscore', 'MetaSVM_score', 'MetaSVM_rankscore', 'MetaSVM_pred','MetaLR_score', 'MetaLR_rankscore', 'MetaLR_pred', 'Reliability_index','M-CAP_score', 'M-CAP_rankscore', 'M-CAP_pred', 'REVEL_score','REVEL_rankscore', 'MutPred_score', 'MutPred_rankscore',\n",
    "       'MutPred_protID', 'MutPred_AAchange', 'MutPred_Top5features',\n",
    "       'MPC_score','MPC_rankscore', 'PrimateAI_score', 'PrimateAI_rankscore', 'PrimateAI_pred',\n",
    "       'DANN_score', 'DANN_rankscore', 'fathmm-MKL_coding_score',\n",
    "       'fathmm-MKL_coding_rankscore', 'fathmm-MKL_coding_pred',\n",
    "       'fathmm-MKL_coding_group', 'fathmm-XF_coding_score',\n",
    "       'fathmm-XF_coding_rankscore', 'fathmm-XF_coding_pred',\n",
    "       'Eigen-raw_coding', 'Eigen-raw_coding_rankscore', 'Eigen-pred_coding',\n",
    "       'Eigen-PC-raw_coding', 'Eigen-PC-raw_coding_rankscore',\n",
    "       'Eigen-PC-phred_coding']\n",
    "simpleD = simpleD[neworder]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(104475, 77)\n"
     ]
    }
   ],
   "source": [
    "print(simpleD.shape)\n",
    "simpleD.to_csv(\"SCORE_dbNSFP_selectcols_detected_CK_104475.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1222911, 385)\n"
     ]
    }
   ],
   "source": [
    "## not detected dbNSFP score file: \n",
    "not_scores = 'SCORE_annotation_3840_CYS_LYS_NOT_allcol_detected.csv'\n",
    "notscore = pd.read_csv(not_scores, low_memory=False)\n",
    "print(notscore.shape)\n",
    "# (1222911, 385)\n",
    "\n",
    "### simplify df with select columns: \n",
    "not_cols = ['pos_id19', 'CADD_phred_hg19', 'CADDdiff_38minus19', 'Amino_acids',\n",
    "       'pos_ID_falseCKtarget', 'aaref',\n",
    "       'aaalt', 'rs_dbSNP151', 'genename', 'Ensembl_transcriptid','cds_strand', 'refcodon', 'codonpos','codon_degeneracy', 'LRT_score',\n",
    "       'LRT_converted_rankscore', 'LRT_pred', 'LRT_Omega',\n",
    "       'MutationTaster_score', 'MutationTaster_converted_rankscore',\n",
    "       'MutationTaster_pred', 'MutationTaster_model', 'MutationTaster_AAE',\n",
    "       'MutationAssessor_score', 'MutationAssessor_rankscore',\n",
    "       'MutationAssessor_pred', 'FATHMM_score', 'FATHMM_converted_rankscore',\n",
    "       'FATHMM_pred', 'PROVEAN_score', 'PROVEAN_converted_rankscore',\n",
    "       'PROVEAN_pred', 'VEST4_score', 'VEST4_rankscore', 'MetaSVM_score',\n",
    "       'MetaSVM_rankscore', 'MetaSVM_pred', 'MetaLR_score', 'MetaLR_rankscore',\n",
    "       'MetaLR_pred', 'Reliability_index', 'M-CAP_score', 'M-CAP_rankscore',\n",
    "       'M-CAP_pred', 'REVEL_score', 'REVEL_rankscore', 'MutPred_score',\n",
    "       'MutPred_rankscore', 'MutPred_protID', 'MutPred_AAchange',\n",
    "       'MutPred_Top5features', 'MPC_score','MPC_rankscore', 'PrimateAI_score', 'PrimateAI_rankscore',\n",
    "       'PrimateAI_pred','CADD_phred','DANN_score', 'DANN_rankscore', 'fathmm-MKL_coding_score',\n",
    "       'fathmm-MKL_coding_rankscore', 'fathmm-MKL_coding_pred',\n",
    "       'fathmm-MKL_coding_group', 'fathmm-XF_coding_score',\n",
    "       'fathmm-XF_coding_rankscore', 'fathmm-XF_coding_pred',\n",
    "       'Eigen-raw_coding', 'Eigen-raw_coding_rankscore', 'Eigen-pred_coding',\n",
    "       'Eigen-PC-raw_coding', 'Eigen-PC-raw_coding_rankscore',\n",
    "       'Eigen-PC-phred_coding', 'matched_UKBID',\n",
    "       'matched_aapos', 'matched_index', 'pos_id38']\n",
    "# filter\n",
    "simpleN = notscore[not_cols].copy()\n",
    "# rename\n",
    "simpleN.columns = ['pos_id19', 'CADD_phred_hg19', 'CADDdiff_38minus19', 'Amino_acids',\n",
    "       'pos_ID', 'aaref',\n",
    "       'aaalt', 'rs_dbSNP151', 'genename', 'Ensembl_transcriptid','cds_strand', 'refcodon', 'codonpos','codon_degeneracy', 'LRT_score',\n",
    "       'LRT_converted_rankscore', 'LRT_pred', 'LRT_Omega',\n",
    "       'MutationTaster_score', 'MutationTaster_converted_rankscore',\n",
    "       'MutationTaster_pred', 'MutationTaster_model', 'MutationTaster_AAE',\n",
    "       'MutationAssessor_score', 'MutationAssessor_rankscore',\n",
    "       'MutationAssessor_pred', 'FATHMM_score', 'FATHMM_converted_rankscore',\n",
    "       'FATHMM_pred', 'PROVEAN_score', 'PROVEAN_converted_rankscore',\n",
    "       'PROVEAN_pred', 'VEST4_score', 'VEST4_rankscore', 'MetaSVM_score',\n",
    "       'MetaSVM_rankscore', 'MetaSVM_pred', 'MetaLR_score', 'MetaLR_rankscore',\n",
    "       'MetaLR_pred', 'Reliability_index', 'M-CAP_score', 'M-CAP_rankscore',\n",
    "       'M-CAP_pred', 'REVEL_score', 'REVEL_rankscore', 'MutPred_score',\n",
    "       'MutPred_rankscore', 'MutPred_protID', 'MutPred_AAchange',\n",
    "       'MutPred_Top5features', 'MPC_score','MPC_rankscore', 'PrimateAI_score', 'PrimateAI_rankscore',\n",
    "       'PrimateAI_pred','CADD_phred_hg38','DANN_score', 'DANN_rankscore', 'fathmm-MKL_coding_score',\n",
    "       'fathmm-MKL_coding_rankscore', 'fathmm-MKL_coding_pred',\n",
    "       'fathmm-MKL_coding_group', 'fathmm-XF_coding_score',\n",
    "       'fathmm-XF_coding_rankscore', 'fathmm-XF_coding_pred',\n",
    "       'Eigen-raw_coding', 'Eigen-raw_coding_rankscore', 'Eigen-pred_coding',\n",
    "       'Eigen-PC-raw_coding', 'Eigen-PC-raw_coding_rankscore',\n",
    "       'Eigen-PC-phred_coding', 'matched_UKBID',\n",
    "       'matched_aapos', 'matched_index', 'pos_id38']\n",
    "# change order\n",
    "neworder = ['pos_id19', 'pos_id38', 'pos_ID', 'matched_UKBID', 'matched_aapos', 'matched_index','Ensembl_transcriptid', 'Amino_acids','aaref','aaalt', \n",
    "'CADD_phred_hg38', 'CADD_phred_hg19', 'CADDdiff_38minus19', 'rs_dbSNP151', 'genename', 'cds_strand', 'refcodon', 'codonpos','codon_degeneracy', 'LRT_score',\n",
    "       'LRT_converted_rankscore', 'LRT_pred', 'LRT_Omega',\n",
    "       'MutationTaster_score', 'MutationTaster_converted_rankscore',\n",
    "       'MutationTaster_pred', 'MutationTaster_model', 'MutationTaster_AAE',\n",
    "       'MutationAssessor_score', 'MutationAssessor_rankscore',\n",
    "       'MutationAssessor_pred', 'FATHMM_score', 'FATHMM_converted_rankscore',\n",
    "       'FATHMM_pred', 'PROVEAN_score', 'PROVEAN_converted_rankscore',\n",
    "       'PROVEAN_pred', 'VEST4_score', 'VEST4_rankscore', 'MetaSVM_score',\n",
    "       'MetaSVM_rankscore', 'MetaSVM_pred', 'MetaLR_score', 'MetaLR_rankscore',\n",
    "       'MetaLR_pred', 'Reliability_index', 'M-CAP_score', 'M-CAP_rankscore',\n",
    "       'M-CAP_pred', 'REVEL_score', 'REVEL_rankscore', 'MutPred_score',\n",
    "       'MutPred_rankscore', 'MutPred_protID', 'MutPred_AAchange',\n",
    "       'MutPred_Top5features', 'MPC_score','MPC_rankscore', 'PrimateAI_score', 'PrimateAI_rankscore',\n",
    "       'PrimateAI_pred','DANN_score', 'DANN_rankscore', 'fathmm-MKL_coding_score',\n",
    "       'fathmm-MKL_coding_rankscore', 'fathmm-MKL_coding_pred',\n",
    "       'fathmm-MKL_coding_group', 'fathmm-XF_coding_score',\n",
    "       'fathmm-XF_coding_rankscore', 'fathmm-XF_coding_pred',\n",
    "       'Eigen-raw_coding', 'Eigen-raw_coding_rankscore', 'Eigen-pred_coding',\n",
    "       'Eigen-PC-raw_coding', 'Eigen-PC-raw_coding_rankscore',\n",
    "       'Eigen-PC-phred_coding']\n",
    "simpleN = simpleN[neworder]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "simpleN.to_csv(\"SCORE_dbNSFP_selectcols_NOT_detected_CK_1222911.csv\", index=False)\n",
    "# shape (1222911, 75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos_id19</th>\n",
       "      <th>pos_id38</th>\n",
       "      <th>pos_ID</th>\n",
       "      <th>matched_UKBID</th>\n",
       "      <th>matched_aapos</th>\n",
       "      <th>matched_index</th>\n",
       "      <th>Ensembl_transcriptid</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>aaref</th>\n",
       "      <th>aaalt</th>\n",
       "      <th>CADD_phred_hg38</th>\n",
       "      <th>CADD_phred_hg19</th>\n",
       "      <th>CADDdiff_38minus19</th>\n",
       "      <th>rs_dbSNP151</th>\n",
       "      <th>genename</th>\n",
       "      <th>cds_strand</th>\n",
       "      <th>refcodon</th>\n",
       "      <th>codonpos</th>\n",
       "      <th>codon_degeneracy</th>\n",
       "      <th>LRT_score</th>\n",
       "      <th>LRT_converted_rankscore</th>\n",
       "      <th>LRT_pred</th>\n",
       "      <th>LRT_Omega</th>\n",
       "      <th>MutationTaster_score</th>\n",
       "      <th>MutationTaster_converted_rankscore</th>\n",
       "      <th>MutationTaster_pred</th>\n",
       "      <th>MutationTaster_model</th>\n",
       "      <th>MutationTaster_AAE</th>\n",
       "      <th>MutationAssessor_score</th>\n",
       "      <th>MutationAssessor_rankscore</th>\n",
       "      <th>MutationAssessor_pred</th>\n",
       "      <th>FATHMM_score</th>\n",
       "      <th>FATHMM_converted_rankscore</th>\n",
       "      <th>FATHMM_pred</th>\n",
       "      <th>PROVEAN_score</th>\n",
       "      <th>PROVEAN_converted_rankscore</th>\n",
       "      <th>PROVEAN_pred</th>\n",
       "      <th>VEST4_score</th>\n",
       "      <th>VEST4_rankscore</th>\n",
       "      <th>MetaSVM_score</th>\n",
       "      <th>MetaSVM_rankscore</th>\n",
       "      <th>MetaSVM_pred</th>\n",
       "      <th>MetaLR_score</th>\n",
       "      <th>MetaLR_rankscore</th>\n",
       "      <th>MetaLR_pred</th>\n",
       "      <th>Reliability_index</th>\n",
       "      <th>M-CAP_score</th>\n",
       "      <th>M-CAP_rankscore</th>\n",
       "      <th>M-CAP_pred</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>REVEL_rankscore</th>\n",
       "      <th>MutPred_score</th>\n",
       "      <th>MutPred_rankscore</th>\n",
       "      <th>MutPred_protID</th>\n",
       "      <th>MutPred_AAchange</th>\n",
       "      <th>MutPred_Top5features</th>\n",
       "      <th>MPC_score</th>\n",
       "      <th>MPC_rankscore</th>\n",
       "      <th>PrimateAI_score</th>\n",
       "      <th>PrimateAI_rankscore</th>\n",
       "      <th>PrimateAI_pred</th>\n",
       "      <th>DANN_score</th>\n",
       "      <th>DANN_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_score</th>\n",
       "      <th>fathmm-MKL_coding_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_pred</th>\n",
       "      <th>fathmm-MKL_coding_group</th>\n",
       "      <th>fathmm-XF_coding_score</th>\n",
       "      <th>fathmm-XF_coding_rankscore</th>\n",
       "      <th>fathmm-XF_coding_pred</th>\n",
       "      <th>Eigen-raw_coding</th>\n",
       "      <th>Eigen-raw_coding_rankscore</th>\n",
       "      <th>Eigen-pred_coding</th>\n",
       "      <th>Eigen-PC-raw_coding</th>\n",
       "      <th>Eigen-PC-raw_coding_rankscore</th>\n",
       "      <th>Eigen-PC-phred_coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>10_000093156_C_A</td>\n",
       "      <td>10_000047216_C_A</td>\n",
       "      <td>Q3ZCM7_K392</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>392</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000564130;ENST00000568866;ENST00000568584</td>\n",
       "      <td>Lys/Asn</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>23.0</td>\n",
       "      <td>22.9</td>\n",
       "      <td>0.1</td>\n",
       "      <td>.</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>AAG</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.654618;0.654618;0.654618;0.654618;1</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D;D;D;D;D</td>\n",
       "      <td>simple_aae;simple_aae;simple_aae;simple_aae;wi...</td>\n",
       "      <td>K358N;K355N;K392N;K320N;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>-1.98;-1.98;-1.98</td>\n",
       "      <td>0.85247</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>-2.77;-2.55;-2.6</td>\n",
       "      <td>0.58733</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>0.579;0.574;0.53</td>\n",
       "      <td>0.6007100000000001</td>\n",
       "      <td>0.3991</td>\n",
       "      <td>0.8908299999999999</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7111</td>\n",
       "      <td>0.90075</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.009489</td>\n",
       "      <td>0.24855</td>\n",
       "      <td>T</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.78035</td>\n",
       "      <td>0.718</td>\n",
       "      <td>0.85313</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>K392N</td>\n",
       "      <td>Loss of methylation at K392 (P = 0.003); Loss ...</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.751451253891</td>\n",
       "      <td>0.74660</td>\n",
       "      <td>T</td>\n",
       "      <td>0.984728</td>\n",
       "      <td>0.41873</td>\n",
       "      <td>0.38681</td>\n",
       "      <td>0.26043</td>\n",
       "      <td>N</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.295876</td>\n",
       "      <td>0.40587</td>\n",
       "      <td>N</td>\n",
       "      <td>0.0676056746461709</td>\n",
       "      <td>0.44955</td>\n",
       "      <td>2.76036</td>\n",
       "      <td>-0.20748312943205102</td>\n",
       "      <td>0.31266</td>\n",
       "      <td>1.767854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>10_000093156_C_G</td>\n",
       "      <td>10_000047216_C_G</td>\n",
       "      <td>Q3ZCM7_K392</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>392</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000564130;ENST00000568866;ENST00000568584</td>\n",
       "      <td>Lys/Asn</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>23.1</td>\n",
       "      <td>22.8</td>\n",
       "      <td>0.3</td>\n",
       "      <td>.</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>AAG</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.654618;0.654618;0.654618;0.654618;1</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D;D;D;D;D</td>\n",
       "      <td>simple_aae;simple_aae;simple_aae;simple_aae;wi...</td>\n",
       "      <td>K358N;K355N;K392N;K320N;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>-1.98;-1.98;-1.98</td>\n",
       "      <td>0.85247</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>-2.77;-2.55;-2.6</td>\n",
       "      <td>0.58733</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>0.579;0.574;0.53</td>\n",
       "      <td>0.6007100000000001</td>\n",
       "      <td>0.3991</td>\n",
       "      <td>0.8908299999999999</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7111</td>\n",
       "      <td>0.90075</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.008795</td>\n",
       "      <td>0.23196</td>\n",
       "      <td>T</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.78035</td>\n",
       "      <td>0.718</td>\n",
       "      <td>0.85313</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>K392N</td>\n",
       "      <td>Loss of methylation at K392 (P = 0.003); Loss ...</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.751451253891</td>\n",
       "      <td>0.74660</td>\n",
       "      <td>T</td>\n",
       "      <td>0.984663</td>\n",
       "      <td>0.41803</td>\n",
       "      <td>0.40406</td>\n",
       "      <td>0.26429</td>\n",
       "      <td>N</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.302556</td>\n",
       "      <td>0.41077</td>\n",
       "      <td>N</td>\n",
       "      <td>0.0676056746461709</td>\n",
       "      <td>0.44955</td>\n",
       "      <td>2.76036</td>\n",
       "      <td>-0.20748312943205102</td>\n",
       "      <td>0.31266</td>\n",
       "      <td>1.767854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>10_000093157_T_A</td>\n",
       "      <td>10_000047217_T_A</td>\n",
       "      <td>Q3ZCM7_K392</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>392</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000564130;ENST00000568866;ENST00000568584</td>\n",
       "      <td>Lys/Met</td>\n",
       "      <td>K</td>\n",
       "      <td>M</td>\n",
       "      <td>23.2</td>\n",
       "      <td>23.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>.</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>AAG</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.820833;0.820833;0.820833;0.820833;1</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D;D;D;D;D</td>\n",
       "      <td>simple_aae;simple_aae;simple_aae;simple_aae;wi...</td>\n",
       "      <td>K358M;K355M;K392M;K320M;.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>-2.08;-2.08;-2.08</td>\n",
       "      <td>0.86010</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>-3.36;-3.08;-3.14</td>\n",
       "      <td>0.66549</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>0.601;0.607;0.568</td>\n",
       "      <td>0.62442</td>\n",
       "      <td>0.5527</td>\n",
       "      <td>0.9127</td>\n",
       "      <td>D</td>\n",
       "      <td>0.8388</td>\n",
       "      <td>0.9461</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.01492</td>\n",
       "      <td>0.35345</td>\n",
       "      <td>T</td>\n",
       "      <td>0.505</td>\n",
       "      <td>0.78906</td>\n",
       "      <td>0.698</td>\n",
       "      <td>0.83491</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>K392M</td>\n",
       "      <td>Loss of methylation at K392 (P = 0.003); Loss ...</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.727143764496</td>\n",
       "      <td>0.71088</td>\n",
       "      <td>T</td>\n",
       "      <td>0.964042</td>\n",
       "      <td>0.29699</td>\n",
       "      <td>0.52831</td>\n",
       "      <td>0.29225</td>\n",
       "      <td>D</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.364324</td>\n",
       "      <td>0.45291</td>\n",
       "      <td>N</td>\n",
       "      <td>0.31076238698785996</td>\n",
       "      <td>0.56701</td>\n",
       "      <td>3.8352209999999998</td>\n",
       "      <td>-0.0235518622390073</td>\n",
       "      <td>0.38654</td>\n",
       "      <td>2.280427</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos_id19          pos_id38       pos_ID matched_UKBID  \\\n",
       "0  10_000093156_C_A  10_000047216_C_A  Q3ZCM7_K392        Q3ZCM7   \n",
       "1  10_000093156_C_G  10_000047216_C_G  Q3ZCM7_K392        Q3ZCM7   \n",
       "2  10_000093157_T_A  10_000047217_T_A  Q3ZCM7_K392        Q3ZCM7   \n",
       "\n",
       "   matched_aapos  matched_index  \\\n",
       "0            392              2   \n",
       "1            392              2   \n",
       "2            392              2   \n",
       "\n",
       "                              Ensembl_transcriptid Amino_acids aaref aaalt  \\\n",
       "0  ENST00000564130;ENST00000568866;ENST00000568584     Lys/Asn     K     N   \n",
       "1  ENST00000564130;ENST00000568866;ENST00000568584     Lys/Asn     K     N   \n",
       "2  ENST00000564130;ENST00000568866;ENST00000568584     Lys/Met     K     M   \n",
       "\n",
       "   CADD_phred_hg38  CADD_phred_hg19  CADDdiff_38minus19 rs_dbSNP151 genename  \\\n",
       "0             23.0             22.9                 0.1           .    TUBB8   \n",
       "1             23.1             22.8                 0.3           .    TUBB8   \n",
       "2             23.2             23.2                 0.0           .    TUBB8   \n",
       "\n",
       "  cds_strand refcodon  codonpos  codon_degeneracy LRT_score  \\\n",
       "0          -      AAG         3                 2  0.000010   \n",
       "1          -      AAG         3                 2  0.000010   \n",
       "2          -      AAG         2                 0  0.000010   \n",
       "\n",
       "  LRT_converted_rankscore LRT_pred LRT_Omega  \\\n",
       "0                 0.62929        U  0.000000   \n",
       "1                 0.62929        U  0.000000   \n",
       "2                 0.62929        U  0.000000   \n",
       "\n",
       "                    MutationTaster_score MutationTaster_converted_rankscore  \\\n",
       "0  0.654618;0.654618;0.654618;0.654618;1                            0.81001   \n",
       "1  0.654618;0.654618;0.654618;0.654618;1                            0.81001   \n",
       "2  0.820833;0.820833;0.820833;0.820833;1                            0.81001   \n",
       "\n",
       "  MutationTaster_pred                               MutationTaster_model  \\\n",
       "0           D;D;D;D;D  simple_aae;simple_aae;simple_aae;simple_aae;wi...   \n",
       "1           D;D;D;D;D  simple_aae;simple_aae;simple_aae;simple_aae;wi...   \n",
       "2           D;D;D;D;D  simple_aae;simple_aae;simple_aae;simple_aae;wi...   \n",
       "\n",
       "          MutationTaster_AAE MutationAssessor_score  \\\n",
       "0  K358N;K355N;K392N;K320N;.                  .;.;.   \n",
       "1  K358N;K355N;K392N;K320N;.                  .;.;.   \n",
       "2  K358M;K355M;K392M;K320M;.                  .;.;.   \n",
       "\n",
       "  MutationAssessor_rankscore MutationAssessor_pred       FATHMM_score  \\\n",
       "0                          .                 .;.;.  -1.98;-1.98;-1.98   \n",
       "1                          .                 .;.;.  -1.98;-1.98;-1.98   \n",
       "2                          .                 .;.;.  -2.08;-2.08;-2.08   \n",
       "\n",
       "  FATHMM_converted_rankscore FATHMM_pred      PROVEAN_score  \\\n",
       "0                    0.85247       D;D;D   -2.77;-2.55;-2.6   \n",
       "1                    0.85247       D;D;D   -2.77;-2.55;-2.6   \n",
       "2                    0.86010       D;D;D  -3.36;-3.08;-3.14   \n",
       "\n",
       "  PROVEAN_converted_rankscore PROVEAN_pred        VEST4_score  \\\n",
       "0                     0.58733        D;D;D   0.579;0.574;0.53   \n",
       "1                     0.58733        D;D;D   0.579;0.574;0.53   \n",
       "2                     0.66549        D;D;D  0.601;0.607;0.568   \n",
       "\n",
       "      VEST4_rankscore MetaSVM_score   MetaSVM_rankscore MetaSVM_pred  \\\n",
       "0  0.6007100000000001        0.3991  0.8908299999999999            D   \n",
       "1  0.6007100000000001        0.3991  0.8908299999999999            D   \n",
       "2             0.62442        0.5527              0.9127            D   \n",
       "\n",
       "  MetaLR_score MetaLR_rankscore MetaLR_pred Reliability_index M-CAP_score  \\\n",
       "0       0.7111          0.90075           D                 9    0.009489   \n",
       "1       0.7111          0.90075           D                 9    0.008795   \n",
       "2       0.8388           0.9461           D                 9     0.01492   \n",
       "\n",
       "  M-CAP_rankscore M-CAP_pred REVEL_score REVEL_rankscore MutPred_score  \\\n",
       "0         0.24855          T       0.491         0.78035         0.718   \n",
       "1         0.23196          T       0.491         0.78035         0.718   \n",
       "2         0.35345          T       0.505         0.78906         0.698   \n",
       "\n",
       "  MutPred_rankscore MutPred_protID MutPred_AAchange  \\\n",
       "0           0.85313         Q3ZCM7            K392N   \n",
       "1           0.85313         Q3ZCM7            K392N   \n",
       "2           0.83491         Q3ZCM7            K392M   \n",
       "\n",
       "                                MutPred_Top5features MPC_score MPC_rankscore  \\\n",
       "0  Loss of methylation at K392 (P = 0.003); Loss ...     .;.;.             .   \n",
       "1  Loss of methylation at K392 (P = 0.003); Loss ...     .;.;.             .   \n",
       "2  Loss of methylation at K392 (P = 0.003); Loss ...     .;.;.             .   \n",
       "\n",
       "  PrimateAI_score PrimateAI_rankscore PrimateAI_pred  DANN_score  \\\n",
       "0  0.751451253891             0.74660              T    0.984728   \n",
       "1  0.751451253891             0.74660              T    0.984663   \n",
       "2  0.727143764496             0.71088              T    0.964042   \n",
       "\n",
       "   DANN_rankscore  fathmm-MKL_coding_score  fathmm-MKL_coding_rankscore  \\\n",
       "0         0.41873                  0.38681                      0.26043   \n",
       "1         0.41803                  0.40406                      0.26429   \n",
       "2         0.29699                  0.52831                      0.29225   \n",
       "\n",
       "  fathmm-MKL_coding_pred fathmm-MKL_coding_group fathmm-XF_coding_score  \\\n",
       "0                      N                   AEFBI               0.295876   \n",
       "1                      N                   AEFBI               0.302556   \n",
       "2                      D                   AEFBI               0.364324   \n",
       "\n",
       "  fathmm-XF_coding_rankscore fathmm-XF_coding_pred     Eigen-raw_coding  \\\n",
       "0                    0.40587                     N   0.0676056746461709   \n",
       "1                    0.41077                     N   0.0676056746461709   \n",
       "2                    0.45291                     N  0.31076238698785996   \n",
       "\n",
       "  Eigen-raw_coding_rankscore   Eigen-pred_coding   Eigen-PC-raw_coding  \\\n",
       "0                    0.44955             2.76036  -0.20748312943205102   \n",
       "1                    0.44955             2.76036  -0.20748312943205102   \n",
       "2                    0.56701  3.8352209999999998   -0.0235518622390073   \n",
       "\n",
       "  Eigen-PC-raw_coding_rankscore Eigen-PC-phred_coding  \n",
       "0                       0.31266              1.767854  \n",
       "1                       0.31266              1.767854  \n",
       "2                       0.38654              2.280427  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simpleN.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos_id19</th>\n",
       "      <th>pos_id38</th>\n",
       "      <th>pos_ID</th>\n",
       "      <th>matched_UKBID</th>\n",
       "      <th>pos_dict</th>\n",
       "      <th>matched_target</th>\n",
       "      <th>matched_aapos</th>\n",
       "      <th>matched_index</th>\n",
       "      <th>Ensembl_transcriptid</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>aaref</th>\n",
       "      <th>aaalt</th>\n",
       "      <th>CADD_phred_hg38</th>\n",
       "      <th>CADD_phred_hg19</th>\n",
       "      <th>CADDdiff_38minus19</th>\n",
       "      <th>Cys_reactivity</th>\n",
       "      <th>Cys_react_threshold</th>\n",
       "      <th>Cys_target_label</th>\n",
       "      <th>Lys_reactivity</th>\n",
       "      <th>Lys_react_threshold</th>\n",
       "      <th>Lys_target_label</th>\n",
       "      <th>rs_dbSNP151</th>\n",
       "      <th>genename</th>\n",
       "      <th>cds_strand</th>\n",
       "      <th>refcodon</th>\n",
       "      <th>codonpos</th>\n",
       "      <th>codon_degeneracy</th>\n",
       "      <th>LRT_score</th>\n",
       "      <th>LRT_converted_rankscore</th>\n",
       "      <th>LRT_pred</th>\n",
       "      <th>LRT_Omega</th>\n",
       "      <th>MutationTaster_score</th>\n",
       "      <th>FATHMM_score</th>\n",
       "      <th>FATHMM_converted_rankscore</th>\n",
       "      <th>FATHMM_pred</th>\n",
       "      <th>PROVEAN_score</th>\n",
       "      <th>PROVEAN_converted_rankscore</th>\n",
       "      <th>PROVEAN_pred</th>\n",
       "      <th>VEST4_score</th>\n",
       "      <th>VEST4_rankscore</th>\n",
       "      <th>MetaSVM_score</th>\n",
       "      <th>MetaSVM_rankscore</th>\n",
       "      <th>MetaSVM_pred</th>\n",
       "      <th>MetaLR_score</th>\n",
       "      <th>MetaLR_rankscore</th>\n",
       "      <th>MetaLR_pred</th>\n",
       "      <th>Reliability_index</th>\n",
       "      <th>M-CAP_score</th>\n",
       "      <th>M-CAP_rankscore</th>\n",
       "      <th>M-CAP_pred</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>REVEL_rankscore</th>\n",
       "      <th>MutPred_score</th>\n",
       "      <th>MutPred_rankscore</th>\n",
       "      <th>MutPred_protID</th>\n",
       "      <th>MutPred_AAchange</th>\n",
       "      <th>MutPred_Top5features</th>\n",
       "      <th>MPC_score</th>\n",
       "      <th>MPC_rankscore</th>\n",
       "      <th>PrimateAI_score</th>\n",
       "      <th>PrimateAI_rankscore</th>\n",
       "      <th>PrimateAI_pred</th>\n",
       "      <th>DANN_score</th>\n",
       "      <th>DANN_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_score</th>\n",
       "      <th>fathmm-MKL_coding_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_pred</th>\n",
       "      <th>fathmm-MKL_coding_group</th>\n",
       "      <th>fathmm-XF_coding_score</th>\n",
       "      <th>fathmm-XF_coding_rankscore</th>\n",
       "      <th>fathmm-XF_coding_pred</th>\n",
       "      <th>Eigen-raw_coding</th>\n",
       "      <th>Eigen-raw_coding_rankscore</th>\n",
       "      <th>Eigen-pred_coding</th>\n",
       "      <th>Eigen-PC-raw_coding</th>\n",
       "      <th>Eigen-PC-raw_coding_rankscore</th>\n",
       "      <th>Eigen-PC-phred_coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>10_000093270_A_C</td>\n",
       "      <td>10_000047330_A_C</td>\n",
       "      <td>Q3ZCM7_C354</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>C354</td>\n",
       "      <td>354</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000564130;ENST00000568866;ENST00000568584</td>\n",
       "      <td>Cys/Trp</td>\n",
       "      <td>C</td>\n",
       "      <td>W</td>\n",
       "      <td>22.6</td>\n",
       "      <td>22.30</td>\n",
       "      <td>0.30</td>\n",
       "      <td>5.37</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.999882;0.999882;0.999882;0.999882;1</td>\n",
       "      <td>-1.98;-1.98;-1.98</td>\n",
       "      <td>0.85247</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>-7.21;-6.69;-6.84</td>\n",
       "      <td>0.94249</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>0.753;0.756;0.751</td>\n",
       "      <td>0.75466</td>\n",
       "      <td>0.3131</td>\n",
       "      <td>0.87764</td>\n",
       "      <td>D</td>\n",
       "      <td>0.8180</td>\n",
       "      <td>0.93873</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.020603</td>\n",
       "      <td>0.43227</td>\n",
       "      <td>T</td>\n",
       "      <td>0.437</td>\n",
       "      <td>0.74446</td>\n",
       "      <td>0.868</td>\n",
       "      <td>0.96212</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>C354W</td>\n",
       "      <td>Gain of MoRF binding (P = 0.0706); Loss of she...</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.76919400692</td>\n",
       "      <td>0.77304</td>\n",
       "      <td>T</td>\n",
       "      <td>0.851523</td>\n",
       "      <td>0.15717</td>\n",
       "      <td>0.19276</td>\n",
       "      <td>0.20597</td>\n",
       "      <td>N</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.422079</td>\n",
       "      <td>0.48823</td>\n",
       "      <td>N</td>\n",
       "      <td>0.0406181179914726</td>\n",
       "      <td>0.43713</td>\n",
       "      <td>2.658815</td>\n",
       "      <td>-0.363070660159128</td>\n",
       "      <td>0.26108</td>\n",
       "      <td>1.439964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>10_000093271_C_A</td>\n",
       "      <td>10_000047331_C_A</td>\n",
       "      <td>Q3ZCM7_C354</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>C354</td>\n",
       "      <td>354</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000564130;ENST00000568866;ENST00000568584</td>\n",
       "      <td>Cys/Phe</td>\n",
       "      <td>C</td>\n",
       "      <td>F</td>\n",
       "      <td>22.6</td>\n",
       "      <td>21.80</td>\n",
       "      <td>0.80</td>\n",
       "      <td>5.37</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.986821;0.986821;0.986821;0.986821;1</td>\n",
       "      <td>-1.95;-1.95;-1.95</td>\n",
       "      <td>0.85003</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>-7.2;-6.67;-6.83</td>\n",
       "      <td>0.94223</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>0.779;0.782;0.782</td>\n",
       "      <td>0.77883</td>\n",
       "      <td>0.4271</td>\n",
       "      <td>0.89500</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7835</td>\n",
       "      <td>0.92649</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.015962</td>\n",
       "      <td>0.36975</td>\n",
       "      <td>T</td>\n",
       "      <td>0.374</td>\n",
       "      <td>0.69594</td>\n",
       "      <td>0.808</td>\n",
       "      <td>0.92509</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>C354F</td>\n",
       "      <td>Gain of sheet (P = 0.0827); Loss of methylatio...</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.727896571159</td>\n",
       "      <td>0.71199</td>\n",
       "      <td>T</td>\n",
       "      <td>0.895918</td>\n",
       "      <td>0.18929</td>\n",
       "      <td>0.57914</td>\n",
       "      <td>0.30502</td>\n",
       "      <td>D</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.537556</td>\n",
       "      <td>0.55531</td>\n",
       "      <td>D</td>\n",
       "      <td>0.0988420731411684</td>\n",
       "      <td>0.46409</td>\n",
       "      <td>2.88164</td>\n",
       "      <td>-0.263815814260368</td>\n",
       "      <td>0.29300</td>\n",
       "      <td>1.640415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>10_000093271_C_G</td>\n",
       "      <td>10_000047331_C_G</td>\n",
       "      <td>Q3ZCM7_C354</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>C354</td>\n",
       "      <td>354</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000564130;ENST00000568866;ENST00000568584</td>\n",
       "      <td>Cys/Ser</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>22.4</td>\n",
       "      <td>19.67</td>\n",
       "      <td>2.73</td>\n",
       "      <td>5.37</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.930492;0.930492;0.930492;0.930492;1</td>\n",
       "      <td>-1.87;-1.87;-1.87</td>\n",
       "      <td>0.84415</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>-6.55;-6.06;-6.21</td>\n",
       "      <td>0.91827</td>\n",
       "      <td>D;D;D</td>\n",
       "      <td>0.74;0.748;0.739</td>\n",
       "      <td>0.74735</td>\n",
       "      <td>0.4621</td>\n",
       "      <td>0.90007</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7769</td>\n",
       "      <td>0.92419</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.022348</td>\n",
       "      <td>0.45230</td>\n",
       "      <td>T</td>\n",
       "      <td>0.377</td>\n",
       "      <td>0.69846</td>\n",
       "      <td>0.781</td>\n",
       "      <td>0.90537</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>C354S</td>\n",
       "      <td>Gain of disorder (P = 0.0216); Gain of loop (P...</td>\n",
       "      <td>.;.;.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.713165700436</td>\n",
       "      <td>0.69052</td>\n",
       "      <td>T</td>\n",
       "      <td>0.799045</td>\n",
       "      <td>0.12954</td>\n",
       "      <td>0.59480</td>\n",
       "      <td>0.30927</td>\n",
       "      <td>D</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.537556</td>\n",
       "      <td>0.55531</td>\n",
       "      <td>D</td>\n",
       "      <td>0.0966724300357427</td>\n",
       "      <td>0.46307</td>\n",
       "      <td>2.873107</td>\n",
       "      <td>-0.265973651554631</td>\n",
       "      <td>0.29227</td>\n",
       "      <td>1.635795</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos_id19          pos_id38       pos_ID matched_UKBID  \\\n",
       "0  10_000093270_A_C  10_000047330_A_C  Q3ZCM7_C354        Q3ZCM7   \n",
       "1  10_000093271_C_A  10_000047331_C_A  Q3ZCM7_C354        Q3ZCM7   \n",
       "2  10_000093271_C_G  10_000047331_C_G  Q3ZCM7_C354        Q3ZCM7   \n",
       "\n",
       "                                  pos_dict matched_target  matched_aapos  \\\n",
       "0  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}           C354            354   \n",
       "1  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}           C354            354   \n",
       "2  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}           C354            354   \n",
       "\n",
       "   matched_index                             Ensembl_transcriptid Amino_acids  \\\n",
       "0              2  ENST00000564130;ENST00000568866;ENST00000568584     Cys/Trp   \n",
       "1              2  ENST00000564130;ENST00000568866;ENST00000568584     Cys/Phe   \n",
       "2              2  ENST00000564130;ENST00000568866;ENST00000568584     Cys/Ser   \n",
       "\n",
       "  aaref aaalt  CADD_phred_hg38  CADD_phred_hg19  CADDdiff_38minus19  \\\n",
       "0     C     W             22.6            22.30                0.30   \n",
       "1     C     F             22.6            21.80                0.80   \n",
       "2     C     S             22.4            19.67                2.73   \n",
       "\n",
       "   Cys_reactivity Cys_react_threshold Cys_target_label  Lys_reactivity  \\\n",
       "0            5.37                 Low              NaN             NaN   \n",
       "1            5.37                 Low              NaN             NaN   \n",
       "2            5.37                 Low              NaN             NaN   \n",
       "\n",
       "  Lys_react_threshold Lys_target_label rs_dbSNP151 genename cds_strand  \\\n",
       "0                 NaN              NaN           .    TUBB8          -   \n",
       "1                 NaN              NaN           .    TUBB8          -   \n",
       "2                 NaN              NaN           .    TUBB8          -   \n",
       "\n",
       "  refcodon  codonpos  codon_degeneracy LRT_score LRT_converted_rankscore  \\\n",
       "0      TGT         3                 2  0.000003                 0.62929   \n",
       "1      TGT         2                 0  0.000003                 0.62929   \n",
       "2      TGT         2                 0  0.000003                 0.62929   \n",
       "\n",
       "  LRT_pred LRT_Omega                   MutationTaster_score  \\\n",
       "0        U  0.000000  0.999882;0.999882;0.999882;0.999882;1   \n",
       "1        U  0.000000  0.986821;0.986821;0.986821;0.986821;1   \n",
       "2        U  0.000000  0.930492;0.930492;0.930492;0.930492;1   \n",
       "\n",
       "        FATHMM_score FATHMM_converted_rankscore FATHMM_pred  \\\n",
       "0  -1.98;-1.98;-1.98                    0.85247       D;D;D   \n",
       "1  -1.95;-1.95;-1.95                    0.85003       D;D;D   \n",
       "2  -1.87;-1.87;-1.87                    0.84415       D;D;D   \n",
       "\n",
       "       PROVEAN_score PROVEAN_converted_rankscore PROVEAN_pred  \\\n",
       "0  -7.21;-6.69;-6.84                     0.94249        D;D;D   \n",
       "1   -7.2;-6.67;-6.83                     0.94223        D;D;D   \n",
       "2  -6.55;-6.06;-6.21                     0.91827        D;D;D   \n",
       "\n",
       "         VEST4_score VEST4_rankscore MetaSVM_score MetaSVM_rankscore  \\\n",
       "0  0.753;0.756;0.751         0.75466        0.3131           0.87764   \n",
       "1  0.779;0.782;0.782         0.77883        0.4271           0.89500   \n",
       "2   0.74;0.748;0.739         0.74735        0.4621           0.90007   \n",
       "\n",
       "  MetaSVM_pred MetaLR_score MetaLR_rankscore MetaLR_pred Reliability_index  \\\n",
       "0            D       0.8180          0.93873           D                 9   \n",
       "1            D       0.7835          0.92649           D                 9   \n",
       "2            D       0.7769          0.92419           D                 9   \n",
       "\n",
       "  M-CAP_score M-CAP_rankscore M-CAP_pred REVEL_score REVEL_rankscore  \\\n",
       "0    0.020603         0.43227          T       0.437         0.74446   \n",
       "1    0.015962         0.36975          T       0.374         0.69594   \n",
       "2    0.022348         0.45230          T       0.377         0.69846   \n",
       "\n",
       "  MutPred_score MutPred_rankscore MutPred_protID MutPred_AAchange  \\\n",
       "0         0.868           0.96212         Q3ZCM7            C354W   \n",
       "1         0.808           0.92509         Q3ZCM7            C354F   \n",
       "2         0.781           0.90537         Q3ZCM7            C354S   \n",
       "\n",
       "                                MutPred_Top5features MPC_score MPC_rankscore  \\\n",
       "0  Gain of MoRF binding (P = 0.0706); Loss of she...     .;.;.             .   \n",
       "1  Gain of sheet (P = 0.0827); Loss of methylatio...     .;.;.             .   \n",
       "2  Gain of disorder (P = 0.0216); Gain of loop (P...     .;.;.             .   \n",
       "\n",
       "  PrimateAI_score PrimateAI_rankscore PrimateAI_pred  DANN_score  \\\n",
       "0   0.76919400692             0.77304              T    0.851523   \n",
       "1  0.727896571159             0.71199              T    0.895918   \n",
       "2  0.713165700436             0.69052              T    0.799045   \n",
       "\n",
       "   DANN_rankscore  fathmm-MKL_coding_score  fathmm-MKL_coding_rankscore  \\\n",
       "0         0.15717                  0.19276                      0.20597   \n",
       "1         0.18929                  0.57914                      0.30502   \n",
       "2         0.12954                  0.59480                      0.30927   \n",
       "\n",
       "  fathmm-MKL_coding_pred fathmm-MKL_coding_group fathmm-XF_coding_score  \\\n",
       "0                      N                   AEFBI               0.422079   \n",
       "1                      D                   AEFBI               0.537556   \n",
       "2                      D                   AEFBI               0.537556   \n",
       "\n",
       "  fathmm-XF_coding_rankscore fathmm-XF_coding_pred    Eigen-raw_coding  \\\n",
       "0                    0.48823                     N  0.0406181179914726   \n",
       "1                    0.55531                     D  0.0988420731411684   \n",
       "2                    0.55531                     D  0.0966724300357427   \n",
       "\n",
       "  Eigen-raw_coding_rankscore Eigen-pred_coding Eigen-PC-raw_coding  \\\n",
       "0                    0.43713          2.658815  -0.363070660159128   \n",
       "1                    0.46409           2.88164  -0.263815814260368   \n",
       "2                    0.46307          2.873107  -0.265973651554631   \n",
       "\n",
       "  Eigen-PC-raw_coding_rankscore Eigen-PC-phred_coding  \n",
       "0                       0.26108              1.439964  \n",
       "1                       0.29300              1.640415  \n",
       "2                       0.29227              1.635795  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simpleD.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [3] filter CADDv1.4 files with dbNSFP CK positions for MISSENSE consequence only\n",
    "\n",
    "dbNSFP SCORE file column names (will map to these files using columns shared with CADD files:\n",
    "\n",
    "**columns detected file:**\n",
    "    - Amino_acids\n",
    "    - pos_ID 'Q3ZCM7_C354'\n",
    "    - pos_id19 or pos_id38_x\n",
    "\n",
    "**columns not detected file:**\n",
    "    - pos_id19 or pos_id38\n",
    "    - Amino_acids\n",
    "    - pos_ID_falseCKtarget\n",
    "\n",
    "\n",
    "```python\n",
    "# !/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "# Pmap M local ipynb code, Pmap_missense_annotations_QC.py\n",
    "# markdown M local QC of positions from dbNSFP overlapped with CADD37 or 38 annotations\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "def create_coordinate_id(df, chrr, pos, ref, alt, assembly):\n",
    "    if assembly == 37:\n",
    "        df.loc[:,'pos_id19'] = df[chrr].astype(str) + '_' + \\\n",
    "                df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "                '_' + df[alt].astype(str)\n",
    "    if assembly == 38:\n",
    "        df.loc[:,'pos_id38'] = df[chrr].astype(str) + '_' + \\\n",
    "                df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "                '_' + df[alt].astype(str)\n",
    "    return df\n",
    "\n",
    "\n",
    "def format_missense_triple(df, oaacol, naacol):\n",
    "    #  A|A turns to Ala/Ala\n",
    "    amino_dict = dict([('A', 'Ala'),('G', 'Gly'), ('I','Ile'), ('L','Leu'), ('P', 'Pro'), ('V','Val'), ('F','Phe'),('W', 'Trp'), ('Y', 'Tyr'), ('D','Asp'),('E','Glu'), ('R','Arg'),('H','His'), ('K','Lys'), ('S','Ser'), ('T', 'Thr'), ('C', 'Cys'), ('M', 'Met'), ('N', 'Asn'), ('Q','Gln')])\n",
    "    df[oaacol].replace(amino_dict, inplace=True)\n",
    "    df[naacol].replace(amino_dict, inplace=True)\n",
    "    ccopy = df[naacol].copy()\n",
    "    df['Amino_acids'] = df[oaacol].str.cat(ccopy, sep='/')\n",
    "    return df\n",
    "\n",
    "\n",
    "def filter_cadd_overlap(df, assembly):\n",
    "    # [1] filter for missense only\n",
    "    miss = df[df['Consequence'] == 'NON_SYNONYMOUS'].copy()\n",
    "    # [2] new pos_id(assembly) to files\n",
    "    if assembly == 37:\n",
    "        miss = create_coordinate_id(miss, 'chr', 'pos_hg19', 'Ref', 'Alt', assembly)\n",
    "    if assembly == 38:\n",
    "        miss = create_coordinate_id(miss, 'chr', 'pos_hg38', 'Ref', 'Alt', assembly)\n",
    "    # [3] new missense type column in 3 letter format with '/' sep {oAA, nAA}\n",
    "    miss = format_missense_triple(miss, 'oAA', 'nAA')\n",
    "\n",
    "    return miss\n",
    "\n",
    "    # [4] concat all files from 37: DECT or NOT ... 38: DECT or NOT\n",
    "\n",
    "\n",
    "def main():\n",
    "    os.chdir('/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/RESULT_pos_overlap_dbNSFPcoordinates')\n",
    "\n",
    "    chrls = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,'X','Y']\n",
    "\n",
    "    # GRCh37\n",
    "    for order in chrls:\n",
    "        chrID = 'chr{}'.format(order)\n",
    "        file1 = '{}_CADD_GRCh37_DETECTED_CK.csv'.format(chrID)\n",
    "        out1 = 'MISSENSE_{}_CADD_GRCh37_DETECTED_CK.csv'.format(chrID)\n",
    "        file2 = '{}_CADD_GRCh37_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        out2 = 'MISSENSE_{}_CADD_GRCh37_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        df1 = pd.read_csv(file1, low_memory=False)\n",
    "        df2 = pd.read_csv(file2, low_memory=False)\n",
    "        df1out = filter_cadd_overlap(df1, 37)\n",
    "        df2out = filter_cadd_overlap(df2, 37)\n",
    "        print(\"saving detected and not detected GRCh37 \", chrID)\n",
    "        print()\n",
    "        df1out.to_csv(out1, index=False)\n",
    "        df2out.to_csv(out2, index=False)\n",
    "\n",
    "    # GRCh38\n",
    "    for order in chrls:\n",
    "        chrID = 'chr{}'.format(order)\n",
    "        file1 = '{}_CADD_GRCh38_DETECTED_CK.csv'.format(chrID)\n",
    "        out1 = 'MISSENSE_{}_CADD_GRCh38_DETECTED_CK.csv'.format(chrID)\n",
    "        file2 = '{}_CADD_GRCh38_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        out2 = 'MISSENSE_{}_CADD_GRCh38_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        df1 = pd.read_csv(file1)\n",
    "        df2 = pd.read_csv(file2)\n",
    "        df1out = filter_cadd_overlap(df1, 38)\n",
    "        df2out = filter_cadd_overlap(df2, 38)\n",
    "        print(\"saving detected and not detected GRCh38 \", chrID)\n",
    "        print()\n",
    "        df1out.to_csv(out1, index=False)\n",
    "        df2out.to_csv(out2, index=False)\n",
    "\n",
    "main()\n",
    "\n",
    "```\n",
    "\n",
    "# RE running GRCh38 sex chrs after fixing cadd source files on hoffman\n",
    "\n",
    "```python\n",
    "# !/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "# Pmap M local ipynb code, Pmap_missense_annotations_QC.py\n",
    "# markdown M local QC of positions from dbNSFP overlapped with CADD37 or 38 annotations\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "\"\"\" \n",
    "dbNSFP SCORE file column names:\n",
    "\n",
    "columns detected file:\n",
    "    Amino_acids\n",
    "    pos_ID 'Q3ZCM7_C354'\n",
    "    pos_id19 or pos_id38_x\n",
    "\n",
    "columns not detected file:\n",
    "    pos_id19 or pos_id38\n",
    "    Amino_acids,\n",
    "    pos_ID_falseCKtarget\n",
    "\"\"\"\n",
    "\n",
    "def create_coordinate_id(df, chrr, pos, ref, alt, assembly):\n",
    "    if assembly == 37:\n",
    "        df.loc[:,'pos_id19'] = df[chrr].astype(str) + '_' + \\\n",
    "                df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "                '_' + df[alt].astype(str)\n",
    "    if assembly == 38:\n",
    "        df.loc[:,'pos_id38'] = df[chrr].astype(str) + '_' + \\\n",
    "                df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "                '_' + df[alt].astype(str)\n",
    "    return df\n",
    "\n",
    "\n",
    "def format_missense_triple(df, oaacol, naacol):\n",
    "    #  A|A turns to Ala/Ala\n",
    "    amino_dict = dict([('A', 'Ala'),('G', 'Gly'), ('I','Ile'), ('L','Leu'), ('P', 'Pro'), ('V','Val'), ('F','Phe'),('W', 'Trp'), ('Y', 'Tyr'), ('D','Asp'),('E','Glu'), ('R','Arg'),('H','His'), ('K','Lys'), ('S','Ser'), ('T', 'Thr'), ('C', 'Cys'), ('M', 'Met'), ('N', 'Asn'), ('Q','Gln')])\n",
    "    df[oaacol].replace(amino_dict, inplace=True)\n",
    "    df[naacol].replace(amino_dict, inplace=True)\n",
    "    ccopy = df[naacol].copy()\n",
    "    df['Amino_acids'] = df[oaacol].str.cat(ccopy, sep='/')\n",
    "    return df\n",
    "\n",
    "\n",
    "def filter_cadd_overlap(df, assembly):\n",
    "    # [1] filter for missense only\n",
    "    miss = df[df['Consequence'] == 'NON_SYNONYMOUS'].copy()\n",
    "    # [2] new pos_id(assembly) to files\n",
    "    if assembly == 37:\n",
    "        miss = create_coordinate_id(miss, 'chr', 'pos_hg19', 'Ref', 'Alt', assembly)\n",
    "    if assembly == 38:\n",
    "        miss = create_coordinate_id(miss, 'chr', 'pos_hg38', 'Ref', 'Alt', assembly)\n",
    "    # [3] new missense type column in 3 letter format with '/' sep {oAA, nAA}\n",
    "    miss = format_missense_triple(miss, 'oAA', 'nAA')\n",
    "\n",
    "    return miss\n",
    "\n",
    "    # [4] concat all files from 37: DECT or NOT ... 38: DECT or NOT\n",
    "\n",
    "\n",
    "def main():\n",
    "    os.chdir('/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/RESULT_pos_overlap_dbNSFPcoordinates')\n",
    "    chrls = ['X', 'Y']\n",
    "    # GRCh38\n",
    "    for order in chrls:\n",
    "        chrID = 'chr{}'.format(order)\n",
    "        file1 = '{}_CADD_GRCh38_DETECTED_CK.csv'.format(chrID)\n",
    "        out1 = 'MISSENSE_{}_CADD_GRCh38_DETECTED_CK.csv'.format(chrID)\n",
    "        file2 = '{}_CADD_GRCh38_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        out2 = 'MISSENSE_{}_CADD_GRCh38_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        df1 = pd.read_csv(file1)\n",
    "        df2 = pd.read_csv(file2)\n",
    "        df1out = filter_cadd_overlap(df1, 38)\n",
    "        df2out = filter_cadd_overlap(df2, 38)\n",
    "        print(\"saving detected and not detected GRCh38 \", chrID)\n",
    "        print()\n",
    "        df1out.to_csv(out1, index=False)\n",
    "        df2out.to_csv(out2, index=False)\n",
    "\n",
    "main()\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
