{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "---\n",
    "---\n",
    "\n",
    "# MAP CADDv1.4 annotations (for detected & non-detected CK pos) to dbNSFP score files\n",
    "\n",
    "1.02.20 mfpfox on local, other M part on hoffman- origianl version with MISSENSE CADD file filtering from 12/27/19\n",
    "\n",
    "### VEP VERSION NOTE: \n",
    "    - CADDv1.4 using VEPv92\n",
    "    - dbNSFP using VEPv94\n",
    "\n",
    "**hoffman PATH : /u/home/m/mfpalafo/project-arboleda/CADD/PULL_annotations**\n",
    "    * DIR downloaded locally for part 3 QC \"RESULT_pos_overlap_dbNSFPcoordinates/\"\n",
    "    * hoffman files have all consequences\n",
    "\n",
    "**hoffman M markdown has code for:** \n",
    "- [1] chr chunks\n",
    "- [2] search chromosome position against CADD files\n",
    "\n",
    "**this M markdwon has code for:** \n",
    "- [3] CADDv1.4 files format AA col and pos_id col with leading 0's\n",
    "- [4] simplify dbNSFP cols and parse list columm values based on 'matched_index' column\n",
    "- [5] merge cadd chr files on pos_id column from dbNSFP\n",
    "    - detected = 104,475 \n",
    "    - not detect = 1,222,911\n",
    "- [6] merge CADD chr files\n",
    "- [7] CADD+dbNSFP annotation files\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [3] adding {pos_id19 | pos_id38} {Amino_acids} col to CADD chr files filtered for hg19 | hg38 position\n",
    "\n",
    "PATH = '/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/RESULT_CADDv14_pos_overlap_dbNSFPcoordinates/ALL_CONSEQUENCES/'\n",
    "\n",
    "- initially used MISSENSE filtered file and recovered 20k less detected positions.\n",
    "- low yeild do to CADD using VEP v92 while dbNSFP changes and annotations from v94 VEP\n",
    "- redoing CADD file parsing using ALL consequences (not filtering on Consequence column which happens in later code of this markdown)\n",
    "\n",
    "```python\n",
    "\n",
    "# !/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "# Pmap M local ipynb code, Pmap_missense_annotations_QC.py\n",
    "# markdown M local QC of positions from dbNSFP overlapped with CADD37 or 38 annotations\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "def create_coordinate_id(df, chrr, pos, ref, alt, assembly):\n",
    "    if assembly == 37:\n",
    "        df.loc[:,'pos_id19'] = df[chrr].astype(str) + '_' + \\\n",
    "                df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "                '_' + df[alt].astype(str)\n",
    "    if assembly == 38:\n",
    "        df.loc[:,'pos_id38'] = df[chrr].astype(str) + '_' + \\\n",
    "                df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "                '_' + df[alt].astype(str)\n",
    "    return df\n",
    "\n",
    "\n",
    "def format_missense_triple(df, oaacol, naacol):\n",
    "    #  A|A turns to Ala/Ala\n",
    "    amino_dict = dict([('A', 'Ala'),('G', 'Gly'), ('I','Ile'), ('L','Leu'), ('P', 'Pro'), ('V','Val'), ('F','Phe'),('W', 'Trp'), ('Y', 'Tyr'), ('D','Asp'),('E','Glu'), ('R','Arg'),('H','His'), ('K','Lys'), ('S','Ser'), ('T', 'Thr'), ('C', 'Cys'), ('M', 'Met'), ('N', 'Asn'), ('Q','Gln')])\n",
    "    df[oaacol].replace(amino_dict, inplace=True)\n",
    "    df[naacol].replace(amino_dict, inplace=True)\n",
    "    ccopy = df[naacol].copy()\n",
    "    df['Amino_acids'] = df[oaacol].str.cat(ccopy, sep='/')\n",
    "    return df\n",
    "\n",
    "\n",
    "def filter_cadd_overlap(df, assembly):\n",
    "    # deleted [1] section that did missense filtering step\n",
    "    # [2] new pos_id(assembly) to files\n",
    "    if assembly == 37:\n",
    "        df = create_coordinate_id(df, 'chr', 'pos_hg19', 'Ref', 'Alt', assembly)\n",
    "    if assembly == 38:\n",
    "        df = create_coordinate_id(df, 'chr', 'pos_hg38', 'Ref', 'Alt', assembly)\n",
    "    # [3] new missense type column in 3 letter format with '/' sep {oAA, nAA}\n",
    "    df = format_missense_triple(df, 'oAA', 'nAA')\n",
    "    return df\n",
    "\n",
    "\n",
    "def main():\n",
    "    os.chdir('/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/RESULT_CADDv14_pos_overlap_dbNSFPcoordinates/ALL_CONSEQUENCES/')\n",
    "\n",
    "    chrls = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,'X','Y']\n",
    "\n",
    "    # GRCh37\n",
    "    for order in chrls:\n",
    "        chrID = 'chr{}'.format(order)\n",
    "        file1 = '{}_CADD_GRCh37_DETECTED_CK.csv'.format(chrID)\n",
    "        out1 = 'ALLCON_{}_CADD_GRCh37_DETECTED_CK.csv'.format(chrID)\n",
    "        file2 = '{}_CADD_GRCh37_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        out2 = 'ALLCON_{}_CADD_GRCh37_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        df1 = pd.read_csv(file1, low_memory=False, converters={'pos_hg19': '{:0>9}'.format})\n",
    "        df2 = pd.read_csv(file2, low_memory=False, converters={'pos_hg19': '{:0>9}'.format})\n",
    "        df1out = filter_cadd_overlap(df1, 37)\n",
    "        df2out = filter_cadd_overlap(df2, 37)\n",
    "        print(\"saving detected and not detected GRCh37 \", chrID)\n",
    "        print()\n",
    "        df1out.to_csv(out1, index=False)\n",
    "        df2out.to_csv(out2, index=False)\n",
    "\n",
    "    # GRCh38\n",
    "    for order in chrls:\n",
    "        chrID = 'chr{}'.format(order)\n",
    "        file1 = '{}_CADD_GRCh38_DETECTED_CK.csv'.format(chrID)\n",
    "        out1 = 'ALLCON_{}_CADD_GRCh38_DETECTED_CK.csv'.format(chrID)\n",
    "        file2 = '{}_CADD_GRCh38_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        out2 = 'ALLCON_{}_CADD_GRCh38_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        df1 = pd.read_csv(file1, low_memory=False, converters={'pos_hg38': '{:0>9}'.format})\n",
    "        df2 = pd.read_csv(file2, low_memory=False, converters={'pos_hg38': '{:0>9}'.format})\n",
    "        df1out = filter_cadd_overlap(df1, 38)\n",
    "        df2out = filter_cadd_overlap(df2, 38)\n",
    "        print(\"saving detected and not detected GRCh38 \", chrID)\n",
    "        print()\n",
    "        df1out.to_csv(out1, index=False)\n",
    "        df2out.to_csv(out2, index=False)\n",
    "\n",
    "main()\n",
    "```\n",
    "\n",
    "# dbNSFP pos_id extracted as single column and used next to filter CADD ALL ANNOTATION files on inner merge \n",
    "\n",
    "path = /Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/BUG_FIX\n",
    "```bash\n",
    "#!/bin/bash\n",
    "# keeps header and appends all rows together\n",
    "# for Pmap dbNSFP CADD mapped detected and non detected chr merge\n",
    "\n",
    "cut -d',' -f1 < SCORE_dbNSFP_selectcols_NOT_detected_CK_1222911.csv > col_notdet_posid19.csv\n",
    "\n",
    "cut -d',' -f2 < SCORE_dbNSFP_selectcols_NOT_detected_CK_1222911.csv > col_notdet_posid38.csv\n",
    "\n",
    "cut -d',' -f1 < SCORE_dbNSFP_selectcols_detected_CK_104475.csv > col_det_posid19.csv\n",
    "\n",
    "cut -d',' -f2 < SCORE_dbNSFP_selectcols_detected_CK_104475.csv > col_det_posid38.csv\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "---\n",
    "---\n",
    "\n",
    "\n",
    "\n",
    "# [4] simplifying dbNSFP score files by COLUMNS OF INTEREST\n",
    "\n",
    "rules: \n",
    "1. merge on pos_idcoordinates\n",
    "2. match on Amino_acids 'A/A'\n",
    "3. protein positions match- not required since most annotations refer to coordinates for annotation\n",
    "\n",
    "\n",
    "```python\n",
    "os.chdir(\"/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/BUG_FIX/\")\n",
    "\n",
    "##  detected dbNSFP score file: \n",
    "det_scores = 'SCORE_annotation_3840_CYS_LYS_detected.csv'\n",
    "detscore = pd.read_csv(det_scores, low_memory=False)\n",
    "print(detscore.shape)\n",
    "# (104475, 400)\n",
    "\n",
    "### simplify df with select columns: \n",
    "det_cols = ['pos_id19', 'pos_id38_x', 'aaref_x', 'aaalt_x', 'matched_aapos_x',\n",
    "       'CADD_phred_hg38', 'matched_UKBID_x', 'pos_dict', 'matched_target',\n",
    "       'CADD_phred_hg19', 'CADDdiff_38minus19', 'Amino_acids', 'pos_ID',\n",
    "       'Cys_reactivity', 'Cys_react_threshold', 'Cys_target_label',\n",
    "       'Lys_reactivity', 'Lys_react_threshold', 'Lys_target_label', 'rs_dbSNP151', 'Ensembl_transcriptid','cds_strand', 'refcodon', 'codonpos', 'codon_degeneracy',\n",
    "       'SIFT_score', 'SIFT_converted_rankscore', 'SIFT_pred', 'SIFT4G_score',\n",
    "       'SIFT4G_converted_rankscore', 'SIFT4G_pred', 'Polyphen2_HDIV_score',\n",
    "       'Polyphen2_HDIV_rankscore', 'Polyphen2_HDIV_pred',\n",
    "       'Polyphen2_HVAR_score', 'Polyphen2_HVAR_rankscore',\n",
    "       'Polyphen2_HVAR_pred','LRT_score', 'LRT_converted_rankscore','LRT_pred', 'LRT_Omega', 'MutationTaster_score','MutationTaster_converted_rankscore', 'MutationTaster_pred',\n",
    "       'MutationTaster_model', 'MutationTaster_AAE', 'MutationAssessor_score',\n",
    "       'MutationAssessor_rankscore', 'MutationAssessor_pred',\n",
    "       'FATHMM_score','FATHMM_converted_rankscore', 'FATHMM_pred', 'PROVEAN_score','PROVEAN_converted_rankscore', 'PROVEAN_pred', 'VEST4_score','VEST4_rankscore', 'MetaSVM_score', 'MetaSVM_rankscore', 'MetaSVM_pred','MetaLR_score', 'MetaLR_rankscore', 'MetaLR_pred', 'Reliability_index','M-CAP_score', 'M-CAP_rankscore', 'M-CAP_pred', 'REVEL_score','REVEL_rankscore', 'MutPred_score', 'MutPred_rankscore',\n",
    "       'MutPred_protID', 'MutPred_AAchange', 'MutPred_Top5features',\n",
    "        'MPC_score','MPC_rankscore','PrimateAI_score', 'PrimateAI_rankscore', 'PrimateAI_pred','DANN_score', 'DANN_rankscore', 'fathmm-MKL_coding_score','fathmm-MKL_coding_rankscore', 'fathmm-MKL_coding_pred','fathmm-MKL_coding_group', 'fathmm-XF_coding_score',\n",
    "       'fathmm-XF_coding_rankscore', 'fathmm-XF_coding_pred',\n",
    "       'Eigen-raw_coding', 'Eigen-raw_coding_rankscore', 'Eigen-pred_coding',\n",
    "       'Eigen-PC-raw_coding', 'Eigen-PC-raw_coding_rankscore',\n",
    "       'Eigen-PC-phred_coding', 'matched_index']\n",
    "# filter\n",
    "simpleD = detscore[det_cols].copy()\n",
    "\n",
    "rename = ['pos_id19', 'pos_id38', 'aaref', 'aaalt', 'matched_aapos',\n",
    "       'CADD_phred_hg38', 'matched_UKBID', 'pos_dict', 'matched_target',\n",
    "       'CADD_phred_hg19', 'CADDdiff_38minus19', 'Amino_acids', 'pos_ID',\n",
    "       'Cys_reactivity', 'Cys_react_threshold', 'Cys_target_label',\n",
    "       'Lys_reactivity', 'Lys_react_threshold', 'Lys_target_label', 'rs_dbSNP151', 'Ensembl_transcriptid','cds_strand', 'refcodon', 'codonpos', 'codon_degeneracy','SIFT_score', 'SIFT_converted_rankscore', 'SIFT_pred', 'SIFT4G_score',\n",
    "       'SIFT4G_converted_rankscore', 'SIFT4G_pred', 'Polyphen2_HDIV_score',\n",
    "       'Polyphen2_HDIV_rankscore', 'Polyphen2_HDIV_pred',\n",
    "       'Polyphen2_HVAR_score', 'Polyphen2_HVAR_rankscore',\n",
    "       'Polyphen2_HVAR_pred','LRT_score', 'LRT_converted_rankscore','LRT_pred', 'LRT_Omega', 'MutationTaster_score','MutationTaster_converted_rankscore', 'MutationTaster_pred','MutationTaster_model', 'MutationTaster_AAE', 'MutationAssessor_score','MutationAssessor_rankscore', 'MutationAssessor_pred',\n",
    "       'FATHMM_score','FATHMM_converted_rankscore', 'FATHMM_pred', 'PROVEAN_score','PROVEAN_converted_rankscore', 'PROVEAN_pred', 'VEST4_score','VEST4_rankscore', 'MetaSVM_score', 'MetaSVM_rankscore', 'MetaSVM_pred','MetaLR_score', 'MetaLR_rankscore', 'MetaLR_pred', 'Reliability_index','M-CAP_score', 'M-CAP_rankscore', 'M-CAP_pred', 'REVEL_score','REVEL_rankscore', 'MutPred_score', 'MutPred_rankscore',\n",
    "       'MutPred_protID', 'MutPred_AAchange', 'MutPred_Top5features',\n",
    "        'MPC_score','MPC_rankscore','PrimateAI_score', 'PrimateAI_rankscore', 'PrimateAI_pred','DANN_score', 'DANN_rankscore', 'fathmm-MKL_coding_score','fathmm-MKL_coding_rankscore', 'fathmm-MKL_coding_pred','fathmm-MKL_coding_group', 'fathmm-XF_coding_score','fathmm-XF_coding_rankscore', 'fathmm-XF_coding_pred','Eigen-raw_coding', 'Eigen-raw_coding_rankscore', 'Eigen-pred_coding',\n",
    "       'Eigen-PC-raw_coding', 'Eigen-PC-raw_coding_rankscore',\n",
    "       'Eigen-PC-phred_coding', 'matched_index']\n",
    "\n",
    "simpleD.columns = rename\n",
    "\n",
    "# change order\n",
    "neworder = ['pos_id19', 'pos_id38', 'pos_ID', 'matched_aapos', 'matched_target', 'matched_UKBID', 'matched_index', 'pos_dict',\n",
    "       'aaref', 'aaalt','Amino_acids',\n",
    "       'CADD_phred_hg38','CADD_phred_hg19', 'CADDdiff_38minus19', \n",
    "       'Cys_reactivity', 'Cys_react_threshold', 'Cys_target_label',\n",
    "       'Lys_reactivity', 'Lys_react_threshold', 'Lys_target_label', 'rs_dbSNP151', 'Ensembl_transcriptid','cds_strand', 'refcodon', 'codonpos', 'codon_degeneracy','SIFT_score', 'SIFT_converted_rankscore', 'SIFT_pred', 'SIFT4G_score',\n",
    "       'SIFT4G_converted_rankscore', 'SIFT4G_pred', 'Polyphen2_HDIV_score',\n",
    "       'Polyphen2_HDIV_rankscore', 'Polyphen2_HDIV_pred',\n",
    "       'Polyphen2_HVAR_score', 'Polyphen2_HVAR_rankscore',\n",
    "       'Polyphen2_HVAR_pred','LRT_score', 'LRT_converted_rankscore','LRT_pred', 'LRT_Omega', 'MutationTaster_score','MutationTaster_converted_rankscore', 'MutationTaster_pred','MutationTaster_model', 'MutationTaster_AAE', 'MutationAssessor_score','MutationAssessor_rankscore', 'MutationAssessor_pred',\n",
    "       'FATHMM_score','FATHMM_converted_rankscore', 'FATHMM_pred', 'PROVEAN_score','PROVEAN_converted_rankscore', 'PROVEAN_pred', 'VEST4_score','VEST4_rankscore', 'MetaSVM_score', 'MetaSVM_rankscore', 'MetaSVM_pred','MetaLR_score', 'MetaLR_rankscore', 'MetaLR_pred', 'Reliability_index','M-CAP_score', 'M-CAP_rankscore', 'M-CAP_pred', 'REVEL_score','REVEL_rankscore', 'MutPred_score', 'MutPred_rankscore',\n",
    "       'MutPred_protID', 'MutPred_AAchange', 'MutPred_Top5features',\n",
    "        'MPC_score','MPC_rankscore','PrimateAI_score', 'PrimateAI_rankscore', 'PrimateAI_pred','DANN_score', 'DANN_rankscore', 'fathmm-MKL_coding_score','fathmm-MKL_coding_rankscore', 'fathmm-MKL_coding_pred','fathmm-MKL_coding_group', 'fathmm-XF_coding_score','fathmm-XF_coding_rankscore', 'fathmm-XF_coding_pred','Eigen-raw_coding', 'Eigen-raw_coding_rankscore', 'Eigen-pred_coding',\n",
    "       'Eigen-PC-raw_coding', 'Eigen-PC-raw_coding_rankscore',\n",
    "       'Eigen-PC-phred_coding']\n",
    "\n",
    "simpleD = simpleD[neworder]\n",
    "simpleD.to_csv(\"SCORE_dbNSFP_selectcols_detected_CK_104475.csv\", index=False)\n",
    "\n",
    "\n",
    "\n",
    "## not detected dbNSFP score file: \n",
    "not_scores = 'SCORE_annotation_3840_CYS_LYS_NOT_allcol_detected.csv'\n",
    "notscore = pd.read_csv(not_scores, low_memory=False)\n",
    "print(notscore.shape)\n",
    "# (1222911, 385)\n",
    "\n",
    "### simplify df with select columns: \n",
    "not_cols = ['pos_id19', 'CADD_phred_hg19', 'CADDdiff_38minus19', 'Amino_acids',\n",
    "       'pos_ID_falseCKtarget', 'aaref',\n",
    "       'aaalt', 'rs_dbSNP151', 'genename', 'Ensembl_transcriptid','cds_strand', 'refcodon', 'codonpos','codon_degeneracy', 'SIFT_score', 'SIFT_converted_rankscore', 'SIFT_pred', 'SIFT4G_score',\n",
    "       'SIFT4G_converted_rankscore', 'SIFT4G_pred', 'Polyphen2_HDIV_score',\n",
    "       'Polyphen2_HDIV_rankscore', 'Polyphen2_HDIV_pred',\n",
    "       'Polyphen2_HVAR_score', 'Polyphen2_HVAR_rankscore',\n",
    "       'Polyphen2_HVAR_pred','LRT_score',\n",
    "       'LRT_converted_rankscore', 'LRT_pred', 'LRT_Omega',\n",
    "       'MutationTaster_score', 'MutationTaster_converted_rankscore',\n",
    "       'MutationTaster_pred', 'MutationTaster_model', 'MutationTaster_AAE',\n",
    "       'MutationAssessor_score', 'MutationAssessor_rankscore',\n",
    "       'MutationAssessor_pred', 'FATHMM_score', 'FATHMM_converted_rankscore',\n",
    "       'FATHMM_pred', 'PROVEAN_score', 'PROVEAN_converted_rankscore',\n",
    "       'PROVEAN_pred', 'VEST4_score', 'VEST4_rankscore', 'MetaSVM_score',\n",
    "       'MetaSVM_rankscore', 'MetaSVM_pred', 'MetaLR_score', 'MetaLR_rankscore',\n",
    "       'MetaLR_pred', 'Reliability_index', 'M-CAP_score', 'M-CAP_rankscore',\n",
    "       'M-CAP_pred', 'REVEL_score', 'REVEL_rankscore', 'MutPred_score',\n",
    "       'MutPred_rankscore', 'MutPred_protID', 'MutPred_AAchange',\n",
    "       'MutPred_Top5features', 'MPC_score','MPC_rankscore', 'PrimateAI_score', 'PrimateAI_rankscore',\n",
    "       'PrimateAI_pred','CADD_phred','DANN_score', 'DANN_rankscore', 'fathmm-MKL_coding_score',\n",
    "       'fathmm-MKL_coding_rankscore', 'fathmm-MKL_coding_pred',\n",
    "       'fathmm-MKL_coding_group', 'fathmm-XF_coding_score',\n",
    "       'fathmm-XF_coding_rankscore', 'fathmm-XF_coding_pred',\n",
    "       'Eigen-raw_coding', 'Eigen-raw_coding_rankscore', 'Eigen-pred_coding',\n",
    "       'Eigen-PC-raw_coding', 'Eigen-PC-raw_coding_rankscore',\n",
    "       'Eigen-PC-phred_coding', 'matched_UKBID',\n",
    "       'matched_aapos', 'matched_index', 'pos_id38']\n",
    "# filter\n",
    "simpleN = notscore[not_cols].copy()\n",
    "# rename\n",
    "simpleN.columns = ['pos_id19', 'CADD_phred_hg19', 'CADDdiff_38minus19', 'Amino_acids',\n",
    "       'pos_ID', 'aaref',\n",
    "       'aaalt', 'rs_dbSNP151', 'genename', 'Ensembl_transcriptid','cds_strand', 'refcodon', 'codonpos','codon_degeneracy', 'SIFT_score', 'SIFT_converted_rankscore', 'SIFT_pred', 'SIFT4G_score',\n",
    "       'SIFT4G_converted_rankscore', 'SIFT4G_pred', 'Polyphen2_HDIV_score',\n",
    "       'Polyphen2_HDIV_rankscore', 'Polyphen2_HDIV_pred',\n",
    "       'Polyphen2_HVAR_score', 'Polyphen2_HVAR_rankscore',\n",
    "       'Polyphen2_HVAR_pred','LRT_score',\n",
    "       'LRT_converted_rankscore', 'LRT_pred', 'LRT_Omega',\n",
    "       'MutationTaster_score', 'MutationTaster_converted_rankscore',\n",
    "       'MutationTaster_pred', 'MutationTaster_model', 'MutationTaster_AAE',\n",
    "       'MutationAssessor_score', 'MutationAssessor_rankscore',\n",
    "       'MutationAssessor_pred', 'FATHMM_score', 'FATHMM_converted_rankscore',\n",
    "       'FATHMM_pred', 'PROVEAN_score', 'PROVEAN_converted_rankscore',\n",
    "       'PROVEAN_pred', 'VEST4_score', 'VEST4_rankscore', 'MetaSVM_score',\n",
    "       'MetaSVM_rankscore', 'MetaSVM_pred', 'MetaLR_score', 'MetaLR_rankscore',\n",
    "       'MetaLR_pred', 'Reliability_index', 'M-CAP_score', 'M-CAP_rankscore',\n",
    "       'M-CAP_pred', 'REVEL_score', 'REVEL_rankscore', 'MutPred_score',\n",
    "       'MutPred_rankscore', 'MutPred_protID', 'MutPred_AAchange',\n",
    "       'MutPred_Top5features', 'MPC_score','MPC_rankscore', 'PrimateAI_score', 'PrimateAI_rankscore',\n",
    "       'PrimateAI_pred','CADD_phred_hg38','DANN_score', 'DANN_rankscore', 'fathmm-MKL_coding_score',\n",
    "       'fathmm-MKL_coding_rankscore', 'fathmm-MKL_coding_pred',\n",
    "       'fathmm-MKL_coding_group', 'fathmm-XF_coding_score',\n",
    "       'fathmm-XF_coding_rankscore', 'fathmm-XF_coding_pred',\n",
    "       'Eigen-raw_coding', 'Eigen-raw_coding_rankscore', 'Eigen-pred_coding',\n",
    "       'Eigen-PC-raw_coding', 'Eigen-PC-raw_coding_rankscore',\n",
    "       'Eigen-PC-phred_coding', 'matched_UKBID',\n",
    "       'matched_aapos', 'matched_index', 'pos_id38']\n",
    "       \n",
    "# change order\n",
    "neworder = ['pos_id19', 'pos_id38', 'pos_ID', 'matched_UKBID', 'matched_aapos', 'matched_index','Ensembl_transcriptid', 'Amino_acids','aaref','aaalt', \n",
    "'CADD_phred_hg38', 'CADD_phred_hg19', 'CADDdiff_38minus19', 'rs_dbSNP151', 'genename', 'cds_strand', 'refcodon', 'codonpos','codon_degeneracy', 'SIFT_score', 'SIFT_converted_rankscore', 'SIFT_pred', 'SIFT4G_score',\n",
    "       'SIFT4G_converted_rankscore', 'SIFT4G_pred', 'Polyphen2_HDIV_score',\n",
    "       'Polyphen2_HDIV_rankscore', 'Polyphen2_HDIV_pred',\n",
    "       'Polyphen2_HVAR_score', 'Polyphen2_HVAR_rankscore',\n",
    "       'Polyphen2_HVAR_pred','LRT_score',\n",
    "       'LRT_converted_rankscore', 'LRT_pred', 'LRT_Omega',\n",
    "       'MutationTaster_score', 'MutationTaster_converted_rankscore',\n",
    "       'MutationTaster_pred', 'MutationTaster_model', 'MutationTaster_AAE',\n",
    "       'MutationAssessor_score', 'MutationAssessor_rankscore',\n",
    "       'MutationAssessor_pred', 'FATHMM_score', 'FATHMM_converted_rankscore',\n",
    "       'FATHMM_pred', 'PROVEAN_score', 'PROVEAN_converted_rankscore',\n",
    "       'PROVEAN_pred', 'VEST4_score', 'VEST4_rankscore', 'MetaSVM_score',\n",
    "       'MetaSVM_rankscore', 'MetaSVM_pred', 'MetaLR_score', 'MetaLR_rankscore',\n",
    "       'MetaLR_pred', 'Reliability_index', 'M-CAP_score', 'M-CAP_rankscore',\n",
    "       'M-CAP_pred', 'REVEL_score', 'REVEL_rankscore', 'MutPred_score',\n",
    "       'MutPred_rankscore', 'MutPred_protID', 'MutPred_AAchange',\n",
    "       'MutPred_Top5features', 'MPC_score','MPC_rankscore', 'PrimateAI_score', 'PrimateAI_rankscore',\n",
    "       'PrimateAI_pred','DANN_score', 'DANN_rankscore', 'fathmm-MKL_coding_score',\n",
    "       'fathmm-MKL_coding_rankscore', 'fathmm-MKL_coding_pred',\n",
    "       'fathmm-MKL_coding_group', 'fathmm-XF_coding_score',\n",
    "       'fathmm-XF_coding_rankscore', 'fathmm-XF_coding_pred',\n",
    "       'Eigen-raw_coding', 'Eigen-raw_coding_rankscore', 'Eigen-pred_coding',\n",
    "       'Eigen-PC-raw_coding', 'Eigen-PC-raw_coding_rankscore',\n",
    "       'Eigen-PC-phred_coding']\n",
    "simpleN = simpleN[neworder]\n",
    "simpleN.to_csv(\"SCORE_dbNSFP_selectcols_NOT_detected_CK_1222911.csv\", index=False)\n",
    "```\n",
    "\n",
    "```python \n",
    "header = simpleD.columns\n",
    "header2 = simpleN.columns\n",
    "d = {header[i] : i for i in range(0, len(header))}\n",
    "# print(d) # 'matched_index': 7, 'Ensembl_transcriptid': 8,\n",
    "d2 = {header2[i] : i for i in range(0, len(header2))}\n",
    "# print(d2) # 'matched_index': 5, 'Ensembl_transcriptid': 6, \n",
    "```\n",
    "\n",
    "\n",
    "## columns parsed: \n",
    "```\n",
    "{DETECTED\n",
    " 'matched_index': 6,\n",
    " 'Ensembl_transcriptid': 21,\n",
    " 'SIFT_score': 26,\n",
    " 'SIFT_pred': 28,\n",
    " 'SIFT4G_score': 29,\n",
    " 'SIFT4G_pred': 31,\n",
    " 'Polyphen2_HDIV_score': 32,\n",
    " 'Polyphen2_HDIV_pred': 34,\n",
    " 'Polyphen2_HVAR_score': 35,\n",
    " 'Polyphen2_HVAR_pred': 37,\n",
    " 'MutationTaster_score': 42,\n",
    " 'MutationTaster_pred': 44,\n",
    " 'MutationTaster_model': 45,\n",
    " 'MutationTaster_AAE': 46,\n",
    " 'MutationAssessor_score': 47,\n",
    " 'MutationAssessor_pred': 49,\n",
    " 'FATHMM_score': 50,\n",
    " 'FATHMM_pred': 52,\n",
    " 'PROVEAN_score': 53,\n",
    " 'PROVEAN_pred': 55,\n",
    " 'VEST4_score': 56,\n",
    " 'MutPred_Top5features': 74,\n",
    " 'MPC_score': 75,\n",
    " }\n",
    "\n",
    "NOT DETECTED\n",
    "{\n",
    " 'matched_index': 5,\n",
    " 'Ensembl_transcriptid': 6,\n",
    " 'SIFT_score': 19,\n",
    " 'SIFT_pred': 21,\n",
    " 'SIFT4G_score': 22,\n",
    " 'SIFT4G_pred': 24,\n",
    " 'Polyphen2_HDIV_score': 25,\n",
    " 'Polyphen2_HDIV_pred': 27,\n",
    " 'Polyphen2_HVAR_score': 28,\n",
    " 'Polyphen2_HVAR_pred': 30,\n",
    " 'MutationTaster_score': 35,\n",
    " 'MutationTaster_pred': 37,\n",
    " 'MutationTaster_model': 38,\n",
    " 'MutationTaster_AAE': 39,\n",
    " 'MutationAssessor_score': 40,\n",
    " 'MutationAssessor_pred': 42,\n",
    " 'FATHMM_score': 43,\n",
    " 'FATHMM_pred': 45,\n",
    " 'PROVEAN_score': 46,\n",
    " 'PROVEAN_pred': 48,\n",
    " 'VEST4_score': 49,\n",
    " 'MutPred_Top5features': 67,\n",
    " 'MPC_score': 68,\n",
    "}\n",
    "```\n",
    "\n",
    "# (done)  parsing list values in columns of dbNSFP files, \n",
    "```python\n",
    "\n",
    "def feature_id_col(filename, outfile, subset):\n",
    "    if subset == 'dect':\n",
    "        # adding feature_ID column from parsing enst list col from dbNSFP with matched_index col\n",
    "        # modeled after function from Pmap_parseID_correction.py\n",
    "        with open(filename, newline='') as file:\n",
    "            # read in file, save header\n",
    "            csvReader = csv.reader(file)\n",
    "            header = next(csvReader)\n",
    "            # create and write to outfile\n",
    "            os.system(\"touch %s\" % (outfile))\n",
    "            with open(outfile, 'w') as out:\n",
    "                csvWriter = csv.writer(out)\n",
    "                csvWriter.writerow(header)\n",
    "            # loop over rows\n",
    "            for row in csvReader:\n",
    "                matchI = int(row[6])\n",
    "                changeindex = [21,26,28,29,31,32,34,35,37,42,44,45,46,47,49,50,52,53,55,56,74,75]\n",
    "                for coli in changeindex:\n",
    "                    rowvalue = row[coli]\n",
    "                    splitval = rowvalue.split(\";\")\n",
    "                    if matchI < len(splitval):\n",
    "                        newval = splitval[matchI]\n",
    "                        row[coli] = newval\n",
    "                with open(outfile, 'a') as out:\n",
    "                        csvWriter = csv.writer(out)\n",
    "                        csvWriter.writerow(row)\n",
    "        print(\"done with : \", outfile)\n",
    "    if subset == 'notdect':\n",
    "        # adding feature_ID column from parsing enst list col from dbNSFP with matched_index col\n",
    "        # modeled after function from Pmap_parseID_correction.py\n",
    "        with open(filename, newline='') as file:\n",
    "            # read in file, save header\n",
    "            csvReader = csv.reader(file)\n",
    "            header = next(csvReader)\n",
    "            # create and write to outfile\n",
    "            os.system(\"touch %s\" % (outfile))\n",
    "            with open(outfile, 'w') as out:\n",
    "                csvWriter = csv.writer(out)\n",
    "                csvWriter.writerow(header)\n",
    "            # loop over rows\n",
    "            for row in csvReader:\n",
    "                matchI = int(row[5])\n",
    "                changeindex = [6,19,21,22,24,25,27,28,30,35,37,38,39,40,42,43,45,46,48,49,67,68]\n",
    "                for coli in changeindex:\n",
    "                    rowvalue = row[coli]\n",
    "                    splitval = rowvalue.split(\";\")\n",
    "                    if matchI < len(splitval):\n",
    "                        newval = splitval[matchI]\n",
    "                        row[coli] = newval\n",
    "                with open(outfile, 'a') as out:\n",
    "                        csvWriter = csv.writer(out)\n",
    "                        csvWriter.writerow(row)\n",
    "        print(\"done with : \", outfile)\n",
    "        \n",
    "        \n",
    "# calling function\n",
    "feature_id_col('SCORE_dbNSFP_selectcols_detected_CK_104475.csv', 'SCORE2_dbNSFP_parsedcols_detected_CK_104475.csv', 'dect')\n",
    "feature_id_col('SCORE_dbNSFP_selectcols_NOT_detected_CK_1222911.csv', 'SCORE2_dbNSFP_parsedcols_NOT_detected_CK_1222911.csv', 'notdect')\n",
    "```\n",
    "\n",
    "## result of feature_id_col() =\n",
    "- all dbnsfp file rows have 1 ENST ID matching index of canonical ukb ID\n",
    "\n",
    "description files for new featureID column added created with below code:\n",
    "```python\n",
    "saveColumnValues(dbnsfp, 'FeatureID', 'dbnsfp_featureID_detected_description.csv')\n",
    "saveColumnValues(notdbnsfp, 'FeatureID', 'dbnsfp_featureID_notdetected_description.csv')\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container {width:90% !important;}</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import argparse\n",
    "import pandas as pd\n",
    "import csv\n",
    "import numpy as np\n",
    "\n",
    "from IPython.display import display, HTML\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.options.display.max_seq_items = 2000\n",
    "display(HTML(\"<style>.container {width:90% !important;}</style>\"))\n",
    "sys.path.append(\"/Users/mariapalafox/Desktop/Toolbox\")\n",
    "from all_funx import *\n",
    "from maplib import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [5] MERGE CADD chr files with new pos_id column with dbNSFP column of pos_id19 or pos_id38 for DETECTED and NONDETECTED positions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5a. MERGE DETECTED 37 CADD ALL CONSEQUENCE with pos_ids from dbNSFP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(22764, 65)\n",
      "merge shape : \n",
      "(16832, 42)\n",
      "unique pos_id: \n",
      "10556\n",
      "done with  1\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(16746, 65)\n",
      "merge shape : \n",
      "(12556, 42)\n",
      "unique pos_id: \n",
      "7693\n",
      "done with  2\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(11871, 65)\n",
      "merge shape : \n",
      "(8449, 42)\n",
      "unique pos_id: \n",
      "5110\n",
      "done with  3\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(8217, 65)\n",
      "merge shape : \n",
      "(6252, 42)\n",
      "unique pos_id: \n",
      "3878\n",
      "done with  4\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(9306, 65)\n",
      "merge shape : \n",
      "(6685, 42)\n",
      "unique pos_id: \n",
      "4249\n",
      "done with  5\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(12342, 65)\n",
      "merge shape : \n",
      "(8799, 42)\n",
      "unique pos_id: \n",
      "4501\n",
      "done with  6\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(12852, 65)\n",
      "merge shape : \n",
      "(9128, 42)\n",
      "unique pos_id: \n",
      "5460\n",
      "done with  7\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(6879, 65)\n",
      "merge shape : \n",
      "(5121, 42)\n",
      "unique pos_id: \n",
      "3269\n",
      "done with  8\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(8424, 65)\n",
      "merge shape : \n",
      "(6033, 42)\n",
      "unique pos_id: \n",
      "3885\n",
      "done with  9\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(9531, 65)\n",
      "merge shape : \n",
      "(7301, 42)\n",
      "unique pos_id: \n",
      "5026\n",
      "done with  10\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(20895, 65)\n",
      "merge shape : \n",
      "(14847, 42)\n",
      "unique pos_id: \n",
      "7959\n",
      "done with  11\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(13485, 65)\n",
      "merge shape : \n",
      "(9975, 42)\n",
      "unique pos_id: \n",
      "5292\n",
      "done with  12\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(2907, 65)\n",
      "merge shape : \n",
      "(2009, 42)\n",
      "unique pos_id: \n",
      "1519\n",
      "done with  13\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(8982, 65)\n",
      "merge shape : \n",
      "(6482, 42)\n",
      "unique pos_id: \n",
      "3619\n",
      "done with  14\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(6720, 65)\n",
      "merge shape : \n",
      "(4659, 42)\n",
      "unique pos_id: \n",
      "2842\n",
      "done with  15\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(9633, 65)\n",
      "merge shape : \n",
      "(6641, 42)\n",
      "unique pos_id: \n",
      "3381\n",
      "done with  16\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(17583, 65)\n",
      "merge shape : \n",
      "(12183, 42)\n",
      "unique pos_id: \n",
      "6461\n",
      "done with  17\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(3222, 65)\n",
      "merge shape : \n",
      "(2156, 42)\n",
      "unique pos_id: \n",
      "1267\n",
      "done with  18\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(19488, 65)\n",
      "merge shape : \n",
      "(14112, 42)\n",
      "unique pos_id: \n",
      "6986\n",
      "done with  19\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(7491, 65)\n",
      "merge shape : \n",
      "(5309, 42)\n",
      "unique pos_id: \n",
      "3073\n",
      "done with  20\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(1926, 65)\n",
      "merge shape : \n",
      "(1379, 42)\n",
      "unique pos_id: \n",
      "903\n",
      "done with  21\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(6303, 65)\n",
      "merge shape : \n",
      "(4510, 42)\n",
      "unique pos_id: \n",
      "2989\n",
      "done with  22\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(8355, 65)\n",
      "merge shape : \n",
      "(6406, 42)\n",
      "unique pos_id: \n",
      "4529\n",
      "done with  X\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(36, 65)\n",
      "merge shape : \n",
      "(28, 42)\n",
      "unique pos_id: \n",
      "28\n",
      "done with  Y\n"
     ]
    }
   ],
   "source": [
    "# !/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "# Pmap_dbNSFP_CADD_merge_DECT19.py\n",
    "'''\n",
    "filter files: \n",
    "\n",
    "col_det_posid19.csv --- ALLCON_{}_CADD_GRCh37_DETECTED_CK.csv\n",
    "\n",
    "'''\n",
    "\n",
    "def main():\n",
    "    os.chdir('/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/RESULT_CADDv14_pos_overlap_dbNSFPcoordinates/ALL_CONSEQUENCES/37det/')\n",
    "\n",
    "    chrls = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,'X','Y']\n",
    "\n",
    "    for order in chrls:\n",
    "        chrID = 'chr{}'.format(order)\n",
    "        file1 = 'ALLCON_{}_CADD_GRCh37_DETECTED_CK.csv'.format(chrID)\n",
    "        out1 = 'FILTERONposid_{}_CADD_GRCh37_DETECTED_CK.csv'.format(chrID)\n",
    "        df1 = pd.read_csv(file1, low_memory=False)\n",
    "\n",
    "        # filter file\n",
    "        fildf = pd.read_csv('col_det_posid19.csv')\n",
    "\n",
    "        print(\"filter file shape: \")\n",
    "        print(fildf.shape)\n",
    "        print(\"chr cadd file shape: \")\n",
    "        print(df1.shape)\n",
    "\n",
    "        dropme = ['Type', 'Length', 'AnnoType','ConsScore', 'ConsDetail','oAA', 'nAA', 'GeneID','FeatureID', 'GeneName', 'CCDS', 'Intron', 'Exon', 'cDNApos','relcDNApos', 'CDSpos', 'relCDSpos', 'relProtPos', 'Domain','Dst2Splice', 'Dst2SplType', 'minDistTSS', 'minDistTSE']\n",
    "\n",
    "        df1.drop(dropme, axis=1, inplace=True)\n",
    "        df1.drop_duplicates(keep='first', inplace = True)\n",
    "\n",
    "        mer = pd.merge(df1, fildf, how='inner', on=['pos_id19'])\n",
    "        print(\"merge shape : \")\n",
    "        print(mer.shape)\n",
    "        mer.to_csv(out1, index=False)\n",
    "        print(\"unique pos_id: \")\n",
    "        print(len(set(mer.pos_id19)))\n",
    "        print(\"done with \", order)\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5b. MERGE DETECTED 38 CADD ALL CONSEQUENCE with pos_ids from dbNSFP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(23163, 65)\n",
      "merge shape : \n",
      "(17093, 42)\n",
      "done with  1\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(16806, 65)\n",
      "merge shape : \n",
      "(12469, 42)\n",
      "done with  2\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(11985, 65)\n",
      "merge shape : \n",
      "(8510, 42)\n",
      "done with  3\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(8145, 65)\n",
      "merge shape : \n",
      "(6188, 42)\n",
      "done with  4\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(9522, 65)\n",
      "merge shape : \n",
      "(6864, 42)\n",
      "done with  5\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(12711, 65)\n",
      "merge shape : \n",
      "(8974, 42)\n",
      "done with  6\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(13095, 65)\n",
      "merge shape : \n",
      "(9289, 42)\n",
      "done with  7\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(7620, 65)\n",
      "merge shape : \n",
      "(5638, 42)\n",
      "done with  8\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(8772, 65)\n",
      "merge shape : \n",
      "(6333, 42)\n",
      "done with  9\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(9585, 65)\n",
      "merge shape : \n",
      "(7371, 42)\n",
      "done with  10\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(21417, 65)\n",
      "merge shape : \n",
      "(15228, 42)\n",
      "done with  11\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(14049, 65)\n",
      "merge shape : \n",
      "(10290, 42)\n",
      "done with  12\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(2997, 65)\n",
      "merge shape : \n",
      "(2051, 42)\n",
      "done with  13\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(9228, 65)\n",
      "merge shape : \n",
      "(6560, 42)\n",
      "done with  14\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(6981, 65)\n",
      "merge shape : \n",
      "(4758, 42)\n",
      "done with  15\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(10320, 65)\n",
      "merge shape : \n",
      "(7042, 42)\n",
      "done with  16\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(18396, 65)\n",
      "merge shape : \n",
      "(12773, 42)\n",
      "done with  17\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(3339, 65)\n",
      "merge shape : \n",
      "(2163, 42)\n",
      "done with  18\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(20142, 65)\n",
      "merge shape : \n",
      "(14541, 42)\n",
      "done with  19\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(7470, 65)\n",
      "merge shape : \n",
      "(5320, 42)\n",
      "done with  20\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(1962, 65)\n",
      "merge shape : \n",
      "(1407, 42)\n",
      "done with  21\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(6612, 65)\n",
      "merge shape : \n",
      "(4738, 42)\n",
      "done with  22\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(8478, 65)\n",
      "merge shape : \n",
      "(6483, 42)\n",
      "done with  X\n",
      "filter file shape: \n",
      "(104475, 1)\n",
      "chr cadd file shape: \n",
      "(36, 65)\n",
      "merge shape : \n",
      "(28, 42)\n",
      "done with  Y\n"
     ]
    }
   ],
   "source": [
    "# !/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "# Pmap_dbNSFP_CADD_merge_DECT19.py\n",
    "'''\n",
    "filter files: \n",
    "\n",
    "col_det_posid38.csv --- ALLCON_{}_CADD_GRCh38_DETECTED_CK.csv\n",
    "\n",
    "'''\n",
    "\n",
    "def main():\n",
    "    # CHANGE PATH HERE\n",
    "    os.chdir('/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/RESULT_CADDv14_pos_overlap_dbNSFPcoordinates/ALL_CONSEQUENCES/38det/')\n",
    "\n",
    "    chrls = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,'X','Y']\n",
    "\n",
    "    for order in chrls:\n",
    "        chrID = 'chr{}'.format(order)\n",
    "        \n",
    "        # CHANGE FILE NAMES HERE\n",
    "        file1 = 'ALLCON_{}_CADD_GRCh38_DETECTED_CK.csv'.format(chrID)\n",
    "        out1 = 'FILTERONposid_{}_CADD_GRCh38_DETECTED_CK.csv'.format(chrID)\n",
    "        df1 = pd.read_csv(file1, low_memory=False)\n",
    "\n",
    "        # CHANGE FILTER FILE NAME HERE\n",
    "        fildf = pd.read_csv('col_det_posid38.csv')\n",
    "\n",
    "        print(\"filter file shape: \")\n",
    "        print(fildf.shape)\n",
    "        print(\"chr cadd file shape: \")\n",
    "        print(df1.shape)\n",
    "\n",
    "        dropme = ['Type', 'Length', 'AnnoType','ConsScore', 'ConsDetail','oAA', 'nAA', 'GeneID','FeatureID', 'GeneName', 'CCDS', 'Intron', 'Exon', 'cDNApos','relcDNApos', 'CDSpos', 'relCDSpos', 'relProtPos', 'Domain','Dst2Splice', 'Dst2SplType', 'minDistTSS', 'minDistTSE']\n",
    "\n",
    "        df1.drop(dropme, axis=1, inplace=True)\n",
    "        df1.drop_duplicates(keep='first', inplace = True)\n",
    "\n",
    "        mer = pd.merge(df1, fildf, how='inner', on=['pos_id38'])\n",
    "        print(\"merge shape : \")\n",
    "        print(mer.shape)\n",
    "        mer.to_csv(out1, index=False)\n",
    "        print(\"unique pos_id: \")\n",
    "        print(len(set(mer.pos_id38)))\n",
    "        print(\"done with \", order)\n",
    "        print()\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5c. MERGE NOT DETECTED 37 CADD ALL CONSEQUENCE with pos_ids from dbNSFP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(219981, 65)\n",
      "merge shape : \n",
      "(165555, 42)\n",
      "done with  1\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(189372, 65)\n",
      "merge shape : \n",
      "(144152, 42)\n",
      "done with  2\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(146550, 65)\n",
      "merge shape : \n",
      "(107666, 42)\n",
      "done with  3\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(78240, 65)\n",
      "merge shape : \n",
      "(60230, 42)\n",
      "done with  4\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(125283, 65)\n",
      "merge shape : \n",
      "(93395, 42)\n",
      "done with  5\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(132042, 65)\n",
      "merge shape : \n",
      "(98700, 42)\n",
      "done with  6\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(105234, 65)\n",
      "merge shape : \n",
      "(77000, 42)\n",
      "done with  7\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(75387, 65)\n",
      "merge shape : \n",
      "(55634, 42)\n",
      "done with  8\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(88968, 65)\n",
      "merge shape : \n",
      "(66091, 42)\n",
      "done with  9\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(115812, 65)\n",
      "merge shape : \n",
      "(88582, 42)\n",
      "done with  10\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(168588, 65)\n",
      "merge shape : \n",
      "(124925, 42)\n",
      "done with  11\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(122262, 65)\n",
      "merge shape : \n",
      "(92233, 42)\n",
      "done with  12\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(32232, 65)\n",
      "merge shape : \n",
      "(24719, 42)\n",
      "done with  13\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(100668, 65)\n",
      "merge shape : \n",
      "(74012, 42)\n",
      "done with  14\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(92526, 65)\n",
      "merge shape : \n",
      "(69328, 42)\n",
      "done with  15\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(105336, 65)\n",
      "merge shape : \n",
      "(78417, 42)\n",
      "done with  16\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(173523, 65)\n",
      "merge shape : \n",
      "(128297, 42)\n",
      "done with  17\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(34899, 65)\n",
      "merge shape : \n",
      "(26556, 42)\n",
      "done with  18\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(134772, 65)\n",
      "merge shape : \n",
      "(99198, 42)\n",
      "done with  19\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(64713, 65)\n",
      "merge shape : \n",
      "(46182, 42)\n",
      "done with  20\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(15234, 65)\n",
      "merge shape : \n",
      "(11303, 42)\n",
      "done with  21\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(54189, 65)\n",
      "merge shape : \n",
      "(39499, 42)\n",
      "done with  22\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(73185, 65)\n",
      "merge shape : \n",
      "(56161, 42)\n",
      "done with  X\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(144, 65)\n",
      "merge shape : \n",
      "(112, 42)\n",
      "done with  Y\n"
     ]
    }
   ],
   "source": [
    "# !/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "# Pmap_dbNSFP_CADD_merge_DECT19.py\n",
    "'''\n",
    "filter files: \n",
    "\n",
    "col_notdet_posid19.csv --- ALLCON_{}_CADD_GRCh37_NOT_DETECTED_CK.csv\n",
    "\n",
    "'''\n",
    "\n",
    "def main():\n",
    "    os.chdir('/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/RESULT_CADDv14_pos_overlap_dbNSFPcoordinates/ALL_CONSEQUENCES/37not/')\n",
    "\n",
    "    chrls = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,'X','Y']\n",
    "\n",
    "    for order in chrls:\n",
    "        chrID = 'chr{}'.format(order)\n",
    "        file1 = 'ALLCON_{}_CADD_GRCh37_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        out1 = 'FILTERONposid_{}_CADD_GRCh37_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        df1 = pd.read_csv(file1, low_memory=False)\n",
    "\n",
    "        # filter file\n",
    "        fildf = pd.read_csv('col_notdet_posid19.csv')\n",
    "\n",
    "        print(\"filter file shape: \")\n",
    "        print(fildf.shape)\n",
    "        print(\"chr cadd file shape: \")\n",
    "        print(df1.shape)\n",
    "\n",
    "        dropme = ['Type', 'Length', 'AnnoType','ConsScore', 'ConsDetail','oAA', 'nAA', 'GeneID','FeatureID', 'GeneName', 'CCDS', 'Intron', 'Exon', 'cDNApos','relcDNApos', 'CDSpos', 'relCDSpos', 'relProtPos', 'Domain','Dst2Splice', 'Dst2SplType', 'minDistTSS', 'minDistTSE']\n",
    "\n",
    "        df1.drop(dropme, axis=1, inplace=True)\n",
    "        df1.drop_duplicates(keep='first', inplace = True)\n",
    "\n",
    "        mer = pd.merge(df1, fildf, how='inner', on=['pos_id19'])\n",
    "        print(\"merge shape : \")\n",
    "        print(mer.shape)\n",
    "        mer.to_csv(out1, index=False)\n",
    "        print(\"done with \", order)\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5d. MERGE NOT DETECTED 38 CADD ALL CONSEQUENCE with pos_ids from dbNSFP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(220821, 65)\n",
      "merge shape : \n",
      "(166751, 42)\n",
      "done with  1\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(188106, 65)\n",
      "merge shape : \n",
      "(143637, 42)\n",
      "done with  2\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(147309, 65)\n",
      "merge shape : \n",
      "(107811, 42)\n",
      "done with  3\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(78663, 65)\n",
      "merge shape : \n",
      "(60484, 42)\n",
      "done with  4\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(128820, 65)\n",
      "merge shape : \n",
      "(96473, 42)\n",
      "done with  5\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(135543, 65)\n",
      "merge shape : \n",
      "(101513, 42)\n",
      "done with  6\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(106776, 65)\n",
      "merge shape : \n",
      "(78054, 42)\n",
      "done with  7\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(78834, 65)\n",
      "merge shape : \n",
      "(57927, 42)\n",
      "done with  8\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(90018, 65)\n",
      "merge shape : \n",
      "(67467, 42)\n",
      "done with  9\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(117798, 65)\n",
      "merge shape : \n",
      "(90356, 42)\n",
      "done with  10\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(171489, 65)\n",
      "merge shape : \n",
      "(126582, 42)\n",
      "done with  11\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(126039, 65)\n",
      "merge shape : \n",
      "(94189, 42)\n",
      "done with  12\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(33063, 65)\n",
      "merge shape : \n",
      "(25346, 42)\n",
      "done with  13\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(102327, 65)\n",
      "merge shape : \n",
      "(75245, 42)\n",
      "done with  14\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(97233, 65)\n",
      "merge shape : \n",
      "(72034, 42)\n",
      "done with  15\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(108228, 65)\n",
      "merge shape : \n",
      "(80038, 42)\n",
      "done with  16\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(178893, 65)\n",
      "merge shape : \n",
      "(131209, 42)\n",
      "done with  17\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(35346, 65)\n",
      "merge shape : \n",
      "(26784, 42)\n",
      "done with  18\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(137367, 65)\n",
      "merge shape : \n",
      "(100783, 42)\n",
      "done with  19\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(65487, 65)\n",
      "merge shape : \n",
      "(46589, 42)\n",
      "done with  20\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(15990, 65)\n",
      "merge shape : \n",
      "(11757, 42)\n",
      "done with  21\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(57522, 65)\n",
      "merge shape : \n",
      "(41808, 42)\n",
      "done with  22\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(75861, 65)\n",
      "merge shape : \n",
      "(57803, 42)\n",
      "done with  X\n",
      "filter file shape: \n",
      "(1222911, 1)\n",
      "chr cadd file shape: \n",
      "(126, 65)\n",
      "merge shape : \n",
      "(98, 42)\n",
      "done with  Y\n"
     ]
    }
   ],
   "source": [
    "# !/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "# Pmap_dbNSFP_CADD_merge_DECT19.py\n",
    "'''\n",
    "filter files: \n",
    "\n",
    "col_notdet_posid38.csv --- ALLCON_{}_CADD_GRCh38_NOT_DETECTED_CK.csv\n",
    "\n",
    "'''\n",
    "\n",
    "def main():\n",
    "    # CHANGE PATH HERE\n",
    "    os.chdir('/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/RESULT_CADDv14_pos_overlap_dbNSFPcoordinates/ALL_CONSEQUENCES/38not/')\n",
    "\n",
    "    chrls = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,'X','Y']\n",
    "\n",
    "    for order in chrls:\n",
    "        chrID = 'chr{}'.format(order)\n",
    "        \n",
    "        # CHANGE FILE NAMES HERE\n",
    "        file1 = 'ALLCON_{}_CADD_GRCh38_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        out1 = 'FILTERONposid_{}_CADD_GRCh38_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        df1 = pd.read_csv(file1, low_memory=False)\n",
    "\n",
    "        # CHANGE FILTER FILE NAME HERE\n",
    "        fildf = pd.read_csv('col_notdet_posid38.csv')\n",
    "\n",
    "        print(\"filter file shape: \")\n",
    "        print(fildf.shape)\n",
    "        print(\"chr cadd file shape: \")\n",
    "        print(df1.shape)\n",
    "\n",
    "        dropme = ['Type', 'Length', 'AnnoType','ConsScore', 'ConsDetail','oAA', 'nAA', 'GeneID','FeatureID', 'GeneName', 'CCDS', 'Intron', 'Exon', 'cDNApos','relcDNApos', 'CDSpos', 'relCDSpos', 'relProtPos', 'Domain','Dst2Splice', 'Dst2SplType', 'minDistTSS', 'minDistTSE']\n",
    "\n",
    "        df1.drop(dropme, axis=1, inplace=True)\n",
    "        df1.drop_duplicates(keep='first', inplace = True)\n",
    "\n",
    "        mer = pd.merge(df1, fildf, how='inner', on=['pos_id38'])\n",
    "        print(\"merge shape : \")\n",
    "        print(mer.shape)\n",
    "        mer.to_csv(out1, index=False)\n",
    "        print(\"done with \", order)\n",
    "main()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [6] merge CADD chr files\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "# keeps header and appends all rows together\n",
    "# for Pmap dbNSFP CADD mapped detected and non detected chr merge\n",
    "\n",
    "awk -F ',' 'FNR==1 && NR!=1 { while (/^pos_hg19/) getline; } 1 {print}' FILTERONposid_chr*_CADD_GRCh37_DETECTED_CK.csv > SCORE_FILTERallcon_CADD_37_DETECTED.csv\n",
    "\n",
    "awk -F ',' 'FNR==1 && NR!=1 { while (/^pos_hg19/) getline; } 1 {print}' FILTERONposid_chr*_CADD_GRCh37_NOT_DETECTED_CK.csv >SCORE_FILTERallcon_CADD_37_NOT_DETECTED.csv\n",
    "\n",
    "awk -F ',' 'FNR==1 && NR!=1 { while (/^pos_hg38/) getline; } 1 {print}' FILTERONposid_chr*_CADD_GRCh38_DETECTED_CK.csv > SCORE_FILTERallcon_CADD_38_DETECTED.csv\n",
    "\n",
    "awk -F ',' 'FNR==1 && NR!=1 { while (/^pos_hg38/) getline; } 1 {print}' FILTERONposid_chr*_CADD_GRCh38_NOT_DETECTED_CK.csv >SCORE_FILTERallcon_CADD_38_NOT_DETECTED.csv\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [6a] prep for dbNSFP merge, pos_id19 or pos_id38 in CADD files should all be unique\n",
    "- make nonredundant rows by dropping columns that are not needed and then dropping duplicate rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/RESULT_CADDv14_pos_overlap_dbNSFPcoordinates/ALL_CONSEQUENCES/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "not37 = pd.read_csv(\"SCORE_FILTERallcon_CADD_37_NOT_DETECTED.csv\", low_memory=False)\n",
    "not38 = pd.read_csv(\"SCORE_FILTERallcon_CADD_38_NOT_DETECTED.csv\", low_memory=False)\n",
    "det37 = pd.read_csv(\"SCORE_FILTERallcon_CADD_37_DETECTED.csv\", low_memory=False)\n",
    "det38 = pd.read_csv(\"SCORE_FILTERallcon_CADD_38_DETECTED.csv\", low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "177852\n",
      "104475\n",
      "\n",
      "1827947\n",
      "1222911\n",
      "\n",
      "182111\n",
      "104475\n",
      "\n",
      "1860738\n",
      "1222911\n",
      "\n",
      "Unique number of uniprot and AA IDs, equivalent to everlabeled counts for detected\n",
      "14925\n",
      "\n",
      "174702\n",
      "\n",
      "14925\n",
      "\n",
      "174702\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# POS ID unique count check\n",
    "dfls = [det37, not37]\n",
    "for i in dfls:\n",
    "    print(len(list(i.pos_id19)))\n",
    "    print(len(set(i.pos_id19)))\n",
    "    print()\n",
    "    \n",
    "dfls = [det38, not38]\n",
    "for i in dfls:\n",
    "    print(len(list(i.pos_id38)))\n",
    "    print(len(set(i.pos_id38)))\n",
    "    print()\n",
    "    \n",
    "print(\"Unique number of uniprot and AA IDs, equivalent to everlabeled counts for detected\")\n",
    "dfls = [det37, not37, det38, not38]\n",
    "for i in dfls:\n",
    "    print(len(set(i.pos_ID)))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**MATCHES ACCOUNTING #'s for dbNSFP pos_id...therefore ALL dbNSFP pos are in CADD files, now filter for only neccessary columns, drop dup rows, and save**\n",
    "\n",
    "| Unique column counts | DETECTED pos_ids | NOT DETECTED pos_ids |\n",
    "|---------|------------|-----------|\n",
    "| pos_id19 or pos_id38 | 104,475 | 1,222,911 | \n",
    "| pos_ID ukbAA | 14,925 | 174,702 | \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['pos_hg19', 'chr', 'pos_ID', 'Ref', 'Alt', 'Consequence', 'GC', 'CpG',\n",
       "       'motifECount', 'motifEName', 'motifEHIPos', 'motifEScoreChng',\n",
       "       'protPos', 'SIFTcat', 'SIFTval', 'PolyPhenCat', 'PolyPhenVal',\n",
       "       'priPhCons', 'mamPhCons', 'verPhCons', 'priPhyloP', 'mamPhyloP',\n",
       "       'verPhyloP', 'GerpRS', 'GerpRSpval', 'GerpN', 'GerpS', 'Grantham',\n",
       "       'Dist2Mutation', 'Freq100bp', 'Rare100bp', 'Sngl100bp', 'Freq1000bp',\n",
       "       'Rare1000bp', 'Sngl1000bp', 'Freq10000bp', 'Rare10000bp', 'Sngl10000bp',\n",
       "       'RawScore', 'PHRED', 'pos_id19', 'Amino_acids'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "det37.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dropping CADD columns that may casuse redundancy (multimapping of pos_id's). Columns of interest are only related to genomic coordinates so differences in missense annotation do not matter\n",
    "\n",
    "**goal is CADD files with non redundant pos_id column to merge with dbNSFP annotations, 1:1 merge**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(177852, 25)\n",
      "(104475, 25)\n",
      "\n",
      "(1827947, 25)\n",
      "(1222911, 25)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dropme = ['pos_hg19','chr', 'pos_ID', 'Ref', 'Alt', 'Consequence', 'motifECount', 'motifEName', 'motifEHIPos', 'motifEScoreChng',\n",
    "       'protPos', 'SIFTcat', 'SIFTval', 'PolyPhenCat', 'PolyPhenVal','Grantham', 'Amino_acids']\n",
    "\n",
    "dfls = [det37, not37]\n",
    "for i in dfls:\n",
    "    i.drop(dropme, axis=1, inplace=True)\n",
    "    print(i.shape)\n",
    "    i.drop_duplicates(keep='first', inplace = True)\n",
    "    print(i.shape)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(182111, 25)\n",
      "(104475, 25)\n",
      "\n",
      "(1860738, 25)\n",
      "(1222911, 25)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dropme = ['pos_hg38','chr', 'pos_ID', 'Ref', 'Alt', 'Consequence', 'motifECount', 'motifEName', 'motifEHIPos', 'motifEScoreChng',\n",
    "       'protPos', 'SIFTcat', 'SIFTval', 'PolyPhenCat', 'PolyPhenVal','Grantham', 'Amino_acids']\n",
    "\n",
    "dfls = [det38, not38]\n",
    "for i in dfls:\n",
    "    i.drop(dropme, axis=1, inplace=True)\n",
    "    print(i.shape)\n",
    "    i.drop_duplicates(keep='first', inplace = True)\n",
    "    print(i.shape)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prep col names for merging with dbNSFP file\n",
    "det37.columns = ['GC_hg19', 'CpG_hg19', 'priPhCons_hg19', 'mamPhCons_hg19', 'verPhCons_hg19',\n",
    "       'priPhyloP_hg19', 'mamPhyloP_hg19', 'verPhyloP_hg19', 'GerpRS_hg19', 'GerpRSpval_hg19', 'GerpN_hg19',\n",
    "       'GerpS_hg19', 'Dist2Mutation_hg19', 'Freq100bp_hg19', 'Rare100bp_hg19', 'Sngl100bp_hg19',\n",
    "       'Freq1000bp_hg19', 'Rare1000bp_hg19', 'Sngl1000bp_hg19', 'Freq10000bp_hg19', 'Rare10000bp_hg19',\n",
    "       'Sngl10000bp_hg19', 'RawScore_hg19', 'PHRED_hg19', 'pos_id19']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "det38.columns = ['GC_hg38', 'CpG_hg38', 'priPhCons_hg38', 'mamPhCons_hg38', 'verPhCons_hg38', 'priPhyloP_hg38','mamPhyloP_hg38', 'verPhyloP_hg38', 'GerpRS_hg38', 'GerpRSpval_hg38', 'GerpN_hg38', 'GerpS_hg38','Dist2Mutation_hg38', 'Freq100bp_hg38', 'Rare100bp_hg38', 'Sngl100bp_hg38', 'Freq1000bp_hg38','Rare1000bp_hg38', 'Sngl1000bp_hg38', 'Freq10000bp_hg38', 'Rare10000bp_hg38', 'Sngl10000bp_hg38','RawScore_hg38', 'PHRED_hg38', 'pos_id38']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "not37.columns = ['GC_hg19', 'CpG_hg19', 'priPhCons_hg19', 'mamPhCons_hg19', 'verPhCons_hg19', 'priPhyloP_hg19','mamPhyloP_hg19', 'verPhyloP_hg19', 'GerpRS_hg19', 'GerpRSpval_hg19', 'GerpN_hg19', 'GerpS_hg19','Dist2Mutation_hg19', 'Freq100bp_hg19', 'Rare100bp_hg19', 'Sngl100bp_hg19', 'Freq1000bp_hg19','Rare1000bp_hg19', 'Sngl1000bp_hg19', 'Freq10000bp_hg19', 'Rare10000bp_hg19', 'Sngl10000bp_hg19','RawScore_hg19', 'PHRED_hg19', 'pos_id19']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "not38.columns = ['GC_hg38', 'CpG_hg38', 'priPhCons_hg38', 'mamPhCons_hg38', 'verPhCons_hg38', 'priPhyloP_hg38', 'mamPhyloP_hg38', 'verPhyloP_hg38', 'GerpRS_hg38', 'GerpRSpval_hg38', 'GerpN_hg38', 'GerpS_hg38','Dist2Mutation_hg38', 'Freq100bp_hg38', 'Rare100bp_hg38', 'Sngl100bp_hg38', 'Freq1000bp_hg38', 'Rare1000bp_hg38', 'Sngl1000bp_hg38', 'Freq10000bp_hg38', 'Rare10000bp_hg38', 'Sngl10000bp_hg38','RawScore_hg38', 'PHRED_hg38', 'pos_id38']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GC_hg19</th>\n",
       "      <th>CpG_hg19</th>\n",
       "      <th>priPhCons_hg19</th>\n",
       "      <th>mamPhCons_hg19</th>\n",
       "      <th>verPhCons_hg19</th>\n",
       "      <th>priPhyloP_hg19</th>\n",
       "      <th>mamPhyloP_hg19</th>\n",
       "      <th>verPhyloP_hg19</th>\n",
       "      <th>GerpRS_hg19</th>\n",
       "      <th>GerpRSpval_hg19</th>\n",
       "      <th>GerpN_hg19</th>\n",
       "      <th>GerpS_hg19</th>\n",
       "      <th>Dist2Mutation_hg19</th>\n",
       "      <th>Freq100bp_hg19</th>\n",
       "      <th>Rare100bp_hg19</th>\n",
       "      <th>Sngl100bp_hg19</th>\n",
       "      <th>Freq1000bp_hg19</th>\n",
       "      <th>Rare1000bp_hg19</th>\n",
       "      <th>Sngl1000bp_hg19</th>\n",
       "      <th>Freq10000bp_hg19</th>\n",
       "      <th>Rare10000bp_hg19</th>\n",
       "      <th>Sngl10000bp_hg19</th>\n",
       "      <th>RawScore_hg19</th>\n",
       "      <th>PHRED_hg19</th>\n",
       "      <th>pos_id19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0.470199</td>\n",
       "      <td>0.053333</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.084</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.094</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.877</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>2.381671</td>\n",
       "      <td>22.30</td>\n",
       "      <td>10_000093270_A_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0.463576</td>\n",
       "      <td>0.053333</td>\n",
       "      <td>0.083</td>\n",
       "      <td>0.082</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.109</td>\n",
       "      <td>3.127</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>2.267094</td>\n",
       "      <td>21.80</td>\n",
       "      <td>10_000093271_C_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0.463576</td>\n",
       "      <td>0.053333</td>\n",
       "      <td>0.083</td>\n",
       "      <td>0.082</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.109</td>\n",
       "      <td>3.127</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>2.045093</td>\n",
       "      <td>19.67</td>\n",
       "      <td>10_000093271_C_G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    GC_hg19  CpG_hg19  priPhCons_hg19  mamPhCons_hg19  verPhCons_hg19  \\\n",
       "0  0.470199  0.053333           0.084           0.084             1.0   \n",
       "2  0.463576  0.053333           0.083           0.082             1.0   \n",
       "4  0.463576  0.053333           0.083           0.082             1.0   \n",
       "\n",
       "   priPhyloP_hg19  mamPhyloP_hg19  verPhyloP_hg19  GerpRS_hg19  \\\n",
       "0           0.094           0.095           0.877          NaN   \n",
       "2           0.110           0.109           3.127          NaN   \n",
       "4           0.110           0.109           3.127          NaN   \n",
       "\n",
       "   GerpRSpval_hg19  GerpN_hg19  GerpS_hg19  Dist2Mutation_hg19  \\\n",
       "0              NaN         0.0         0.0             38936.0   \n",
       "2              NaN         0.0         0.0             38936.0   \n",
       "4              NaN         0.0         0.0             38936.0   \n",
       "\n",
       "   Freq100bp_hg19  Rare100bp_hg19  Sngl100bp_hg19  Freq1000bp_hg19  \\\n",
       "0             NaN             NaN             NaN              NaN   \n",
       "2             NaN             NaN             NaN              NaN   \n",
       "4             NaN             NaN             NaN              NaN   \n",
       "\n",
       "   Rare1000bp_hg19  Sngl1000bp_hg19  Freq10000bp_hg19  Rare10000bp_hg19  \\\n",
       "0              NaN              NaN               5.0              24.0   \n",
       "2              NaN              NaN               5.0              24.0   \n",
       "4              NaN              NaN               5.0              24.0   \n",
       "\n",
       "   Sngl10000bp_hg19  RawScore_hg19  PHRED_hg19          pos_id19  \n",
       "0             137.0       2.381671       22.30  10_000093270_A_C  \n",
       "2             137.0       2.267094       21.80  10_000093271_C_A  \n",
       "4             137.0       2.045093       19.67  10_000093271_C_G  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "det37.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GC_hg19</th>\n",
       "      <th>CpG_hg19</th>\n",
       "      <th>priPhCons_hg19</th>\n",
       "      <th>mamPhCons_hg19</th>\n",
       "      <th>verPhCons_hg19</th>\n",
       "      <th>priPhyloP_hg19</th>\n",
       "      <th>mamPhyloP_hg19</th>\n",
       "      <th>verPhyloP_hg19</th>\n",
       "      <th>GerpRS_hg19</th>\n",
       "      <th>GerpRSpval_hg19</th>\n",
       "      <th>GerpN_hg19</th>\n",
       "      <th>GerpS_hg19</th>\n",
       "      <th>Dist2Mutation_hg19</th>\n",
       "      <th>Freq100bp_hg19</th>\n",
       "      <th>Rare100bp_hg19</th>\n",
       "      <th>Sngl100bp_hg19</th>\n",
       "      <th>Freq1000bp_hg19</th>\n",
       "      <th>Rare1000bp_hg19</th>\n",
       "      <th>Sngl1000bp_hg19</th>\n",
       "      <th>Freq10000bp_hg19</th>\n",
       "      <th>Rare10000bp_hg19</th>\n",
       "      <th>Sngl10000bp_hg19</th>\n",
       "      <th>RawScore_hg19</th>\n",
       "      <th>PHRED_hg19</th>\n",
       "      <th>pos_id19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0.536424</td>\n",
       "      <td>0.106667</td>\n",
       "      <td>0.550</td>\n",
       "      <td>0.546</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.109</td>\n",
       "      <td>1.700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>2.636298</td>\n",
       "      <td>22.9</td>\n",
       "      <td>10_000093156_C_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0.536424</td>\n",
       "      <td>0.106667</td>\n",
       "      <td>0.550</td>\n",
       "      <td>0.546</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.109</td>\n",
       "      <td>1.700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>2.627006</td>\n",
       "      <td>22.8</td>\n",
       "      <td>10_000093156_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0.529801</td>\n",
       "      <td>0.106667</td>\n",
       "      <td>0.546</td>\n",
       "      <td>0.542</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.094</td>\n",
       "      <td>0.095</td>\n",
       "      <td>2.564</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>2.819085</td>\n",
       "      <td>23.2</td>\n",
       "      <td>10_000093157_T_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    GC_hg19  CpG_hg19  priPhCons_hg19  mamPhCons_hg19  verPhCons_hg19  \\\n",
       "0  0.536424  0.106667           0.550           0.546             1.0   \n",
       "2  0.536424  0.106667           0.550           0.546             1.0   \n",
       "4  0.529801  0.106667           0.546           0.542             1.0   \n",
       "\n",
       "   priPhyloP_hg19  mamPhyloP_hg19  verPhyloP_hg19  GerpRS_hg19  \\\n",
       "0           0.110           0.109           1.700          NaN   \n",
       "2           0.110           0.109           1.700          NaN   \n",
       "4           0.094           0.095           2.564          NaN   \n",
       "\n",
       "   GerpRSpval_hg19  GerpN_hg19  GerpS_hg19  Dist2Mutation_hg19  \\\n",
       "0              NaN         0.0         0.0             38936.0   \n",
       "2              NaN         0.0         0.0             38936.0   \n",
       "4              NaN         0.0         0.0             38936.0   \n",
       "\n",
       "   Freq100bp_hg19  Rare100bp_hg19  Sngl100bp_hg19  Freq1000bp_hg19  \\\n",
       "0             NaN             NaN             NaN              NaN   \n",
       "2             NaN             NaN             NaN              NaN   \n",
       "4             NaN             NaN             NaN              NaN   \n",
       "\n",
       "   Rare1000bp_hg19  Sngl1000bp_hg19  Freq10000bp_hg19  Rare10000bp_hg19  \\\n",
       "0              NaN              NaN               5.0              24.0   \n",
       "2              NaN              NaN               5.0              24.0   \n",
       "4              NaN              NaN               5.0              24.0   \n",
       "\n",
       "   Sngl10000bp_hg19  RawScore_hg19  PHRED_hg19          pos_id19  \n",
       "0             137.0       2.636298        22.9  10_000093156_C_A  \n",
       "2             137.0       2.627006        22.8  10_000093156_C_G  \n",
       "4             137.0       2.819085        23.2  10_000093157_T_A  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not37.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## importing dbNSFP files with select/parse column values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "detdbnsfp = pd.read_csv(\"SCORE2_dbNSFP_parsedcols_detected_CK_104475.csv\", low_memory=False)\n",
    "\n",
    "notdbnsfp = pd.read_csv(\"SCORE2_dbNSFP_parsedcols_NOT_detected_CK_1222911.csv\", low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos_id19</th>\n",
       "      <th>pos_id38</th>\n",
       "      <th>pos_ID</th>\n",
       "      <th>matched_aapos</th>\n",
       "      <th>matched_target</th>\n",
       "      <th>matched_UKBID</th>\n",
       "      <th>matched_index</th>\n",
       "      <th>pos_dict</th>\n",
       "      <th>aaref</th>\n",
       "      <th>aaalt</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>CADD_phred_hg38</th>\n",
       "      <th>CADD_phred_hg19</th>\n",
       "      <th>CADDdiff_38minus19</th>\n",
       "      <th>Cys_reactivity</th>\n",
       "      <th>Cys_react_threshold</th>\n",
       "      <th>Cys_target_label</th>\n",
       "      <th>Lys_reactivity</th>\n",
       "      <th>Lys_react_threshold</th>\n",
       "      <th>Lys_target_label</th>\n",
       "      <th>rs_dbSNP151</th>\n",
       "      <th>Ensembl_transcriptid</th>\n",
       "      <th>cds_strand</th>\n",
       "      <th>refcodon</th>\n",
       "      <th>codonpos</th>\n",
       "      <th>codon_degeneracy</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>SIFT_converted_rankscore</th>\n",
       "      <th>SIFT_pred</th>\n",
       "      <th>SIFT4G_score</th>\n",
       "      <th>SIFT4G_converted_rankscore</th>\n",
       "      <th>SIFT4G_pred</th>\n",
       "      <th>Polyphen2_HDIV_score</th>\n",
       "      <th>Polyphen2_HDIV_rankscore</th>\n",
       "      <th>Polyphen2_HDIV_pred</th>\n",
       "      <th>Polyphen2_HVAR_score</th>\n",
       "      <th>Polyphen2_HVAR_rankscore</th>\n",
       "      <th>Polyphen2_HVAR_pred</th>\n",
       "      <th>LRT_score</th>\n",
       "      <th>LRT_converted_rankscore</th>\n",
       "      <th>LRT_pred</th>\n",
       "      <th>LRT_Omega</th>\n",
       "      <th>MutationTaster_score</th>\n",
       "      <th>MutationTaster_converted_rankscore</th>\n",
       "      <th>MutationTaster_pred</th>\n",
       "      <th>MutationTaster_model</th>\n",
       "      <th>MutationTaster_AAE</th>\n",
       "      <th>MutationAssessor_score</th>\n",
       "      <th>MutationAssessor_rankscore</th>\n",
       "      <th>MutationAssessor_pred</th>\n",
       "      <th>FATHMM_score</th>\n",
       "      <th>FATHMM_converted_rankscore</th>\n",
       "      <th>FATHMM_pred</th>\n",
       "      <th>PROVEAN_score</th>\n",
       "      <th>PROVEAN_converted_rankscore</th>\n",
       "      <th>PROVEAN_pred</th>\n",
       "      <th>VEST4_score</th>\n",
       "      <th>VEST4_rankscore</th>\n",
       "      <th>MetaSVM_score</th>\n",
       "      <th>MetaSVM_rankscore</th>\n",
       "      <th>MetaSVM_pred</th>\n",
       "      <th>MetaLR_score</th>\n",
       "      <th>MetaLR_rankscore</th>\n",
       "      <th>MetaLR_pred</th>\n",
       "      <th>Reliability_index</th>\n",
       "      <th>M-CAP_score</th>\n",
       "      <th>M-CAP_rankscore</th>\n",
       "      <th>M-CAP_pred</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>REVEL_rankscore</th>\n",
       "      <th>MutPred_score</th>\n",
       "      <th>MutPred_rankscore</th>\n",
       "      <th>MutPred_protID</th>\n",
       "      <th>MutPred_AAchange</th>\n",
       "      <th>MutPred_Top5features</th>\n",
       "      <th>MPC_score</th>\n",
       "      <th>MPC_rankscore</th>\n",
       "      <th>PrimateAI_score</th>\n",
       "      <th>PrimateAI_rankscore</th>\n",
       "      <th>PrimateAI_pred</th>\n",
       "      <th>DANN_score</th>\n",
       "      <th>DANN_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_score</th>\n",
       "      <th>fathmm-MKL_coding_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_pred</th>\n",
       "      <th>fathmm-MKL_coding_group</th>\n",
       "      <th>fathmm-XF_coding_score</th>\n",
       "      <th>fathmm-XF_coding_rankscore</th>\n",
       "      <th>fathmm-XF_coding_pred</th>\n",
       "      <th>Eigen-raw_coding</th>\n",
       "      <th>Eigen-raw_coding_rankscore</th>\n",
       "      <th>Eigen-pred_coding</th>\n",
       "      <th>Eigen-PC-raw_coding</th>\n",
       "      <th>Eigen-PC-raw_coding_rankscore</th>\n",
       "      <th>Eigen-PC-phred_coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>10_000093270_A_C</td>\n",
       "      <td>10_000047330_A_C</td>\n",
       "      <td>Q3ZCM7_C354</td>\n",
       "      <td>354</td>\n",
       "      <td>C354</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>2</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>C</td>\n",
       "      <td>W</td>\n",
       "      <td>Cys/Trp</td>\n",
       "      <td>22.6</td>\n",
       "      <td>22.30</td>\n",
       "      <td>0.30</td>\n",
       "      <td>5.37</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000568584</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>0.91255</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.92824</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.90584</td>\n",
       "      <td>D</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.88582</td>\n",
       "      <td>D</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.999882</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D</td>\n",
       "      <td>simple_aae</td>\n",
       "      <td>C354W</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1.98</td>\n",
       "      <td>0.85247</td>\n",
       "      <td>D</td>\n",
       "      <td>-6.84</td>\n",
       "      <td>0.94249</td>\n",
       "      <td>D</td>\n",
       "      <td>0.751</td>\n",
       "      <td>0.75466</td>\n",
       "      <td>0.3131</td>\n",
       "      <td>0.87764</td>\n",
       "      <td>D</td>\n",
       "      <td>0.8180</td>\n",
       "      <td>0.93873</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.020603</td>\n",
       "      <td>0.43227</td>\n",
       "      <td>T</td>\n",
       "      <td>0.437</td>\n",
       "      <td>0.74446</td>\n",
       "      <td>0.868</td>\n",
       "      <td>0.96212</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>C354W</td>\n",
       "      <td>Loss of methylation at K350 (P = 0.0838)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.76919400692</td>\n",
       "      <td>0.77304</td>\n",
       "      <td>T</td>\n",
       "      <td>0.851523</td>\n",
       "      <td>0.15717</td>\n",
       "      <td>0.19276</td>\n",
       "      <td>0.20597</td>\n",
       "      <td>N</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.422079</td>\n",
       "      <td>0.48823</td>\n",
       "      <td>N</td>\n",
       "      <td>0.0406181179914726</td>\n",
       "      <td>0.43713</td>\n",
       "      <td>2.658815</td>\n",
       "      <td>-0.363070660159128</td>\n",
       "      <td>0.26108</td>\n",
       "      <td>1.439964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>10_000093271_C_A</td>\n",
       "      <td>10_000047331_C_A</td>\n",
       "      <td>Q3ZCM7_C354</td>\n",
       "      <td>354</td>\n",
       "      <td>C354</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>2</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>C</td>\n",
       "      <td>F</td>\n",
       "      <td>Cys/Phe</td>\n",
       "      <td>22.6</td>\n",
       "      <td>21.80</td>\n",
       "      <td>0.80</td>\n",
       "      <td>5.37</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000568584</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>0.78490</td>\n",
       "      <td>.</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.66756</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.90584</td>\n",
       "      <td>D</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.88582</td>\n",
       "      <td>D</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.986821</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D</td>\n",
       "      <td>simple_aae</td>\n",
       "      <td>C354F</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1.95</td>\n",
       "      <td>0.85003</td>\n",
       "      <td>D</td>\n",
       "      <td>-6.83</td>\n",
       "      <td>0.94223</td>\n",
       "      <td>D</td>\n",
       "      <td>0.782</td>\n",
       "      <td>0.77883</td>\n",
       "      <td>0.4271</td>\n",
       "      <td>0.89500</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7835</td>\n",
       "      <td>0.92649</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.015962</td>\n",
       "      <td>0.36975</td>\n",
       "      <td>T</td>\n",
       "      <td>0.374</td>\n",
       "      <td>0.69594</td>\n",
       "      <td>0.808</td>\n",
       "      <td>0.92509</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>C354F</td>\n",
       "      <td>Gain of catalytic residue at C354 (P = 0.1094)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.727896571159</td>\n",
       "      <td>0.71199</td>\n",
       "      <td>T</td>\n",
       "      <td>0.895918</td>\n",
       "      <td>0.18929</td>\n",
       "      <td>0.57914</td>\n",
       "      <td>0.30502</td>\n",
       "      <td>D</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.537556</td>\n",
       "      <td>0.55531</td>\n",
       "      <td>D</td>\n",
       "      <td>0.0988420731411684</td>\n",
       "      <td>0.46409</td>\n",
       "      <td>2.88164</td>\n",
       "      <td>-0.263815814260368</td>\n",
       "      <td>0.29300</td>\n",
       "      <td>1.640415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>10_000093271_C_G</td>\n",
       "      <td>10_000047331_C_G</td>\n",
       "      <td>Q3ZCM7_C354</td>\n",
       "      <td>354</td>\n",
       "      <td>C354</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>2</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>Cys/Ser</td>\n",
       "      <td>22.4</td>\n",
       "      <td>19.67</td>\n",
       "      <td>2.73</td>\n",
       "      <td>5.37</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000568584</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>0.91255</td>\n",
       "      <td>.</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.74150</td>\n",
       "      <td>D</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.70673</td>\n",
       "      <td>D</td>\n",
       "      <td>0.968</td>\n",
       "      <td>0.71741</td>\n",
       "      <td>D</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.930492</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D</td>\n",
       "      <td>simple_aae</td>\n",
       "      <td>C354S</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1.87</td>\n",
       "      <td>0.84415</td>\n",
       "      <td>D</td>\n",
       "      <td>-6.21</td>\n",
       "      <td>0.91827</td>\n",
       "      <td>D</td>\n",
       "      <td>0.739</td>\n",
       "      <td>0.74735</td>\n",
       "      <td>0.4621</td>\n",
       "      <td>0.90007</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7769</td>\n",
       "      <td>0.92419</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.022348</td>\n",
       "      <td>0.45230</td>\n",
       "      <td>T</td>\n",
       "      <td>0.377</td>\n",
       "      <td>0.69846</td>\n",
       "      <td>0.781</td>\n",
       "      <td>0.90537</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>C354S</td>\n",
       "      <td>Loss of sheet (P = 0.0817)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.713165700436</td>\n",
       "      <td>0.69052</td>\n",
       "      <td>T</td>\n",
       "      <td>0.799045</td>\n",
       "      <td>0.12954</td>\n",
       "      <td>0.59480</td>\n",
       "      <td>0.30927</td>\n",
       "      <td>D</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.537556</td>\n",
       "      <td>0.55531</td>\n",
       "      <td>D</td>\n",
       "      <td>0.0966724300357427</td>\n",
       "      <td>0.46307</td>\n",
       "      <td>2.873107</td>\n",
       "      <td>-0.265973651554631</td>\n",
       "      <td>0.29227</td>\n",
       "      <td>1.635795</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos_id19          pos_id38       pos_ID  matched_aapos  \\\n",
       "0  10_000093270_A_C  10_000047330_A_C  Q3ZCM7_C354            354   \n",
       "1  10_000093271_C_A  10_000047331_C_A  Q3ZCM7_C354            354   \n",
       "2  10_000093271_C_G  10_000047331_C_G  Q3ZCM7_C354            354   \n",
       "\n",
       "  matched_target matched_UKBID  matched_index  \\\n",
       "0           C354        Q3ZCM7              2   \n",
       "1           C354        Q3ZCM7              2   \n",
       "2           C354        Q3ZCM7              2   \n",
       "\n",
       "                                  pos_dict aaref aaalt Amino_acids  \\\n",
       "0  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}     C     W     Cys/Trp   \n",
       "1  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}     C     F     Cys/Phe   \n",
       "2  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}     C     S     Cys/Ser   \n",
       "\n",
       "   CADD_phred_hg38  CADD_phred_hg19  CADDdiff_38minus19  Cys_reactivity  \\\n",
       "0             22.6            22.30                0.30            5.37   \n",
       "1             22.6            21.80                0.80            5.37   \n",
       "2             22.4            19.67                2.73            5.37   \n",
       "\n",
       "  Cys_react_threshold Cys_target_label  Lys_reactivity Lys_react_threshold  \\\n",
       "0                 Low              NaN             NaN                 NaN   \n",
       "1                 Low              NaN             NaN                 NaN   \n",
       "2                 Low              NaN             NaN                 NaN   \n",
       "\n",
       "  Lys_target_label rs_dbSNP151 Ensembl_transcriptid cds_strand refcodon  \\\n",
       "0              NaN           .      ENST00000568584          -      TGT   \n",
       "1              NaN           .      ENST00000568584          -      TGT   \n",
       "2              NaN           .      ENST00000568584          -      TGT   \n",
       "\n",
       "   codonpos  codon_degeneracy SIFT_score SIFT_converted_rankscore SIFT_pred  \\\n",
       "0         3                 2          .                  0.91255         .   \n",
       "1         2                 0          .                  0.78490         .   \n",
       "2         2                 0          .                  0.91255         .   \n",
       "\n",
       "  SIFT4G_score SIFT4G_converted_rankscore SIFT4G_pred Polyphen2_HDIV_score  \\\n",
       "0          0.0                    0.92824           D                  1.0   \n",
       "1        0.013                    0.66756           D                  1.0   \n",
       "2        0.026                    0.74150           D                0.997   \n",
       "\n",
       "  Polyphen2_HDIV_rankscore Polyphen2_HDIV_pred Polyphen2_HVAR_score  \\\n",
       "0                  0.90584                   D                0.998   \n",
       "1                  0.90584                   D                0.998   \n",
       "2                  0.70673                   D                0.968   \n",
       "\n",
       "  Polyphen2_HVAR_rankscore Polyphen2_HVAR_pred LRT_score  \\\n",
       "0                  0.88582                   D  0.000003   \n",
       "1                  0.88582                   D  0.000003   \n",
       "2                  0.71741                   D  0.000003   \n",
       "\n",
       "  LRT_converted_rankscore LRT_pred LRT_Omega MutationTaster_score  \\\n",
       "0                 0.62929        U  0.000000             0.999882   \n",
       "1                 0.62929        U  0.000000             0.986821   \n",
       "2                 0.62929        U  0.000000             0.930492   \n",
       "\n",
       "  MutationTaster_converted_rankscore MutationTaster_pred MutationTaster_model  \\\n",
       "0                            0.81001                   D           simple_aae   \n",
       "1                            0.81001                   D           simple_aae   \n",
       "2                            0.81001                   D           simple_aae   \n",
       "\n",
       "  MutationTaster_AAE MutationAssessor_score MutationAssessor_rankscore  \\\n",
       "0              C354W                      .                          .   \n",
       "1              C354F                      .                          .   \n",
       "2              C354S                      .                          .   \n",
       "\n",
       "  MutationAssessor_pred FATHMM_score FATHMM_converted_rankscore FATHMM_pred  \\\n",
       "0                     .        -1.98                    0.85247           D   \n",
       "1                     .        -1.95                    0.85003           D   \n",
       "2                     .        -1.87                    0.84415           D   \n",
       "\n",
       "  PROVEAN_score PROVEAN_converted_rankscore PROVEAN_pred VEST4_score  \\\n",
       "0         -6.84                     0.94249            D       0.751   \n",
       "1         -6.83                     0.94223            D       0.782   \n",
       "2         -6.21                     0.91827            D       0.739   \n",
       "\n",
       "  VEST4_rankscore MetaSVM_score MetaSVM_rankscore MetaSVM_pred MetaLR_score  \\\n",
       "0         0.75466        0.3131           0.87764            D       0.8180   \n",
       "1         0.77883        0.4271           0.89500            D       0.7835   \n",
       "2         0.74735        0.4621           0.90007            D       0.7769   \n",
       "\n",
       "  MetaLR_rankscore MetaLR_pred Reliability_index M-CAP_score M-CAP_rankscore  \\\n",
       "0          0.93873           D                 9    0.020603         0.43227   \n",
       "1          0.92649           D                 9    0.015962         0.36975   \n",
       "2          0.92419           D                 9    0.022348         0.45230   \n",
       "\n",
       "  M-CAP_pred REVEL_score REVEL_rankscore MutPred_score MutPred_rankscore  \\\n",
       "0          T       0.437         0.74446         0.868           0.96212   \n",
       "1          T       0.374         0.69594         0.808           0.92509   \n",
       "2          T       0.377         0.69846         0.781           0.90537   \n",
       "\n",
       "  MutPred_protID MutPred_AAchange  \\\n",
       "0         Q3ZCM7            C354W   \n",
       "1         Q3ZCM7            C354F   \n",
       "2         Q3ZCM7            C354S   \n",
       "\n",
       "                              MutPred_Top5features MPC_score MPC_rankscore  \\\n",
       "0         Loss of methylation at K350 (P = 0.0838)         .             .   \n",
       "1   Gain of catalytic residue at C354 (P = 0.1094)         .             .   \n",
       "2                       Loss of sheet (P = 0.0817)         .             .   \n",
       "\n",
       "  PrimateAI_score PrimateAI_rankscore PrimateAI_pred  DANN_score  \\\n",
       "0   0.76919400692             0.77304              T    0.851523   \n",
       "1  0.727896571159             0.71199              T    0.895918   \n",
       "2  0.713165700436             0.69052              T    0.799045   \n",
       "\n",
       "   DANN_rankscore  fathmm-MKL_coding_score  fathmm-MKL_coding_rankscore  \\\n",
       "0         0.15717                  0.19276                      0.20597   \n",
       "1         0.18929                  0.57914                      0.30502   \n",
       "2         0.12954                  0.59480                      0.30927   \n",
       "\n",
       "  fathmm-MKL_coding_pred fathmm-MKL_coding_group fathmm-XF_coding_score  \\\n",
       "0                      N                   AEFBI               0.422079   \n",
       "1                      D                   AEFBI               0.537556   \n",
       "2                      D                   AEFBI               0.537556   \n",
       "\n",
       "  fathmm-XF_coding_rankscore fathmm-XF_coding_pred    Eigen-raw_coding  \\\n",
       "0                    0.48823                     N  0.0406181179914726   \n",
       "1                    0.55531                     D  0.0988420731411684   \n",
       "2                    0.55531                     D  0.0966724300357427   \n",
       "\n",
       "  Eigen-raw_coding_rankscore Eigen-pred_coding Eigen-PC-raw_coding  \\\n",
       "0                    0.43713          2.658815  -0.363070660159128   \n",
       "1                    0.46409           2.88164  -0.263815814260368   \n",
       "2                    0.46307          2.873107  -0.265973651554631   \n",
       "\n",
       "  Eigen-PC-raw_coding_rankscore Eigen-PC-phred_coding  \n",
       "0                       0.26108              1.439964  \n",
       "1                       0.29300              1.640415  \n",
       "2                       0.29227              1.635795  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "detdbnsfp.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos_id19</th>\n",
       "      <th>pos_id38</th>\n",
       "      <th>pos_ID</th>\n",
       "      <th>matched_UKBID</th>\n",
       "      <th>matched_aapos</th>\n",
       "      <th>matched_index</th>\n",
       "      <th>Ensembl_transcriptid</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>aaref</th>\n",
       "      <th>aaalt</th>\n",
       "      <th>CADD_phred_hg38</th>\n",
       "      <th>CADD_phred_hg19</th>\n",
       "      <th>CADDdiff_38minus19</th>\n",
       "      <th>rs_dbSNP151</th>\n",
       "      <th>genename</th>\n",
       "      <th>cds_strand</th>\n",
       "      <th>refcodon</th>\n",
       "      <th>codonpos</th>\n",
       "      <th>codon_degeneracy</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>SIFT_converted_rankscore</th>\n",
       "      <th>SIFT_pred</th>\n",
       "      <th>SIFT4G_score</th>\n",
       "      <th>SIFT4G_converted_rankscore</th>\n",
       "      <th>SIFT4G_pred</th>\n",
       "      <th>Polyphen2_HDIV_score</th>\n",
       "      <th>Polyphen2_HDIV_rankscore</th>\n",
       "      <th>Polyphen2_HDIV_pred</th>\n",
       "      <th>Polyphen2_HVAR_score</th>\n",
       "      <th>Polyphen2_HVAR_rankscore</th>\n",
       "      <th>Polyphen2_HVAR_pred</th>\n",
       "      <th>LRT_score</th>\n",
       "      <th>LRT_converted_rankscore</th>\n",
       "      <th>LRT_pred</th>\n",
       "      <th>LRT_Omega</th>\n",
       "      <th>MutationTaster_score</th>\n",
       "      <th>MutationTaster_converted_rankscore</th>\n",
       "      <th>MutationTaster_pred</th>\n",
       "      <th>MutationTaster_model</th>\n",
       "      <th>MutationTaster_AAE</th>\n",
       "      <th>MutationAssessor_score</th>\n",
       "      <th>MutationAssessor_rankscore</th>\n",
       "      <th>MutationAssessor_pred</th>\n",
       "      <th>FATHMM_score</th>\n",
       "      <th>FATHMM_converted_rankscore</th>\n",
       "      <th>FATHMM_pred</th>\n",
       "      <th>PROVEAN_score</th>\n",
       "      <th>PROVEAN_converted_rankscore</th>\n",
       "      <th>PROVEAN_pred</th>\n",
       "      <th>VEST4_score</th>\n",
       "      <th>VEST4_rankscore</th>\n",
       "      <th>MetaSVM_score</th>\n",
       "      <th>MetaSVM_rankscore</th>\n",
       "      <th>MetaSVM_pred</th>\n",
       "      <th>MetaLR_score</th>\n",
       "      <th>MetaLR_rankscore</th>\n",
       "      <th>MetaLR_pred</th>\n",
       "      <th>Reliability_index</th>\n",
       "      <th>M-CAP_score</th>\n",
       "      <th>M-CAP_rankscore</th>\n",
       "      <th>M-CAP_pred</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>REVEL_rankscore</th>\n",
       "      <th>MutPred_score</th>\n",
       "      <th>MutPred_rankscore</th>\n",
       "      <th>MutPred_protID</th>\n",
       "      <th>MutPred_AAchange</th>\n",
       "      <th>MutPred_Top5features</th>\n",
       "      <th>MPC_score</th>\n",
       "      <th>MPC_rankscore</th>\n",
       "      <th>PrimateAI_score</th>\n",
       "      <th>PrimateAI_rankscore</th>\n",
       "      <th>PrimateAI_pred</th>\n",
       "      <th>DANN_score</th>\n",
       "      <th>DANN_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_score</th>\n",
       "      <th>fathmm-MKL_coding_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_pred</th>\n",
       "      <th>fathmm-MKL_coding_group</th>\n",
       "      <th>fathmm-XF_coding_score</th>\n",
       "      <th>fathmm-XF_coding_rankscore</th>\n",
       "      <th>fathmm-XF_coding_pred</th>\n",
       "      <th>Eigen-raw_coding</th>\n",
       "      <th>Eigen-raw_coding_rankscore</th>\n",
       "      <th>Eigen-pred_coding</th>\n",
       "      <th>Eigen-PC-raw_coding</th>\n",
       "      <th>Eigen-PC-raw_coding_rankscore</th>\n",
       "      <th>Eigen-PC-phred_coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>10_000093156_C_A</td>\n",
       "      <td>10_000047216_C_A</td>\n",
       "      <td>Q3ZCM7_K392</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>392</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000568584</td>\n",
       "      <td>Lys/Asn</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>23.0</td>\n",
       "      <td>22.9</td>\n",
       "      <td>0.1</td>\n",
       "      <td>.</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>AAG</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>0.91255</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.92824</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.90584</td>\n",
       "      <td>D</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.9235899999999999</td>\n",
       "      <td>D</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.654618</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D</td>\n",
       "      <td>simple_aae</td>\n",
       "      <td>K392N</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1.98</td>\n",
       "      <td>0.85247</td>\n",
       "      <td>D</td>\n",
       "      <td>-2.6</td>\n",
       "      <td>0.58733</td>\n",
       "      <td>D</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.6007100000000001</td>\n",
       "      <td>0.3991</td>\n",
       "      <td>0.8908299999999999</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7111</td>\n",
       "      <td>0.90075</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.009489</td>\n",
       "      <td>0.24855</td>\n",
       "      <td>T</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.78035</td>\n",
       "      <td>0.718</td>\n",
       "      <td>0.85313</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>K392N</td>\n",
       "      <td>Loss of MoRF binding (P = 0.0523)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.751451253891</td>\n",
       "      <td>0.74660</td>\n",
       "      <td>T</td>\n",
       "      <td>0.984728</td>\n",
       "      <td>0.41873</td>\n",
       "      <td>0.38681</td>\n",
       "      <td>0.26043</td>\n",
       "      <td>N</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.295876</td>\n",
       "      <td>0.40587</td>\n",
       "      <td>N</td>\n",
       "      <td>0.0676056746461709</td>\n",
       "      <td>0.44955</td>\n",
       "      <td>2.76036</td>\n",
       "      <td>-0.20748312943205102</td>\n",
       "      <td>0.31266</td>\n",
       "      <td>1.767854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>10_000093156_C_G</td>\n",
       "      <td>10_000047216_C_G</td>\n",
       "      <td>Q3ZCM7_K392</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>392</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000568584</td>\n",
       "      <td>Lys/Asn</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>23.1</td>\n",
       "      <td>22.8</td>\n",
       "      <td>0.3</td>\n",
       "      <td>.</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>AAG</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>0.91255</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.92824</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.90584</td>\n",
       "      <td>D</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.9235899999999999</td>\n",
       "      <td>D</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.654618</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D</td>\n",
       "      <td>simple_aae</td>\n",
       "      <td>K392N</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1.98</td>\n",
       "      <td>0.85247</td>\n",
       "      <td>D</td>\n",
       "      <td>-2.6</td>\n",
       "      <td>0.58733</td>\n",
       "      <td>D</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.6007100000000001</td>\n",
       "      <td>0.3991</td>\n",
       "      <td>0.8908299999999999</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7111</td>\n",
       "      <td>0.90075</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.008795</td>\n",
       "      <td>0.23196</td>\n",
       "      <td>T</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.78035</td>\n",
       "      <td>0.718</td>\n",
       "      <td>0.85313</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>K392N</td>\n",
       "      <td>Loss of MoRF binding (P = 0.0523)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.751451253891</td>\n",
       "      <td>0.74660</td>\n",
       "      <td>T</td>\n",
       "      <td>0.984663</td>\n",
       "      <td>0.41803</td>\n",
       "      <td>0.40406</td>\n",
       "      <td>0.26429</td>\n",
       "      <td>N</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.302556</td>\n",
       "      <td>0.41077</td>\n",
       "      <td>N</td>\n",
       "      <td>0.0676056746461709</td>\n",
       "      <td>0.44955</td>\n",
       "      <td>2.76036</td>\n",
       "      <td>-0.20748312943205102</td>\n",
       "      <td>0.31266</td>\n",
       "      <td>1.767854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>10_000093157_T_A</td>\n",
       "      <td>10_000047217_T_A</td>\n",
       "      <td>Q3ZCM7_K392</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>392</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000568584</td>\n",
       "      <td>Lys/Met</td>\n",
       "      <td>K</td>\n",
       "      <td>M</td>\n",
       "      <td>23.2</td>\n",
       "      <td>23.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>.</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>AAG</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>0.91255</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.92824</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.90584</td>\n",
       "      <td>D</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.9235899999999999</td>\n",
       "      <td>D</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.820833</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D</td>\n",
       "      <td>simple_aae</td>\n",
       "      <td>K392M</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-2.08</td>\n",
       "      <td>0.86010</td>\n",
       "      <td>D</td>\n",
       "      <td>-3.14</td>\n",
       "      <td>0.66549</td>\n",
       "      <td>D</td>\n",
       "      <td>0.568</td>\n",
       "      <td>0.62442</td>\n",
       "      <td>0.5527</td>\n",
       "      <td>0.9127</td>\n",
       "      <td>D</td>\n",
       "      <td>0.8388</td>\n",
       "      <td>0.9461</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.01492</td>\n",
       "      <td>0.35345</td>\n",
       "      <td>T</td>\n",
       "      <td>0.505</td>\n",
       "      <td>0.78906</td>\n",
       "      <td>0.698</td>\n",
       "      <td>0.83491</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>K392M</td>\n",
       "      <td>Loss of ubiquitination at K392 (P = 0.0326)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.727143764496</td>\n",
       "      <td>0.71088</td>\n",
       "      <td>T</td>\n",
       "      <td>0.964042</td>\n",
       "      <td>0.29699</td>\n",
       "      <td>0.52831</td>\n",
       "      <td>0.29225</td>\n",
       "      <td>D</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.364324</td>\n",
       "      <td>0.45291</td>\n",
       "      <td>N</td>\n",
       "      <td>0.31076238698785996</td>\n",
       "      <td>0.56701</td>\n",
       "      <td>3.8352209999999998</td>\n",
       "      <td>-0.0235518622390073</td>\n",
       "      <td>0.38654</td>\n",
       "      <td>2.280427</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos_id19          pos_id38       pos_ID matched_UKBID  \\\n",
       "0  10_000093156_C_A  10_000047216_C_A  Q3ZCM7_K392        Q3ZCM7   \n",
       "1  10_000093156_C_G  10_000047216_C_G  Q3ZCM7_K392        Q3ZCM7   \n",
       "2  10_000093157_T_A  10_000047217_T_A  Q3ZCM7_K392        Q3ZCM7   \n",
       "\n",
       "   matched_aapos  matched_index Ensembl_transcriptid Amino_acids aaref aaalt  \\\n",
       "0            392              2      ENST00000568584     Lys/Asn     K     N   \n",
       "1            392              2      ENST00000568584     Lys/Asn     K     N   \n",
       "2            392              2      ENST00000568584     Lys/Met     K     M   \n",
       "\n",
       "   CADD_phred_hg38  CADD_phred_hg19  CADDdiff_38minus19 rs_dbSNP151 genename  \\\n",
       "0             23.0             22.9                 0.1           .    TUBB8   \n",
       "1             23.1             22.8                 0.3           .    TUBB8   \n",
       "2             23.2             23.2                 0.0           .    TUBB8   \n",
       "\n",
       "  cds_strand refcodon  codonpos  codon_degeneracy SIFT_score  \\\n",
       "0          -      AAG         3                 2          .   \n",
       "1          -      AAG         3                 2          .   \n",
       "2          -      AAG         2                 0          .   \n",
       "\n",
       "  SIFT_converted_rankscore SIFT_pred SIFT4G_score SIFT4G_converted_rankscore  \\\n",
       "0                  0.91255         .          0.0                    0.92824   \n",
       "1                  0.91255         .          0.0                    0.92824   \n",
       "2                  0.91255         .          0.0                    0.92824   \n",
       "\n",
       "  SIFT4G_pred Polyphen2_HDIV_score Polyphen2_HDIV_rankscore  \\\n",
       "0           D                  1.0                  0.90584   \n",
       "1           D                  1.0                  0.90584   \n",
       "2           D                  1.0                  0.90584   \n",
       "\n",
       "  Polyphen2_HDIV_pred Polyphen2_HVAR_score Polyphen2_HVAR_rankscore  \\\n",
       "0                   D                0.999       0.9235899999999999   \n",
       "1                   D                0.999       0.9235899999999999   \n",
       "2                   D                0.999       0.9235899999999999   \n",
       "\n",
       "  Polyphen2_HVAR_pred LRT_score LRT_converted_rankscore LRT_pred LRT_Omega  \\\n",
       "0                   D  0.000010                 0.62929        U  0.000000   \n",
       "1                   D  0.000010                 0.62929        U  0.000000   \n",
       "2                   D  0.000010                 0.62929        U  0.000000   \n",
       "\n",
       "  MutationTaster_score MutationTaster_converted_rankscore MutationTaster_pred  \\\n",
       "0             0.654618                            0.81001                   D   \n",
       "1             0.654618                            0.81001                   D   \n",
       "2             0.820833                            0.81001                   D   \n",
       "\n",
       "  MutationTaster_model MutationTaster_AAE MutationAssessor_score  \\\n",
       "0           simple_aae              K392N                      .   \n",
       "1           simple_aae              K392N                      .   \n",
       "2           simple_aae              K392M                      .   \n",
       "\n",
       "  MutationAssessor_rankscore MutationAssessor_pred FATHMM_score  \\\n",
       "0                          .                     .        -1.98   \n",
       "1                          .                     .        -1.98   \n",
       "2                          .                     .        -2.08   \n",
       "\n",
       "  FATHMM_converted_rankscore FATHMM_pred PROVEAN_score  \\\n",
       "0                    0.85247           D          -2.6   \n",
       "1                    0.85247           D          -2.6   \n",
       "2                    0.86010           D         -3.14   \n",
       "\n",
       "  PROVEAN_converted_rankscore PROVEAN_pred VEST4_score     VEST4_rankscore  \\\n",
       "0                     0.58733            D        0.53  0.6007100000000001   \n",
       "1                     0.58733            D        0.53  0.6007100000000001   \n",
       "2                     0.66549            D       0.568             0.62442   \n",
       "\n",
       "  MetaSVM_score   MetaSVM_rankscore MetaSVM_pred MetaLR_score  \\\n",
       "0        0.3991  0.8908299999999999            D       0.7111   \n",
       "1        0.3991  0.8908299999999999            D       0.7111   \n",
       "2        0.5527              0.9127            D       0.8388   \n",
       "\n",
       "  MetaLR_rankscore MetaLR_pred Reliability_index M-CAP_score M-CAP_rankscore  \\\n",
       "0          0.90075           D                 9    0.009489         0.24855   \n",
       "1          0.90075           D                 9    0.008795         0.23196   \n",
       "2           0.9461           D                 9     0.01492         0.35345   \n",
       "\n",
       "  M-CAP_pred REVEL_score REVEL_rankscore MutPred_score MutPred_rankscore  \\\n",
       "0          T       0.491         0.78035         0.718           0.85313   \n",
       "1          T       0.491         0.78035         0.718           0.85313   \n",
       "2          T       0.505         0.78906         0.698           0.83491   \n",
       "\n",
       "  MutPred_protID MutPred_AAchange  \\\n",
       "0         Q3ZCM7            K392N   \n",
       "1         Q3ZCM7            K392N   \n",
       "2         Q3ZCM7            K392M   \n",
       "\n",
       "                           MutPred_Top5features MPC_score MPC_rankscore  \\\n",
       "0             Loss of MoRF binding (P = 0.0523)         .             .   \n",
       "1             Loss of MoRF binding (P = 0.0523)         .             .   \n",
       "2   Loss of ubiquitination at K392 (P = 0.0326)         .             .   \n",
       "\n",
       "  PrimateAI_score PrimateAI_rankscore PrimateAI_pred  DANN_score  \\\n",
       "0  0.751451253891             0.74660              T    0.984728   \n",
       "1  0.751451253891             0.74660              T    0.984663   \n",
       "2  0.727143764496             0.71088              T    0.964042   \n",
       "\n",
       "   DANN_rankscore  fathmm-MKL_coding_score  fathmm-MKL_coding_rankscore  \\\n",
       "0         0.41873                  0.38681                      0.26043   \n",
       "1         0.41803                  0.40406                      0.26429   \n",
       "2         0.29699                  0.52831                      0.29225   \n",
       "\n",
       "  fathmm-MKL_coding_pred fathmm-MKL_coding_group fathmm-XF_coding_score  \\\n",
       "0                      N                   AEFBI               0.295876   \n",
       "1                      N                   AEFBI               0.302556   \n",
       "2                      D                   AEFBI               0.364324   \n",
       "\n",
       "  fathmm-XF_coding_rankscore fathmm-XF_coding_pred     Eigen-raw_coding  \\\n",
       "0                    0.40587                     N   0.0676056746461709   \n",
       "1                    0.41077                     N   0.0676056746461709   \n",
       "2                    0.45291                     N  0.31076238698785996   \n",
       "\n",
       "  Eigen-raw_coding_rankscore   Eigen-pred_coding   Eigen-PC-raw_coding  \\\n",
       "0                    0.44955             2.76036  -0.20748312943205102   \n",
       "1                    0.44955             2.76036  -0.20748312943205102   \n",
       "2                    0.56701  3.8352209999999998   -0.0235518622390073   \n",
       "\n",
       "  Eigen-PC-raw_coding_rankscore Eigen-PC-phred_coding  \n",
       "0                       0.31266              1.767854  \n",
       "1                       0.31266              1.767854  \n",
       "2                       0.38654              2.280427  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "notdbnsfp.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [6b] MERGING CADD37 and 38 DETECTED with dbNSFP detected df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "merdet37 = pd.merge(detdbnsfp, det37, how='inner', on=['pos_id19'])\n",
    "\n",
    "mergeDET = pd.merge(merdet37, det38, how='inner', on=['pos_id38'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos_id19</th>\n",
       "      <th>pos_id38</th>\n",
       "      <th>pos_ID</th>\n",
       "      <th>matched_aapos</th>\n",
       "      <th>matched_target</th>\n",
       "      <th>matched_UKBID</th>\n",
       "      <th>matched_index</th>\n",
       "      <th>pos_dict</th>\n",
       "      <th>aaref</th>\n",
       "      <th>aaalt</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>CADD_phred_hg38</th>\n",
       "      <th>CADD_phred_hg19</th>\n",
       "      <th>CADDdiff_38minus19</th>\n",
       "      <th>Cys_reactivity</th>\n",
       "      <th>Cys_react_threshold</th>\n",
       "      <th>Cys_target_label</th>\n",
       "      <th>Lys_reactivity</th>\n",
       "      <th>Lys_react_threshold</th>\n",
       "      <th>Lys_target_label</th>\n",
       "      <th>rs_dbSNP151</th>\n",
       "      <th>Ensembl_transcriptid</th>\n",
       "      <th>cds_strand</th>\n",
       "      <th>refcodon</th>\n",
       "      <th>codonpos</th>\n",
       "      <th>codon_degeneracy</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>SIFT_converted_rankscore</th>\n",
       "      <th>SIFT_pred</th>\n",
       "      <th>SIFT4G_score</th>\n",
       "      <th>SIFT4G_converted_rankscore</th>\n",
       "      <th>SIFT4G_pred</th>\n",
       "      <th>Polyphen2_HDIV_score</th>\n",
       "      <th>Polyphen2_HDIV_rankscore</th>\n",
       "      <th>Polyphen2_HDIV_pred</th>\n",
       "      <th>Polyphen2_HVAR_score</th>\n",
       "      <th>Polyphen2_HVAR_rankscore</th>\n",
       "      <th>Polyphen2_HVAR_pred</th>\n",
       "      <th>LRT_score</th>\n",
       "      <th>LRT_converted_rankscore</th>\n",
       "      <th>LRT_pred</th>\n",
       "      <th>LRT_Omega</th>\n",
       "      <th>MutationTaster_score</th>\n",
       "      <th>MutationTaster_converted_rankscore</th>\n",
       "      <th>MutationTaster_pred</th>\n",
       "      <th>MutationTaster_model</th>\n",
       "      <th>MutationTaster_AAE</th>\n",
       "      <th>MutationAssessor_score</th>\n",
       "      <th>MutationAssessor_rankscore</th>\n",
       "      <th>MutationAssessor_pred</th>\n",
       "      <th>FATHMM_score</th>\n",
       "      <th>FATHMM_converted_rankscore</th>\n",
       "      <th>FATHMM_pred</th>\n",
       "      <th>PROVEAN_score</th>\n",
       "      <th>PROVEAN_converted_rankscore</th>\n",
       "      <th>PROVEAN_pred</th>\n",
       "      <th>VEST4_score</th>\n",
       "      <th>VEST4_rankscore</th>\n",
       "      <th>MetaSVM_score</th>\n",
       "      <th>MetaSVM_rankscore</th>\n",
       "      <th>MetaSVM_pred</th>\n",
       "      <th>MetaLR_score</th>\n",
       "      <th>MetaLR_rankscore</th>\n",
       "      <th>MetaLR_pred</th>\n",
       "      <th>Reliability_index</th>\n",
       "      <th>M-CAP_score</th>\n",
       "      <th>M-CAP_rankscore</th>\n",
       "      <th>M-CAP_pred</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>REVEL_rankscore</th>\n",
       "      <th>MutPred_score</th>\n",
       "      <th>MutPred_rankscore</th>\n",
       "      <th>MutPred_protID</th>\n",
       "      <th>MutPred_AAchange</th>\n",
       "      <th>MutPred_Top5features</th>\n",
       "      <th>MPC_score</th>\n",
       "      <th>MPC_rankscore</th>\n",
       "      <th>PrimateAI_score</th>\n",
       "      <th>PrimateAI_rankscore</th>\n",
       "      <th>PrimateAI_pred</th>\n",
       "      <th>DANN_score</th>\n",
       "      <th>DANN_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_score</th>\n",
       "      <th>fathmm-MKL_coding_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_pred</th>\n",
       "      <th>fathmm-MKL_coding_group</th>\n",
       "      <th>fathmm-XF_coding_score</th>\n",
       "      <th>fathmm-XF_coding_rankscore</th>\n",
       "      <th>fathmm-XF_coding_pred</th>\n",
       "      <th>Eigen-raw_coding</th>\n",
       "      <th>Eigen-raw_coding_rankscore</th>\n",
       "      <th>Eigen-pred_coding</th>\n",
       "      <th>Eigen-PC-raw_coding</th>\n",
       "      <th>Eigen-PC-raw_coding_rankscore</th>\n",
       "      <th>Eigen-PC-phred_coding</th>\n",
       "      <th>GC_hg19</th>\n",
       "      <th>CpG_hg19</th>\n",
       "      <th>priPhCons_hg19</th>\n",
       "      <th>mamPhCons_hg19</th>\n",
       "      <th>verPhCons_hg19</th>\n",
       "      <th>priPhyloP_hg19</th>\n",
       "      <th>mamPhyloP_hg19</th>\n",
       "      <th>verPhyloP_hg19</th>\n",
       "      <th>GerpRS_hg19</th>\n",
       "      <th>GerpRSpval_hg19</th>\n",
       "      <th>GerpN_hg19</th>\n",
       "      <th>GerpS_hg19</th>\n",
       "      <th>Dist2Mutation_hg19</th>\n",
       "      <th>Freq100bp_hg19</th>\n",
       "      <th>Rare100bp_hg19</th>\n",
       "      <th>Sngl100bp_hg19</th>\n",
       "      <th>Freq1000bp_hg19</th>\n",
       "      <th>Rare1000bp_hg19</th>\n",
       "      <th>Sngl1000bp_hg19</th>\n",
       "      <th>Freq10000bp_hg19</th>\n",
       "      <th>Rare10000bp_hg19</th>\n",
       "      <th>Sngl10000bp_hg19</th>\n",
       "      <th>RawScore_hg19</th>\n",
       "      <th>PHRED_hg19</th>\n",
       "      <th>GC_hg38</th>\n",
       "      <th>CpG_hg38</th>\n",
       "      <th>priPhCons_hg38</th>\n",
       "      <th>mamPhCons_hg38</th>\n",
       "      <th>verPhCons_hg38</th>\n",
       "      <th>priPhyloP_hg38</th>\n",
       "      <th>mamPhyloP_hg38</th>\n",
       "      <th>verPhyloP_hg38</th>\n",
       "      <th>GerpRS_hg38</th>\n",
       "      <th>GerpRSpval_hg38</th>\n",
       "      <th>GerpN_hg38</th>\n",
       "      <th>GerpS_hg38</th>\n",
       "      <th>Dist2Mutation_hg38</th>\n",
       "      <th>Freq100bp_hg38</th>\n",
       "      <th>Rare100bp_hg38</th>\n",
       "      <th>Sngl100bp_hg38</th>\n",
       "      <th>Freq1000bp_hg38</th>\n",
       "      <th>Rare1000bp_hg38</th>\n",
       "      <th>Sngl1000bp_hg38</th>\n",
       "      <th>Freq10000bp_hg38</th>\n",
       "      <th>Rare10000bp_hg38</th>\n",
       "      <th>Sngl10000bp_hg38</th>\n",
       "      <th>RawScore_hg38</th>\n",
       "      <th>PHRED_hg38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>10_000093270_A_C</td>\n",
       "      <td>10_000047330_A_C</td>\n",
       "      <td>Q3ZCM7_C354</td>\n",
       "      <td>354</td>\n",
       "      <td>C354</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>2</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>C</td>\n",
       "      <td>W</td>\n",
       "      <td>Cys/Trp</td>\n",
       "      <td>22.6</td>\n",
       "      <td>22.30</td>\n",
       "      <td>0.30</td>\n",
       "      <td>5.37</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000568584</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>0.91255</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.92824</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.90584</td>\n",
       "      <td>D</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.88582</td>\n",
       "      <td>D</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.999882</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D</td>\n",
       "      <td>simple_aae</td>\n",
       "      <td>C354W</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1.98</td>\n",
       "      <td>0.85247</td>\n",
       "      <td>D</td>\n",
       "      <td>-6.84</td>\n",
       "      <td>0.94249</td>\n",
       "      <td>D</td>\n",
       "      <td>0.751</td>\n",
       "      <td>0.75466</td>\n",
       "      <td>0.3131</td>\n",
       "      <td>0.87764</td>\n",
       "      <td>D</td>\n",
       "      <td>0.8180</td>\n",
       "      <td>0.93873</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.020603</td>\n",
       "      <td>0.43227</td>\n",
       "      <td>T</td>\n",
       "      <td>0.437</td>\n",
       "      <td>0.74446</td>\n",
       "      <td>0.868</td>\n",
       "      <td>0.96212</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>C354W</td>\n",
       "      <td>Loss of methylation at K350 (P = 0.0838)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.76919400692</td>\n",
       "      <td>0.77304</td>\n",
       "      <td>T</td>\n",
       "      <td>0.851523</td>\n",
       "      <td>0.15717</td>\n",
       "      <td>0.19276</td>\n",
       "      <td>0.20597</td>\n",
       "      <td>N</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.422079</td>\n",
       "      <td>0.48823</td>\n",
       "      <td>N</td>\n",
       "      <td>0.0406181179914726</td>\n",
       "      <td>0.43713</td>\n",
       "      <td>2.658815</td>\n",
       "      <td>-0.363070660159128</td>\n",
       "      <td>0.26108</td>\n",
       "      <td>1.439964</td>\n",
       "      <td>0.470199</td>\n",
       "      <td>0.053333</td>\n",
       "      <td>0.084</td>\n",
       "      <td>0.084</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.094</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.877</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>2.381671</td>\n",
       "      <td>22.30</td>\n",
       "      <td>0.470199</td>\n",
       "      <td>0.053333</td>\n",
       "      <td>0.075</td>\n",
       "      <td>0.075</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-0.918</td>\n",
       "      <td>-0.893</td>\n",
       "      <td>0.492</td>\n",
       "      <td>8890.65</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.7</td>\n",
       "      <td>13.7</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>944.0</td>\n",
       "      <td>2.794161</td>\n",
       "      <td>22.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>10_000093271_C_A</td>\n",
       "      <td>10_000047331_C_A</td>\n",
       "      <td>Q3ZCM7_C354</td>\n",
       "      <td>354</td>\n",
       "      <td>C354</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>2</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>C</td>\n",
       "      <td>F</td>\n",
       "      <td>Cys/Phe</td>\n",
       "      <td>22.6</td>\n",
       "      <td>21.80</td>\n",
       "      <td>0.80</td>\n",
       "      <td>5.37</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000568584</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>0.78490</td>\n",
       "      <td>.</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.66756</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.90584</td>\n",
       "      <td>D</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.88582</td>\n",
       "      <td>D</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.986821</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D</td>\n",
       "      <td>simple_aae</td>\n",
       "      <td>C354F</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1.95</td>\n",
       "      <td>0.85003</td>\n",
       "      <td>D</td>\n",
       "      <td>-6.83</td>\n",
       "      <td>0.94223</td>\n",
       "      <td>D</td>\n",
       "      <td>0.782</td>\n",
       "      <td>0.77883</td>\n",
       "      <td>0.4271</td>\n",
       "      <td>0.89500</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7835</td>\n",
       "      <td>0.92649</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.015962</td>\n",
       "      <td>0.36975</td>\n",
       "      <td>T</td>\n",
       "      <td>0.374</td>\n",
       "      <td>0.69594</td>\n",
       "      <td>0.808</td>\n",
       "      <td>0.92509</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>C354F</td>\n",
       "      <td>Gain of catalytic residue at C354 (P = 0.1094)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.727896571159</td>\n",
       "      <td>0.71199</td>\n",
       "      <td>T</td>\n",
       "      <td>0.895918</td>\n",
       "      <td>0.18929</td>\n",
       "      <td>0.57914</td>\n",
       "      <td>0.30502</td>\n",
       "      <td>D</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.537556</td>\n",
       "      <td>0.55531</td>\n",
       "      <td>D</td>\n",
       "      <td>0.0988420731411684</td>\n",
       "      <td>0.46409</td>\n",
       "      <td>2.88164</td>\n",
       "      <td>-0.263815814260368</td>\n",
       "      <td>0.29300</td>\n",
       "      <td>1.640415</td>\n",
       "      <td>0.463576</td>\n",
       "      <td>0.053333</td>\n",
       "      <td>0.083</td>\n",
       "      <td>0.082</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.109</td>\n",
       "      <td>3.127</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>2.267094</td>\n",
       "      <td>21.80</td>\n",
       "      <td>0.463576</td>\n",
       "      <td>0.053333</td>\n",
       "      <td>0.078</td>\n",
       "      <td>0.077</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.247</td>\n",
       "      <td>0.249</td>\n",
       "      <td>5.260</td>\n",
       "      <td>8890.65</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>945.0</td>\n",
       "      <td>2.814052</td>\n",
       "      <td>22.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>10_000093271_C_G</td>\n",
       "      <td>10_000047331_C_G</td>\n",
       "      <td>Q3ZCM7_C354</td>\n",
       "      <td>354</td>\n",
       "      <td>C354</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>2</td>\n",
       "      <td>{12: 'C', 239: 'C', 303: 'C', 354: 'C'}</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>Cys/Ser</td>\n",
       "      <td>22.4</td>\n",
       "      <td>19.67</td>\n",
       "      <td>2.73</td>\n",
       "      <td>5.37</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000568584</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>0.91255</td>\n",
       "      <td>.</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.74150</td>\n",
       "      <td>D</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.70673</td>\n",
       "      <td>D</td>\n",
       "      <td>0.968</td>\n",
       "      <td>0.71741</td>\n",
       "      <td>D</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.930492</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D</td>\n",
       "      <td>simple_aae</td>\n",
       "      <td>C354S</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1.87</td>\n",
       "      <td>0.84415</td>\n",
       "      <td>D</td>\n",
       "      <td>-6.21</td>\n",
       "      <td>0.91827</td>\n",
       "      <td>D</td>\n",
       "      <td>0.739</td>\n",
       "      <td>0.74735</td>\n",
       "      <td>0.4621</td>\n",
       "      <td>0.90007</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7769</td>\n",
       "      <td>0.92419</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.022348</td>\n",
       "      <td>0.45230</td>\n",
       "      <td>T</td>\n",
       "      <td>0.377</td>\n",
       "      <td>0.69846</td>\n",
       "      <td>0.781</td>\n",
       "      <td>0.90537</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>C354S</td>\n",
       "      <td>Loss of sheet (P = 0.0817)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.713165700436</td>\n",
       "      <td>0.69052</td>\n",
       "      <td>T</td>\n",
       "      <td>0.799045</td>\n",
       "      <td>0.12954</td>\n",
       "      <td>0.59480</td>\n",
       "      <td>0.30927</td>\n",
       "      <td>D</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.537556</td>\n",
       "      <td>0.55531</td>\n",
       "      <td>D</td>\n",
       "      <td>0.0966724300357427</td>\n",
       "      <td>0.46307</td>\n",
       "      <td>2.873107</td>\n",
       "      <td>-0.265973651554631</td>\n",
       "      <td>0.29227</td>\n",
       "      <td>1.635795</td>\n",
       "      <td>0.463576</td>\n",
       "      <td>0.053333</td>\n",
       "      <td>0.083</td>\n",
       "      <td>0.082</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.109</td>\n",
       "      <td>3.127</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>2.045093</td>\n",
       "      <td>19.67</td>\n",
       "      <td>0.463576</td>\n",
       "      <td>0.053333</td>\n",
       "      <td>0.078</td>\n",
       "      <td>0.077</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.247</td>\n",
       "      <td>0.249</td>\n",
       "      <td>5.260</td>\n",
       "      <td>8890.65</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>945.0</td>\n",
       "      <td>2.697199</td>\n",
       "      <td>22.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos_id19          pos_id38       pos_ID  matched_aapos  \\\n",
       "0  10_000093270_A_C  10_000047330_A_C  Q3ZCM7_C354            354   \n",
       "1  10_000093271_C_A  10_000047331_C_A  Q3ZCM7_C354            354   \n",
       "2  10_000093271_C_G  10_000047331_C_G  Q3ZCM7_C354            354   \n",
       "\n",
       "  matched_target matched_UKBID  matched_index  \\\n",
       "0           C354        Q3ZCM7              2   \n",
       "1           C354        Q3ZCM7              2   \n",
       "2           C354        Q3ZCM7              2   \n",
       "\n",
       "                                  pos_dict aaref aaalt Amino_acids  \\\n",
       "0  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}     C     W     Cys/Trp   \n",
       "1  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}     C     F     Cys/Phe   \n",
       "2  {12: 'C', 239: 'C', 303: 'C', 354: 'C'}     C     S     Cys/Ser   \n",
       "\n",
       "   CADD_phred_hg38  CADD_phred_hg19  CADDdiff_38minus19  Cys_reactivity  \\\n",
       "0             22.6            22.30                0.30            5.37   \n",
       "1             22.6            21.80                0.80            5.37   \n",
       "2             22.4            19.67                2.73            5.37   \n",
       "\n",
       "  Cys_react_threshold Cys_target_label  Lys_reactivity Lys_react_threshold  \\\n",
       "0                 Low              NaN             NaN                 NaN   \n",
       "1                 Low              NaN             NaN                 NaN   \n",
       "2                 Low              NaN             NaN                 NaN   \n",
       "\n",
       "  Lys_target_label rs_dbSNP151 Ensembl_transcriptid cds_strand refcodon  \\\n",
       "0              NaN           .      ENST00000568584          -      TGT   \n",
       "1              NaN           .      ENST00000568584          -      TGT   \n",
       "2              NaN           .      ENST00000568584          -      TGT   \n",
       "\n",
       "   codonpos  codon_degeneracy SIFT_score SIFT_converted_rankscore SIFT_pred  \\\n",
       "0         3                 2          .                  0.91255         .   \n",
       "1         2                 0          .                  0.78490         .   \n",
       "2         2                 0          .                  0.91255         .   \n",
       "\n",
       "  SIFT4G_score SIFT4G_converted_rankscore SIFT4G_pred Polyphen2_HDIV_score  \\\n",
       "0          0.0                    0.92824           D                  1.0   \n",
       "1        0.013                    0.66756           D                  1.0   \n",
       "2        0.026                    0.74150           D                0.997   \n",
       "\n",
       "  Polyphen2_HDIV_rankscore Polyphen2_HDIV_pred Polyphen2_HVAR_score  \\\n",
       "0                  0.90584                   D                0.998   \n",
       "1                  0.90584                   D                0.998   \n",
       "2                  0.70673                   D                0.968   \n",
       "\n",
       "  Polyphen2_HVAR_rankscore Polyphen2_HVAR_pred LRT_score  \\\n",
       "0                  0.88582                   D  0.000003   \n",
       "1                  0.88582                   D  0.000003   \n",
       "2                  0.71741                   D  0.000003   \n",
       "\n",
       "  LRT_converted_rankscore LRT_pred LRT_Omega MutationTaster_score  \\\n",
       "0                 0.62929        U  0.000000             0.999882   \n",
       "1                 0.62929        U  0.000000             0.986821   \n",
       "2                 0.62929        U  0.000000             0.930492   \n",
       "\n",
       "  MutationTaster_converted_rankscore MutationTaster_pred MutationTaster_model  \\\n",
       "0                            0.81001                   D           simple_aae   \n",
       "1                            0.81001                   D           simple_aae   \n",
       "2                            0.81001                   D           simple_aae   \n",
       "\n",
       "  MutationTaster_AAE MutationAssessor_score MutationAssessor_rankscore  \\\n",
       "0              C354W                      .                          .   \n",
       "1              C354F                      .                          .   \n",
       "2              C354S                      .                          .   \n",
       "\n",
       "  MutationAssessor_pred FATHMM_score FATHMM_converted_rankscore FATHMM_pred  \\\n",
       "0                     .        -1.98                    0.85247           D   \n",
       "1                     .        -1.95                    0.85003           D   \n",
       "2                     .        -1.87                    0.84415           D   \n",
       "\n",
       "  PROVEAN_score PROVEAN_converted_rankscore PROVEAN_pred VEST4_score  \\\n",
       "0         -6.84                     0.94249            D       0.751   \n",
       "1         -6.83                     0.94223            D       0.782   \n",
       "2         -6.21                     0.91827            D       0.739   \n",
       "\n",
       "  VEST4_rankscore MetaSVM_score MetaSVM_rankscore MetaSVM_pred MetaLR_score  \\\n",
       "0         0.75466        0.3131           0.87764            D       0.8180   \n",
       "1         0.77883        0.4271           0.89500            D       0.7835   \n",
       "2         0.74735        0.4621           0.90007            D       0.7769   \n",
       "\n",
       "  MetaLR_rankscore MetaLR_pred Reliability_index M-CAP_score M-CAP_rankscore  \\\n",
       "0          0.93873           D                 9    0.020603         0.43227   \n",
       "1          0.92649           D                 9    0.015962         0.36975   \n",
       "2          0.92419           D                 9    0.022348         0.45230   \n",
       "\n",
       "  M-CAP_pred REVEL_score REVEL_rankscore MutPred_score MutPred_rankscore  \\\n",
       "0          T       0.437         0.74446         0.868           0.96212   \n",
       "1          T       0.374         0.69594         0.808           0.92509   \n",
       "2          T       0.377         0.69846         0.781           0.90537   \n",
       "\n",
       "  MutPred_protID MutPred_AAchange  \\\n",
       "0         Q3ZCM7            C354W   \n",
       "1         Q3ZCM7            C354F   \n",
       "2         Q3ZCM7            C354S   \n",
       "\n",
       "                              MutPred_Top5features MPC_score MPC_rankscore  \\\n",
       "0         Loss of methylation at K350 (P = 0.0838)         .             .   \n",
       "1   Gain of catalytic residue at C354 (P = 0.1094)         .             .   \n",
       "2                       Loss of sheet (P = 0.0817)         .             .   \n",
       "\n",
       "  PrimateAI_score PrimateAI_rankscore PrimateAI_pred  DANN_score  \\\n",
       "0   0.76919400692             0.77304              T    0.851523   \n",
       "1  0.727896571159             0.71199              T    0.895918   \n",
       "2  0.713165700436             0.69052              T    0.799045   \n",
       "\n",
       "   DANN_rankscore  fathmm-MKL_coding_score  fathmm-MKL_coding_rankscore  \\\n",
       "0         0.15717                  0.19276                      0.20597   \n",
       "1         0.18929                  0.57914                      0.30502   \n",
       "2         0.12954                  0.59480                      0.30927   \n",
       "\n",
       "  fathmm-MKL_coding_pred fathmm-MKL_coding_group fathmm-XF_coding_score  \\\n",
       "0                      N                   AEFBI               0.422079   \n",
       "1                      D                   AEFBI               0.537556   \n",
       "2                      D                   AEFBI               0.537556   \n",
       "\n",
       "  fathmm-XF_coding_rankscore fathmm-XF_coding_pred    Eigen-raw_coding  \\\n",
       "0                    0.48823                     N  0.0406181179914726   \n",
       "1                    0.55531                     D  0.0988420731411684   \n",
       "2                    0.55531                     D  0.0966724300357427   \n",
       "\n",
       "  Eigen-raw_coding_rankscore Eigen-pred_coding Eigen-PC-raw_coding  \\\n",
       "0                    0.43713          2.658815  -0.363070660159128   \n",
       "1                    0.46409           2.88164  -0.263815814260368   \n",
       "2                    0.46307          2.873107  -0.265973651554631   \n",
       "\n",
       "  Eigen-PC-raw_coding_rankscore Eigen-PC-phred_coding   GC_hg19  CpG_hg19  \\\n",
       "0                       0.26108              1.439964  0.470199  0.053333   \n",
       "1                       0.29300              1.640415  0.463576  0.053333   \n",
       "2                       0.29227              1.635795  0.463576  0.053333   \n",
       "\n",
       "   priPhCons_hg19  mamPhCons_hg19  verPhCons_hg19  priPhyloP_hg19  \\\n",
       "0           0.084           0.084             1.0           0.094   \n",
       "1           0.083           0.082             1.0           0.110   \n",
       "2           0.083           0.082             1.0           0.110   \n",
       "\n",
       "   mamPhyloP_hg19  verPhyloP_hg19  GerpRS_hg19  GerpRSpval_hg19  GerpN_hg19  \\\n",
       "0           0.095           0.877          NaN              NaN         0.0   \n",
       "1           0.109           3.127          NaN              NaN         0.0   \n",
       "2           0.109           3.127          NaN              NaN         0.0   \n",
       "\n",
       "   GerpS_hg19  Dist2Mutation_hg19  Freq100bp_hg19  Rare100bp_hg19  \\\n",
       "0         0.0             38936.0             NaN             NaN   \n",
       "1         0.0             38936.0             NaN             NaN   \n",
       "2         0.0             38936.0             NaN             NaN   \n",
       "\n",
       "   Sngl100bp_hg19  Freq1000bp_hg19  Rare1000bp_hg19  Sngl1000bp_hg19  \\\n",
       "0             NaN              NaN              NaN              NaN   \n",
       "1             NaN              NaN              NaN              NaN   \n",
       "2             NaN              NaN              NaN              NaN   \n",
       "\n",
       "   Freq10000bp_hg19  Rare10000bp_hg19  Sngl10000bp_hg19  RawScore_hg19  \\\n",
       "0               5.0              24.0             137.0       2.381671   \n",
       "1               5.0              24.0             137.0       2.267094   \n",
       "2               5.0              24.0             137.0       2.045093   \n",
       "\n",
       "   PHRED_hg19   GC_hg38  CpG_hg38  priPhCons_hg38  mamPhCons_hg38  \\\n",
       "0       22.30  0.470199  0.053333           0.075           0.075   \n",
       "1       21.80  0.463576  0.053333           0.078           0.077   \n",
       "2       19.67  0.463576  0.053333           0.078           0.077   \n",
       "\n",
       "   verPhCons_hg38  priPhyloP_hg38  mamPhyloP_hg38  verPhyloP_hg38  \\\n",
       "0           0.999          -0.918          -0.893           0.492   \n",
       "1           1.000           0.247           0.249           5.260   \n",
       "2           1.000           0.247           0.249           5.260   \n",
       "\n",
       "   GerpRS_hg38  GerpRSpval_hg38  GerpN_hg38  GerpS_hg38  Dist2Mutation_hg38  \\\n",
       "0      8890.65              0.0        13.7        13.7                12.0   \n",
       "1      8890.65              0.0         NaN         NaN                 6.0   \n",
       "2      8890.65              0.0         NaN         NaN                 6.0   \n",
       "\n",
       "   Freq100bp_hg38  Rare100bp_hg38  Sngl100bp_hg38  Freq1000bp_hg38  \\\n",
       "0             0.0             0.0            12.0              0.0   \n",
       "1             0.0             0.0            12.0              0.0   \n",
       "2             0.0             0.0            12.0              0.0   \n",
       "\n",
       "   Rare1000bp_hg38  Sngl1000bp_hg38  Freq10000bp_hg38  Rare10000bp_hg38  \\\n",
       "0              0.0             95.0              12.0              31.0   \n",
       "1              0.0             95.0              12.0              31.0   \n",
       "2              0.0             95.0              12.0              31.0   \n",
       "\n",
       "   Sngl10000bp_hg38  RawScore_hg38  PHRED_hg38  \n",
       "0             944.0       2.794161        22.6  \n",
       "1             945.0       2.814052        22.6  \n",
       "2             945.0       2.697199        22.4  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergeDET.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [6c] MERGING CADD37 and 38 NOT_DETECTED with dbNSFP not_detected df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "mernot37 = pd.merge(notdbnsfp, not37, how='inner', on=['pos_id19'])\n",
    "\n",
    "mergeNOT = pd.merge(mernot37, not38, how='inner', on=['pos_id38'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos_id19</th>\n",
       "      <th>pos_id38</th>\n",
       "      <th>pos_ID</th>\n",
       "      <th>matched_UKBID</th>\n",
       "      <th>matched_aapos</th>\n",
       "      <th>matched_index</th>\n",
       "      <th>Ensembl_transcriptid</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>aaref</th>\n",
       "      <th>aaalt</th>\n",
       "      <th>CADD_phred_hg38</th>\n",
       "      <th>CADD_phred_hg19</th>\n",
       "      <th>CADDdiff_38minus19</th>\n",
       "      <th>rs_dbSNP151</th>\n",
       "      <th>genename</th>\n",
       "      <th>cds_strand</th>\n",
       "      <th>refcodon</th>\n",
       "      <th>codonpos</th>\n",
       "      <th>codon_degeneracy</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>SIFT_converted_rankscore</th>\n",
       "      <th>SIFT_pred</th>\n",
       "      <th>SIFT4G_score</th>\n",
       "      <th>SIFT4G_converted_rankscore</th>\n",
       "      <th>SIFT4G_pred</th>\n",
       "      <th>Polyphen2_HDIV_score</th>\n",
       "      <th>Polyphen2_HDIV_rankscore</th>\n",
       "      <th>Polyphen2_HDIV_pred</th>\n",
       "      <th>Polyphen2_HVAR_score</th>\n",
       "      <th>Polyphen2_HVAR_rankscore</th>\n",
       "      <th>Polyphen2_HVAR_pred</th>\n",
       "      <th>LRT_score</th>\n",
       "      <th>LRT_converted_rankscore</th>\n",
       "      <th>LRT_pred</th>\n",
       "      <th>LRT_Omega</th>\n",
       "      <th>MutationTaster_score</th>\n",
       "      <th>MutationTaster_converted_rankscore</th>\n",
       "      <th>MutationTaster_pred</th>\n",
       "      <th>MutationTaster_model</th>\n",
       "      <th>MutationTaster_AAE</th>\n",
       "      <th>MutationAssessor_score</th>\n",
       "      <th>MutationAssessor_rankscore</th>\n",
       "      <th>MutationAssessor_pred</th>\n",
       "      <th>FATHMM_score</th>\n",
       "      <th>FATHMM_converted_rankscore</th>\n",
       "      <th>FATHMM_pred</th>\n",
       "      <th>PROVEAN_score</th>\n",
       "      <th>PROVEAN_converted_rankscore</th>\n",
       "      <th>PROVEAN_pred</th>\n",
       "      <th>VEST4_score</th>\n",
       "      <th>VEST4_rankscore</th>\n",
       "      <th>MetaSVM_score</th>\n",
       "      <th>MetaSVM_rankscore</th>\n",
       "      <th>MetaSVM_pred</th>\n",
       "      <th>MetaLR_score</th>\n",
       "      <th>MetaLR_rankscore</th>\n",
       "      <th>MetaLR_pred</th>\n",
       "      <th>Reliability_index</th>\n",
       "      <th>M-CAP_score</th>\n",
       "      <th>M-CAP_rankscore</th>\n",
       "      <th>M-CAP_pred</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>REVEL_rankscore</th>\n",
       "      <th>MutPred_score</th>\n",
       "      <th>MutPred_rankscore</th>\n",
       "      <th>MutPred_protID</th>\n",
       "      <th>MutPred_AAchange</th>\n",
       "      <th>MutPred_Top5features</th>\n",
       "      <th>MPC_score</th>\n",
       "      <th>MPC_rankscore</th>\n",
       "      <th>PrimateAI_score</th>\n",
       "      <th>PrimateAI_rankscore</th>\n",
       "      <th>PrimateAI_pred</th>\n",
       "      <th>DANN_score</th>\n",
       "      <th>DANN_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_score</th>\n",
       "      <th>fathmm-MKL_coding_rankscore</th>\n",
       "      <th>fathmm-MKL_coding_pred</th>\n",
       "      <th>fathmm-MKL_coding_group</th>\n",
       "      <th>fathmm-XF_coding_score</th>\n",
       "      <th>fathmm-XF_coding_rankscore</th>\n",
       "      <th>fathmm-XF_coding_pred</th>\n",
       "      <th>Eigen-raw_coding</th>\n",
       "      <th>Eigen-raw_coding_rankscore</th>\n",
       "      <th>Eigen-pred_coding</th>\n",
       "      <th>Eigen-PC-raw_coding</th>\n",
       "      <th>Eigen-PC-raw_coding_rankscore</th>\n",
       "      <th>Eigen-PC-phred_coding</th>\n",
       "      <th>GC_hg19</th>\n",
       "      <th>CpG_hg19</th>\n",
       "      <th>priPhCons_hg19</th>\n",
       "      <th>mamPhCons_hg19</th>\n",
       "      <th>verPhCons_hg19</th>\n",
       "      <th>priPhyloP_hg19</th>\n",
       "      <th>mamPhyloP_hg19</th>\n",
       "      <th>verPhyloP_hg19</th>\n",
       "      <th>GerpRS_hg19</th>\n",
       "      <th>GerpRSpval_hg19</th>\n",
       "      <th>GerpN_hg19</th>\n",
       "      <th>GerpS_hg19</th>\n",
       "      <th>Dist2Mutation_hg19</th>\n",
       "      <th>Freq100bp_hg19</th>\n",
       "      <th>Rare100bp_hg19</th>\n",
       "      <th>Sngl100bp_hg19</th>\n",
       "      <th>Freq1000bp_hg19</th>\n",
       "      <th>Rare1000bp_hg19</th>\n",
       "      <th>Sngl1000bp_hg19</th>\n",
       "      <th>Freq10000bp_hg19</th>\n",
       "      <th>Rare10000bp_hg19</th>\n",
       "      <th>Sngl10000bp_hg19</th>\n",
       "      <th>RawScore_hg19</th>\n",
       "      <th>PHRED_hg19</th>\n",
       "      <th>GC_hg38</th>\n",
       "      <th>CpG_hg38</th>\n",
       "      <th>priPhCons_hg38</th>\n",
       "      <th>mamPhCons_hg38</th>\n",
       "      <th>verPhCons_hg38</th>\n",
       "      <th>priPhyloP_hg38</th>\n",
       "      <th>mamPhyloP_hg38</th>\n",
       "      <th>verPhyloP_hg38</th>\n",
       "      <th>GerpRS_hg38</th>\n",
       "      <th>GerpRSpval_hg38</th>\n",
       "      <th>GerpN_hg38</th>\n",
       "      <th>GerpS_hg38</th>\n",
       "      <th>Dist2Mutation_hg38</th>\n",
       "      <th>Freq100bp_hg38</th>\n",
       "      <th>Rare100bp_hg38</th>\n",
       "      <th>Sngl100bp_hg38</th>\n",
       "      <th>Freq1000bp_hg38</th>\n",
       "      <th>Rare1000bp_hg38</th>\n",
       "      <th>Sngl1000bp_hg38</th>\n",
       "      <th>Freq10000bp_hg38</th>\n",
       "      <th>Rare10000bp_hg38</th>\n",
       "      <th>Sngl10000bp_hg38</th>\n",
       "      <th>RawScore_hg38</th>\n",
       "      <th>PHRED_hg38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>10_000093156_C_A</td>\n",
       "      <td>10_000047216_C_A</td>\n",
       "      <td>Q3ZCM7_K392</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>392</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000568584</td>\n",
       "      <td>Lys/Asn</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>23.0</td>\n",
       "      <td>22.9</td>\n",
       "      <td>0.1</td>\n",
       "      <td>.</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>AAG</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>0.91255</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.92824</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.90584</td>\n",
       "      <td>D</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.9235899999999999</td>\n",
       "      <td>D</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.654618</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D</td>\n",
       "      <td>simple_aae</td>\n",
       "      <td>K392N</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1.98</td>\n",
       "      <td>0.85247</td>\n",
       "      <td>D</td>\n",
       "      <td>-2.6</td>\n",
       "      <td>0.58733</td>\n",
       "      <td>D</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.6007100000000001</td>\n",
       "      <td>0.3991</td>\n",
       "      <td>0.8908299999999999</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7111</td>\n",
       "      <td>0.90075</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.009489</td>\n",
       "      <td>0.24855</td>\n",
       "      <td>T</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.78035</td>\n",
       "      <td>0.718</td>\n",
       "      <td>0.85313</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>K392N</td>\n",
       "      <td>Loss of MoRF binding (P = 0.0523)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.751451253891</td>\n",
       "      <td>0.74660</td>\n",
       "      <td>T</td>\n",
       "      <td>0.984728</td>\n",
       "      <td>0.41873</td>\n",
       "      <td>0.38681</td>\n",
       "      <td>0.26043</td>\n",
       "      <td>N</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.295876</td>\n",
       "      <td>0.40587</td>\n",
       "      <td>N</td>\n",
       "      <td>0.0676056746461709</td>\n",
       "      <td>0.44955</td>\n",
       "      <td>2.76036</td>\n",
       "      <td>-0.20748312943205102</td>\n",
       "      <td>0.31266</td>\n",
       "      <td>1.767854</td>\n",
       "      <td>0.536424</td>\n",
       "      <td>0.106667</td>\n",
       "      <td>0.550</td>\n",
       "      <td>0.546</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.109</td>\n",
       "      <td>1.700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>2.636298</td>\n",
       "      <td>22.9</td>\n",
       "      <td>0.536424</td>\n",
       "      <td>0.106667</td>\n",
       "      <td>0.046</td>\n",
       "      <td>0.046</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.247</td>\n",
       "      <td>0.249</td>\n",
       "      <td>2.730</td>\n",
       "      <td>8890.65</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.7</td>\n",
       "      <td>13.7</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>930.0</td>\n",
       "      <td>3.029607</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>10_000093156_C_G</td>\n",
       "      <td>10_000047216_C_G</td>\n",
       "      <td>Q3ZCM7_K392</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>392</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000568584</td>\n",
       "      <td>Lys/Asn</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>23.1</td>\n",
       "      <td>22.8</td>\n",
       "      <td>0.3</td>\n",
       "      <td>.</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>AAG</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>0.91255</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.92824</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.90584</td>\n",
       "      <td>D</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.9235899999999999</td>\n",
       "      <td>D</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.654618</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D</td>\n",
       "      <td>simple_aae</td>\n",
       "      <td>K392N</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1.98</td>\n",
       "      <td>0.85247</td>\n",
       "      <td>D</td>\n",
       "      <td>-2.6</td>\n",
       "      <td>0.58733</td>\n",
       "      <td>D</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.6007100000000001</td>\n",
       "      <td>0.3991</td>\n",
       "      <td>0.8908299999999999</td>\n",
       "      <td>D</td>\n",
       "      <td>0.7111</td>\n",
       "      <td>0.90075</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.008795</td>\n",
       "      <td>0.23196</td>\n",
       "      <td>T</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.78035</td>\n",
       "      <td>0.718</td>\n",
       "      <td>0.85313</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>K392N</td>\n",
       "      <td>Loss of MoRF binding (P = 0.0523)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.751451253891</td>\n",
       "      <td>0.74660</td>\n",
       "      <td>T</td>\n",
       "      <td>0.984663</td>\n",
       "      <td>0.41803</td>\n",
       "      <td>0.40406</td>\n",
       "      <td>0.26429</td>\n",
       "      <td>N</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.302556</td>\n",
       "      <td>0.41077</td>\n",
       "      <td>N</td>\n",
       "      <td>0.0676056746461709</td>\n",
       "      <td>0.44955</td>\n",
       "      <td>2.76036</td>\n",
       "      <td>-0.20748312943205102</td>\n",
       "      <td>0.31266</td>\n",
       "      <td>1.767854</td>\n",
       "      <td>0.536424</td>\n",
       "      <td>0.106667</td>\n",
       "      <td>0.550</td>\n",
       "      <td>0.546</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.109</td>\n",
       "      <td>1.700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>2.627006</td>\n",
       "      <td>22.8</td>\n",
       "      <td>0.536424</td>\n",
       "      <td>0.106667</td>\n",
       "      <td>0.046</td>\n",
       "      <td>0.046</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.247</td>\n",
       "      <td>0.249</td>\n",
       "      <td>2.730</td>\n",
       "      <td>8890.65</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.7</td>\n",
       "      <td>13.7</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>930.0</td>\n",
       "      <td>3.055876</td>\n",
       "      <td>23.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>10_000093157_T_A</td>\n",
       "      <td>10_000047217_T_A</td>\n",
       "      <td>Q3ZCM7_K392</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>392</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000568584</td>\n",
       "      <td>Lys/Met</td>\n",
       "      <td>K</td>\n",
       "      <td>M</td>\n",
       "      <td>23.2</td>\n",
       "      <td>23.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>.</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>AAG</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>0.91255</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.92824</td>\n",
       "      <td>D</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.90584</td>\n",
       "      <td>D</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.9235899999999999</td>\n",
       "      <td>D</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.62929</td>\n",
       "      <td>U</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.820833</td>\n",
       "      <td>0.81001</td>\n",
       "      <td>D</td>\n",
       "      <td>simple_aae</td>\n",
       "      <td>K392M</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-2.08</td>\n",
       "      <td>0.86010</td>\n",
       "      <td>D</td>\n",
       "      <td>-3.14</td>\n",
       "      <td>0.66549</td>\n",
       "      <td>D</td>\n",
       "      <td>0.568</td>\n",
       "      <td>0.62442</td>\n",
       "      <td>0.5527</td>\n",
       "      <td>0.9127</td>\n",
       "      <td>D</td>\n",
       "      <td>0.8388</td>\n",
       "      <td>0.9461</td>\n",
       "      <td>D</td>\n",
       "      <td>9</td>\n",
       "      <td>0.01492</td>\n",
       "      <td>0.35345</td>\n",
       "      <td>T</td>\n",
       "      <td>0.505</td>\n",
       "      <td>0.78906</td>\n",
       "      <td>0.698</td>\n",
       "      <td>0.83491</td>\n",
       "      <td>Q3ZCM7</td>\n",
       "      <td>K392M</td>\n",
       "      <td>Loss of ubiquitination at K392 (P = 0.0326)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.727143764496</td>\n",
       "      <td>0.71088</td>\n",
       "      <td>T</td>\n",
       "      <td>0.964042</td>\n",
       "      <td>0.29699</td>\n",
       "      <td>0.52831</td>\n",
       "      <td>0.29225</td>\n",
       "      <td>D</td>\n",
       "      <td>AEFBI</td>\n",
       "      <td>0.364324</td>\n",
       "      <td>0.45291</td>\n",
       "      <td>N</td>\n",
       "      <td>0.31076238698785996</td>\n",
       "      <td>0.56701</td>\n",
       "      <td>3.8352209999999998</td>\n",
       "      <td>-0.0235518622390073</td>\n",
       "      <td>0.38654</td>\n",
       "      <td>2.280427</td>\n",
       "      <td>0.529801</td>\n",
       "      <td>0.106667</td>\n",
       "      <td>0.546</td>\n",
       "      <td>0.542</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.094</td>\n",
       "      <td>0.095</td>\n",
       "      <td>2.564</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>2.819085</td>\n",
       "      <td>23.2</td>\n",
       "      <td>0.529801</td>\n",
       "      <td>0.106667</td>\n",
       "      <td>0.042</td>\n",
       "      <td>0.042</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.274</td>\n",
       "      <td>0.275</td>\n",
       "      <td>6.207</td>\n",
       "      <td>8890.65</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>930.0</td>\n",
       "      <td>3.100150</td>\n",
       "      <td>23.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos_id19          pos_id38       pos_ID matched_UKBID  \\\n",
       "0  10_000093156_C_A  10_000047216_C_A  Q3ZCM7_K392        Q3ZCM7   \n",
       "1  10_000093156_C_G  10_000047216_C_G  Q3ZCM7_K392        Q3ZCM7   \n",
       "2  10_000093157_T_A  10_000047217_T_A  Q3ZCM7_K392        Q3ZCM7   \n",
       "\n",
       "   matched_aapos  matched_index Ensembl_transcriptid Amino_acids aaref aaalt  \\\n",
       "0            392              2      ENST00000568584     Lys/Asn     K     N   \n",
       "1            392              2      ENST00000568584     Lys/Asn     K     N   \n",
       "2            392              2      ENST00000568584     Lys/Met     K     M   \n",
       "\n",
       "   CADD_phred_hg38  CADD_phred_hg19  CADDdiff_38minus19 rs_dbSNP151 genename  \\\n",
       "0             23.0             22.9                 0.1           .    TUBB8   \n",
       "1             23.1             22.8                 0.3           .    TUBB8   \n",
       "2             23.2             23.2                 0.0           .    TUBB8   \n",
       "\n",
       "  cds_strand refcodon  codonpos  codon_degeneracy SIFT_score  \\\n",
       "0          -      AAG         3                 2          .   \n",
       "1          -      AAG         3                 2          .   \n",
       "2          -      AAG         2                 0          .   \n",
       "\n",
       "  SIFT_converted_rankscore SIFT_pred SIFT4G_score SIFT4G_converted_rankscore  \\\n",
       "0                  0.91255         .          0.0                    0.92824   \n",
       "1                  0.91255         .          0.0                    0.92824   \n",
       "2                  0.91255         .          0.0                    0.92824   \n",
       "\n",
       "  SIFT4G_pred Polyphen2_HDIV_score Polyphen2_HDIV_rankscore  \\\n",
       "0           D                  1.0                  0.90584   \n",
       "1           D                  1.0                  0.90584   \n",
       "2           D                  1.0                  0.90584   \n",
       "\n",
       "  Polyphen2_HDIV_pred Polyphen2_HVAR_score Polyphen2_HVAR_rankscore  \\\n",
       "0                   D                0.999       0.9235899999999999   \n",
       "1                   D                0.999       0.9235899999999999   \n",
       "2                   D                0.999       0.9235899999999999   \n",
       "\n",
       "  Polyphen2_HVAR_pred LRT_score LRT_converted_rankscore LRT_pred LRT_Omega  \\\n",
       "0                   D  0.000010                 0.62929        U  0.000000   \n",
       "1                   D  0.000010                 0.62929        U  0.000000   \n",
       "2                   D  0.000010                 0.62929        U  0.000000   \n",
       "\n",
       "  MutationTaster_score MutationTaster_converted_rankscore MutationTaster_pred  \\\n",
       "0             0.654618                            0.81001                   D   \n",
       "1             0.654618                            0.81001                   D   \n",
       "2             0.820833                            0.81001                   D   \n",
       "\n",
       "  MutationTaster_model MutationTaster_AAE MutationAssessor_score  \\\n",
       "0           simple_aae              K392N                      .   \n",
       "1           simple_aae              K392N                      .   \n",
       "2           simple_aae              K392M                      .   \n",
       "\n",
       "  MutationAssessor_rankscore MutationAssessor_pred FATHMM_score  \\\n",
       "0                          .                     .        -1.98   \n",
       "1                          .                     .        -1.98   \n",
       "2                          .                     .        -2.08   \n",
       "\n",
       "  FATHMM_converted_rankscore FATHMM_pred PROVEAN_score  \\\n",
       "0                    0.85247           D          -2.6   \n",
       "1                    0.85247           D          -2.6   \n",
       "2                    0.86010           D         -3.14   \n",
       "\n",
       "  PROVEAN_converted_rankscore PROVEAN_pred VEST4_score     VEST4_rankscore  \\\n",
       "0                     0.58733            D        0.53  0.6007100000000001   \n",
       "1                     0.58733            D        0.53  0.6007100000000001   \n",
       "2                     0.66549            D       0.568             0.62442   \n",
       "\n",
       "  MetaSVM_score   MetaSVM_rankscore MetaSVM_pred MetaLR_score  \\\n",
       "0        0.3991  0.8908299999999999            D       0.7111   \n",
       "1        0.3991  0.8908299999999999            D       0.7111   \n",
       "2        0.5527              0.9127            D       0.8388   \n",
       "\n",
       "  MetaLR_rankscore MetaLR_pred Reliability_index M-CAP_score M-CAP_rankscore  \\\n",
       "0          0.90075           D                 9    0.009489         0.24855   \n",
       "1          0.90075           D                 9    0.008795         0.23196   \n",
       "2           0.9461           D                 9     0.01492         0.35345   \n",
       "\n",
       "  M-CAP_pred REVEL_score REVEL_rankscore MutPred_score MutPred_rankscore  \\\n",
       "0          T       0.491         0.78035         0.718           0.85313   \n",
       "1          T       0.491         0.78035         0.718           0.85313   \n",
       "2          T       0.505         0.78906         0.698           0.83491   \n",
       "\n",
       "  MutPred_protID MutPred_AAchange  \\\n",
       "0         Q3ZCM7            K392N   \n",
       "1         Q3ZCM7            K392N   \n",
       "2         Q3ZCM7            K392M   \n",
       "\n",
       "                           MutPred_Top5features MPC_score MPC_rankscore  \\\n",
       "0             Loss of MoRF binding (P = 0.0523)         .             .   \n",
       "1             Loss of MoRF binding (P = 0.0523)         .             .   \n",
       "2   Loss of ubiquitination at K392 (P = 0.0326)         .             .   \n",
       "\n",
       "  PrimateAI_score PrimateAI_rankscore PrimateAI_pred  DANN_score  \\\n",
       "0  0.751451253891             0.74660              T    0.984728   \n",
       "1  0.751451253891             0.74660              T    0.984663   \n",
       "2  0.727143764496             0.71088              T    0.964042   \n",
       "\n",
       "   DANN_rankscore  fathmm-MKL_coding_score  fathmm-MKL_coding_rankscore  \\\n",
       "0         0.41873                  0.38681                      0.26043   \n",
       "1         0.41803                  0.40406                      0.26429   \n",
       "2         0.29699                  0.52831                      0.29225   \n",
       "\n",
       "  fathmm-MKL_coding_pred fathmm-MKL_coding_group fathmm-XF_coding_score  \\\n",
       "0                      N                   AEFBI               0.295876   \n",
       "1                      N                   AEFBI               0.302556   \n",
       "2                      D                   AEFBI               0.364324   \n",
       "\n",
       "  fathmm-XF_coding_rankscore fathmm-XF_coding_pred     Eigen-raw_coding  \\\n",
       "0                    0.40587                     N   0.0676056746461709   \n",
       "1                    0.41077                     N   0.0676056746461709   \n",
       "2                    0.45291                     N  0.31076238698785996   \n",
       "\n",
       "  Eigen-raw_coding_rankscore   Eigen-pred_coding   Eigen-PC-raw_coding  \\\n",
       "0                    0.44955             2.76036  -0.20748312943205102   \n",
       "1                    0.44955             2.76036  -0.20748312943205102   \n",
       "2                    0.56701  3.8352209999999998   -0.0235518622390073   \n",
       "\n",
       "  Eigen-PC-raw_coding_rankscore Eigen-PC-phred_coding   GC_hg19  CpG_hg19  \\\n",
       "0                       0.31266              1.767854  0.536424  0.106667   \n",
       "1                       0.31266              1.767854  0.536424  0.106667   \n",
       "2                       0.38654              2.280427  0.529801  0.106667   \n",
       "\n",
       "   priPhCons_hg19  mamPhCons_hg19  verPhCons_hg19  priPhyloP_hg19  \\\n",
       "0           0.550           0.546             1.0           0.110   \n",
       "1           0.550           0.546             1.0           0.110   \n",
       "2           0.546           0.542             1.0           0.094   \n",
       "\n",
       "   mamPhyloP_hg19  verPhyloP_hg19  GerpRS_hg19  GerpRSpval_hg19  GerpN_hg19  \\\n",
       "0           0.109           1.700          NaN              NaN         0.0   \n",
       "1           0.109           1.700          NaN              NaN         0.0   \n",
       "2           0.095           2.564          NaN              NaN         0.0   \n",
       "\n",
       "   GerpS_hg19  Dist2Mutation_hg19  Freq100bp_hg19  Rare100bp_hg19  \\\n",
       "0         0.0             38936.0             NaN             NaN   \n",
       "1         0.0             38936.0             NaN             NaN   \n",
       "2         0.0             38936.0             NaN             NaN   \n",
       "\n",
       "   Sngl100bp_hg19  Freq1000bp_hg19  Rare1000bp_hg19  Sngl1000bp_hg19  \\\n",
       "0             NaN              NaN              NaN              NaN   \n",
       "1             NaN              NaN              NaN              NaN   \n",
       "2             NaN              NaN              NaN              NaN   \n",
       "\n",
       "   Freq10000bp_hg19  Rare10000bp_hg19  Sngl10000bp_hg19  RawScore_hg19  \\\n",
       "0               5.0              24.0             137.0       2.636298   \n",
       "1               5.0              24.0             137.0       2.627006   \n",
       "2               5.0              24.0             137.0       2.819085   \n",
       "\n",
       "   PHRED_hg19   GC_hg38  CpG_hg38  priPhCons_hg38  mamPhCons_hg38  \\\n",
       "0        22.9  0.536424  0.106667           0.046           0.046   \n",
       "1        22.8  0.536424  0.106667           0.046           0.046   \n",
       "2        23.2  0.529801  0.106667           0.042           0.042   \n",
       "\n",
       "   verPhCons_hg38  priPhyloP_hg38  mamPhyloP_hg38  verPhyloP_hg38  \\\n",
       "0             1.0           0.247           0.249           2.730   \n",
       "1             1.0           0.247           0.249           2.730   \n",
       "2             1.0           0.274           0.275           6.207   \n",
       "\n",
       "   GerpRS_hg38  GerpRSpval_hg38  GerpN_hg38  GerpS_hg38  Dist2Mutation_hg38  \\\n",
       "0      8890.65              0.0        13.7        13.7                40.0   \n",
       "1      8890.65              0.0        13.7        13.7                40.0   \n",
       "2      8890.65              0.0         NaN         NaN                40.0   \n",
       "\n",
       "   Freq100bp_hg38  Rare100bp_hg38  Sngl100bp_hg38  Freq1000bp_hg38  \\\n",
       "0             0.0             0.0             8.0              0.0   \n",
       "1             0.0             0.0             8.0              0.0   \n",
       "2             0.0             0.0             8.0              0.0   \n",
       "\n",
       "   Rare1000bp_hg38  Sngl1000bp_hg38  Freq10000bp_hg38  Rare10000bp_hg38  \\\n",
       "0              0.0             91.0              12.0              31.0   \n",
       "1              0.0             91.0              12.0              31.0   \n",
       "2              0.0             91.0              12.0              31.0   \n",
       "\n",
       "   Sngl10000bp_hg38  RawScore_hg38  PHRED_hg38  \n",
       "0             930.0       3.029607        23.0  \n",
       "1             930.0       3.055876        23.1  \n",
       "2             930.0       3.100150        23.2  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergeNOT.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving files\n",
    "mergeDET.to_csv(\"MERGE_dbNSFP_CADD_DETECTED_CK_104475.csv\", index=False)\n",
    "\n",
    "mergeNOT.to_csv(\"MERGE_dbNSFP_CADD_NOT_DETECTED_CK_1222911.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(104475, 143)\n",
      "(1222911, 136)\n"
     ]
    }
   ],
   "source": [
    "print(mergeDET.shape)\n",
    "print(mergeNOT.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariapalafox/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "isnaDET = mergeDET.isna().sum()\n",
    "isnaDET.to_csv(\"MERGE_DETECTED_isna_sum.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariapalafox/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "isnaNOT = mergeNOT.isna().sum()\n",
    "isnaNOT.to_csv(\"MERGE_NOTDETECTED_isna_sum.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "---\n",
    "\n",
    "\n",
    "# (done) CADD file correcting pos_id # format (add leading 0's)\n",
    "- 9 digits with leading zeros for pos_id19 or pos_id38\n",
    "\n",
    "```python \n",
    "def create_pos_id19(df, chrr, pos, ref, alt):\n",
    "    # variables are colnames of df\n",
    "    df.loc[:,'pos_id19'] = df[chrr].astype(str) + '_' + \\\n",
    "    df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "    '_' + df[alt].astype(str)\n",
    "    return df\n",
    "\n",
    "def create_pos_id38(df, chrr, pos, ref, alt):\n",
    "    # variables are colnames of df\n",
    "    df.loc[:,'pos_id38'] = df[chrr].astype(str) + '_' + \\\n",
    "    df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "    '_' + df[alt].astype(str)\n",
    "    return df\n",
    "\n",
    "# dbNSFP FILES\n",
    "dbnsfp = pd.read_csv('SCORE_dbNSFP_selectcols_detected_CK_104475.csv', low_memory=False)\n",
    "notdbnsfp = pd.read_csv('SCORE_dbNSFP_selectcols_NOT_detected_CK_1222911.csv', low_memory=False)\n",
    "\n",
    "# CADD FILES\n",
    "det19 = pd.read_csv('SCORE_CADD_37_detected_CK.csv', low_memory=False, converters={'pos_hg19': '{:0>9}'.format})\n",
    "det38 = pd.read_csv('SCORE_CADD_38_detected_CK.csv', low_memory=False, converters={'pos_hg38': '{:0>9}'.format})\n",
    "not19 = pd.read_csv('SCORE_CADD_37_NOT_detected_CK.csv', low_memory=False, converters={'pos_hg19': '{:0>9}'.format})\n",
    "not38 = pd.read_csv('SCORE_CADD_38_NOT_detected_CK.csv', low_memory=False, converters={'pos_hg38': '{:0>9}'.format})\n",
    "```\n",
    "\n",
    "## calling function to create pos id with correct # format:\n",
    "```python\n",
    "# cadd detected files\n",
    "det19 = create_pos_id19(det19, 'chr', 'pos_hg19', 'Ref', 'Alt')\n",
    "det38 = create_pos_id38(det38, 'chr', 'pos_hg38', 'Ref', 'Alt')\n",
    "\n",
    "# cadd NOTdetected files\n",
    "not19 = create_pos_id19(not19, 'chr', 'pos_hg19', 'Ref', 'Alt')\n",
    "not38 = create_pos_id38(not38, 'chr', 'pos_hg38', 'Ref', 'Alt')\n",
    "\n",
    "# saving files with correct pos_id\n",
    "det19.to_csv('SCORE_CADD_37_detected_CK.csv', index=False)\n",
    "det38.to_csv('SCORE_CADD_38_detected_CK.csv',  index=False)\n",
    "not19.to_csv('SCORE_CADD_37_NOT_detected_CK.csv', index=False)\n",
    "not38.to_csv('SCORE_CADD_38_NOT_detected_CK.csv',  index=False)\n",
    "```\n",
    "\n",
    "---\n",
    "---\n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# (initial version, see other M html for more details)\n",
    "- [3] filter CADDv1.4 files with dbNSFP CK positions for MISSENSE consequence only\n",
    "\n",
    "dbNSFP SCORE file column names (will map to these files using columns shared with CADD files:\n",
    "\n",
    "**columns detected file:**\n",
    "    - Amino_acids\n",
    "    - pos_ID 'Q3ZCM7_C354'\n",
    "    - pos_id19 or pos_id38_x\n",
    "\n",
    "**columns not detected file:**\n",
    "    - pos_id19 or pos_id38\n",
    "    - Amino_acids\n",
    "    - pos_ID_falseCKtarget\n",
    "\n",
    "\n",
    "```python\n",
    "# !/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "# Pmap M local ipynb code, Pmap_missense_annotations_QC.py\n",
    "# markdown M local QC of positions from dbNSFP overlapped with CADD37 or 38 annotations\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "def create_coordinate_id(df, chrr, pos, ref, alt, assembly):\n",
    "    if assembly == 37:\n",
    "        df.loc[:,'pos_id19'] = df[chrr].astype(str) + '_' + \\\n",
    "                df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "                '_' + df[alt].astype(str)\n",
    "    if assembly == 38:\n",
    "        df.loc[:,'pos_id38'] = df[chrr].astype(str) + '_' + \\\n",
    "                df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "                '_' + df[alt].astype(str)\n",
    "    return df\n",
    "\n",
    "\n",
    "def format_missense_triple(df, oaacol, naacol):\n",
    "    #  A|A turns to Ala/Ala\n",
    "    amino_dict = dict([('A', 'Ala'),('G', 'Gly'), ('I','Ile'), ('L','Leu'), ('P', 'Pro'), ('V','Val'), ('F','Phe'),('W', 'Trp'), ('Y', 'Tyr'), ('D','Asp'),('E','Glu'), ('R','Arg'),('H','His'), ('K','Lys'), ('S','Ser'), ('T', 'Thr'), ('C', 'Cys'), ('M', 'Met'), ('N', 'Asn'), ('Q','Gln')])\n",
    "    df[oaacol].replace(amino_dict, inplace=True)\n",
    "    df[naacol].replace(amino_dict, inplace=True)\n",
    "    ccopy = df[naacol].copy()\n",
    "    df['Amino_acids'] = df[oaacol].str.cat(ccopy, sep='/')\n",
    "    return df\n",
    "\n",
    "\n",
    "def filter_cadd_overlap(df, assembly):\n",
    "    # [1] filter for missense only\n",
    "    miss = df[df['Consequence'] == 'NON_SYNONYMOUS'].copy()\n",
    "    # [2] new pos_id(assembly) to files\n",
    "    if assembly == 37:\n",
    "        miss = create_coordinate_id(miss, 'chr', 'pos_hg19', 'Ref', 'Alt', assembly)\n",
    "    if assembly == 38:\n",
    "        miss = create_coordinate_id(miss, 'chr', 'pos_hg38', 'Ref', 'Alt', assembly)\n",
    "    # [3] new missense type column in 3 letter format with '/' sep {oAA, nAA}\n",
    "    miss = format_missense_triple(miss, 'oAA', 'nAA')\n",
    "\n",
    "    return miss\n",
    "\n",
    "    # [4] concat all files from 37: DECT or NOT ... 38: DECT or NOT\n",
    "\n",
    "\n",
    "def main():\n",
    "    os.chdir('/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/RESULT_pos_overlap_dbNSFPcoordinates')\n",
    "\n",
    "    chrls = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,'X','Y']\n",
    "\n",
    "    # GRCh37\n",
    "    for order in chrls:\n",
    "        chrID = 'chr{}'.format(order)\n",
    "        file1 = '{}_CADD_GRCh37_DETECTED_CK.csv'.format(chrID)\n",
    "        out1 = 'MISSENSE_{}_CADD_GRCh37_DETECTED_CK.csv'.format(chrID)\n",
    "        file2 = '{}_CADD_GRCh37_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        out2 = 'MISSENSE_{}_CADD_GRCh37_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        df1 = pd.read_csv(file1, low_memory=False)\n",
    "        df2 = pd.read_csv(file2, low_memory=False)\n",
    "        df1out = filter_cadd_overlap(df1, 37)\n",
    "        df2out = filter_cadd_overlap(df2, 37)\n",
    "        print(\"saving detected and not detected GRCh37 \", chrID)\n",
    "        print()\n",
    "        df1out.to_csv(out1, index=False)\n",
    "        df2out.to_csv(out2, index=False)\n",
    "\n",
    "    # GRCh38\n",
    "    for order in chrls:\n",
    "        chrID = 'chr{}'.format(order)\n",
    "        file1 = '{}_CADD_GRCh38_DETECTED_CK.csv'.format(chrID)\n",
    "        out1 = 'MISSENSE_{}_CADD_GRCh38_DETECTED_CK.csv'.format(chrID)\n",
    "        file2 = '{}_CADD_GRCh38_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        out2 = 'MISSENSE_{}_CADD_GRCh38_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        df1 = pd.read_csv(file1)\n",
    "        df2 = pd.read_csv(file2)\n",
    "        df1out = filter_cadd_overlap(df1, 38)\n",
    "        df2out = filter_cadd_overlap(df2, 38)\n",
    "        print(\"saving detected and not detected GRCh38 \", chrID)\n",
    "        print()\n",
    "        df1out.to_csv(out1, index=False)\n",
    "        df2out.to_csv(out2, index=False)\n",
    "\n",
    "main()\n",
    "\n",
    "```\n",
    "\n",
    "# RE running GRCh38 sex chrs after fixing cadd source files on hoffman\n",
    "\n",
    "```python\n",
    "# !/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "# Pmap M local ipynb code, Pmap_missense_annotations_QC.py\n",
    "# markdown M local QC of positions from dbNSFP overlapped with CADD37 or 38 annotations\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "\"\"\" \n",
    "dbNSFP SCORE file column names:\n",
    "\n",
    "columns detected file:\n",
    "    Amino_acids\n",
    "    pos_ID 'Q3ZCM7_C354'\n",
    "    pos_id19 or pos_id38_x\n",
    "\n",
    "columns not detected file:\n",
    "    pos_id19 or pos_id38\n",
    "    Amino_acids,\n",
    "    pos_ID_falseCKtarget\n",
    "\"\"\"\n",
    "\n",
    "def create_coordinate_id(df, chrr, pos, ref, alt, assembly):\n",
    "    if assembly == 37:\n",
    "        df.loc[:,'pos_id19'] = df[chrr].astype(str) + '_' + \\\n",
    "                df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "                '_' + df[alt].astype(str)\n",
    "    if assembly == 38:\n",
    "        df.loc[:,'pos_id38'] = df[chrr].astype(str) + '_' + \\\n",
    "                df[pos].astype(str) + '_' + df[ref].astype(str) + \\\n",
    "                '_' + df[alt].astype(str)\n",
    "    return df\n",
    "\n",
    "\n",
    "def format_missense_triple(df, oaacol, naacol):\n",
    "    #  A|A turns to Ala/Ala\n",
    "    amino_dict = dict([('A', 'Ala'),('G', 'Gly'), ('I','Ile'), ('L','Leu'), ('P', 'Pro'), ('V','Val'), ('F','Phe'),('W', 'Trp'), ('Y', 'Tyr'), ('D','Asp'),('E','Glu'), ('R','Arg'),('H','His'), ('K','Lys'), ('S','Ser'), ('T', 'Thr'), ('C', 'Cys'), ('M', 'Met'), ('N', 'Asn'), ('Q','Gln')])\n",
    "    df[oaacol].replace(amino_dict, inplace=True)\n",
    "    df[naacol].replace(amino_dict, inplace=True)\n",
    "    ccopy = df[naacol].copy()\n",
    "    df['Amino_acids'] = df[oaacol].str.cat(ccopy, sep='/')\n",
    "    return df\n",
    "\n",
    "\n",
    "def filter_cadd_overlap(df, assembly):\n",
    "    # [1] filter for missense only\n",
    "    miss = df[df['Consequence'] == 'NON_SYNONYMOUS'].copy()\n",
    "    # [2] new pos_id(assembly) to files\n",
    "    if assembly == 37:\n",
    "        miss = create_coordinate_id(miss, 'chr', 'pos_hg19', 'Ref', 'Alt', assembly)\n",
    "    if assembly == 38:\n",
    "        miss = create_coordinate_id(miss, 'chr', 'pos_hg38', 'Ref', 'Alt', assembly)\n",
    "    # [3] new missense type column in 3 letter format with '/' sep {oAA, nAA}\n",
    "    miss = format_missense_triple(miss, 'oAA', 'nAA')\n",
    "\n",
    "    return miss\n",
    "\n",
    "    # [4] concat all files from 37: DECT or NOT ... 38: DECT or NOT\n",
    "\n",
    "\n",
    "def main():\n",
    "    os.chdir('/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CADDmapped/RESULT_pos_overlap_dbNSFPcoordinates')\n",
    "    chrls = ['X', 'Y']\n",
    "    # GRCh38\n",
    "    for order in chrls:\n",
    "        chrID = 'chr{}'.format(order)\n",
    "        file1 = '{}_CADD_GRCh38_DETECTED_CK.csv'.format(chrID)\n",
    "        out1 = 'MISSENSE_{}_CADD_GRCh38_DETECTED_CK.csv'.format(chrID)\n",
    "        file2 = '{}_CADD_GRCh38_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        out2 = 'MISSENSE_{}_CADD_GRCh38_NOT_DETECTED_CK.csv'.format(chrID)\n",
    "        df1 = pd.read_csv(file1)\n",
    "        df2 = pd.read_csv(file2)\n",
    "        df1out = filter_cadd_overlap(df1, 38)\n",
    "        df2out = filter_cadd_overlap(df2, 38)\n",
    "        print(\"saving detected and not detected GRCh38 \", chrID)\n",
    "        print()\n",
    "        df1out.to_csv(out1, index=False)\n",
    "        df2out.to_csv(out2, index=False)\n",
    "\n",
    "main()\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
