{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mapping backwards PROTEIN SEQUENCE --> TRANSCRIPT SEQ --> GENE SEQUENCE\n",
    "\n",
    "## ALT: PRO SEQ --> REF GENE SEQ --> CANONICAL TX POSITION\n",
    "\n",
    "### - project for confirming that Uniprot ID (included in cys_position annotation unique identifier) can be used to map to a position on canonical transcript in Exac where variance is. \n",
    "######################################################################################################\n",
    "## questions:\n",
    "\n",
    "00) Using UNIPROTKB_all (swiss and trembl) 170k-used for proteomic mapping***scripts lab used a data file that was closer to 20k enteries for mass spec mapping to reduce redundancy problem** || KB_proteome (swiss and trembl) 70k  || KB_SWISS 20386k || KB_SWISS 20373k\n",
    "    - proteome: genome assembly and annotation == GCA_000001405.27 from Ensembl https://www.uniprot.org/proteomes/UP000005640\n",
    "    \n",
    "0) do we need Index of /pub/databases/uniprot/current_release/knowledgebase/proteomics_mapping/\n",
    "ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/proteomics_mapping/\n",
    "\n",
    "    a) nonUniquePeptides.tsv and uniquePeptides.tsv\n",
    "    UniProt is running an analysis pipeline to provide a mapping from publicly\n",
    "available mass spectrometry-based data to UniProtKB sequences. The pipeline\n",
    "consists of the following steps:\n",
    "\n",
    "## #proteomic mapping:\n",
    "\n",
    "    I) Species-specific lists of identified peptides are retrieved from the mass\n",
    "spectrometry (MS) proteomics repositories indicated in the output files. These\n",
    "peptides are then filtered based on quality metrics metadata provided by the MS\n",
    "proteomics repositories.\n",
    "\n",
    "    II) All sequences (canonical sequences plus isoforms) from the UniProtKB\n",
    "reference proteome of a specific taxonomic identifier are in-silico digested\n",
    "with up to nine cleaving rules, allowing for up to two missed cleavages and with\n",
    "or without initiator methionine cleavage.\n",
    "\n",
    "    III) The in-silico generated peptides are then filtered in order to remove\n",
    "peptides which are less than seven amino acids in length and to remove X, B or\n",
    "Z-containing peptide sequences.\n",
    "\n",
    "    IV) Unicity of the in-silico generated peptides is evaluated according to gene\n",
    "groups underlying the UniProtKB reference proteomes where protein sequences are\n",
    "grouped based on the gene(s) encoding them. Each gene group is constituted by\n",
    "one or more UniProtKB sequences. A peptide is considered unique if it belongs to\n",
    "only one gene group. Two types of in-silico generated peptides are therefore\n",
    "obtained: unique and non-unique.\n",
    "\n",
    "    V) The two types of in-silico generated peptides are compared to the\n",
    "corresponding species lists of identified peptides from the MS proteomics\n",
    "repositories and those with a 100% sequence match are exported to the\n",
    "corresponding output files that are described below. UniProtKB identifiers with\n",
    "only one seven, eight or nine-amino acids long mapped peptide are removed.\n",
    "\n",
    "\n",
    "\n",
    "1) uniprotKB is canonical seq. they additionally have isoforms in fasta file format.  The Exac files have all variants (can filter for canon transcript only variants). ???Easy to map from canonical tx exac file to KB AC??? no because the uniprot def of canonical is different from ExAC......, but how can I map from not canonical tx exac to KB AC???\n",
    "\n",
    "    a) do the uniprot sequence annotations only refer to the canonical?  **YES**\n",
    "    \n",
    "    - Sequence annotations (position-specific annotations) used to be found in the ‘Sequence annotation (Features)’ section in the previous version of the UniProtKB entry view. The flat file and XML formats still group all position-specific annotation together in a “feature table” (FT, <feature>). Each sequence annotation consists of a “feature key”, “from” and “to” positions as well as a short description.\n",
    "    \n",
    "    b) if the features only pertain to the canonical position...then a variant on non canonical transcript can not have added on proteomic features....you can always go to genomic features no matter what transcript variant is on...but we are only going to use canoncial transcript from exac which is based on **Ensemble v75 Gencode v19**\n",
    "\n",
    "2) MS mapping back to uniprot, do you use KB (all proteome has 70k, all KB has 150k) or KB Swiss-- redundancy problem comes with using anything other than KBSwiss\n",
    "\n",
    "**NOTES:**\n",
    "\n",
    "\n",
    "    • UniProtKB/TrEMBL: one record for 100% identical full-length sequences in one species;\n",
    "    \n",
    "    • UniProtKB/Swiss-Prot: one record per gene in one species;\n",
    "    \n",
    " ## what does non-redundancy mean?\n",
    " \n",
    "- **UniProtKB/TrEMBL** is ‘non-redundant’ in the sense that all identical, full-length protein sequences, provided they come from the same species, are represented in a single record. Fragments, isoforms, variants and so on, encoded by the same gene, are stored in separate entries.\n",
    "\n",
    "\n",
    "\n",
    "- **UniProtKB/Swiss-Prot** is ‘non-redundant’ in the sense that all protein products encoded by **one gene in a given species are represented in a single record.** This includes alternative splicing isoforms, fragments, polymorphisms, sequence conflicts, etc. Differences between sequence reports are analyzed, fully documented and reported in the entry. Cross-references to the original submissions to EMBL-Bank/GenBank/DDBJ databases are kept (see for instance, Q9BXB7).\n",
    "\n",
    "- UniProtKB entries in  (Flat Text, XML and RDF/XML) formats **each contain only one protein sequence, the so-called ‘canonical’ sequence**, additional manually curated isoform sequences in UniProtKB/Swiss-Prot FASTA\n",
    "\n",
    "\n",
    "\n",
    "3) How many variants are there in Exac_filtered compared to how many are in Exac_filtered_canonical?? (hoffman delay, email about working at 5pm, then failed at night) **from release 3.1 the ALL file is 7G the canonical is 4.9G, so a substantial amount is lost** \n",
    "\n",
    "ISSUES: if the disease relevant variant is not on the canonical transcript- I can find it in the fasta file with isoform information, however all the annotations including domains will not be there.  The files that have all the annotations are only referring to the canonical sequence. ??? \n",
    "\n",
    "- does trembl have annotations for non canonical forms of protein?  \n",
    "\n",
    "4) should i remake figures using the human_proteome_KBSWISS files: the current table used to make all figures is 20,386 non proteome filter, the proteome filter brings protein count down to 20,373?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# mapping question links \n",
    "\n",
    "## natural variant help:\n",
    "\n",
    "https://www.uniprot.org/help/variant\n",
    "\n",
    "## Index of manually curated Human polymorphisms and disease mutations from UniProtKB/Swiss-Prot.\n",
    "\n",
    "https://www.uniprot.org/docs/humsavar\n",
    "- key is FTId, because all variants have this values which links to entry AC, but not all have dbSNPs NCBI\n",
    "\n",
    "## uniprot sequence features help\n",
    "\n",
    "https://www.uniprot.org/help/sequence_annotation\n",
    "-FT ID is feature table ID, how do we link that uniprot ID to variant in exac?  \n",
    "\n",
    " # fasta headers\n",
    " \n",
    " - https://www.uniprot.org/help/fasta-headers\n",
    " \n",
    "## Amino acid altering variants imported from Ensembl Variation databases uniprot_sprot_human.dat.gz\n",
    "\n",
    "- pwd project-arboleda/public_files/localUNIPROT/complete\n",
    "\n",
    "## lists Swiss-Prot  human  protein  sequence  entries  that  contain information on at least one sequence variant (polymorphism or disease-linked mutation)\n",
    "\n",
    "https://www.uniprot.org/docs/humpvar\n",
    "\n",
    "- NOTE Variations due to different sequence reports are considered in Swiss-Prot\n",
    "   as \"conflicts\" and are not considered here. It is only when a conflict is\n",
    "   proved to be a polymorphism that it is noted as a \"variant\".\n",
    "   \n",
    "## What is the canonical sequence? Are all isoforms described in one entry?\n",
    "\n",
    "https://www.uniprot.org/help/canonical_and_isoforms\n",
    "\n",
    "## MAPPING functional domains to DNA\n",
    "\n",
    "ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/genome_annotation_tracks/\n",
    "\n",
    "## exac VEP annotations from ReadME - Vep is ensembl tool.  the tool was used with ExAC.sites.vcf file for all variants\n",
    "\n",
    "- http://uswest.ensembl.org/info/docs/tools/vep/vep_formats.html\n",
    "- http://www.ensembl.org/info/docs/tools/vep/index.html\n",
    "## ExAC files, the raw files with .tbi \n",
    "- TABIX and VCF files: https://genome.ucsc.edu/goldenpath/help/vcf.html\n",
    "\n",
    "## ensembl guide for futher annotations of missense variants on protein function\n",
    "\n",
    "http://uswest.ensembl.org/info/genome/variation/prediction/protein_function.html\n",
    "\n",
    "# These files hold the coding sequences corresponding to Ensembl gene predictions. CDS does not contain UTR or intronic sequence.\n",
    "\n",
    "## ensembl readme\n",
    "ftp://ftp.ensembl.org/pub/grch37/update/fasta/homo_sapiens/cds/README\n",
    "\n",
    "\n",
    "http://uswest.ensembl.org/info/genome/genebuild/annotation_sources.html\n",
    "\n",
    "## manuscript data\n",
    "\n",
    "ftp://ftp.broadinstitute.org/pub/ExAC_release/release1/manuscript_data/\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "-"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **uniprot notes**\n",
    "- proteome ID  ‎UP000005640\t\n",
    "- taxonomy human 9606\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Files in HOFFMAN\n",
    "\n",
    "- pwd : /u/home/m/mfpalafo/project-arboleda/public_files/localUNIPROT\n",
    "----------------------------------------------------------------------------\n",
    "1) **doc_humCHROMtext**\n",
    "\n",
    "- humchr01.txt  humchr04.txt  humchr07.txt  humchr10.txt  humchr13.txt  humchr16.txt  humchr19.txt  humchr22.txt  humdisease.txt\n",
    "humchr02.txt  humchr05.txt  humchr08.txt  humchr11.txt  humchr14.txt  humchr17.txt  humchr20.txt  humchrx.txt   humpvar.txt\n",
    "humchr03.txt  humchr06.txt  humchr09.txt  humchr12.txt  humchr15.txt  humchr18.txt  humchr21.txt  humchry.txt   humsavar.txt\n",
    "\n",
    "*less humsavar.txt* == \n",
    "\n",
    "Description: Human polymorphisms and disease mutations: index\n",
    "Name:        humsavar.txt\n",
    "Release:     2018_07 of 18-Jul-2018\n",
    "\n",
    "----------------------------------------------------------------------------\n",
    "This file lists **all missense variants annotated in human UniProtKB/Swiss-Prot \n",
    "entries.** It provides a variant classification which is intended for research\n",
    "purposes only, not for clinical and diagnostic use.\n",
    "\n",
    " **some of these variants are experimentally generated so they are not the same as the Exac file variants **\n",
    " \n",
    "*less humchr01.txt* == \n",
    "\n",
    "Description: Human chromosome 1: entries, gene names and\n",
    "             cross-references to MIM\n",
    "Name:        humchr01.txt\n",
    "Release:     2018_07 of 18-Jul-2018\n",
    "\n",
    "----------------------------------------------------------------------------\n",
    "\n",
    "This documents lists all the human protein sequence entries whose genes\n",
    "are known to be encoded on chromosome 1 in this release of UniProtKB/Swiss-Prot.\n",
    "\n",
    "Number of UniProtKB/Swiss-Prot entries encoded on chromosome 1: 2068\n",
    "\n",
    "Reference for the chromosome sequence:\n",
    "Nature 441:315-321(2006).\n",
    "PubMed=16710414; DOI=10.1038/nature04727;\n",
    "\n",
    "For more information on chromosome 1 see:\n",
    "\n",
    "Sanger  : http://www.sanger.ac.uk/about/history/hgp/chr1.html\n",
    "\n",
    "Ensembl : http://www.ensembl.org/Homo_sapiens/mapview?chr=1\n",
    "\n",
    "NCBI    : https://www.ncbi.nlm.nih.gov/mapview/maps.cgi?taxid=9606&chr=1\n",
    "\n",
    "OMIM    : https://www.ncbi.nlm.nih.gov/Omim/getmap.cgi?chromosome=1\n",
    "\n",
    "DOE     : http://www.ornl.gov/sci/techresources/Human_Genome/launchpad/chrom01.shtml\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------------\n",
    "\n",
    "2) **complete**\n",
    "\n",
    "\n",
    "\n",
    "- humsavar.txt    == all missense variants annotated in human UniProtKB/Swiss-Prot \n",
    "entries. **Amino acid altering variants imported from Ensembl Variation databases, tab-delimited download** == http://uswest.ensembl.org/info/genome/variation/species/sources_documentation.html the Ensembl V db does not contain ExAC. Variants are manually annotated in UniProtKB/Swiss-Prot.  These do not include all the variants in ExAC, and some of these are experimentally generated...so do not rely on this to represent nature\n",
    "\n",
    "- uniprot_sprot_human.dat.gz    ==       contains all human entries   \n",
    "\n",
    "\n",
    "- uniprot_sprot.dat.gz  uniprot_sprot.fasta.gz  uniprot_sprot.xml.gz isoforms_curated_uniprot_SWISSprot_varsplic.fasta.gz == complete SWISSKB (not exclusively human)\n",
    "\n",
    "\n",
    "- uniprot_trembl.dat.gz uniprot_trembl.xml.gz uniprot_trembl.fasta.gz  == all of trembl\n",
    "\n",
    "\n",
    "- ALLHumKB_isoform&canonical_uniprot-filtered-organism%3A_Homo+sapiens+%28Human%29+%5B9606%5D_.fasta.gz == Human KB all isoforms and canonical fasta\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------------\n",
    "### subdirectories\n",
    "\n",
    "#### proteomes - human: https://www.uniprot.org/proteomes/UP000005640\n",
    "\n",
    "- **proteome, i.e. of a set of proteins thought to be expressed by organisms whose genomes have been completely sequenced.**\n",
    "\n",
    "    - proteome_ALL *UniProtKB/TrEMBL (unreviewed) contains protein sequences associated with computationally generated annotation and large-scale functional characterization. UniProtKB/Swiss-Prot (reviewed) is a high quality manually annotated and non-redundant protein sequence database, which brings together experimental results, computed features and scientific conclusions.*\n",
    "\n",
    "\n",
    "    a) **human_proteome_ALL_trembl_swiss**\n",
    "\n",
    "- KBproteomeALL_uniprot-filtered-proteome%3AUP000005640+AND+organism%3A_Homo+sapiens+%28Human%.canonical.fasta.gz\n",
    "\n",
    "- KBproteomeALL_uniprot-filtered-proteome%3AUP000005640+AND+organism%3A_Homo+sapiens+%28Human%.isoform&canonical.fasta.gz\n",
    "\n",
    "- KBproteomeALL_uniprot-filtered-proteome%3AUP000005640+AND+organism%3A_Homo+sapiens+%28Human%--.tab.gz\n",
    "\n",
    "- KBproteomeALL_uniprot-filtered-proteome%3AUP000005640+AND+organism%3A_Homo+sapiens+%28Human%--.xml.gz\n",
    "\n",
    "\n",
    "    b) **human_proteome_KBSWISS**\n",
    "\n",
    "- uniprot-proteome%3AUP000005640+reviewed%3Ayes.canonical.fasta.gz\n",
    "\n",
    "- uniprot-proteome%3AUP000005640+reviewed%3Ayes.tab.gz\n",
    "\n",
    "- uniprot-proteome%3AUP000005640+reviewed%3Ayes.isoform&canonical.fasta.gz  uniprot-proteome%3AUP000005640+reviewed%3Ayes.xml.gz\n",
    "\n",
    "- uniprot-proteome%3AUP000005640+reviewed%3Ayes.rdf.gz\n",
    "\n",
    "\n",
    "    c) **human_nonPROTEOME_KBSWISS**\n",
    "\n",
    "NOTES: what figures are created from\n",
    "- 20386_uniprot-filtered-organism%3A_Homo+sapiens+%28Human%29+%5B9606%5D_+AND+review--.tab.gz\n",
    "\n",
    "- 20386uniprot-filtered-organism%3A_Homo+sapiens+%28Human%29+%5B9606%5D_+AND+review--.xml.gz\n",
    "\n",
    "- uniprot-filtered-organism%3A_Homo+sapiens+%28Human%29+%5B9606%5D_+AND+review.canonical.fasta.gz\n",
    "\n",
    "- uniprot-filtered-organism%3A_Homo+sapiens+%28Human%29+%5B9606%5D_+AND+review.isoform&canonical.fasta.gz\n",
    "\n",
    "----------------------------------------------------------------------------\n",
    "\n",
    "3)  **genome_annotation_track**\n",
    "\n",
    "NOTES: way to visually map UNIPROTKB sequence featrues to genomic context.  BED && BigBed files av from uniprot-- use with genome browser\n",
    "\n",
    "\n",
    "*GENOME BROWSERS*\n",
    "A) https://uswest.ensembl.org/Homo_sapiens/Info/Index (recommended)\n",
    "B) https://genome.ucsc.edu/\n",
    "\n",
    "----------------------------------------------------------------------------\n",
    "\n",
    "4) ID_mapping\n",
    "\n",
    "- HUMAN_9606_idmapping.dat == This file has three columns, delimited by tab:\n",
    "\n",
    "        1. UniProtKB-AC \n",
    "        2. ID_type \n",
    "        3. ID\n",
    "        \n",
    "        where ID_type is the database name as appearing in UniProtKB cross-references, \n",
    "        and as supported by the ID mapping tool on the UniProt web site, \n",
    "        http://www.uniprot.org/mapping and where ID is the identifier in \n",
    "        that cross-referenced database.\n",
    "\n",
    "- HUMAN_9606_idmapping_selected.tab  == We also provide this tab-delimited table which includes\n",
    "the following mappings delimited by tab:\n",
    "\n",
    "    1. UniProtKB-AC\n",
    "    2. UniProtKB-ID\n",
    "    3. GeneID (EntrezGene)\n",
    "    4. RefSeq\n",
    "    5. GI\n",
    "    6. PDB\n",
    "    7. GO\n",
    "    8. UniRef100\n",
    "    9. UniRef90\n",
    "    10. UniRef50\n",
    "    11. UniParc\n",
    "    12. PIR\n",
    "    13. NCBI-taxon\n",
    "    14. MIM\n",
    "    15. UniGene\n",
    "    16. PubMed\n",
    "    17. EMBL\n",
    "    18. EMBL-CDS\n",
    "    19. Ensembl\n",
    "    20. Ensembl_TRS\n",
    "    21. Ensembl_PRO\n",
    "    22. Additional PubMed\n",
    "\n",
    "\n",
    "- README  == databases/uniprot/current_release/knowledgebase/idmapping/,\n",
    "contains the idmapping data files which are updated in conjunction with the UniProt \n",
    "Knowledgebase (UniProtKB). Whenever available the mappings are extracted from the \n",
    "UniProtKB records. \n",
    "\n",
    "----------------------------------------------------------------------------\n",
    "\n",
    "----------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RDF notes from uniprot files\n",
    "NOTES: UniProt collects information from the scientific literature and other databases and provides links to over one hundred and fifty biological resources. Such links between different databases are an important basis for data integration, but the lack of a common standard to represent and link information makes data integration an expensive business. One way to tackle this problem at UniProt is by using the Resource Description Framework to represent our data. https://www.uniprot.org/format/uniprot_rdf\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
