{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container {width:90% !important;}</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import argparse\n",
    "import pandas as pd\n",
    "import csv\n",
    "import numpy as np\n",
    "from IPython.display import display, HTML\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.options.display.max_seq_items = 2000\n",
    "display(HTML(\"<style>.container {width:90% !important;}</style>\"))\n",
    "sys.path.append(\"/Users/mariapalafox/Desktop/Toolbox\")\n",
    "from all_funx import *\n",
    "\n",
    "# mfp 3/30/20 before first exam, subaim 2.3 prioritize and validate CpDAA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# VEP output for input .bed detected cys and lysine (map paper)\n",
    "- files in VEP dir on desktop\n",
    "- uniprot proteins in dataset = 3840\n",
    "\n",
    "```\n",
    "## ENSEMBL VARIANT EFFECT PREDICTOR v96.0\n",
    "## Output produced at 2019-12-23 16:28:16\n",
    "## Using cache in /Users/mariapalafox/.vep/homo_sapiens/96_GRCh37\n",
    "## Using API version 96, DB version ?\n",
    "## ensembl-funcgen version 96.d901739\n",
    "## ensembl-variation version 96.617872b\n",
    "## ensembl-io version 96.6e65b30\n",
    "## ensembl version 96.af6c2b8\n",
    "## regbuild version 1.0\n",
    "## gnomAD version r2.1\n",
    "## ESP version 20141103\n",
    "## sift version sift5.2.2\n",
    "## ClinVar version 201810\n",
    "## polyphen version 2.2.2\n",
    "## genebuild version 2011-04\n",
    "## HGMD-PUBLIC version 20174\n",
    "## gencode version GENCODE 19\n",
    "## 1000genomes version phase3\n",
    "## dbSNP version 151\n",
    "## assembly version GRCh37.p13\n",
    "## COSMIC version 86\n",
    "## Column descriptions:\n",
    "## Uploaded_variation : Identifier of uploaded variant\n",
    "## Location : Location of variant in standard coordinate format (chr:start or chr:start-end)\n",
    "## Allele : The variant allele used to calculate the consequence\n",
    "## Gene : Stable ID of affected gene\n",
    "## Feature : Stable ID of feature\n",
    "## Feature_type : Type of feature - Transcript, RegulatoryFeature or MotifFeature\n",
    "## Consequence : Consequence type\n",
    "## cDNA_position : Relative position of base pair in cDNA sequence\n",
    "## CDS_position : Relative position of base pair in coding sequence\n",
    "## Protein_position : Relative position of amino acid in protein\n",
    "## Amino_acids : Reference and variant amino acids\n",
    "## Codons : Reference and variant codon sequence\n",
    "## Existing_variation : Identifier(s) of co-located known variants\n",
    "## IMPACT : Subjective impact classification of consequence type\n",
    "## DISTANCE : Shortest distance from variant to transcript\n",
    "## STRAND : Strand of the feature (1/-1)\n",
    "## FLAGS : Transcript quality flags\n",
    "## PICK : Indicates if this consequence has been picked as the most severe\n",
    "## SYMBOL : Gene symbol (e.g. HGNC)\n",
    "## SYMBOL_SOURCE : Source of gene symbol\n",
    "## HGNC_ID : Stable identifer of HGNC gene symbol\n",
    "## BIOTYPE : Biotype of transcript or regulatory feature\n",
    "## CANONICAL : Indicates if transcript is canonical for this gene\n",
    "## TSL : Transcript support level\n",
    "## APPRIS : Annotates alternatively spliced transcripts as primary or alternate based on a range of computational methods\n",
    "## CCDS : Indicates if transcript is a CCDS transcript\n",
    "## ENSP : Protein identifer\n",
    "## SWISSPROT : UniProtKB/Swiss-Prot accession\n",
    "## TREMBL : UniProtKB/TrEMBL accession\n",
    "## UNIPARC : UniParc accession\n",
    "## RefSeq : ?\n",
    "## GENE_PHENO : Indicates if gene is associated with a phenotype, disease or trait\n",
    "## EXON : Exon number(s) / total\n",
    "## INTRON : Intron number(s) / total\n",
    "## DOMAINS : The source and identifer of any overlapping protein domains\n",
    "## AF : Frequency of existing variant in 1000 Genomes combined population\n",
    "## AFR_AF : Frequency of existing variant in 1000 Genomes combined African population\n",
    "## AMR_AF : Frequency of existing variant in 1000 Genomes combined American population\n",
    "## EAS_AF : Frequency of existing variant in 1000 Genomes combined East Asian population\n",
    "## EUR_AF : Frequency of existing variant in 1000 Genomes combined European population\n",
    "## SAS_AF : Frequency of existing variant in 1000 Genomes combined South Asian population\n",
    "## gnomAD_AF : Frequency of existing variant in gnomAD exomes combined population\n",
    "## gnomAD_AFR_AF : Frequency of existing variant in gnomAD exomes African/American population\n",
    "## gnomAD_AMR_AF : Frequency of existing variant in gnomAD exomes American population\n",
    "## gnomAD_ASJ_AF : Frequency of existing variant in gnomAD exomes Ashkenazi Jewish population\n",
    "## gnomAD_EAS_AF : Frequency of existing variant in gnomAD exomes East Asian population\n",
    "## gnomAD_FIN_AF : Frequency of existing variant in gnomAD exomes Finnish population\n",
    "## gnomAD_NFE_AF : Frequency of existing variant in gnomAD exomes Non-Finnish European population\n",
    "## gnomAD_OTH_AF : Frequency of existing variant in gnomAD exomes other combined populations\n",
    "## gnomAD_SAS_AF : Frequency of existing variant in gnomAD exomes South Asian population\n",
    "## MAX_AF : Maximum observed allele frequency in 1000 Genomes, ESP and ExAC/gnomAD\n",
    "## MAX_AF_POPS : Populations in which maximum allele frequency was observed\n",
    "## CLIN_SIG : ClinVar clinical significance of the dbSNP variant\n",
    "## SOMATIC : Somatic status of existing variant\n",
    "## PHENO : Indicates if existing variant(s) is associated with a phenotype, disease or trait; multiple values correspond to multiple variants\n",
    "## MOTIF_NAME : The source and identifier of a transcription factor binding profile (TFBP) aligned at this position\n",
    "## MOTIF_POS : The relative position of the variation in the aligned TFBP\n",
    "## HIGH_INF_POS : A flag indicating if the variant falls in a high information position of the TFBP\n",
    "## MOTIF_SCORE_CHANGE : The difference in motif score of the reference and variant sequences for the TFBP\n",
    "## BLOSUM62 : BLOSUM62 substitution score for the reference and alternative amino acids\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/Users/mariapalafox/Desktop/ensembl-vep/CpDCysLys_mapping\")\n",
    "vep = \"VEP_detected_filtered_missenseonly_header_removed.txt\"\n",
    "vep = pd.read_csv(vep, delimiter=\"\\t\", low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "disease = vep[['#Uploaded_variation', 'Location', 'Allele', 'Gene', 'Feature',\n",
    "       'Feature_type', 'Consequence', 'cDNA_position', 'CDS_position',\n",
    "       'Protein_position', 'Amino_acids', 'Codons', 'SYMBOL', 'Existing_variation',\n",
    "       'IMPACT','GENE_PHENO','MAX_AF', 'CLIN_SIG', 'SOMATIC', 'PHENO']].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>GENE_PHENO</th>\n",
       "      <th>MAX_AF</th>\n",
       "      <th>CLIN_SIG</th>\n",
       "      <th>SOMATIC</th>\n",
       "      <th>PHENO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>10_000093270_A_C;Q3ZCM7_C354;Cys/Trp</td>\n",
       "      <td>10:93270</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000173876</td>\n",
       "      <td>ENST00000309812</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1125/1563</td>\n",
       "      <td>1062/1335</td>\n",
       "      <td>354/444</td>\n",
       "      <td>C/W</td>\n",
       "      <td>tgT/tgG</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>10_000093270_A_C;Q3ZCM7_C354;Cys/Trp</td>\n",
       "      <td>10:93270</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000173876</td>\n",
       "      <td>ENST00000447903</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1162/1604</td>\n",
       "      <td>846/1119</td>\n",
       "      <td>282/372</td>\n",
       "      <td>C/W</td>\n",
       "      <td>tgT/tgG</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>10_000093271_C_A;Q3ZCM7_C354;Cys/Phe</td>\n",
       "      <td>10:93271</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000173876</td>\n",
       "      <td>ENST00000309812</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1124/1563</td>\n",
       "      <td>1061/1335</td>\n",
       "      <td>354/444</td>\n",
       "      <td>C/F</td>\n",
       "      <td>tGt/tTt</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>-</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    #Uploaded_variation  Location Allele             Gene  \\\n",
       "0  10_000093270_A_C;Q3ZCM7_C354;Cys/Trp  10:93270      G  ENSG00000173876   \n",
       "1  10_000093270_A_C;Q3ZCM7_C354;Cys/Trp  10:93270      G  ENSG00000173876   \n",
       "2  10_000093271_C_A;Q3ZCM7_C354;Cys/Phe  10:93271      T  ENSG00000173876   \n",
       "\n",
       "           Feature Feature_type       Consequence cDNA_position CDS_position  \\\n",
       "0  ENST00000309812   Transcript  missense_variant     1125/1563    1062/1335   \n",
       "1  ENST00000447903   Transcript  missense_variant     1162/1604     846/1119   \n",
       "2  ENST00000309812   Transcript  missense_variant     1124/1563    1061/1335   \n",
       "\n",
       "  Protein_position Amino_acids   Codons SYMBOL Existing_variation    IMPACT  \\\n",
       "0          354/444         C/W  tgT/tgG  TUBB8                  -  MODERATE   \n",
       "1          282/372         C/W  tgT/tgG  TUBB8                  -  MODERATE   \n",
       "2          354/444         C/F  tGt/tTt  TUBB8                  -  MODERATE   \n",
       "\n",
       "  GENE_PHENO MAX_AF CLIN_SIG SOMATIC PHENO  \n",
       "0          -      -        -       -     -  \n",
       "1          -      -        -       -     -  \n",
       "2          -      -        -       -     -  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'MODERATE': 314238}\n"
     ]
    }
   ],
   "source": [
    "colvalues(disease, 'IMPACT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len of col:  314238\n",
      "len of col set:  20\n",
      "\n",
      "\n",
      "{'-': 313590, 'uncertain_significance': 322, 'likely_pathogenic': 68, 'benign': 7, 'pathogenic': 77, 'likely_benign': 54, 'likely_pathogenic,pathogenic': 15, 'uncertain_significance,not_provided': 4, 'likely_benign,other': 1, 'other': 7, 'uncertain_significance,likely_benign': 21, 'uncertain_significance,likely_pathogenic': 11, 'uncertain_significance,pathogenic': 15, 'not_provided': 15, 'not_provided,pathogenic': 18, 'uncertain_significance,likely_benign,likely_pathogenic': 2, 'uncertain_significance,likely_pathogenic,pathogenic': 1, 'benign,likely_pathogenic': 6, 'not_provided,likely_pathogenic': 1, 'pathogenic,other': 3}\n"
     ]
    }
   ],
   "source": [
    "# CLINVAR VARIANTS\n",
    "uniqueCount(disease, 'CLIN_SIG')\n",
    "print()\n",
    "colvalues(disease, 'CLIN_SIG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>GENE_PHENO</th>\n",
       "      <th>MAX_AF</th>\n",
       "      <th>CLIN_SIG</th>\n",
       "      <th>SOMATIC</th>\n",
       "      <th>PHENO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>6391</td>\n",
       "      <td>10_075854066_A_C;P18206_K464;Lys/Gln</td>\n",
       "      <td>10:75854066</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000035403</td>\n",
       "      <td>ENST00000211998</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1484/5482</td>\n",
       "      <td>1390/3405</td>\n",
       "      <td>464/1134</td>\n",
       "      <td>K/Q</td>\n",
       "      <td>Aaa/Caa</td>\n",
       "      <td>VCL</td>\n",
       "      <td>rs200710859,COSM1964803</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>0,1</td>\n",
       "      <td>0,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6392</td>\n",
       "      <td>10_075854066_A_C;P18206_K464;Lys/Gln</td>\n",
       "      <td>10:75854066</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000035403</td>\n",
       "      <td>ENST00000372755</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1484/5286</td>\n",
       "      <td>1390/3201</td>\n",
       "      <td>464/1066</td>\n",
       "      <td>K/Q</td>\n",
       "      <td>Aaa/Caa</td>\n",
       "      <td>VCL</td>\n",
       "      <td>rs200710859,COSM1964803</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>0,1</td>\n",
       "      <td>0,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6393</td>\n",
       "      <td>10_075854066_A_G;P18206_K464;Lys/Glu</td>\n",
       "      <td>10:75854066</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000035403</td>\n",
       "      <td>ENST00000211998</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1484/5482</td>\n",
       "      <td>1390/3405</td>\n",
       "      <td>464/1134</td>\n",
       "      <td>K/E</td>\n",
       "      <td>Aaa/Gaa</td>\n",
       "      <td>VCL</td>\n",
       "      <td>rs200710859,COSM1964803</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0005981</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>0,1</td>\n",
       "      <td>0,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6394</td>\n",
       "      <td>10_075854066_A_G;P18206_K464;Lys/Glu</td>\n",
       "      <td>10:75854066</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000035403</td>\n",
       "      <td>ENST00000372755</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1484/5286</td>\n",
       "      <td>1390/3201</td>\n",
       "      <td>464/1066</td>\n",
       "      <td>K/E</td>\n",
       "      <td>Aaa/Gaa</td>\n",
       "      <td>VCL</td>\n",
       "      <td>rs200710859,COSM1964803</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0005981</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>0,1</td>\n",
       "      <td>0,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6405</td>\n",
       "      <td>10_075857037_A_C;P18206_K607;Lys/Gln</td>\n",
       "      <td>10:75857037</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000035403</td>\n",
       "      <td>ENST00000211998</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1913/5482</td>\n",
       "      <td>1819/3405</td>\n",
       "      <td>607/1134</td>\n",
       "      <td>K/Q</td>\n",
       "      <td>Aag/Cag</td>\n",
       "      <td>VCL</td>\n",
       "      <td>rs794729192</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>313617</td>\n",
       "      <td>X_153760913_T_C;P11413_K386;Lys/Glu</td>\n",
       "      <td>X:153760913</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000160211</td>\n",
       "      <td>ENST00000393562</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1630/2631</td>\n",
       "      <td>1246/1638</td>\n",
       "      <td>416/545</td>\n",
       "      <td>K/E</td>\n",
       "      <td>Aag/Gag</td>\n",
       "      <td>G6PD</td>\n",
       "      <td>rs137852320,CM930278</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>pathogenic,other</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>313618</td>\n",
       "      <td>X_153760913_T_C;P11413_K386;Lys/Glu</td>\n",
       "      <td>X:153760913</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000160211</td>\n",
       "      <td>ENST00000393564</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1269/1661</td>\n",
       "      <td>1156/1548</td>\n",
       "      <td>386/515</td>\n",
       "      <td>K/E</td>\n",
       "      <td>Aag/Gag</td>\n",
       "      <td>G6PD</td>\n",
       "      <td>rs137852320,CM930278</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>pathogenic,other</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>313637</td>\n",
       "      <td>X_153760916_A_G;P11413_C385;Cys/Arg</td>\n",
       "      <td>X:153760916</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000160211</td>\n",
       "      <td>ENST00000369620</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1404/1799</td>\n",
       "      <td>1291/1686</td>\n",
       "      <td>431/561</td>\n",
       "      <td>C/R</td>\n",
       "      <td>Tgc/Cgc</td>\n",
       "      <td>G6PD</td>\n",
       "      <td>rs137852322,CM930277</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>other</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>313638</td>\n",
       "      <td>X_153760916_A_G;P11413_C385;Cys/Arg</td>\n",
       "      <td>X:153760916</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000160211</td>\n",
       "      <td>ENST00000393562</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1627/2631</td>\n",
       "      <td>1243/1638</td>\n",
       "      <td>415/545</td>\n",
       "      <td>C/R</td>\n",
       "      <td>Tgc/Cgc</td>\n",
       "      <td>G6PD</td>\n",
       "      <td>rs137852322,CM930277</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>other</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>313639</td>\n",
       "      <td>X_153760916_A_G;P11413_C385;Cys/Arg</td>\n",
       "      <td>X:153760916</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000160211</td>\n",
       "      <td>ENST00000393564</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1266/1661</td>\n",
       "      <td>1153/1548</td>\n",
       "      <td>385/515</td>\n",
       "      <td>C/R</td>\n",
       "      <td>Tgc/Cgc</td>\n",
       "      <td>G6PD</td>\n",
       "      <td>rs137852322,CM930277</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>other</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>648 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         #Uploaded_variation     Location Allele  \\\n",
       "6391    10_075854066_A_C;P18206_K464;Lys/Gln  10:75854066      C   \n",
       "6392    10_075854066_A_C;P18206_K464;Lys/Gln  10:75854066      C   \n",
       "6393    10_075854066_A_G;P18206_K464;Lys/Glu  10:75854066      G   \n",
       "6394    10_075854066_A_G;P18206_K464;Lys/Glu  10:75854066      G   \n",
       "6405    10_075857037_A_C;P18206_K607;Lys/Gln  10:75857037      C   \n",
       "...                                      ...          ...    ...   \n",
       "313617   X_153760913_T_C;P11413_K386;Lys/Glu  X:153760913      G   \n",
       "313618   X_153760913_T_C;P11413_K386;Lys/Glu  X:153760913      G   \n",
       "313637   X_153760916_A_G;P11413_C385;Cys/Arg  X:153760916      C   \n",
       "313638   X_153760916_A_G;P11413_C385;Cys/Arg  X:153760916      C   \n",
       "313639   X_153760916_A_G;P11413_C385;Cys/Arg  X:153760916      C   \n",
       "\n",
       "                   Gene          Feature Feature_type       Consequence  \\\n",
       "6391    ENSG00000035403  ENST00000211998   Transcript  missense_variant   \n",
       "6392    ENSG00000035403  ENST00000372755   Transcript  missense_variant   \n",
       "6393    ENSG00000035403  ENST00000211998   Transcript  missense_variant   \n",
       "6394    ENSG00000035403  ENST00000372755   Transcript  missense_variant   \n",
       "6405    ENSG00000035403  ENST00000211998   Transcript  missense_variant   \n",
       "...                 ...              ...          ...               ...   \n",
       "313617  ENSG00000160211  ENST00000393562   Transcript  missense_variant   \n",
       "313618  ENSG00000160211  ENST00000393564   Transcript  missense_variant   \n",
       "313637  ENSG00000160211  ENST00000369620   Transcript  missense_variant   \n",
       "313638  ENSG00000160211  ENST00000393562   Transcript  missense_variant   \n",
       "313639  ENSG00000160211  ENST00000393564   Transcript  missense_variant   \n",
       "\n",
       "       cDNA_position CDS_position Protein_position Amino_acids   Codons  \\\n",
       "6391       1484/5482    1390/3405         464/1134         K/Q  Aaa/Caa   \n",
       "6392       1484/5286    1390/3201         464/1066         K/Q  Aaa/Caa   \n",
       "6393       1484/5482    1390/3405         464/1134         K/E  Aaa/Gaa   \n",
       "6394       1484/5286    1390/3201         464/1066         K/E  Aaa/Gaa   \n",
       "6405       1913/5482    1819/3405         607/1134         K/Q  Aag/Cag   \n",
       "...              ...          ...              ...         ...      ...   \n",
       "313617     1630/2631    1246/1638          416/545         K/E  Aag/Gag   \n",
       "313618     1269/1661    1156/1548          386/515         K/E  Aag/Gag   \n",
       "313637     1404/1799    1291/1686          431/561         C/R  Tgc/Cgc   \n",
       "313638     1627/2631    1243/1638          415/545         C/R  Tgc/Cgc   \n",
       "313639     1266/1661    1153/1548          385/515         C/R  Tgc/Cgc   \n",
       "\n",
       "       SYMBOL       Existing_variation    IMPACT GENE_PHENO     MAX_AF  \\\n",
       "6391      VCL  rs200710859,COSM1964803  MODERATE          1          -   \n",
       "6392      VCL  rs200710859,COSM1964803  MODERATE          1          -   \n",
       "6393      VCL  rs200710859,COSM1964803  MODERATE          1  0.0005981   \n",
       "6394      VCL  rs200710859,COSM1964803  MODERATE          1  0.0005981   \n",
       "6405      VCL              rs794729192  MODERATE          1          -   \n",
       "...       ...                      ...       ...        ...        ...   \n",
       "313617   G6PD     rs137852320,CM930278  MODERATE          1          -   \n",
       "313618   G6PD     rs137852320,CM930278  MODERATE          1          -   \n",
       "313637   G6PD     rs137852322,CM930277  MODERATE          1          -   \n",
       "313638   G6PD     rs137852322,CM930277  MODERATE          1          -   \n",
       "313639   G6PD     rs137852322,CM930277  MODERATE          1          -   \n",
       "\n",
       "                      CLIN_SIG SOMATIC PHENO  \n",
       "6391    uncertain_significance     0,1   0,1  \n",
       "6392    uncertain_significance     0,1   0,1  \n",
       "6393    uncertain_significance     0,1   0,1  \n",
       "6394    uncertain_significance     0,1   0,1  \n",
       "6405    uncertain_significance       -     -  \n",
       "...                        ...     ...   ...  \n",
       "313617        pathogenic,other       -   1,1  \n",
       "313618        pathogenic,other       -   1,1  \n",
       "313637                   other       -   1,1  \n",
       "313638                   other       -   1,1  \n",
       "313639                   other       -   1,1  \n",
       "\n",
       "[648 rows x 20 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filtering out '-' rows in 'CLIN_SIG'\n",
    "clinvar = disease[disease['CLIN_SIG'] != '-'].copy()\n",
    "clinvar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>GENE_PHENO</th>\n",
       "      <th>MAX_AF</th>\n",
       "      <th>CLIN_SIG</th>\n",
       "      <th>SOMATIC</th>\n",
       "      <th>PHENO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>6391</td>\n",
       "      <td>10_075854066_A_C;P18206_K464;Lys/Gln</td>\n",
       "      <td>10:75854066</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000035403</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>K/Q</td>\n",
       "      <td>Aaa/Caa</td>\n",
       "      <td>VCL</td>\n",
       "      <td>rs200710859,COSM1964803</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>0,1</td>\n",
       "      <td>0,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>6393</td>\n",
       "      <td>10_075854066_A_G;P18206_K464;Lys/Glu</td>\n",
       "      <td>10:75854066</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000035403</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>K/E</td>\n",
       "      <td>Aaa/Gaa</td>\n",
       "      <td>VCL</td>\n",
       "      <td>rs200710859,COSM1964803</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0005981</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>0,1</td>\n",
       "      <td>0,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>6405</td>\n",
       "      <td>10_075857037_A_C;P18206_K607;Lys/Gln</td>\n",
       "      <td>10:75857037</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000035403</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>K/Q</td>\n",
       "      <td>Aag/Cag</td>\n",
       "      <td>VCL</td>\n",
       "      <td>rs794729192</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>8902</td>\n",
       "      <td>10_097373888_A_G;P54886_C546;Cys/Arg</td>\n",
       "      <td>10:97373888</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000059573</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>C/R</td>\n",
       "      <td>Tgc/Cgc</td>\n",
       "      <td>ALDH18A1</td>\n",
       "      <td>rs753157831</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0002977</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>9010</td>\n",
       "      <td>10_097396926_C_T;P54886_C161;Cys/Tyr</td>\n",
       "      <td>10:97396926</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000059573</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>C/Y</td>\n",
       "      <td>tGt/tAt</td>\n",
       "      <td>ALDH18A1</td>\n",
       "      <td>rs1057518541</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>likely_pathogenic</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>196</td>\n",
       "      <td>312908</td>\n",
       "      <td>X_153591114_T_A;P21333_K773;Lys/Asn</td>\n",
       "      <td>X:153591114</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000196924</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>K/N</td>\n",
       "      <td>aaA/aaT</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>rs1060500719</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>197</td>\n",
       "      <td>313248</td>\n",
       "      <td>X_153596100_C_A;P21333_C210;Cys/Phe</td>\n",
       "      <td>X:153596100</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000196924</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>C/F</td>\n",
       "      <td>tGt/tTt</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>rs137853318,CM071770</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>198</td>\n",
       "      <td>313360</td>\n",
       "      <td>X_153608077_A_T;P50402_K37;Lys/Met</td>\n",
       "      <td>X:153608077</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000102119</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>K/M</td>\n",
       "      <td>aAg/aTg</td>\n",
       "      <td>EMD</td>\n",
       "      <td>rs1085307681,CD084833</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>0,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>199</td>\n",
       "      <td>313616</td>\n",
       "      <td>X_153760913_T_C;P11413_K386;Lys/Glu</td>\n",
       "      <td>X:153760913</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000160211</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>K/E</td>\n",
       "      <td>Aag/Gag</td>\n",
       "      <td>G6PD</td>\n",
       "      <td>rs137852320,CM930278</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>pathogenic,other</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>200</td>\n",
       "      <td>313637</td>\n",
       "      <td>X_153760916_A_G;P11413_C385;Cys/Arg</td>\n",
       "      <td>X:153760916</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000160211</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>C/R</td>\n",
       "      <td>Tgc/Cgc</td>\n",
       "      <td>G6PD</td>\n",
       "      <td>rs137852322,CM930277</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>other</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>201 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index                   #Uploaded_variation     Location Allele  \\\n",
       "0      6391  10_075854066_A_C;P18206_K464;Lys/Gln  10:75854066      C   \n",
       "1      6393  10_075854066_A_G;P18206_K464;Lys/Glu  10:75854066      G   \n",
       "2      6405  10_075857037_A_C;P18206_K607;Lys/Gln  10:75857037      C   \n",
       "3      8902  10_097373888_A_G;P54886_C546;Cys/Arg  10:97373888      C   \n",
       "4      9010  10_097396926_C_T;P54886_C161;Cys/Tyr  10:97396926      A   \n",
       "..      ...                                   ...          ...    ...   \n",
       "196  312908   X_153591114_T_A;P21333_K773;Lys/Asn  X:153591114      T   \n",
       "197  313248   X_153596100_C_A;P21333_C210;Cys/Phe  X:153596100      T   \n",
       "198  313360    X_153608077_A_T;P50402_K37;Lys/Met  X:153608077      T   \n",
       "199  313616   X_153760913_T_C;P11413_K386;Lys/Glu  X:153760913      G   \n",
       "200  313637   X_153760916_A_G;P11413_C385;Cys/Arg  X:153760916      C   \n",
       "\n",
       "                Gene Feature_type       Consequence Amino_acids   Codons  \\\n",
       "0    ENSG00000035403   Transcript  missense_variant         K/Q  Aaa/Caa   \n",
       "1    ENSG00000035403   Transcript  missense_variant         K/E  Aaa/Gaa   \n",
       "2    ENSG00000035403   Transcript  missense_variant         K/Q  Aag/Cag   \n",
       "3    ENSG00000059573   Transcript  missense_variant         C/R  Tgc/Cgc   \n",
       "4    ENSG00000059573   Transcript  missense_variant         C/Y  tGt/tAt   \n",
       "..               ...          ...               ...         ...      ...   \n",
       "196  ENSG00000196924   Transcript  missense_variant         K/N  aaA/aaT   \n",
       "197  ENSG00000196924   Transcript  missense_variant         C/F  tGt/tTt   \n",
       "198  ENSG00000102119   Transcript  missense_variant         K/M  aAg/aTg   \n",
       "199  ENSG00000160211   Transcript  missense_variant         K/E  Aag/Gag   \n",
       "200  ENSG00000160211   Transcript  missense_variant         C/R  Tgc/Cgc   \n",
       "\n",
       "       SYMBOL       Existing_variation    IMPACT GENE_PHENO     MAX_AF  \\\n",
       "0         VCL  rs200710859,COSM1964803  MODERATE          1          -   \n",
       "1         VCL  rs200710859,COSM1964803  MODERATE          1  0.0005981   \n",
       "2         VCL              rs794729192  MODERATE          1          -   \n",
       "3    ALDH18A1              rs753157831  MODERATE          1  0.0002977   \n",
       "4    ALDH18A1             rs1057518541  MODERATE          1          -   \n",
       "..        ...                      ...       ...        ...        ...   \n",
       "196      FLNA             rs1060500719  MODERATE          1          -   \n",
       "197      FLNA     rs137853318,CM071770  MODERATE          1          -   \n",
       "198       EMD    rs1085307681,CD084833  MODERATE          1          -   \n",
       "199      G6PD     rs137852320,CM930278  MODERATE          1          -   \n",
       "200      G6PD     rs137852322,CM930277  MODERATE          1          -   \n",
       "\n",
       "                   CLIN_SIG SOMATIC PHENO  \n",
       "0    uncertain_significance     0,1   0,1  \n",
       "1    uncertain_significance     0,1   0,1  \n",
       "2    uncertain_significance       -     -  \n",
       "3    uncertain_significance       -     -  \n",
       "4         likely_pathogenic       -     -  \n",
       "..                      ...     ...   ...  \n",
       "196  uncertain_significance       -     1  \n",
       "197              pathogenic       -   1,1  \n",
       "198  uncertain_significance       -   0,1  \n",
       "199        pathogenic,other       -   1,1  \n",
       "200                   other       -   1,1  \n",
       "\n",
       "[201 rows x 17 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dropping Feature column then deleting all duplicate rows\n",
    "clinvar.drop(['Feature','cDNA_position','CDS_position','Protein_position'], axis=1, inplace=True)\n",
    "\n",
    "clinvar.drop_duplicates(inplace=True)\n",
    "clinvar.reset_index(inplace=True)\n",
    "clinvar"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### filtering note: \n",
    "- dropped columns related to TX ID, cDNA CDS and protein position,these were causing redundancy in CLIN_SIG columns...needed to drop redundant rows to get accurate count of overlapping clinvar variants of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'uncertain_significance': 97, 'likely_pathogenic': 19, 'benign': 5, 'pathogenic': 29, 'likely_benign': 17, 'likely_pathogenic,pathogenic': 4, 'uncertain_significance,not_provided': 1, 'likely_benign,other': 1, 'other': 5, 'uncertain_significance,likely_benign': 6, 'uncertain_significance,likely_pathogenic': 2, 'uncertain_significance,pathogenic': 3, 'not_provided': 3, 'not_provided,pathogenic': 2, 'uncertain_significance,likely_benign,likely_pathogenic': 1, 'uncertain_significance,likely_pathogenic,pathogenic': 1, 'benign,likely_pathogenic': 3, 'not_provided,likely_pathogenic': 1, 'pathogenic,other': 1}\n"
     ]
    }
   ],
   "source": [
    "colvalues(clinvar, 'CLIN_SIG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>GENE_PHENO</th>\n",
       "      <th>MAX_AF</th>\n",
       "      <th>CLIN_SIG</th>\n",
       "      <th>SOMATIC</th>\n",
       "      <th>PHENO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>6391</td>\n",
       "      <td>10_075854066_A_C;P18206_K464;Lys/Gln</td>\n",
       "      <td>10:75854066</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000035403</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>K/Q</td>\n",
       "      <td>Aaa/Caa</td>\n",
       "      <td>VCL</td>\n",
       "      <td>rs200710859,COSM1964803</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>0,1</td>\n",
       "      <td>0,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>6393</td>\n",
       "      <td>10_075854066_A_G;P18206_K464;Lys/Glu</td>\n",
       "      <td>10:75854066</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000035403</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>K/E</td>\n",
       "      <td>Aaa/Gaa</td>\n",
       "      <td>VCL</td>\n",
       "      <td>rs200710859,COSM1964803</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0005981</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>0,1</td>\n",
       "      <td>0,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>6405</td>\n",
       "      <td>10_075857037_A_C;P18206_K607;Lys/Gln</td>\n",
       "      <td>10:75857037</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000035403</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>K/Q</td>\n",
       "      <td>Aag/Cag</td>\n",
       "      <td>VCL</td>\n",
       "      <td>rs794729192</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>8902</td>\n",
       "      <td>10_097373888_A_G;P54886_C546;Cys/Arg</td>\n",
       "      <td>10:97373888</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000059573</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>C/R</td>\n",
       "      <td>Tgc/Cgc</td>\n",
       "      <td>ALDH18A1</td>\n",
       "      <td>rs753157831</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0002977</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>9010</td>\n",
       "      <td>10_097396926_C_T;P54886_C161;Cys/Tyr</td>\n",
       "      <td>10:97396926</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000059573</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>C/Y</td>\n",
       "      <td>tGt/tAt</td>\n",
       "      <td>ALDH18A1</td>\n",
       "      <td>rs1057518541</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>likely_pathogenic</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>196</td>\n",
       "      <td>312908</td>\n",
       "      <td>X_153591114_T_A;P21333_K773;Lys/Asn</td>\n",
       "      <td>X:153591114</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000196924</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>K/N</td>\n",
       "      <td>aaA/aaT</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>rs1060500719</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>197</td>\n",
       "      <td>313248</td>\n",
       "      <td>X_153596100_C_A;P21333_C210;Cys/Phe</td>\n",
       "      <td>X:153596100</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000196924</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>C/F</td>\n",
       "      <td>tGt/tTt</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>rs137853318,CM071770</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>198</td>\n",
       "      <td>313360</td>\n",
       "      <td>X_153608077_A_T;P50402_K37;Lys/Met</td>\n",
       "      <td>X:153608077</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000102119</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>K/M</td>\n",
       "      <td>aAg/aTg</td>\n",
       "      <td>EMD</td>\n",
       "      <td>rs1085307681,CD084833</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>0,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>199</td>\n",
       "      <td>313616</td>\n",
       "      <td>X_153760913_T_C;P11413_K386;Lys/Glu</td>\n",
       "      <td>X:153760913</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000160211</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>K/E</td>\n",
       "      <td>Aag/Gag</td>\n",
       "      <td>G6PD</td>\n",
       "      <td>rs137852320,CM930278</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>pathogenic,other</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>200</td>\n",
       "      <td>313637</td>\n",
       "      <td>X_153760916_A_G;P11413_C385;Cys/Arg</td>\n",
       "      <td>X:153760916</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000160211</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>C/R</td>\n",
       "      <td>Tgc/Cgc</td>\n",
       "      <td>G6PD</td>\n",
       "      <td>rs137852322,CM930277</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>other</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>201 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index                   #Uploaded_variation     Location Allele  \\\n",
       "0      6391  10_075854066_A_C;P18206_K464;Lys/Gln  10:75854066      C   \n",
       "1      6393  10_075854066_A_G;P18206_K464;Lys/Glu  10:75854066      G   \n",
       "2      6405  10_075857037_A_C;P18206_K607;Lys/Gln  10:75857037      C   \n",
       "3      8902  10_097373888_A_G;P54886_C546;Cys/Arg  10:97373888      C   \n",
       "4      9010  10_097396926_C_T;P54886_C161;Cys/Tyr  10:97396926      A   \n",
       "..      ...                                   ...          ...    ...   \n",
       "196  312908   X_153591114_T_A;P21333_K773;Lys/Asn  X:153591114      T   \n",
       "197  313248   X_153596100_C_A;P21333_C210;Cys/Phe  X:153596100      T   \n",
       "198  313360    X_153608077_A_T;P50402_K37;Lys/Met  X:153608077      T   \n",
       "199  313616   X_153760913_T_C;P11413_K386;Lys/Glu  X:153760913      G   \n",
       "200  313637   X_153760916_A_G;P11413_C385;Cys/Arg  X:153760916      C   \n",
       "\n",
       "                Gene Feature_type       Consequence Amino_acids   Codons  \\\n",
       "0    ENSG00000035403   Transcript  missense_variant         K/Q  Aaa/Caa   \n",
       "1    ENSG00000035403   Transcript  missense_variant         K/E  Aaa/Gaa   \n",
       "2    ENSG00000035403   Transcript  missense_variant         K/Q  Aag/Cag   \n",
       "3    ENSG00000059573   Transcript  missense_variant         C/R  Tgc/Cgc   \n",
       "4    ENSG00000059573   Transcript  missense_variant         C/Y  tGt/tAt   \n",
       "..               ...          ...               ...         ...      ...   \n",
       "196  ENSG00000196924   Transcript  missense_variant         K/N  aaA/aaT   \n",
       "197  ENSG00000196924   Transcript  missense_variant         C/F  tGt/tTt   \n",
       "198  ENSG00000102119   Transcript  missense_variant         K/M  aAg/aTg   \n",
       "199  ENSG00000160211   Transcript  missense_variant         K/E  Aag/Gag   \n",
       "200  ENSG00000160211   Transcript  missense_variant         C/R  Tgc/Cgc   \n",
       "\n",
       "       SYMBOL       Existing_variation    IMPACT GENE_PHENO     MAX_AF  \\\n",
       "0         VCL  rs200710859,COSM1964803  MODERATE          1          -   \n",
       "1         VCL  rs200710859,COSM1964803  MODERATE          1  0.0005981   \n",
       "2         VCL              rs794729192  MODERATE          1          -   \n",
       "3    ALDH18A1              rs753157831  MODERATE          1  0.0002977   \n",
       "4    ALDH18A1             rs1057518541  MODERATE          1          -   \n",
       "..        ...                      ...       ...        ...        ...   \n",
       "196      FLNA             rs1060500719  MODERATE          1          -   \n",
       "197      FLNA     rs137853318,CM071770  MODERATE          1          -   \n",
       "198       EMD    rs1085307681,CD084833  MODERATE          1          -   \n",
       "199      G6PD     rs137852320,CM930278  MODERATE          1          -   \n",
       "200      G6PD     rs137852322,CM930277  MODERATE          1          -   \n",
       "\n",
       "                   CLIN_SIG SOMATIC PHENO  \n",
       "0    uncertain_significance     0,1   0,1  \n",
       "1    uncertain_significance     0,1   0,1  \n",
       "2    uncertain_significance       -     -  \n",
       "3    uncertain_significance       -     -  \n",
       "4         likely_pathogenic       -     -  \n",
       "..                      ...     ...   ...  \n",
       "196  uncertain_significance       -     1  \n",
       "197              pathogenic       -   1,1  \n",
       "198  uncertain_significance       -   0,1  \n",
       "199        pathogenic,other       -   1,1  \n",
       "200                   other       -   1,1  \n",
       "\n",
       "[201 rows x 17 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'K/Q': 11, 'K/E': 29, 'C/R': 25, 'C/Y': 29, 'C/F': 21, 'K/T': 7, 'K/R': 19, 'K/M': 4, 'K/N': 19, 'K/I': 1, 'C/W': 12, 'C/G': 9, 'C/S': 15}\n"
     ]
    }
   ],
   "source": [
    "colvalues(clinvar, 'Amino_acids')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar.to_csv(\"VEP_detectedCK_subaim2.3_CLINVAR_SIG_201missense.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
